{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "114ce4a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<img src onerror=\"\n",
       "                    // get all code cells\n",
       "                    let all_cells = document.querySelectorAll('.code_cell,.jp-CodeCell')\n",
       "                    // define the cell selection\n",
       "                    let selection = Array(3,5,8,9,10,14);\n",
       "                    // select the cells\n",
       "                    let sel_cells = selection.map(x=>all_cells[x]);\n",
       "                    // select the parts of the cells that should be colored (code area)\n",
       "                    let ca = sel_cells.map(e=>Array.from(e.querySelectorAll('.CodeMirror-sizer,.highlight'))).flat();\n",
       "                    // change the background color of all selected elements\n",
       "                    ca.forEach(e=>e.style.background='PowderBlue');\n",
       "                    \" style=\"display:none\">"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/workspace2/jdetlef/.conda/envs/247-tobias/lib/python3.12/site-packages/sctoolbox/utils/jupyter.py:132: UserWarning: The Notebook seems to be outdated (notebook version: N/A). Some functions may not work! Consider updating.\n",
      "  warnings.warn(\"The Notebook seems to be outdated (notebook version: N/A). Some functions may not work! Consider updating.\")\n"
     ]
    }
   ],
   "source": [
    "from sctoolbox.utils.jupyter import bgcolor, _compare_version\n",
    "\n",
    "# change the background of input cells\n",
    "bgcolor(\"PowderBlue\", select=[3, 5, 8, 9, 10, 14])\n",
    "\n",
    "nb_name = \"05_tobias.ipynb\"\n",
    "\n",
    "_compare_version(nb_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d313e854",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "import sctoolbox.tools.bam as bam\n",
    "import sctoolbox.tools.tobias as tobias\n",
    "\n",
    "def prepare_tobias(adata,\n",
    "                   cluster_column,\n",
    "                   barcode_column,\n",
    "                   path_out,\n",
    "                   path_bam,\n",
    "                   barcode_tag='CB',\n",
    "                   path_TOBIAS_fasta=None,\n",
    "                   path_TOBIAS_motifs=None,\n",
    "                   path_TOBIAS_gtf=None,\n",
    "                   path_TOBIAS_blacklist=None,\n",
    "                   TOBIAS_organism='human',\n",
    "                   TOBIAS_yml=\"TOBIAS_config.yml\",\n",
    "                   plot_comparison=True,\n",
    "                   plot_correction=True,\n",
    "                   plot_venn=True,\n",
    "                   coverage=False,\n",
    "                   wilson=False,\n",
    "                   threads=4):\n",
    "    \"\"\"\n",
    "    Split ATAC-seq bamfile by cluster and prepare TOBIAS run.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata : AnnData\n",
    "        Annotated data matrix.\n",
    "    cluster_column : str\n",
    "        Column in adata.obs that contains the cluster information.\n",
    "    barcode_column : str\n",
    "        Column in adata.obs that contains the barcode information.\n",
    "    path_out : str\n",
    "        Path to save the output files.\n",
    "    path_bam : str\n",
    "        Path\n",
    "    barcode_tag : str\n",
    "        Tag to extract the barcode from the read name.\n",
    "    path_TOBIAS_fasta : str\n",
    "        Path to the organism fasta file.\n",
    "    path_TOBIAS_motifs : str\n",
    "        Path to the motifs file.\n",
    "    path_TOBIAS_gtf : str\n",
    "        Path to the organisms gtf file (genes).\n",
    "    path_TOBIAS_blacklist : str\n",
    "        Path to the blacklist file.\n",
    "    TOBIAS_organism : str\n",
    "        Organism for TOBIAS.\n",
    "    TOBIAS_yml : str\n",
    "        Name of TOBIAS config yml. Cannot be named \"config.yml\" or it will be overwritten when running TOBIAS.\n",
    "    plot_comparison : bool\n",
    "        TOBIAS flag for plotting comparison between condition.\n",
    "    plot_correction : bool\n",
    "        TOBIAS flag for plotting correction.\n",
    "    plot_venn : bool\n",
    "        TOBIAS flag for plotting venn diagramms.\n",
    "    coverage : bool\n",
    "        TOBIAS flag for coverage calculation.\n",
    "    wilson : bool\n",
    "        TOBIAS flag for wilson calculation.\n",
    "    threads : int\n",
    "        Number of threads to use.\n",
    "    \"\"\"\n",
    "\n",
    "    # Check if output directory exists\n",
    "    if not os.path.exists(path_out):\n",
    "        raise OSError(f\"The file or directory \\'{path_out}\\' does not exist. Please give valid input.\")\n",
    "\n",
    "    # Make the TOBIAS directory\n",
    "    path_TOBIAS = os.path.join(path_out, 'TOBIAS_run')\n",
    "\n",
    "    # Check if directory for TOBIAS run exists, if not create it\n",
    "    if os.path.exists(path_TOBIAS):\n",
    "        print(f\"WARNING: The directory \\'{path_TOBIAS}\\' already exists. Any files in this directory may be overwritten, which can cause inconsistencies.\")\n",
    "    else:\n",
    "        os.mkdir(path_TOBIAS)\n",
    "\n",
    "    # Get path for TOBIAS input and create directory\n",
    "    path_TOBIAS_in = os.path.join(path_TOBIAS, \"input\", \"\")\n",
    "\n",
    "    if not os.path.exists(path_TOBIAS_in):\n",
    "        os.mkdir(path_TOBIAS_in)\n",
    "\n",
    "    # Get path for TOBIAS output and create directory\n",
    "    path_TOBIAS_out = os.path.join(path_TOBIAS, \"output\")\n",
    "\n",
    "    if not os.path.exists(path_TOBIAS_out):\n",
    "        os.mkdir(path_TOBIAS_out)\n",
    "\n",
    "    # check if the correct barcode tag is used\n",
    "    bam.check_barcode_tag(adata, path_bam, barcode_tag)\n",
    "\n",
    "    # Prepare splitting of the bam file if necessary\n",
    "    if cluster_column not in adata.obs.columns:\n",
    "        raise ValueError(f\"Column \\'{cluster_column}\\' not found in adata.obs. Please give valid input.\")\n",
    "\n",
    "    if barcode_column not in adata.obs.columns:\n",
    "        raise ValueError(f\"Column \\'{barcode_column}\\' not found in adata.obs. Please give valid input.\")\n",
    "\n",
    "    if len(adata.obs[cluster_column].unique()) == 1:\n",
    "        split_bam = False\n",
    "    else:\n",
    "        split_bam = True\n",
    "\n",
    "    # Split bam file by cluster\n",
    "    if split_bam:\n",
    "\n",
    "        if threads > 1:\n",
    "            parallel = True\n",
    "        else:\n",
    "            parallel = False\n",
    "\n",
    "        bam.split_bam_clusters(adata,\n",
    "                               bams=path_bam,\n",
    "                               groupby=cluster_column,\n",
    "                               barcode_col=barcode_column,\n",
    "                               read_tag=barcode_tag,\n",
    "                               output_prefix=path_TOBIAS_in,\n",
    "                               reader_threads=threads,\n",
    "                               writer_threads=threads,\n",
    "                               parallel=parallel,\n",
    "                               pysam_threads=threads)\n",
    "\n",
    "    # Get paths to bam files for TOBIAS run\n",
    "    path_TOBIAS_pseudobams = glob.glob(\"\".join([path_TOBIAS_in, \"*.bam\"]))\n",
    "    path_TOBIAS_pseudobams = [os.path.abspath(f) for f in path_TOBIAS_pseudobams]\n",
    "\n",
    "    # If there is no blacklist file given, create a mock blacklist file\n",
    "    if path_TOBIAS_blacklist is None:\n",
    "        path_TOBIAS_blacklist = os.path.join(path_TOBIAS_out, \"blacklist.bed\")\n",
    "        f = open(path_TOBIAS_blacklist, \"w\")\n",
    "        f.write(\"chr1\\t0\\t1\\n\")\n",
    "        f.close()\n",
    "\n",
    "    print(\"Writing TOBIAS config yaml.\")\n",
    "\n",
    "    # Call function to write TOBIAS config yml\n",
    "    tobias.write_TOBIAS_config(os.path.join(path_TOBIAS_out, TOBIAS_yml),\n",
    "                               bams=path_TOBIAS_pseudobams,\n",
    "                               fasta=path_TOBIAS_fasta,\n",
    "                               gtf=path_TOBIAS_gtf,\n",
    "                               motifs=path_TOBIAS_motifs,\n",
    "                               blacklist=path_TOBIAS_blacklist,\n",
    "                               organism=TOBIAS_organism,\n",
    "                               output=path_TOBIAS_out,\n",
    "                               plot_comparison=plot_comparison,\n",
    "                               plot_correction=plot_correction,\n",
    "                               plot_venn=plot_venn,\n",
    "                               coverage=coverage,\n",
    "                               wilson=wilson)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3734c14e",
   "metadata": {},
   "source": [
    "#  05 - TOBIAS\n",
    "<hr style=\"border:2px solid black\"> </hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a76c304a",
   "metadata": {},
   "source": [
    "## 1 - Description\n",
    "This notebook provides an ATAC-seq based footprinting analysis, to infer transcription factor binding sites. Herefore TOBIAS is embedded to the notebook by customized wrapper functions. TOBIAS provides a collection of command-line bioinformatics tools to perform footprinting analysis including tools for:\n",
    "- Correction of Tn5 insertion bias\n",
    "- Calculation of footprint scores within regulatory regions\n",
    "- Estimation of bound/unbound transcription factor binding sites\n",
    "- Visualization of footprints within and across different conditions\n",
    "\n",
    "### 1.1 Embedding\n",
    "Embeddings are dimension reduction methods to transform high-dimensional data into lower-dimensional representations while preserving the inherent structure and relationships between individual cells.  \n",
    "The sctoolbox supports the [Uniform Manifold Approximation and Projection (UMAP)](https://arxiv.org/abs/1802.03426) and the [t-distributed stochastic neighbor embedding (t-SNE)](https://www.jmlr.org/papers/volume9/vandermaaten08a/vandermaaten08a.pdf) methods for dimension reduction, with UMAP being set as the default value.\n",
    "To learn more about the differences between those methods and get more insight in the parameter selction have a look [here for umap](https://pair-code.github.io/understanding-umap/) and [here for t-SNE](https://distill.pub/2016/misread-tsne/?_ga=2.135835192.888864733.1531353600-1779571267.1531353600).\n",
    "### 1.2 Clustering\n",
    "Single cell clustering is used to group individual cells into clusters based on similarities in their gene expression. The clustering allows to identify distinct cell types and characterize cellular heterogeneity within a population.\n",
    "The sctoolbox supports the [leiden](https://www.nature.com/articles/s41598-019-41695-z) and the [louvain](https://iopscience.iop.org/article/10.1088/1742-5468/2008/10/P10008) clustering methods, with the leiden clustering algorithm being newer and recommended to use."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "088728fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "pd.set_option('display.max_columns', None)  # no limit to the number of columns shown\n",
    "from sctoolbox import settings\n",
    "import sctoolbox.utils as utils\n",
    "import sctoolbox.tools as tools\n",
    "import sctoolbox.utils.decorator as deco"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "36e3aa4e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Created directory: ../adatas/\n",
      "Created directory: ../figures/proportion_analysis/\n"
     ]
    }
   ],
   "source": [
    "# In/output paths\n",
    "settings.adata_input_dir = \"../adatas/\"\n",
    "settings.adata_output_dir = \"../adatas/\"\n",
    "settings.figure_dir = \"../figures/tobias/\"\n",
    "settings.log_file: \"../logs/tobias_analysis_log.txt\"\n",
    "\n",
    "# Input/Output\n",
    "last_notebook_adata = \"anndata_4.h5ad\"\n",
    "output = \"anndata_tobias.h5ad\"\n",
    "plot_suffix = \"scanpro\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "94033894",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'last_notebook_adata' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[5], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m adata \u001b[38;5;241m=\u001b[39m utils\u001b[38;5;241m.\u001b[39madata\u001b[38;5;241m.\u001b[39mload_h5ad(\u001b[43mlast_notebook_adata\u001b[49m)\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m pd\u001b[38;5;241m.\u001b[39moption_context(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdisplay.max.rows\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;241m5\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdisplay.max.columns\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m      4\u001b[0m     display(adata)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'last_notebook_adata' is not defined"
     ]
    }
   ],
   "source": [
    "adata = utils.adata.load_h5ad(last_notebook_adata)\n",
    "\n",
    "with pd.option_context(\"display.max.rows\", 5, \"display.max.columns\", None):\n",
    "    display(adata)\n",
    "    display(adata.obs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d733214b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The path to the ATAC experiment BAM file\n",
    "# Must contain all cells\n",
    "path_bam = \"\"\n",
    "\n",
    "# The path to the FASTA file for the organism\n",
    "path_TOBIAS_fasta = \"\"\n",
    "\n",
    "# The path to the (uncompressed) GTF file for the organism\n",
    "path_TOBIAS_gtf = \"\"\n",
    "\n",
    "# The path to the blacklist file to use in the TOBIAS run (optional)\n",
    "# If 'None', a mock blacklist file will be generated for the run\n",
    "path_TOBIAS_blacklist = None\n",
    "\n",
    "# The Path to the motif file for the organism (JASPAR or MEME)\n",
    "path_TOBIAS_motifs = \"\"\n",
    "\n",
    "# If the ATAC modality already has a column with read tags matching the tags used in the bam file\n",
    "# give the name of the column here\n",
    "# Must match \"ATAC:<name of coulmn in anndata.obs table>\"\n",
    "bam_barcodes = None\n",
    "\n",
    "# If bam_barcodes is None give the name of the column that contains the raw ATAC barcodes\n",
    "# Must match \"ATAC:<name of column in anndata.obs table>\"\n",
    "raw_barcodes_ATAC = \"\"\n",
    "\n",
    "# Name of the organism from which the data stems\n",
    "# options = [\"mouse\", \"human\", \"zebrafish\"]\n",
    "TOBIAS_organism = \"\"\n",
    "\n",
    "# Give the name of the TOBIAS config yaml file in the format of \"<name of file>.yml\"\n",
    "# It cannot be 'config.yml'\n",
    "TOBIAS_yml = \"\"\n",
    "\n",
    "# Give the path to the TOBIAS snakemake repo directory\n",
    "path_TOBIAS_snakemake = \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a37cca5",
   "metadata": {},
   "outputs": [],
   "source": [
    "prepare_tobias(adata,\n",
    "                   cluster_column,\n",
    "                   barcode_column,\n",
    "                   path_out,\n",
    "                   path_bam,\n",
    "                   barcode_tag='CB',\n",
    "                   path_TOBIAS_fasta=None,\n",
    "                   path_TOBIAS_motifs=None,\n",
    "                   path_TOBIAS_gtf=None,\n",
    "                   path_TOBIAS_blacklist=None,\n",
    "                   TOBIAS_organism='human',\n",
    "                   TOBIAS_yml=\"TOBIAS_config.yml\",\n",
    "                   plot_comparison=True,\n",
    "                   plot_correction=True,\n",
    "                   plot_venn=True,\n",
    "                   coverage=False,\n",
    "                   wilson=False,\n",
    "                   threads=4)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sct-tobias",
   "language": "python",
   "name": "sct-tobias"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
