<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>02 - QC and filtering &mdash; SC FRAMEWORK  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../_static/plot_directive.css" />
      <link rel="stylesheet" type="text/css" href="../_static/nbsphinx-code-cells.css" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../_static/doctools.js?v=9a2dae69"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="03 - Batch effect correction and comparisons" href="03_normalization_batch_correction.html" />
    <link rel="prev" title="01 - Assembling or loading anndata object" href="01_assembling.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            SC FRAMEWORK
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../CHANGES.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../rna-notebooks/index.html">RNA notebooks</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">ATAC notebooks</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="01_assembling.html">01 - Assembling or loading anndata object</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">02 - QC and filtering</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#1---Description">1 - Description</a></li>
<li class="toctree-l3"><a class="reference internal" href="#2---Setup">2 - Setup</a></li>
<li class="toctree-l3"><a class="reference internal" href="#3---Load-anndata">3 - Load anndata</a></li>
<li class="toctree-l3"><a class="reference internal" href="#4---QC-and-filtering">4 - QC and filtering</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#4.1---Show-STARsolo-quality-(optional)">4.1 - Show STARsolo quality (optional)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#4.2---Remove-empty-cells-and-features">4.2 - Remove empty cells and features</a></li>
<li class="toctree-l4"><a class="reference internal" href="#4.3---Binarize-matrix">4.3 - Binarize matrix</a></li>
<li class="toctree-l4"><a class="reference internal" href="#4.4-Calculate-and-remove-doublets">4.4 Calculate and remove doublets</a></li>
<li class="toctree-l4"><a class="reference internal" href="#4.6---Cell-filtering">4.6 - Cell filtering</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#5---Feature-processing">5 - Feature processing</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#5.1---Filter-additional-marked-features">5.1 - Filter additional marked features</a></li>
<li class="toctree-l4"><a class="reference internal" href="#5.2---Select-highly-variable-features">5.2 - Select highly variable features</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#6---Save-filtered-adata">6 - Save filtered adata</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="03_normalization_batch_correction.html">03 - Batch effect correction and comparisons</a></li>
<li class="toctree-l2"><a class="reference internal" href="04_clustering.html">04 - Embedding and clustering</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../general-notebooks/index.html">General notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../API/index.html">sctoolbox API reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">SC FRAMEWORK</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">ATAC notebooks</a></li>
      <li class="breadcrumb-item active">02 - QC and filtering</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/atac-notebooks/02_QC.nblink.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">sctoolbox</span>
<span class="kn">from</span> <span class="nn">sctoolbox.utilities</span> <span class="kn">import</span> <span class="n">bgcolor</span>
</pre></div>
</div>
</div>
<section id="02---QC-and-filtering">
<h1>02 - QC and filtering<a class="headerlink" href="#02---QC-and-filtering" title="Link to this heading"></a></h1>
<hr style="border:2px solid black"></hr><section id="1---Description">
<h2>1 - Description<a class="headerlink" href="#1---Description" title="Link to this heading"></a></h2>
<p><strong>Quality control</strong></p>
<p>We must ensure that all cellular barcode data correspond to viable cells.</p>
<p>To ensure this quality control (QC) is mandatory and for ATAC-seq data based on 4 key aspects:</p>
<ol class="arabic simple">
<li><p>The signal-to-noise, either via i) the enrichment of known regions, or ii) the determination of the ratio of fragments in peaks (FRiP).</p></li>
<li><p>The total number of unique fragments, also known as library complexity.</p></li>
<li><p>The fraction of reads derived from mitochondrial DNA vs. nuclear DNA.</p></li>
<li><p>The Fragment Length Distribution.</p></li>
</ol>
<p><strong>DOI: 10.1186/s13059-020-1929-3</strong></p>
<p>On the single cell scale additional aspects, such as multiplets have to be taken into account.</p>
<p><strong>DOI: 10.1038/s41467-021-21583-9</strong></p>
<p>Based on QC related columns stored in the .obs we can filter for high quality cells based on all these aspects in this notebook.</p>
<p><strong>Feature Selection</strong></p>
<p>For subsequent processing steps such as dimension reduction, embedding and clustering, the filtering of features and the selection of highly variable features can be conducive.</p>
<p>Therefore this notebooks provides steps to exclude regions from sex chromosomes and mitochondrial chromosomes and to select highly variabele features.</p>
<hr class="docutils" />
</section>
<section id="2---Setup">
<h2>2 - Setup<a class="headerlink" href="#2---Setup" title="Link to this heading"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># sctoolbox modules</span>
<span class="kn">import</span> <span class="nn">sctoolbox.utils</span> <span class="k">as</span> <span class="nn">utils</span>
<span class="kn">import</span> <span class="nn">sctoolbox.tools</span> <span class="k">as</span> <span class="nn">tools</span>
<span class="kn">import</span> <span class="nn">sctoolbox.plotting</span> <span class="k">as</span> <span class="nn">pl</span>
<span class="kn">import</span> <span class="nn">sctoolbox.tools.qc_filter</span> <span class="k">as</span> <span class="nn">qc</span>

<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">episcanpy</span> <span class="k">as</span> <span class="nn">epi</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">scrublet</span> <span class="k">as</span> <span class="nn">scr</span>

<span class="n">utils</span><span class="o">.</span><span class="n">settings_from_config</span><span class="p">(</span><span class="s2">&quot;config.yaml&quot;</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="s2">&quot;02&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
</section>
<section id="3---Load-anndata">
<h2>3 - Load anndata<a class="headerlink" href="#3---Load-anndata" title="Link to this heading"></a></h2>
<p>Uses the anndata object written by the previous notebook.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">load_h5ad</span><span class="p">(</span><span class="s2">&quot;anndata_1.h5ad&quot;</span><span class="p">)</span>

<span class="k">with</span> <span class="n">pd</span><span class="o">.</span><span class="n">option_context</span><span class="p">(</span><span class="s2">&quot;display.max.rows&quot;</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="s2">&quot;display.max.columns&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">):</span>
    <span class="n">display</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
    <span class="n">display</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">)</span>
    <span class="n">display</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">)</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
</section>
<section id="4---QC-and-filtering">
<h2>4 - QC and filtering<a class="headerlink" href="#4---QC-and-filtering" title="Link to this heading"></a></h2>
<hr style="border:2px solid black"></hr><h1><center><p>⬐ Fill in input data here ⬎</p>
</center></h1><div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">bgcolor</span> PowderBlue

<span class="c1"># Set the column in adata.obs containing the biological condition to evaluate</span>
<span class="n">condition_column</span> <span class="o">=</span> <span class="s2">&quot;sample&quot;</span>

<span class="c1"># Absolute minimum number of features for pre-selection of cells before QC</span>
<span class="n">min_genes</span> <span class="o">=</span> <span class="mi">1</span>

<span class="c1"># Choose whether to binarize the X matrix</span>
<span class="n">binarize_mtx</span> <span class="o">=</span> <span class="kc">True</span>  <span class="c1"># True or False; convert matrix to binary</span>

<span class="c1"># Optional: Plot STARsolo quality if a path is given</span>
<span class="n">quant_folder</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

<span class="c1">#----------------------- Doublet removal ------------------------</span>

<span class="c1"># Decide whether to remove doublets using scrublet (True) or to skip doublet calculation (False)</span>
<span class="n">filter_doublets</span> <span class="o">=</span> <span class="kc">True</span>

<span class="c1"># Use native scrublet or the scanpy wrapper (scanpy: optimized for RNA)</span>
<span class="n">use_native_scrublet</span> <span class="o">=</span> <span class="kc">True</span>

<span class="c1"># Default threshold to apply doublet removal on</span>
<span class="n">doublet_threshold</span> <span class="o">=</span> <span class="mf">0.2</span>

<span class="c1"># Apply doublet filtering on each condition separately</span>
<span class="n">use_condition_column</span> <span class="o">=</span> <span class="kc">False</span>

<span class="c1"># If latter is False choose another column</span>
<span class="n">condition_doublet_removal</span> <span class="o">=</span> <span class="kc">None</span>

<span class="c1"># Available threads</span>
<span class="n">threads</span> <span class="o">=</span> <span class="mi">2</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Ensure that condition column is a category</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">condition_column</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">condition_column</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;category&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<section id="4.1---Show-STARsolo-quality-(optional)">
<h3>4.1 - Show STARsolo quality (optional)<a class="headerlink" href="#4.1---Show-STARsolo-quality-(optional)" title="Link to this heading"></a></h3>
<p>If the data was mapped using STARsolo, use the parameter to set the path to the STARsolo runs and plot quality measures across runs. The path must be a folder, e.g. “path/to/starsolo_output”, which contains folders per condition e.g. “cond1”, “cond2”, etc.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="n">quant_folder</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">plot_starsolo_quality</span><span class="p">(</span><span class="n">quant_folder</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="s2">&quot;starsolo_quality.pdf&quot;</span><span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">plot_starsolo_UMI</span><span class="p">(</span><span class="n">quant_folder</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="s2">&quot;starsolo_cell_selection.pdf&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
</section>
<section id="4.2---Remove-empty-cells-and-features">
<h3>4.2 - Remove empty cells and features<a class="headerlink" href="#4.2---Remove-empty-cells-and-features" title="Link to this heading"></a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;original shape: &#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Removing empty features and cells...&#39;</span><span class="p">)</span>

<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;new shape:&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
</section>
<section id="4.3---Binarize-matrix">
<h3>4.3 - Binarize matrix<a class="headerlink" href="#4.3---Binarize-matrix" title="Link to this heading"></a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># save raw matrix</span>
<span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;raw&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="c1"># binarize</span>
<span class="k">if</span> <span class="n">binarize_mtx</span><span class="p">:</span>
    <span class="n">epi</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">binarize</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
</section>
<section id="4.4-Calculate-and-remove-doublets">
<h3>4.4 Calculate and remove doublets<a class="headerlink" href="#4.4-Calculate-and-remove-doublets" title="Link to this heading"></a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span> <span class="k">if</span> <span class="n">filter_doublets</span><span class="p">:</span>

    <span class="k">if</span> <span class="n">use_condition_column</span><span class="p">:</span>
        <span class="n">condition_doublet_removal</span> <span class="o">=</span> <span class="n">condition_column</span>

    <span class="k">if</span> <span class="n">use_native_scrublet</span><span class="p">:</span>
        <span class="c1"># TODO: Implement Wrapper function for sctoolbox</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;doublet_score&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="s1">&#39;NaN&#39;</span><span class="p">)</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;predicted_doublet&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="n">sample_dict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">condition_column</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Run scrublet for condition: &#39;</span> <span class="o">+</span> <span class="n">sample</span><span class="p">)</span>
            <span class="n">X</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">condition_column</span><span class="p">]</span> <span class="o">==</span> <span class="n">sample</span><span class="p">]</span>
            <span class="n">scrub</span> <span class="o">=</span> <span class="n">scr</span><span class="o">.</span><span class="n">Scrublet</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
            <span class="n">doublet_scores</span><span class="p">,</span> <span class="n">predicted_doublets</span> <span class="o">=</span> <span class="n">scrub</span><span class="o">.</span><span class="n">scrub_doublets</span><span class="p">()</span>
            <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">condition_column</span><span class="p">]</span><span class="o">==</span><span class="n">sample</span><span class="p">,</span> <span class="s1">&#39;doublet_score&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">doublet_scores</span>
            <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">condition_column</span><span class="p">]</span><span class="o">==</span><span class="n">sample</span><span class="p">,</span> <span class="s1">&#39;predicted_doublet&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">predicted_doublets</span>

        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;predicted_doublet&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;predicted_doublet&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">qc</span><span class="o">.</span><span class="n">estimate_doublets</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="n">condition_doublet_removal</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="n">threads</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="n">doublet_threshold</span><span class="p">)</span>

    <span class="c1">#Remove the duplicates from adata</span>
    <span class="n">tools</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;predicted_doublet&quot;</span><span class="p">,</span> <span class="n">remove_bool</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># remove empty features</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
</section>
<section id="4.6---Cell-filtering">
<h3>4.6 - Cell filtering<a class="headerlink" href="#4.6---Cell-filtering" title="Link to this heading"></a></h3>
<hr style="border:1px solid black"></hr><div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Recalculate standard QC metrics (counts...)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">tools</span><span class="o">.</span><span class="n">calculate_qc_metrics</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">var_type</span><span class="o">=</span><span class="s1">&#39;features&#39;</span><span class="p">)</span>

<span class="c1">#Remove genes with 0 count</span>
<span class="n">zero_bool</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;n_cells_by_counts&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span><span class="o">~</span><span class="n">zero_bool</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># available obs columns</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span>
</pre></div>
</div>
</div>
<h1><center><p>⬐ Fill in input data here ⬎</p>
</center></h1><div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">bgcolor</span> PowderBlue

<span class="c1"># Decide whether to estimate thresholds individual per condition (False) or globally (True)</span>
<span class="n">global_threshold</span> <span class="o">=</span> <span class="kc">False</span>

<span class="c1">## Set default filter thresholds</span>

<span class="c1"># This will be applied to all samples - the thresholds can be changed manually when plotted</span>
<span class="n">use_default_thresholds</span> <span class="o">=</span> <span class="kc">True</span>  <span class="c1"># set to False to ignore default_thresholds</span>
<span class="n">default_thresholds</span> <span class="o">=</span> <span class="p">{</span>
                      <span class="s1">&#39;n_features&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;min&#39;</span><span class="p">:</span> <span class="mi">100</span><span class="p">,</span> <span class="s1">&#39;max&#39;</span><span class="p">:</span> <span class="mi">5000</span><span class="p">},</span>
                      <span class="s1">&#39;fld_score_cwt&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;min&#39;</span><span class="p">:</span> <span class="mf">0.01</span><span class="p">,</span> <span class="s1">&#39;max&#39;</span><span class="p">:</span> <span class="mf">0.6</span><span class="p">}</span>
                      <span class="c1"># add additional columns if needed</span>
                     <span class="p">}</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
<section id="4.6.1-Adapt-thresholds">
<h4>4.6.1 Adapt thresholds<a class="headerlink" href="#4.6.1-Adapt-thresholds" title="Link to this heading"></a></h4>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">groupby</span> <span class="o">=</span> <span class="n">condition_column</span> <span class="k">if</span> <span class="n">global_threshold</span> <span class="ow">is</span> <span class="kc">False</span> <span class="k">else</span> <span class="kc">None</span>
<span class="n">initial_thresholds</span> <span class="o">=</span> <span class="n">tools</span><span class="o">.</span><span class="n">get_thresholds_wrapper</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">default_thresholds</span><span class="p">,</span>
                                                  <span class="n">only_automatic_thresholds</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="n">groupby</span><span class="p">)</span>
<span class="n">obs_columns</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">initial_thresholds</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
<span class="n">tools</span><span class="o">.</span><span class="n">thresholds_as_table</span><span class="p">(</span><span class="n">initial_thresholds</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">matplotlib</span> widget

<span class="c1">#Plot violins and sliders</span>
<span class="n">obs_figure</span><span class="p">,</span> <span class="n">obs_slider_dict</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">quality_violin</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">obs_columns</span><span class="p">,</span>
                                                <span class="n">groupby</span><span class="o">=</span><span class="n">condition_column</span><span class="p">,</span>
                                                <span class="n">which</span><span class="o">=</span><span class="s2">&quot;obs&quot;</span><span class="p">,</span>
                                                <span class="n">thresholds</span><span class="o">=</span><span class="n">initial_thresholds</span><span class="p">,</span>
                                                <span class="n">global_threshold</span><span class="o">=</span><span class="n">global_threshold</span><span class="p">,</span>
                                                <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Cell quality control (before)&quot;</span><span class="p">,</span>
                                                <span class="n">save</span><span class="o">=</span><span class="s2">&quot;cell_filtering.png&quot;</span><span class="p">)</span>
<span class="n">obs_figure</span>
</pre></div>
</div>
</div>
</section>
<section id="4.6.2---Apply-gene-filtering">
<h4>4.6.2 - Apply gene filtering<a class="headerlink" href="#4.6.2---Apply-gene-filtering" title="Link to this heading"></a></h4>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get final thresholds</span>
<span class="n">final_thresholds</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">get_slider_thresholds</span><span class="p">(</span><span class="n">obs_slider_dict</span><span class="p">)</span>
<span class="n">tools</span><span class="o">.</span><span class="n">thresholds_as_table</span><span class="p">(</span><span class="n">final_thresholds</span><span class="p">)</span> <span class="c1"># show thresholds</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Show pairwise comparisons of column values w/ thresholds (mean values in case thresholds are grouped)</span>
<span class="o">%</span><span class="k">matplotlib</span> inline
<span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>  <span class="c1"># close previous figure</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">final_thresholds</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
    <span class="n">mean_thresholds</span> <span class="o">=</span> <span class="n">qc</span><span class="o">.</span><span class="n">get_mean_thresholds</span><span class="p">(</span><span class="n">final_thresholds</span><span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">pairwise_scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">obs_columns</span><span class="p">,</span> <span class="n">thresholds</span><span class="o">=</span><span class="n">mean_thresholds</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="s2">&quot;cell_filtering_scatter.pdf&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tools</span><span class="o">.</span><span class="n">apply_qc_thresholds</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">final_thresholds</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="n">groupby</span><span class="p">)</span>

<span class="c1"># remove empty features after cell filtering</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
</section>
<section id="4.6.3---Show-data-after-filtering">
<h4>4.6.3 - Show data after filtering<a class="headerlink" href="#4.6.3---Show-data-after-filtering" title="Link to this heading"></a></h4>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">matplotlib</span> inline

<span class="c1">#Plot violins and sliders</span>
<span class="n">figure</span><span class="p">,</span> <span class="n">slider_dict</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">quality_violin</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">obs_columns</span><span class="p">,</span>
                                        <span class="n">groupby</span><span class="o">=</span><span class="n">condition_column</span><span class="p">,</span>
                                        <span class="n">which</span><span class="o">=</span><span class="s2">&quot;obs&quot;</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
                                        <span class="n">global_threshold</span> <span class="o">=</span> <span class="n">global_threshold</span><span class="p">,</span>
                                        <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Cell quality control (after)&quot;</span><span class="p">,</span>
                                        <span class="n">save</span><span class="o">=</span><span class="s2">&quot;cell_filtering_final.pdf&quot;</span><span class="p">)</span>
<span class="n">figure</span>
</pre></div>
</div>
</div>
</section>
</section>
</section>
<section id="5---Feature-processing">
<h2>5 - Feature processing<a class="headerlink" href="#5---Feature-processing" title="Link to this heading"></a></h2>
<hr style="border:2px solid black"></hr><h1><center><p>⬐ Fill in input data here ⬎</p>
</center></h1><div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">bgcolor</span> PowderBlue

<span class="c1"># Removal of gene subsets</span>
<span class="n">filter_chrM</span> <span class="o">=</span> <span class="kc">True</span>  <span class="c1"># True or False; filtering out chrM</span>
<span class="n">filter_xy</span> <span class="o">=</span> <span class="kc">True</span>    <span class="c1"># True or False; filtering out chrX and chrY</span>

<span class="c1"># Highly Variable Features options</span>
<span class="n">select_highly_variable</span> <span class="o">=</span> <span class="kc">True</span>
<span class="n">min_cells</span> <span class="o">=</span> <span class="mi">5</span> <span class="c1"># This one is mandatory</span>
<span class="n">max_cells</span> <span class="o">=</span> <span class="kc">None</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
<section id="5.1---Filter-additional-marked-features">
<h3>5.1 - Filter additional marked features<a class="headerlink" href="#5.1---Filter-additional-marked-features" title="Link to this heading"></a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="n">filter_chrM</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Removing chromosomal features...&quot;</span><span class="p">)</span>
    <span class="n">non_m</span> <span class="o">=</span> <span class="p">[</span><span class="n">name</span> <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">name</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;chrM&#39;</span><span class="p">)]</span>  <span class="c1"># remove chrM</span>
    <span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">non_m</span><span class="p">]</span>

<span class="k">if</span> <span class="n">filter_xy</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Removing gender related features...&quot;</span><span class="p">)</span>
    <span class="n">non_xy</span> <span class="o">=</span> <span class="p">[</span><span class="n">name</span> <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">name</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;chrY&#39;</span><span class="p">)</span> <span class="o">|</span> <span class="n">name</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;chrX&#39;</span><span class="p">)]</span>
    <span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">non_xy</span><span class="p">]</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
</section>
<section id="5.2---Select-highly-variable-features">
<h3>5.2 - Select highly variable features<a class="headerlink" href="#5.2---Select-highly-variable-features" title="Link to this heading"></a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># update number of cells per feature</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">tools</span><span class="o">.</span><span class="n">calculate_qc_metrics</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">var_type</span><span class="o">=</span><span class="s1">&#39;features&#39;</span><span class="p">)</span>
<span class="k">if</span> <span class="n">select_highly_variable</span><span class="p">:</span>
    <span class="c1"># get highly variable features</span>
    <span class="n">tools</span><span class="o">.</span><span class="n">get_variable_features</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">max_cells</span><span class="p">,</span> <span class="n">min_cells</span><span class="p">)</span>
    <span class="c1">#Number of variable genes selected</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;highly_variable&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="c1"># plot HVF violin</span>
    <span class="n">pl</span><span class="o">.</span><span class="n">violin_HVF_distribution</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
</section>
</section>
<section id="6---Save-filtered-adata">
<h2>6 - Save filtered adata<a class="headerlink" href="#6---Save-filtered-adata" title="Link to this heading"></a></h2>
<hr style="border:2px solid black"></hr><p>Store the final results</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Saving the data</span>
<span class="n">adata_output</span> <span class="o">=</span> <span class="s2">&quot;anndata_2.h5ad&quot;</span>
<span class="n">utils</span><span class="o">.</span><span class="n">save_h5ad</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">adata_output</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sctoolbox</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">close_logfile</span><span class="p">()</span>
</pre></div>
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="01_assembling.html" class="btn btn-neutral float-left" title="01 - Assembling or loading anndata object" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="03_normalization_batch_correction.html" class="btn btn-neutral float-right" title="03 - Batch effect correction and comparisons" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Loosolab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>