<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>04 - Embedding and clustering &mdash; SC FRAMEWORK  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../_static/plot_directive.css" />
      <link rel="stylesheet" type="text/css" href="../_static/nbsphinx-code-cells.css" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../_static/doctools.js?v=9a2dae69"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="0A - Marker gene computation and analysis" href="0A_marker_genes.html" />
    <link rel="prev" title="03 - Batch effect correction and comparisons" href="03_batch.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            SC FRAMEWORK
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../CHANGES.html">Changelog</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">RNA notebooks</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="01_assembling_anndata.html">01 - Assembling or loading anndata object</a></li>
<li class="toctree-l2"><a class="reference internal" href="02_QC_filtering.html">02 - QC and filtering</a></li>
<li class="toctree-l2"><a class="reference internal" href="03_batch.html">03 - Batch effect correction and comparisons</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">04 - Embedding and clustering</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#1---Description">1 - Description</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#1.1-Embedding">1.1 Embedding</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#2---Setup">2 - Setup</a></li>
<li class="toctree-l3"><a class="reference internal" href="#3---Load-anndata">3 - Load anndata</a></li>
<li class="toctree-l3"><a class="reference internal" href="#4---General-input">4 - General input</a></li>
<li class="toctree-l3"><a class="reference internal" href="#5---Calculate-UMAP/TSNE-and-find-best-setting">5 - Calculate UMAP/TSNE and find best setting</a></li>
<li class="toctree-l3"><a class="reference internal" href="#6---Plot-final-embedding-with-quality-measures">6 - Plot final embedding with quality measures</a></li>
<li class="toctree-l3"><a class="reference internal" href="#7---Cell-clustering">7 - Cell clustering</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#7.1---Reclustering-(optional)">7.1 - Reclustering (optional)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#7.2---Final-clustering-of-cells">7.2 - Final clustering of cells</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#8---Plot-distribution-of-cells-across-clusters">8 - Plot distribution of cells across clusters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#9---Generating-3D-Object-with-UMAP-coordinates-in-HTML">9 - Generating 3D Object with UMAP coordinates in HTML</a></li>
<li class="toctree-l3"><a class="reference internal" href="#10---Saving-adata-for-next-notebook">10 - Saving adata for next notebook</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="0A_marker_genes.html">0A - Marker gene computation and analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="0B_ligand_receptor.html">0B - Receptor-Ligand Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="0C_velocity_analysis.html">0C - Velocity Analysis</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../atac-notebooks/index.html">ATAC notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general-notebooks/index.html">General notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../API/index.html">sctoolbox API reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">SC FRAMEWORK</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">RNA notebooks</a></li>
      <li class="breadcrumb-item active">04 - Embedding and clustering</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/rna-notebooks/04_clustering.nblink.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">sctoolbox</span>
<span class="kn">from</span> <span class="nn">sctoolbox.utilities</span> <span class="kn">import</span> <span class="n">bgcolor</span>
</pre></div>
</div>
</div>
<section id="04---Embedding-and-clustering">
<h1>04 - Embedding and clustering<a class="headerlink" href="#04---Embedding-and-clustering" title="Link to this heading"></a></h1>
<hr style="border:2px solid black"></hr><section id="1---Description">
<h2>1 - Description<a class="headerlink" href="#1---Description" title="Link to this heading"></a></h2>
<section id="1.1-Embedding">
<h3>1.1 Embedding<a class="headerlink" href="#1.1-Embedding" title="Link to this heading"></a></h3>
<p>Embeddings are dimension reduction methods to transform high-dimensional data into lower-dimensional representations while preserving the inherent structure and relationships between individual cells. The sctoolbox supports the <a class="reference external" href="https://arxiv.org/abs/1802.03426">Uniform Manifold Approximation and Projection (UMAP)</a> and the <a class="reference external" href="https://www.jmlr.org/papers/volume9/vandermaaten08a/vandermaaten08a.pdf">t-distributed stochastic neighbor embedding (t-SNE)</a> methods for dimension reduction, with UMAP
being set as the default value. To learn more about the differences between those methods and get more insight in the parameter selction have a look <a class="reference external" href="https://pair-code.github.io/understanding-umap/">here for umap</a> and <a class="reference external" href="https://distill.pub/2016/misread-tsne/?_ga=2.135835192.888864733.1531353600-1779571267.1531353600">here for t-SNE</a>. ### 1.2 Clustering Single cell clustering is used to group individual cells into clusters based on similarities in their gene expression. The clustering allows
to identify distinct cell types and characterize cellular heterogeneity within a population. The sctoolbox supports the <a class="reference external" href="https://www.nature.com/articles/s41598-019-41695-z">leiden</a> and the <a class="reference external" href="https://iopscience.iop.org/article/10.1088/1742-5468/2008/10/P10008">louvain</a> clustering methods, with the leiden clustering algorithm being newer and recommended to use.</p>
<hr class="docutils" />
</section>
</section>
<section id="2---Setup">
<h2>2 - Setup<a class="headerlink" href="#2---Setup" title="Link to this heading"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">sctoolbox.utils</span> <span class="k">as</span> <span class="nn">utils</span>
<span class="kn">import</span> <span class="nn">sctoolbox.tools</span> <span class="k">as</span> <span class="nn">tools</span>
<span class="kn">import</span> <span class="nn">sctoolbox.plotting</span> <span class="k">as</span> <span class="nn">pl</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">utils</span><span class="o">.</span><span class="n">settings_from_config</span><span class="p">(</span><span class="s2">&quot;config.yaml&quot;</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="s2">&quot;04&quot;</span><span class="p">)</span>

<span class="c1"># Set additional options for figures</span>
<span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">vector_friendly</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">dpi_save</span><span class="o">=</span><span class="mi">600</span><span class="p">,</span> <span class="n">scanpy</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
</section>
<section id="3---Load-anndata">
<h2>3 - Load anndata<a class="headerlink" href="#3---Load-anndata" title="Link to this heading"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">load_h5ad</span><span class="p">(</span><span class="s2">&quot;anndata_3.h5ad&quot;</span><span class="p">)</span>

<span class="k">with</span> <span class="n">pd</span><span class="o">.</span><span class="n">option_context</span><span class="p">(</span><span class="s2">&quot;display.max.rows&quot;</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="s2">&quot;display.max.columns&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">):</span>
    <span class="n">display</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
    <span class="n">display</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">)</span>
    <span class="n">display</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">)</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
</section>
<section id="4---General-input">
<h2>4 - General input<a class="headerlink" href="#4---General-input" title="Link to this heading"></a></h2>
<h1><center><p>⬐ Fill in input data here ⬎</p>
</center></h1><div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">bgcolor</span> PowderBlue

<span class="c1">#Column to show in UMAPs</span>
<span class="n">condition_column</span> <span class="o">=</span> <span class="s2">&quot;sample&quot;</span>

<span class="c1">#Number of threads to use for multiprocessing</span>
<span class="n">threads</span> <span class="o">=</span> <span class="mi">4</span>

<span class="c1"># Search embedding parameters (or set parameters later)</span>
<span class="n">embedding</span> <span class="o">=</span> <span class="s2">&quot;umap&quot;</span>   <span class="c1">#umap or tsne</span>
<span class="n">search_parameters</span> <span class="o">=</span> <span class="kc">True</span>
<span class="n">dist_range</span> <span class="o">=</span> <span class="p">(</span><span class="mf">0.1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">)</span> <span class="c1"># Set min_dist range for umap</span>
<span class="n">spread_range</span> <span class="o">=</span> <span class="p">(</span><span class="mf">1.0</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">)</span> <span class="c1"># Set spread range for umap</span>
<span class="n">n_components</span> <span class="o">=</span> <span class="mi">2</span> <span class="c1"># Number of components for umap</span>
<span class="n">perplexity_range</span> <span class="o">=</span> <span class="p">(</span><span class="mi">30</span><span class="p">,</span> <span class="mi">60</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span>        <span class="c1"># perplexity range for tsne</span>
<span class="n">learning_rate_range</span> <span class="o">=</span> <span class="p">(</span><span class="mi">600</span><span class="p">,</span> <span class="mi">1000</span><span class="p">,</span> <span class="mi">200</span><span class="p">)</span>   <span class="c1"># learning_rate for tsne</span>


<span class="c1"># Search different clustering resolutions</span>
<span class="n">search_clustering_parameters</span> <span class="o">=</span> <span class="kc">True</span>
<span class="n">clustering_method</span> <span class="o">=</span> <span class="s2">&quot;leiden&quot;</span> <span class="c1">#leiden/louvain</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
</section>
<section id="5---Calculate-UMAP/TSNE-and-find-best-setting">
<h2>5 - Calculate UMAP/TSNE and find best setting<a class="headerlink" href="#5---Calculate-UMAP/TSNE-and-find-best-setting" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>NOTE: min_dist: distances between points to make the plot looks more ‘clustered’</p></li>
<li><p>NOTE: spread: The effective scale of embedded points value be de default is 1</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="n">search_parameters</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">embedding</span> <span class="o">==</span> <span class="s2">&quot;umap&quot;</span><span class="p">:</span>
        <span class="n">pl</span><span class="o">.</span><span class="n">search_umap_parameters</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span>
                                  <span class="n">min_dist_range</span><span class="o">=</span><span class="n">dist_range</span><span class="p">,</span>
                                  <span class="n">spread_range</span><span class="o">=</span><span class="n">spread_range</span><span class="p">,</span>
                                  <span class="n">color</span><span class="o">=</span><span class="n">condition_column</span><span class="p">,</span>
                                  <span class="n">n_components</span><span class="o">=</span><span class="n">n_components</span><span class="p">,</span>
                                  <span class="n">threads</span><span class="o">=</span><span class="n">threads</span><span class="p">,</span>
                                  <span class="n">save</span><span class="o">=</span><span class="s2">&quot;UMAP_parameter_search.pdf&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">embedding</span> <span class="o">==</span> <span class="s2">&quot;tsne&quot;</span><span class="p">:</span>
        <span class="n">pl</span><span class="o">.</span><span class="n">search_tsne_parameters</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span>
                                  <span class="n">perplexity_range</span><span class="o">=</span><span class="n">perplexity_range</span><span class="p">,</span>
                                  <span class="n">learning_rate_range</span><span class="o">=</span><span class="n">learning_rate_range</span><span class="p">,</span>
                                  <span class="n">color</span><span class="o">=</span><span class="n">condition_column</span><span class="p">,</span>
                                  <span class="n">threads</span><span class="o">=</span><span class="n">threads</span><span class="p">,</span>
                                  <span class="n">save</span><span class="o">=</span><span class="s2">&quot;TSNE_parameter_search.pdf&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<h1><center><p>⬐ Fill in input data here ⬎</p>
</center></h1><div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">bgcolor</span> PowderBlue

<span class="c1"># Final choice of spread / dist for umap</span>
<span class="n">min_dist</span> <span class="o">=</span> <span class="mf">0.4</span>
<span class="n">spread</span> <span class="o">=</span> <span class="mf">2.5</span>

<span class="c1"># Final choice of perplexity_range / perplexity_range for tsne</span>
<span class="n">perplexity</span> <span class="o">=</span> <span class="mi">50</span>
<span class="n">learning_rate</span> <span class="o">=</span> <span class="mi">800</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calculate final embedding</span>
<span class="k">if</span> <span class="n">embedding</span> <span class="o">==</span> <span class="s2">&quot;umap&quot;</span><span class="p">:</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_dist</span><span class="o">=</span><span class="n">min_dist</span><span class="p">,</span> <span class="n">spread</span><span class="o">=</span><span class="n">spread</span><span class="p">,</span> <span class="n">n_components</span><span class="o">=</span><span class="n">n_components</span><span class="p">)</span>
<span class="k">elif</span> <span class="n">embedding</span> <span class="o">==</span> <span class="s2">&quot;tsne&quot;</span><span class="p">:</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">tsne</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">perplexity</span><span class="o">=</span><span class="n">perplexity</span><span class="p">,</span> <span class="n">learning_rate</span><span class="o">=</span><span class="n">learning_rate</span><span class="p">)</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
</section>
<section id="6---Plot-final-embedding-with-quality-measures">
<h2>6 - Plot final embedding with quality measures<a class="headerlink" href="#6---Plot-final-embedding-with-quality-measures" title="Link to this heading"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Adjust qc columns to show in plot</span>
<span class="n">qc_columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">condition_column</span><span class="p">]</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;sctoolbox&quot;</span><span class="p">][</span><span class="s2">&quot;obs_metrics&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot final umap/tsne with quality measures</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="n">embedding</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">qc_columns</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">pl</span><span class="o">.</span><span class="n">_save_figure</span><span class="p">(</span><span class="s2">&quot;embedding_quality.pdf&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot distribution of samples in embedding</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">plot_group_embeddings</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="n">condition_column</span><span class="p">,</span> <span class="n">embedding</span><span class="o">=</span><span class="n">embedding</span><span class="p">,</span>
                             <span class="n">save</span><span class="o">=</span><span class="s2">&quot;embedding_sample_distribution.pdf&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
</section>
<section id="7---Cell-clustering">
<h2>7 - Cell clustering<a class="headerlink" href="#7---Cell-clustering" title="Link to this heading"></a></h2>
<hr style="border:2px solid black"></hr><ul class="simple">
<li><p>NOTE: resolution: controls the coarseness of the clustering. Higher values lead to more clusters.</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="n">search_clustering_parameters</span><span class="p">:</span>
    <span class="n">pl</span><span class="o">.</span><span class="n">search_clustering_parameters</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="n">clustering_method</span><span class="p">,</span>
                                    <span class="n">embedding</span><span class="o">=</span><span class="n">embedding</span><span class="p">,</span>
                                    <span class="n">save</span><span class="o">=</span><span class="s2">&quot;clustering_search.png&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<h1><center><p>⬐ Fill in input data here ⬎</p>
</center></h1><div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">bgcolor</span> PowderBlue

<span class="c1">#Choose final resolution</span>
<span class="n">clustering_column</span> <span class="o">=</span> <span class="s2">&quot;leiden_0.5&quot;</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
<section id="7.1---Reclustering-(optional)">
<h3>7.1 - Reclustering (optional)<a class="headerlink" href="#7.1---Reclustering-(optional)" title="Link to this heading"></a></h3>
<p>Here you can use the <code class="docutils literal notranslate"><span class="pre">tools.recluster</span></code> function to iteratively adjust clustering. The two cells below are provided as a template for either <em>joining</em> clusters (first cell) or <em>splitting</em> clusters (second cell) <strong>copy the cells as needed</strong>. Fill in the list at the top of the respective cell to activate <em>joining</em> or <em>splitting</em>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">bgcolor</span> PowderBlue

<span class="c1"># combine multiple clusters into one</span>
<span class="c1"># skipped when empty</span>
<span class="n">combine</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1"># add cluster names</span>

<span class="k">if</span> <span class="n">combine</span><span class="p">:</span>
    <span class="n">tools</span><span class="o">.</span><span class="n">recluster</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span>
                    <span class="n">column</span><span class="o">=</span><span class="n">clustering_column</span><span class="p">,</span>
                    <span class="n">clusters</span><span class="o">=</span><span class="n">combine</span><span class="p">,</span>
                    <span class="n">task</span><span class="o">=</span><span class="s2">&quot;join&quot;</span><span class="p">,</span>
                    <span class="n">embedding</span><span class="o">=</span><span class="n">embedding</span><span class="p">,</span>
                    <span class="n">key_added</span><span class="o">=</span><span class="s2">&quot;recluster&quot;</span>
                   <span class="p">)</span>

    <span class="n">clustering_column</span> <span class="o">=</span> <span class="s2">&quot;recluster&quot;</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">bgcolor</span> PowderBlue

<span class="c1"># split (recluster) one or more clusters</span>
<span class="c1"># skipped when empty</span>
<span class="n">split</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1"># add cluster names</span>
<span class="n">resolution</span><span class="o">=</span><span class="mf">0.15</span> <span class="c1"># 0-1, small values create less clusters</span>

<span class="k">if</span> <span class="n">split</span><span class="p">:</span>
    <span class="n">tools</span><span class="o">.</span><span class="n">recluster</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span>
                    <span class="n">column</span><span class="o">=</span><span class="n">clustering_column</span><span class="p">,</span>
                    <span class="n">clusters</span><span class="o">=</span><span class="n">split</span><span class="p">,</span>
                    <span class="n">task</span><span class="o">=</span><span class="s2">&quot;split&quot;</span><span class="p">,</span>
                    <span class="n">resolution</span><span class="o">=</span><span class="n">resolution</span><span class="p">,</span>
                    <span class="n">embedding</span><span class="o">=</span><span class="n">embedding</span><span class="p">,</span>
                    <span class="n">key_added</span><span class="o">=</span><span class="s2">&quot;recluster&quot;</span>
                   <span class="p">)</span>

    <span class="n">clustering_column</span> <span class="o">=</span> <span class="s2">&quot;recluster&quot;</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create final clustering</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;clustering&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">rename_categories</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clustering_column</span><span class="p">])</span>
<span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;sctoolbox&quot;</span><span class="p">][</span><span class="s2">&quot;obs_metrics&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;sctoolbox&quot;</span><span class="p">][</span><span class="s2">&quot;obs_metrics&quot;</span><span class="p">])</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;clustering&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
</section>
<section id="7.2---Final-clustering-of-cells">
<h3>7.2 - Final clustering of cells<a class="headerlink" href="#7.2---Final-clustering-of-cells" title="Link to this heading"></a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Plot final leiden</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;X_&quot;</span> <span class="o">+</span> <span class="n">embedding</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">condition_column</span><span class="p">,</span> <span class="s2">&quot;clustering&quot;</span><span class="p">],</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">pl</span><span class="o">.</span><span class="n">_save_figure</span><span class="p">(</span><span class="s2">&quot;embedding_clustering.pdf&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
</section>
</section>
<section id="8---Plot-distribution-of-cells-across-clusters">
<h2>8 - Plot distribution of cells across clusters<a class="headerlink" href="#8---Plot-distribution-of-cells-across-clusters" title="Link to this heading"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">n_cells_barplot</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;clustering&quot;</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="n">condition_column</span><span class="p">,</span>
                       <span class="n">save</span><span class="o">=</span><span class="s2">&quot;cell_distribution_barplot.pdf&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
</section>
<section id="9---Generating-3D-Object-with-UMAP-coordinates-in-HTML">
<h2>9 - Generating 3D Object with UMAP coordinates in HTML<a class="headerlink" href="#9---Generating-3D-Object-with-UMAP-coordinates-in-HTML" title="Link to this heading"></a></h2>
<h1><center><p>⬐ Fill in input data here ⬎</p>
</center></h1><div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">bgcolor</span> PowderBlue

<span class="c1">#plot 3D html for the &quot;clustering&quot; adata.obs, change to individual leiden or other columns if needed</span>
<span class="n">column_3d</span> <span class="o">=</span> <span class="s2">&quot;clustering&quot;</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="n">embedding</span> <span class="o">==</span> <span class="s2">&quot;umap&quot;</span> <span class="ow">and</span> <span class="n">n_components</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">:</span>
    <span class="n">pl</span><span class="o">.</span><span class="n">plot_3D_UMAP</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">column_3d</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;umap_3d_</span><span class="si">{</span><span class="n">column_3d</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">html_file</span> <span class="o">=</span> <span class="n">sctoolbox</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">full_figure_prefix</span> <span class="o">+</span> <span class="sa">f</span><span class="s2">&quot;umap_3d_</span><span class="si">{</span><span class="n">column_3d</span><span class="si">}</span><span class="s2">.html&quot;</span>

    <span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">IFrame</span>
    <span class="n">display</span><span class="p">(</span><span class="n">IFrame</span><span class="p">(</span><span class="n">src</span><span class="o">=</span><span class="n">html_file</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mi">800</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">400</span><span class="p">))</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
</section>
<section id="10---Saving-adata-for-next-notebook">
<h2>10 - Saving adata for next notebook<a class="headerlink" href="#10---Saving-adata-for-next-notebook" title="Link to this heading"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">utils</span><span class="o">.</span><span class="n">save_h5ad</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;anndata_4.h5ad&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sctoolbox</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">close_logfile</span><span class="p">()</span>
</pre></div>
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="03_batch.html" class="btn btn-neutral float-left" title="03 - Batch effect correction and comparisons" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="0A_marker_genes.html" class="btn btn-neutral float-right" title="0A - Marker gene computation and analysis" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Loosolab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>