<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>sctoolbox.tools.bam &mdash; SC FRAMEWORK  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../../../_static/plot_directive.css" />

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../../../_static/doctools.js?v=9a2dae69"></script>
        <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            SC FRAMEWORK
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../CHANGES.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../rna-notebooks/index.html">RNA notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../atac-notebooks/index.html">ATAC notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../general-notebooks/index.html">General notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../API/index.html">sctoolbox API reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">SC FRAMEWORK</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">sctoolbox.tools.bam</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for sctoolbox.tools.bam</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;Functionality to split bam files into smaller bam files based on clustering in adata.obs.&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">multiprocessing</span> <span class="k">as</span> <span class="nn">mp</span>
<span class="kn">from</span> <span class="nn">multiprocessing.managers</span> <span class="kn">import</span> <span class="n">BaseProxy</span>
<span class="kn">from</span> <span class="nn">multiprocessing.pool</span> <span class="kn">import</span> <span class="n">ApplyResult</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">from</span> <span class="nn">functools</span> <span class="kn">import</span> <span class="n">partial</span>

<span class="kn">from</span> <span class="nn">beartype.typing</span> <span class="kn">import</span> <span class="n">TYPE_CHECKING</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Literal</span><span class="p">,</span> <span class="n">Any</span><span class="p">,</span> <span class="n">Sequence</span>
<span class="kn">from</span> <span class="nn">beartype</span> <span class="kn">import</span> <span class="n">beartype</span>

<span class="kn">import</span> <span class="nn">sctoolbox.utils</span> <span class="k">as</span> <span class="nn">utils</span>
<span class="kn">import</span> <span class="nn">sctoolbox.utils.decorator</span> <span class="k">as</span> <span class="nn">deco</span>
<span class="kn">from</span> <span class="nn">sctoolbox._settings</span> <span class="kn">import</span> <span class="n">settings</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">settings</span><span class="o">.</span><span class="n">logger</span>

<span class="k">if</span> <span class="n">TYPE_CHECKING</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">pysam</span>


<div class="viewcode-block" id="bam_adata_ov">
<a class="viewcode-back" href="../../../API/tools.html#sctoolbox.tools.bam.bam_adata_ov">[docs]</a>
<span class="nd">@beartype</span>
<span class="k">def</span> <span class="nf">bam_adata_ov</span><span class="p">(</span><span class="n">adata</span><span class="p">:</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span>
                 <span class="n">bamfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                 <span class="n">cb_tag</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;CB&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Check if adata.obs barcodes existing in a column of a bamfile.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    adata : sc.AnnData</span>
<span class="sd">        adata object where adata.obs is stored</span>
<span class="sd">    bamfile : str</span>
<span class="sd">        path of the bamfile to investigate</span>
<span class="sd">    cb_tag : str, default &#39;CB&#39;</span>
<span class="sd">        bamfile column to extract the barcodes from</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    float</span>
<span class="sd">        hitrate of the barcodes in the bamfile</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;calculating barcode overlap between bamfile and adata.obs&quot;</span><span class="p">)</span>
    <span class="n">bam_obj</span> <span class="o">=</span> <span class="n">open_bam</span><span class="p">(</span><span class="n">bamfile</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>

    <span class="n">sample</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">iterations</span> <span class="o">=</span> <span class="mi">1000</span>
    <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">bam_obj</span><span class="p">:</span>
        <span class="n">tag</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="n">cb_tag</span><span class="p">)</span>
        <span class="n">sample</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tag</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">counter</span> <span class="o">==</span> <span class="n">iterations</span><span class="p">:</span>
            <span class="k">break</span>
        <span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="n">barcodes_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
    <span class="n">count_table</span> <span class="o">=</span> <span class="n">barcodes_df</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
    <span class="n">hits</span> <span class="o">=</span> <span class="n">count_table</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="n">hitrate</span> <span class="o">=</span> <span class="n">hits</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">/</span> <span class="n">iterations</span>

    <span class="k">return</span> <span class="n">hitrate</span></div>



<div class="viewcode-block" id="check_barcode_tag">
<a class="viewcode-back" href="../../../API/tools.html#sctoolbox.tools.bam.check_barcode_tag">[docs]</a>
<span class="nd">@deco</span><span class="o">.</span><span class="n">log_anndata</span>
<span class="nd">@beartype</span>
<span class="k">def</span> <span class="nf">check_barcode_tag</span><span class="p">(</span><span class="n">adata</span><span class="p">:</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span>
                      <span class="n">bamfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                      <span class="n">cb_tag</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;CB&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Check for the possibilty that the wrong barcode is used.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    adata : sc.AnnData</span>
<span class="sd">        adata object where adata.obs is stored</span>
<span class="sd">    bamfile : str</span>
<span class="sd">        path of the bamfile to investigate</span>
<span class="sd">    cb_tag : str, default &#39;CB&#39;</span>
<span class="sd">        bamfile column to extract the barcodes from</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    ValueError</span>
<span class="sd">        If barcode hit rate could not be identified</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">hitrate</span> <span class="o">=</span> <span class="n">bam_adata_ov</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">bamfile</span><span class="p">,</span> <span class="n">cb_tag</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">hitrate</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s1">&#39;None of the barcodes from the bamfile found in the .obs table.</span><span class="se">\n</span><span class="s1">&#39;</span>
                       <span class="s1">&#39;Consider if you are using the wrong column cb-tag or bamfile.&#39;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">hitrate</span> <span class="o">&lt;=</span> <span class="mf">0.05</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s1">&#39;Only 5</span><span class="si">% o</span><span class="s1">r less of the barcodes from the bamfile found in the .obs table.</span><span class="se">\n</span><span class="s1">&#39;</span>
                       <span class="s1">&#39;Consider if you are using the wrong column for cb-tag or bamfile.&#39;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">hitrate</span> <span class="o">&gt;</span> <span class="mf">0.05</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Barcode tag: OK&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Could not identify barcode hit rate.&quot;</span><span class="p">)</span></div>


<span class="c1">#####################################################################</span>


<div class="viewcode-block" id="subset_bam">
<a class="viewcode-back" href="../../../API/tools.html#sctoolbox.tools.bam.subset_bam">[docs]</a>
<span class="nd">@beartype</span>
<span class="k">def</span> <span class="nf">subset_bam</span><span class="p">(</span><span class="n">bam_in</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
               <span class="n">bam_out</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
               <span class="n">barcodes</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
               <span class="n">read_tag</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;CB&quot;</span><span class="p">,</span>
               <span class="n">pysam_threads</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span>
               <span class="n">overwrite</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Subset a bam file based on a list of barcodes.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    bam_in : str</span>
<span class="sd">        Path to input bam file.</span>
<span class="sd">    bam_out : str</span>
<span class="sd">        Path to output bam file.</span>
<span class="sd">    barcodes : Iterable[str]</span>
<span class="sd">        List of barcodes to keep.</span>
<span class="sd">    read_tag : str, default &quot;CB&quot;</span>
<span class="sd">        Tag in bam file to use for barcode.</span>
<span class="sd">    pysam_threads : int, default 4</span>
<span class="sd">        Number of threads to use for pysam.</span>
<span class="sd">    overwrite : bool, default False</span>
<span class="sd">        Overwrite output file if it exists.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># check then load modules</span>
    <span class="n">utils</span><span class="o">.</span><span class="n">check_module</span><span class="p">(</span><span class="s2">&quot;tqdm&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">utils</span><span class="o">.</span><span class="n">_is_notebook</span><span class="p">()</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
        <span class="kn">from</span> <span class="nn">tqdm</span> <span class="kn">import</span> <span class="n">tqdm_notebook</span> <span class="k">as</span> <span class="n">tqdm</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="kn">from</span> <span class="nn">tqdm</span> <span class="kn">import</span> <span class="n">tqdm</span>
    <span class="n">utils</span><span class="o">.</span><span class="n">check_module</span><span class="p">(</span><span class="s2">&quot;pysam&quot;</span><span class="p">)</span>

    <span class="c1"># Create output dir if needed</span>
    <span class="n">utils</span><span class="o">.</span><span class="n">create_dir</span><span class="p">(</span><span class="n">bam_out</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">bam_out</span><span class="p">)</span> <span class="ow">and</span> <span class="n">overwrite</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Output file </span><span class="si">{</span><span class="n">bam_out</span><span class="si">}</span><span class="s2"> exists. Skipping.&quot;</span><span class="p">)</span>
        <span class="k">return</span>

    <span class="c1"># Open files</span>
    <span class="n">bam_in_obj</span> <span class="o">=</span> <span class="n">open_bam</span><span class="p">(</span><span class="n">bam_in</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">,</span> <span class="n">verbosity</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="n">pysam_threads</span><span class="p">)</span>
    <span class="n">bam_out_obj</span> <span class="o">=</span> <span class="n">open_bam</span><span class="p">(</span><span class="n">bam_out</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">,</span> <span class="n">template</span><span class="o">=</span><span class="n">bam_in_obj</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="n">pysam_threads</span><span class="p">,</span> <span class="n">verbosity</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="n">barcodes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">barcodes</span><span class="p">)</span>

    <span class="c1"># Update progress based on total number of reads</span>
    <span class="n">total</span> <span class="o">=</span> <span class="n">get_bam_reads</span><span class="p">(</span><span class="n">bam_in_obj</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">flush</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>  <span class="c1"># hack for making progress bars work in notebooks; https://github.com/tqdm/tqdm/issues/485#issuecomment-473338308</span>
    <span class="n">pbar_reading</span> <span class="o">=</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">total</span><span class="o">=</span><span class="n">total</span><span class="p">,</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Reading... &quot;</span><span class="p">,</span> <span class="n">unit</span><span class="o">=</span><span class="s2">&quot;reads&quot;</span><span class="p">)</span>
    <span class="n">pbar_writing</span> <span class="o">=</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">total</span><span class="o">=</span><span class="n">total</span><span class="p">,</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;% written from input&quot;</span><span class="p">,</span> <span class="n">unit</span><span class="o">=</span><span class="s2">&quot;reads&quot;</span><span class="p">)</span>
    <span class="n">step</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">total</span> <span class="o">/</span> <span class="mi">10000</span><span class="p">)</span>  <span class="c1"># 10000 total updates</span>

    <span class="c1"># Iterate over reads</span>
    <span class="n">writing_i</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">reading_i</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">written</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">bam_in_obj</span><span class="p">:</span>
        <span class="n">reading_i</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">has_tag</span><span class="p">(</span><span class="n">read_tag</span><span class="p">):</span>
            <span class="n">bc</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="n">read_tag</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">bc</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="c1"># Update step manually - there is an overhead to update per read with hundreds of million reads</span>
        <span class="k">if</span> <span class="n">reading_i</span> <span class="o">==</span> <span class="n">step</span><span class="p">:</span>
            <span class="n">pbar_reading</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">step</span><span class="p">)</span>
            <span class="n">pbar_reading</span><span class="o">.</span><span class="n">refresh</span><span class="p">()</span>
            <span class="n">reading_i</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="c1"># Write read to output bam if barcode is in barcodes</span>
        <span class="k">if</span> <span class="n">bc</span> <span class="ow">in</span> <span class="n">barcodes</span><span class="p">:</span>
            <span class="n">bam_out_obj</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">read</span><span class="p">)</span>
            <span class="n">written</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="k">if</span> <span class="n">writing_i</span> <span class="o">==</span> <span class="n">step</span><span class="p">:</span>
                <span class="n">pbar_writing</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">step</span><span class="p">)</span>
                <span class="n">pbar_writing</span><span class="o">.</span><span class="n">refresh</span><span class="p">()</span>
                <span class="n">writing_i</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="c1"># close progressbars</span>
    <span class="n">pbar_reading</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">pbar_writing</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Close bamfiles</span>
    <span class="n">bam_in_obj</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">bam_out_obj</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Wrote </span><span class="si">{</span><span class="n">written</span><span class="si">}</span><span class="s2"> reads to output bam&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="split_bam_clusters">
<a class="viewcode-back" href="../../../API/tools.html#sctoolbox.tools.bam.split_bam_clusters">[docs]</a>
<span class="nd">@deco</span><span class="o">.</span><span class="n">log_anndata</span>
<span class="nd">@beartype</span>
<span class="k">def</span> <span class="nf">split_bam_clusters</span><span class="p">(</span><span class="n">adata</span><span class="p">:</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span>
                       <span class="n">bams</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
                       <span class="n">groupby</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                       <span class="n">barcode_col</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                       <span class="n">read_tag</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;CB&quot;</span><span class="p">,</span>
                       <span class="n">output_prefix</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;split_&quot;</span><span class="p">,</span>
                       <span class="n">reader_threads</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
                       <span class="n">writer_threads</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
                       <span class="n">parallel</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
                       <span class="n">pysam_threads</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span>
                       <span class="n">buffer_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10000</span><span class="p">,</span>
                       <span class="n">max_queue_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span>
                       <span class="n">individual_pbars</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
                       <span class="n">sort_bams</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
                       <span class="n">index_bams</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Split BAM files into clusters based on &#39;groupby&#39; from the anndata.obs table.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    adata : sc.Anndata</span>
<span class="sd">        Annotated data matrix containing clustered cells in .obs.</span>
<span class="sd">    bams : str | Iterable[str]</span>
<span class="sd">        One or more BAM files to split into clusters</span>
<span class="sd">    groupby : str</span>
<span class="sd">        Name of a column in adata.obs to cluster the pseudobulks by.</span>
<span class="sd">    barcode_col : str, default None</span>
<span class="sd">        Name of a column in adata.obs to use as barcodes. If None, use the index of .obs.</span>
<span class="sd">    read_tag : str, default &quot;CB&quot;</span>
<span class="sd">        Tag to use to identify the reads to split. Must match the barcodes of the barcode_col.</span>
<span class="sd">    output_prefix : str, default `split_`</span>
<span class="sd">        Prefix to use for the output files.</span>
<span class="sd">    reader_threads : int, default 1</span>
<span class="sd">        Number of threads to use for reading.</span>
<span class="sd">    writer_threads : int, default 1</span>
<span class="sd">        Number of threads to use for writing.</span>
<span class="sd">    parallel : bool, default False</span>
<span class="sd">        Whether to enable parallel processsing.</span>
<span class="sd">    pysam_threads : int, default 4</span>
<span class="sd">        Number of threads for pysam.</span>
<span class="sd">    buffer_size : int, default 10000</span>
<span class="sd">        The size of the buffer between readers and writers.</span>
<span class="sd">    max_queue_size : int, default 1000</span>
<span class="sd">        The maximum size of the queue between readers and writers.</span>
<span class="sd">    individual_pbars : bool, default False</span>
<span class="sd">        Whether to show a progress bar for each individual BAM file and output clusters. Default: False (overall progress bars).</span>
<span class="sd">    sort_bams : bool, default False</span>
<span class="sd">        Sort reads in each output bam</span>
<span class="sd">    index_bams : bool, default False</span>
<span class="sd">        Create an index file for each output bam. Will throw an error if `sort_bams` is False.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    ValueError:</span>
<span class="sd">        1. If groupby column is not in adata.obs</span>
<span class="sd">        2. If barcode column is not in adata.obs</span>
<span class="sd">        3. If index_bams is set and sort_bams is False</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># then load modules</span>
    <span class="n">utils</span><span class="o">.</span><span class="n">check_module</span><span class="p">(</span><span class="s2">&quot;tqdm&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">utils</span><span class="o">.</span><span class="n">_is_notebook</span><span class="p">()</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
        <span class="kn">from</span> <span class="nn">tqdm</span> <span class="kn">import</span> <span class="n">tqdm_notebook</span> <span class="k">as</span> <span class="n">tqdm</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="kn">from</span> <span class="nn">tqdm</span> <span class="kn">import</span> <span class="n">tqdm</span>

    <span class="n">utils</span><span class="o">.</span><span class="n">check_module</span><span class="p">(</span><span class="s2">&quot;pysam&quot;</span><span class="p">)</span>
    <span class="kn">import</span> <span class="nn">pysam</span>

    <span class="c1"># check whether groupby and barcode_col are in adata.obs</span>
    <span class="k">if</span> <span class="n">groupby</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Column &#39;</span><span class="si">{</span><span class="n">groupby</span><span class="si">}</span><span class="s2">&#39; not found in adata.obs!&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">barcode_col</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">barcode_col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Column &#39;</span><span class="si">{</span><span class="n">barcode_col</span><span class="si">}</span><span class="s2">&#39; not found in adata.obs!&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">index_bams</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">sort_bams</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`sort_bams=True` must be set for indexing to be possible.&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">bams</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">bams</span> <span class="o">=</span> <span class="p">[</span><span class="n">bams</span><span class="p">]</span>

    <span class="c1"># create output folder if needed</span>
    <span class="n">utils</span><span class="o">.</span><span class="n">create_dir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">output_prefix</span><span class="p">))</span>  <span class="c1"># upper directory of prefix</span>

    <span class="c1"># Establish clusters from obs</span>
    <span class="n">clusters</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">groupby</span><span class="p">]))</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">clusters</span><span class="p">)</span><span class="si">}</span><span class="s2"> groups in .obs.</span><span class="si">{</span><span class="n">groupby</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">clusters</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">writer_threads</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">clusters</span><span class="p">):</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The number of writers (</span><span class="si">{</span><span class="n">writer_threads</span><span class="si">}</span><span class="s2">) is larger than the number of output clusters (</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">clusters</span><span class="p">)</span><span class="si">}</span><span class="s2">). Limiting writer_threads to the number of clusters.&quot;</span><span class="p">)</span>
        <span class="n">writer_threads</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">clusters</span><span class="p">)</span>

    <span class="c1"># setup barcode &lt;-&gt; cluster dict</span>
    <span class="k">if</span> <span class="n">barcode_col</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">barcode2cluster</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">groupby</span><span class="p">]))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">barcode2cluster</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">barcode_col</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">groupby</span><span class="p">]))</span>

    <span class="c1"># create template used for bam header</span>
    <span class="n">template</span> <span class="o">=</span> <span class="n">open_bam</span><span class="p">(</span><span class="n">bams</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s2">&quot;rb&quot;</span><span class="p">,</span> <span class="n">verbosity</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="c1"># Get number in reads in input bam(s)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reading total number of reads from bams...&quot;</span><span class="p">)</span>
    <span class="n">n_reads</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">path</span> <span class="ow">in</span> <span class="n">bams</span><span class="p">:</span>
        <span class="n">handle</span> <span class="o">=</span> <span class="n">open_bam</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">,</span> <span class="n">verbosity</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">n_reads</span><span class="p">[</span><span class="n">path</span><span class="p">]</span> <span class="o">=</span> <span class="n">get_bam_reads</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
        <span class="n">handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># --------- Start splitting --------- #</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Starting splitting of bams...&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">parallel</span><span class="p">:</span>
        <span class="c1"># ---------- parallel splitting ---------- #</span>

        <span class="c1"># create path for output files</span>
        <span class="n">out_paths</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">output_files</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">clusters</span><span class="p">:</span>
            <span class="c1"># replace special characters in filename with &quot;_&quot; https://stackoverflow.com/a/27647173</span>
            <span class="n">save_cluster_name</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;[</span><span class="se">\\</span><span class="s1">/*?:&quot;&lt;&gt;| ]&#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="n">cluster</span><span class="p">)</span>
            <span class="n">out_paths</span><span class="p">[</span><span class="n">cluster</span><span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">output_prefix</span><span class="si">}{</span><span class="n">save_cluster_name</span><span class="si">}</span><span class="s2">.bam&quot;</span>
            <span class="n">output_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">output_prefix</span><span class="si">}{</span><span class="n">save_cluster_name</span><span class="si">}</span><span class="s2">.bam&quot;</span><span class="p">)</span>

        <span class="c1"># ---- Setup pools and queues ---- #</span>
        <span class="c1"># setup pools</span>
        <span class="n">reader_pool</span> <span class="o">=</span> <span class="n">mp</span><span class="o">.</span><span class="n">Pool</span><span class="p">(</span><span class="n">reader_threads</span><span class="p">)</span>
        <span class="n">writer_pool</span> <span class="o">=</span> <span class="n">mp</span><span class="o">.</span><span class="n">Pool</span><span class="p">(</span><span class="n">writer_threads</span><span class="p">)</span>

        <span class="c1"># setup queues to forward data between processes</span>
        <span class="n">manager</span> <span class="o">=</span> <span class="n">mp</span><span class="o">.</span><span class="n">Manager</span><span class="p">()</span>

        <span class="c1"># Make chunks of writer_threads</span>
        <span class="n">cluster_chunks</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">split_list</span><span class="p">(</span><span class="n">clusters</span><span class="p">,</span> <span class="n">writer_threads</span><span class="p">)</span>
        <span class="n">cluster_queues</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">cluster_chunks</span><span class="p">:</span>
            <span class="n">m</span> <span class="o">=</span> <span class="n">manager</span><span class="o">.</span><span class="n">Queue</span><span class="p">(</span><span class="n">maxsize</span><span class="o">=</span><span class="n">max_queue_size</span><span class="p">)</span>  <span class="c1"># setup queue to get reads from readers</span>
            <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">chunk</span><span class="p">:</span>
                <span class="n">cluster_queues</span><span class="p">[</span><span class="n">cluster</span><span class="p">]</span> <span class="o">=</span> <span class="n">m</span>  <span class="c1"># manager is shared between different clusters if writer_threads &lt; number of clusters</span>

        <span class="c1"># Queue for progress bar</span>
        <span class="n">progress_queue</span> <span class="o">=</span> <span class="n">manager</span><span class="o">.</span><span class="n">Queue</span><span class="p">()</span>

        <span class="c1"># ---- Start process ---- #</span>
        <span class="c1"># start reading bams and add reads into respective cluster queue</span>
        <span class="n">reader_jobs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">bam</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">bams</span><span class="p">):</span>
            <span class="n">reader_jobs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">reader_pool</span><span class="o">.</span><span class="n">apply_async</span><span class="p">(</span><span class="n">_buffered_reader</span><span class="p">,</span> <span class="p">(</span><span class="n">bam</span><span class="p">,</span> <span class="n">cluster_queues</span><span class="p">,</span> <span class="n">barcode2cluster</span><span class="p">,</span> <span class="n">read_tag</span><span class="p">,</span> <span class="n">progress_queue</span><span class="p">,</span> <span class="n">buffer_size</span><span class="p">)))</span>
        <span class="n">reader_pool</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>  <span class="c1"># no more tasks to add</span>

        <span class="c1"># write reads to files; one process per file</span>
        <span class="n">writer_jobs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">cluster_chunks</span><span class="p">:</span>
            <span class="n">queue</span> <span class="o">=</span> <span class="n">cluster_queues</span><span class="p">[</span><span class="n">chunk</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>  <span class="c1"># same queue for the whole chunk</span>
            <span class="n">path_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">cluster</span><span class="p">:</span> <span class="n">out_paths</span><span class="p">[</span><span class="n">cluster</span><span class="p">]</span> <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">chunk</span><span class="p">}</span>  <span class="c1"># subset of out_paths specific for this writer</span>
            <span class="n">writer_jobs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">writer_pool</span><span class="o">.</span><span class="n">apply_async</span><span class="p">(</span><span class="n">_writer</span><span class="p">,</span> <span class="p">(</span><span class="n">queue</span><span class="p">,</span> <span class="n">path_dict</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">template</span><span class="o">.</span><span class="n">header</span><span class="p">),</span> <span class="n">progress_queue</span><span class="p">,</span> <span class="n">pysam_threads</span><span class="p">)))</span>
        <span class="n">writer_pool</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="c1"># ---- End process + cleanup ---- #</span>

        <span class="c1"># wait for readers and writers to finish</span>
        <span class="n">_monitor_progress</span><span class="p">(</span><span class="n">progress_queue</span><span class="p">,</span> <span class="n">cluster_queues</span><span class="p">,</span> <span class="n">reader_jobs</span><span class="p">,</span> <span class="n">writer_jobs</span><span class="p">,</span> <span class="n">n_reads</span><span class="p">,</span> <span class="n">individual_pbars</span><span class="p">)</span>  <span class="c1"># does not return until all readers/writers are finished</span>
        <span class="n">reader_pool</span><span class="o">.</span><span class="n">join</span><span class="p">()</span>
        <span class="n">writer_pool</span><span class="o">.</span><span class="n">join</span><span class="p">()</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># ---------- sequential splitting ---------- #</span>
        <span class="c1"># open output bam files</span>
        <span class="n">handles</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">output_files</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">clusters</span><span class="p">:</span>
            <span class="c1"># replace special characters in filename with &quot;_&quot; https://stackoverflow.com/a/27647173</span>
            <span class="n">save_cluster_name</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;[</span><span class="se">\\</span><span class="s1">/*?:&quot;&lt;&gt;| ]&#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="n">cluster</span><span class="p">)</span>
            <span class="n">f_out</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">output_prefix</span><span class="si">}{</span><span class="n">save_cluster_name</span><span class="si">}</span><span class="s2">.bam&quot;</span>
            <span class="n">handles</span><span class="p">[</span><span class="n">cluster</span><span class="p">]</span> <span class="o">=</span> <span class="n">open_bam</span><span class="p">(</span><span class="n">f_out</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">,</span> <span class="n">template</span><span class="o">=</span><span class="n">template</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="n">pysam_threads</span><span class="p">,</span> <span class="n">verbosity</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

            <span class="n">output_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">f_out</span><span class="p">)</span>

        <span class="c1"># Loop over bamfile(s)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">bam</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">bams</span><span class="p">):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Looping over reads from </span><span class="si">{</span><span class="n">bam</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">bams</span><span class="p">)</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>

            <span class="n">bam_obj</span> <span class="o">=</span> <span class="n">open_bam</span><span class="p">(</span><span class="n">bam</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">,</span> <span class="n">verbosity</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

            <span class="c1"># Update progress based on total number of reads</span>
            <span class="n">total</span> <span class="o">=</span> <span class="n">get_bam_reads</span><span class="p">(</span><span class="n">bam_obj</span><span class="p">)</span>
            <span class="n">pbar</span> <span class="o">=</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">total</span><span class="o">=</span><span class="n">total</span><span class="p">)</span>
            <span class="n">step</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">total</span> <span class="o">/</span> <span class="mi">10000</span><span class="p">)</span>  <span class="c1"># 10000 total updates</span>

            <span class="n">i</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">written</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">bam_obj</span><span class="p">:</span>
                <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">has_tag</span><span class="p">(</span><span class="n">read_tag</span><span class="p">):</span>
                    <span class="n">bc</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="n">read_tag</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">bc</span> <span class="o">=</span> <span class="kc">None</span>

                <span class="c1"># Update step manually - there is an overhead to update per read with hundreds of million reads</span>
                <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="n">step</span><span class="p">:</span>
                    <span class="n">pbar</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">step</span><span class="p">)</span>
                    <span class="n">i</span> <span class="o">=</span> <span class="mi">0</span>

                <span class="k">if</span> <span class="n">bc</span> <span class="ow">in</span> <span class="n">barcode2cluster</span><span class="p">:</span>
                    <span class="n">cluster</span> <span class="o">=</span> <span class="n">barcode2cluster</span><span class="p">[</span><span class="n">bc</span><span class="p">]</span>
                    <span class="n">handles</span><span class="p">[</span><span class="n">cluster</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">read</span><span class="p">)</span>
                    <span class="n">written</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="c1"># close progressbar</span>
            <span class="n">pbar</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Wrote </span><span class="si">{</span><span class="n">written</span><span class="si">}</span><span class="s2"> reads to cluster files&quot;</span><span class="p">)</span>

        <span class="c1"># Close all files</span>
        <span class="k">for</span> <span class="n">handle</span> <span class="ow">in</span> <span class="n">handles</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># ---------- post split functionality ---------- #</span>
    <span class="c1"># sort reads</span>
    <span class="k">if</span> <span class="n">sort_bams</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Sorting output bams...&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">output_files</span><span class="p">,</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Sorting reads&quot;</span><span class="p">,</span> <span class="n">unit</span><span class="o">=</span><span class="s2">&quot;files&quot;</span><span class="p">):</span>
            <span class="n">temp_file</span> <span class="o">=</span> <span class="n">file</span> <span class="o">+</span> <span class="s2">&quot;.tmp&quot;</span>  <span class="c1"># temporary sort file</span>
            <span class="n">pysam</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="s2">&quot;-o&quot;</span><span class="p">,</span> <span class="n">temp_file</span><span class="p">,</span> <span class="n">file</span><span class="p">)</span>
            <span class="n">os</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">temp_file</span><span class="p">,</span> <span class="n">file</span><span class="p">)</span>

    <span class="c1"># index files</span>
    <span class="k">if</span> <span class="n">index_bams</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Indexing output bams...&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">output_files</span><span class="p">,</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Indexing&quot;</span><span class="p">,</span> <span class="n">unit</span><span class="o">=</span><span class="s2">&quot;files&quot;</span><span class="p">):</span>
            <span class="n">pysam</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="s2">&quot;-@&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">pysam_threads</span><span class="p">))</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished splitting bams!&quot;</span><span class="p">)</span></div>



<span class="c1"># ------------------------------------------------------------------ #</span>
<span class="c1"># ---------------------- bam helper functions ---------------------- #</span>
<span class="c1"># ------------------------------------------------------------------ #</span>

<div class="viewcode-block" id="open_bam">
<a class="viewcode-back" href="../../../API/tools.html#sctoolbox.tools.bam.open_bam">[docs]</a>
<span class="nd">@beartype</span>
<span class="k">def</span> <span class="nf">open_bam</span><span class="p">(</span><span class="n">file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
             <span class="n">mode</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
             <span class="n">verbosity</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">]</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="n">Any</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;pysam.AlignmentFile&quot;</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Open bam file with pysam.AlignmentFile. On a specific verbosity level.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    file : str</span>
<span class="sd">        Path to bam file.</span>
<span class="sd">    mode : str</span>
<span class="sd">        Mode to open the file in. See pysam.AlignmentFile</span>
<span class="sd">    verbosity : Literal[0, 1, 2, 3], default 3</span>
<span class="sd">        Set verbosity level. Verbosity level 0 for no messages.</span>
<span class="sd">    **kwargs : Any</span>
<span class="sd">        Forwarded to pysam.AlignmentFile</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    pysam.AlignmentFile</span>
<span class="sd">        Object to work on SAM/BAM files.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># check then load modules</span>
    <span class="n">utils</span><span class="o">.</span><span class="n">check_module</span><span class="p">(</span><span class="s2">&quot;pysam&quot;</span><span class="p">)</span>
    <span class="kn">import</span> <span class="nn">pysam</span>

    <span class="c1"># save verbosity, then set temporary one</span>
    <span class="n">former_verbosity</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">get_verbosity</span><span class="p">()</span>
    <span class="n">pysam</span><span class="o">.</span><span class="n">set_verbosity</span><span class="p">(</span><span class="n">verbosity</span><span class="p">)</span>

    <span class="c1"># open file</span>
    <span class="n">handle</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">mode</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

    <span class="c1"># return to former verbosity</span>
    <span class="n">pysam</span><span class="o">.</span><span class="n">set_verbosity</span><span class="p">(</span><span class="n">former_verbosity</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">handle</span></div>



<div class="viewcode-block" id="get_bam_reads">
<a class="viewcode-back" href="../../../API/tools.html#sctoolbox.tools.bam.get_bam_reads">[docs]</a>
<span class="nd">@beartype</span>
<span class="k">def</span> <span class="nf">get_bam_reads</span><span class="p">(</span><span class="n">bam_obj</span><span class="p">:</span> <span class="s2">&quot;pysam.AlignmentFile&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the number of reads from an open pysam.AlignmentFile.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    bam_obj : pysam.AlignmentFile</span>
<span class="sd">        An open pysam.AlignmentFile object to get the number of reads from.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    int</span>
<span class="sd">        number of reads in the bam file</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># check then load modules</span>
    <span class="n">utils</span><span class="o">.</span><span class="n">check_module</span><span class="p">(</span><span class="s2">&quot;pysam&quot;</span><span class="p">)</span>
    <span class="kn">import</span> <span class="nn">pysam</span>

    <span class="c1"># Get number of reads in bam</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">total</span> <span class="o">=</span> <span class="n">bam_obj</span><span class="o">.</span><span class="n">mapped</span> <span class="o">+</span> <span class="n">bam_obj</span><span class="o">.</span><span class="n">unmapped</span>
    <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>  <span class="c1"># fall back to &quot;samtools view -c file&quot; if bam_obj.mapped is not available</span>
        <span class="n">path</span> <span class="o">=</span> <span class="n">bam_obj</span><span class="o">.</span><span class="n">filename</span>
        <span class="n">total</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">pysam</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="s2">&quot;-c&quot;</span><span class="p">,</span> <span class="n">path</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">total</span></div>



<span class="c1"># ------------------------------------------------------------------ #</span>
<span class="c1"># -------------------- multiprocessing functions ------------------- #</span>
<span class="c1"># ------------------------------------------------------------------ #</span>


<span class="nd">@beartype</span>
<span class="k">def</span> <span class="nf">_monitor_progress</span><span class="p">(</span><span class="n">progress_queue</span><span class="p">:</span> <span class="n">Any</span><span class="p">,</span>
                      <span class="n">cluster_queues</span><span class="p">:</span> <span class="n">Any</span><span class="p">,</span>
                      <span class="n">reader_jobs</span><span class="p">:</span> <span class="n">Any</span><span class="p">,</span>
                      <span class="n">writer_jobs</span><span class="p">:</span> <span class="n">Any</span><span class="p">,</span>
                      <span class="n">total_reads</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">],</span>
                      <span class="n">individual_pbars</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Monitor read/write progress of split_bam_clusters.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    progress_queue : multiprocessing.queues.Queue</span>
<span class="sd">        Queue to send progress updates to.</span>
<span class="sd">    cluster_queues : dict[str, multiprocessing.queues.Queue]</span>
<span class="sd">        Dictionary of queues which reads were sent to.</span>
<span class="sd">    reader_jobs : list[multiprocessing.pool.AsyncResult]</span>
<span class="sd">        List of reader jobs.</span>
<span class="sd">    writer_jobs : list[multiprocessing.pool.AsyncResult]</span>
<span class="sd">        List of writer jobs.</span>
<span class="sd">    total_reads : dict[str, int]</span>
<span class="sd">        Dictionary containing total reads for each input .bam-file. Keys are filenames and values are total reads.</span>
<span class="sd">    individual_pbars : bool, default False</span>
<span class="sd">        If True, show progress bar for each cluster. If False, show one progress bar for all clusters.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    int</span>
<span class="sd">        Returns 0 on success.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Check parameter that beartype cannot cover</span>
    <span class="n">utils</span><span class="o">.</span><span class="n">check_type</span><span class="p">(</span><span class="n">progress_queue</span><span class="p">,</span> <span class="s2">&quot;progress_queue&quot;</span><span class="p">,</span> <span class="n">BaseProxy</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">cluster_queues</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="n">utils</span><span class="o">.</span><span class="n">check_type</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;cluster_queues value&quot;</span><span class="p">,</span> <span class="n">BaseProxy</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">jobs</span> <span class="ow">in</span> <span class="n">reader_jobs</span><span class="p">:</span>
        <span class="n">utils</span><span class="o">.</span><span class="n">check_type</span><span class="p">(</span><span class="n">jobs</span><span class="p">,</span> <span class="s2">&quot;reader_jobs&quot;</span><span class="p">,</span> <span class="n">ApplyResult</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">jobs</span> <span class="ow">in</span> <span class="n">writer_jobs</span><span class="p">:</span>
        <span class="n">utils</span><span class="o">.</span><span class="n">check_type</span><span class="p">(</span><span class="n">jobs</span><span class="p">,</span> <span class="s2">&quot;writer_jobs&quot;</span><span class="p">,</span> <span class="n">ApplyResult</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">utils</span><span class="o">.</span><span class="n">_is_notebook</span><span class="p">()</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
        <span class="kn">from</span> <span class="nn">tqdm</span> <span class="kn">import</span> <span class="n">tqdm_notebook</span> <span class="k">as</span> <span class="n">tqdm</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="kn">from</span> <span class="nn">tqdm</span> <span class="kn">import</span> <span class="n">tqdm</span>

    <span class="n">cluster_names</span> <span class="o">=</span> <span class="n">cluster_queues</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>

    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">flush</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>  <span class="c1"># hack for making progress bars work in notebooks; https://github.com/tqdm/tqdm/issues/485#issuecomment-473338308</span>

    <span class="c1"># Initialize progress bars dependent on individual_pbars true/false</span>
    <span class="n">pbars</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">if</span> <span class="n">individual_pbars</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>  <span class="c1"># individual progress per reader/writer</span>

        <span class="c1"># readers</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">total_reads</span><span class="p">):</span>
            <span class="n">pbars</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">total</span><span class="o">=</span><span class="n">total_reads</span><span class="p">[</span><span class="n">name</span><span class="p">],</span> <span class="n">position</span><span class="o">=</span><span class="n">i</span><span class="p">,</span> <span class="n">desc</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Reading (</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">,</span> <span class="n">unit</span><span class="o">=</span><span class="s2">&quot;reads&quot;</span><span class="p">)</span>

        <span class="c1"># writers</span>
        <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">cluster_names</span><span class="p">):</span>
            <span class="n">pbars</span><span class="p">[</span><span class="n">cluster</span><span class="p">]</span> <span class="o">=</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">total</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">position</span><span class="o">=</span><span class="n">i</span> <span class="o">+</span> <span class="n">j</span><span class="p">,</span> <span class="n">desc</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Writing queued reads (</span><span class="si">{</span><span class="n">cluster</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>  <span class="c1"># total is 1 instead of 0 to trigger progress bar</span>
            <span class="n">pbars</span><span class="p">[</span><span class="n">cluster</span><span class="p">]</span><span class="o">.</span><span class="n">total</span> <span class="o">=</span> <span class="mi">0</span>  <span class="c1"># reset to 0 (bar is still shown)</span>
            <span class="n">pbars</span><span class="p">[</span><span class="n">cluster</span><span class="p">]</span><span class="o">.</span><span class="n">refresh</span><span class="p">()</span>

    <span class="k">else</span><span class="p">:</span>   <span class="c1"># merged progress for reading and writing</span>

        <span class="c1"># readers</span>
        <span class="n">sum_reads</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">total_reads</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
        <span class="n">pbar</span> <span class="o">=</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">total</span><span class="o">=</span><span class="n">sum_reads</span><span class="p">,</span> <span class="n">position</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Reading from bams&quot;</span><span class="p">,</span> <span class="n">unit</span><span class="o">=</span><span class="s2">&quot;reads&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">total_reads</span><span class="p">:</span>
            <span class="n">pbars</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">pbar</span>  <span class="c1"># all names share the same pbar</span>

        <span class="c1"># writers</span>
        <span class="n">pbar</span> <span class="o">=</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">total</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">position</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Writing queued reads&quot;</span><span class="p">,</span> <span class="n">unit</span><span class="o">=</span><span class="s2">&quot;reads&quot;</span><span class="p">)</span>
        <span class="n">pbar</span><span class="o">.</span><span class="n">total</span> <span class="o">=</span> <span class="mi">0</span>  <span class="c1"># reset to 0 (bar is still shown)</span>
        <span class="n">pbar</span><span class="o">.</span><span class="n">refresh</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">cluster_names</span><span class="p">:</span>
            <span class="n">pbars</span><span class="p">[</span><span class="n">cluster</span><span class="p">]</span> <span class="o">=</span> <span class="n">pbar</span>  <span class="c1"># all clusters share the same pbar</span>

    <span class="c1"># Fetch progress from readers/writers</span>
    <span class="n">writers_running</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">writer_jobs</span><span class="p">)</span>
    <span class="n">reading_done</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>

        <span class="n">task</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">value</span> <span class="o">=</span> <span class="n">progress_queue</span><span class="o">.</span><span class="n">get</span><span class="p">()</span>
        <span class="n">pbar</span> <span class="o">=</span> <span class="n">pbars</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>

        <span class="c1"># update pbar depending on the task</span>
        <span class="k">if</span> <span class="n">task</span> <span class="o">==</span> <span class="s2">&quot;read&quot;</span><span class="p">:</span>
            <span class="n">pbar</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
            <span class="n">pbar</span><span class="o">.</span><span class="n">refresh</span><span class="p">()</span>

        <span class="k">elif</span> <span class="n">task</span> <span class="o">==</span> <span class="s2">&quot;sent&quot;</span><span class="p">:</span>
            <span class="n">pbar</span><span class="o">.</span><span class="n">total</span> <span class="o">+=</span> <span class="n">value</span>
            <span class="n">pbar</span><span class="o">.</span><span class="n">refresh</span><span class="p">()</span>

        <span class="k">elif</span> <span class="n">task</span> <span class="o">==</span> <span class="s2">&quot;written&quot;</span><span class="p">:</span>
            <span class="n">pbar</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
            <span class="n">pbar</span><span class="o">.</span><span class="n">refresh</span><span class="p">()</span>

        <span class="k">elif</span> <span class="n">task</span> <span class="o">==</span> <span class="s2">&quot;done&quot;</span><span class="p">:</span>
            <span class="n">writers_running</span> <span class="o">-=</span> <span class="mi">1</span>

        <span class="c1"># Check if all reads were read by comparing to pbar total</span>
        <span class="n">reader_pbars</span> <span class="o">=</span> <span class="p">[</span><span class="n">pbars</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">total_reads</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">reading_done</span> <span class="ow">is</span> <span class="kc">False</span> <span class="ow">and</span> <span class="p">(</span><span class="nb">sum</span><span class="p">([</span><span class="n">pbar</span><span class="o">.</span><span class="n">total</span> <span class="k">for</span> <span class="n">pbar</span> <span class="ow">in</span> <span class="n">reader_pbars</span><span class="p">])</span> <span class="o">&gt;=</span> <span class="nb">sum</span><span class="p">([</span><span class="n">pbar</span><span class="o">.</span><span class="n">n</span> <span class="k">for</span> <span class="n">pbar</span> <span class="ow">in</span> <span class="n">reader_pbars</span><span class="p">])):</span>
            <span class="n">reading_done</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">_</span> <span class="o">=</span> <span class="p">[</span><span class="n">reader_job</span><span class="o">.</span><span class="n">get</span><span class="p">()</span> <span class="k">for</span> <span class="n">reader_job</span> <span class="ow">in</span> <span class="n">reader_jobs</span><span class="p">]</span>  <span class="c1"># wait for all writers to finish</span>

            <span class="k">for</span> <span class="n">queue</span> <span class="ow">in</span> <span class="n">cluster_queues</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
                <span class="n">queue</span><span class="o">.</span><span class="n">put</span><span class="p">((</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">))</span>  <span class="c1"># Insert None into queue to signal end of reads</span>

        <span class="c1"># Check if all writers are done</span>
        <span class="k">if</span> <span class="n">writers_running</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">_</span> <span class="o">=</span> <span class="p">[</span><span class="n">writer_job</span><span class="o">.</span><span class="n">get</span><span class="p">()</span> <span class="k">for</span> <span class="n">writer_job</span> <span class="ow">in</span> <span class="n">writer_jobs</span><span class="p">]</span>  <span class="c1"># wait for all readers to finish</span>
            <span class="k">break</span>

    <span class="k">return</span> <span class="mi">0</span>  <span class="c1"># success</span>


<span class="c1"># @beartype beartype seems to not work with this function</span>
<span class="k">def</span> <span class="nf">_buffered_reader</span><span class="p">(</span><span class="n">path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                     <span class="n">out_queues</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
                     <span class="n">bc2cluster</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">],</span>
                     <span class="n">tag</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                     <span class="n">progress_queue</span><span class="p">:</span> <span class="n">Any</span><span class="p">,</span>
                     <span class="n">buffer_size</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="mi">10000</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Open bam file and add reads to respective output queue.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    path : str</span>
<span class="sd">        Path to bam file.</span>
<span class="sd">    out_queues : dict[str | int, multiprocessing.queues.Queue]</span>
<span class="sd">        Dict of multiprocesssing.Queues with cluster as key.</span>
<span class="sd">    bc2cluster : dict[str | int, str | int]</span>
<span class="sd">        Dict of clusters with barcode as key.</span>
<span class="sd">    tag : str</span>
<span class="sd">        Read tag that should be used for queue assignment.</span>
<span class="sd">    progress_queue : multiprocessing.queues.Queue</span>
<span class="sd">        Queue to send progress updates to.</span>
<span class="sd">    buffer_size : int, default 10000</span>
<span class="sd">        Size of buffer (number of reads) for each queue to collect before writing.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    int</span>
<span class="sd">        Returns 0 on success.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    Exception</span>
<span class="sd">        If buffered reader failes.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Test parameter types not covered by beartype</span>
    <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">out_queues</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="n">utils</span><span class="o">.</span><span class="n">check_type</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;out_queues.values&quot;</span><span class="p">,</span> <span class="n">BaseProxy</span><span class="p">)</span>
    <span class="n">utils</span><span class="o">.</span><span class="n">check_type</span><span class="p">(</span><span class="n">progress_queue</span><span class="p">,</span> <span class="s2">&quot;progress_queue&quot;</span><span class="p">,</span> <span class="n">BaseProxy</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># open bam</span>
        <span class="n">bam</span> <span class="o">=</span> <span class="n">open_bam</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">,</span> <span class="n">verbosity</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

        <span class="c1"># Setup read buffer per cluster</span>
        <span class="n">read_buffer</span> <span class="o">=</span> <span class="p">{</span><span class="n">cluster</span><span class="p">:</span> <span class="p">[]</span> <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">bc2cluster</span><span class="o">.</span><span class="n">values</span><span class="p">())}</span>

        <span class="c1"># put each read into correct queue</span>
        <span class="n">step</span> <span class="o">=</span> <span class="mi">100000</span>  <span class="c1"># progress every in hundred thousand reads</span>
        <span class="n">n_reads_step</span> <span class="o">=</span> <span class="mi">0</span>  <span class="c1"># count of reads read from bam per step</span>
        <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">bam</span><span class="p">:</span>

            <span class="c1"># get barcode</span>
            <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">has_tag</span><span class="p">(</span><span class="n">tag</span><span class="p">):</span>
                <span class="n">bc</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="n">tag</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">bc</span> <span class="o">=</span> <span class="kc">None</span>

            <span class="c1"># put read into matching cluster queue</span>
            <span class="k">if</span> <span class="n">bc</span> <span class="ow">in</span> <span class="n">bc2cluster</span><span class="p">:</span>
                <span class="n">cluster</span> <span class="o">=</span> <span class="n">bc2cluster</span><span class="p">[</span><span class="n">bc</span><span class="p">]</span>
                <span class="n">read_buffer</span><span class="p">[</span><span class="n">cluster</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">to_string</span><span class="p">())</span>

                <span class="c1"># Send reads to buffer when buffer size is reached</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">read_buffer</span><span class="p">[</span><span class="n">cluster</span><span class="p">])</span> <span class="o">==</span> <span class="n">buffer_size</span><span class="p">:</span>
                    <span class="n">out_queues</span><span class="p">[</span><span class="n">cluster</span><span class="p">]</span><span class="o">.</span><span class="n">put</span><span class="p">((</span><span class="n">cluster</span><span class="p">,</span> <span class="n">read_buffer</span><span class="p">[</span><span class="n">cluster</span><span class="p">]))</span>  <span class="c1"># send the tuple of (clustername, read_buffer) to queue</span>
                    <span class="n">progress_queue</span><span class="o">.</span><span class="n">put</span><span class="p">((</span><span class="s2">&quot;sent&quot;</span><span class="p">,</span> <span class="n">cluster</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">read_buffer</span><span class="p">[</span><span class="n">cluster</span><span class="p">])))</span>
                    <span class="n">read_buffer</span><span class="p">[</span><span class="n">cluster</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

            <span class="c1"># update progress bar for reading</span>
            <span class="n">n_reads_step</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="n">n_reads_step</span> <span class="o">==</span> <span class="n">step</span><span class="p">:</span>
                <span class="n">progress_queue</span><span class="o">.</span><span class="n">put</span><span class="p">((</span><span class="s2">&quot;read&quot;</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="n">n_reads_step</span><span class="p">))</span>
                <span class="n">n_reads_step</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="c1"># all reads have been read; add remaining reads to progress</span>
        <span class="n">progress_queue</span><span class="o">.</span><span class="n">put</span><span class="p">((</span><span class="s2">&quot;read&quot;</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="n">n_reads_step</span><span class="p">))</span>

        <span class="c1"># Send remaining reads to buffer</span>
        <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">read_buffer</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">read_buffer</span><span class="p">[</span><span class="n">cluster</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">out_queues</span><span class="p">[</span><span class="n">cluster</span><span class="p">]</span><span class="o">.</span><span class="n">put</span><span class="p">((</span><span class="n">cluster</span><span class="p">,</span> <span class="n">read_buffer</span><span class="p">[</span><span class="n">cluster</span><span class="p">]))</span>
                <span class="n">progress_queue</span><span class="o">.</span><span class="n">put</span><span class="p">((</span><span class="s2">&quot;sent&quot;</span><span class="p">,</span> <span class="n">cluster</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">read_buffer</span><span class="p">[</span><span class="n">cluster</span><span class="p">])))</span>

        <span class="k">return</span> <span class="mi">0</span>  <span class="c1"># success</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">message</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">e</span>


<span class="nd">@beartype</span>
<span class="k">def</span> <span class="nf">_writer</span><span class="p">(</span><span class="n">read_queue</span><span class="p">:</span> <span class="n">Any</span><span class="p">,</span>
            <span class="n">out_paths</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span>
            <span class="n">bam_header</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
            <span class="n">progress_queue</span><span class="p">:</span> <span class="n">Any</span><span class="p">,</span>
            <span class="n">pysam_threads</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">4</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Write reads to given file.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    read_queue : multiprocessing.queues.Queue</span>
<span class="sd">        Queue of reads to be written into file.</span>
<span class="sd">    out_paths : dict</span>
<span class="sd">        Path to output files for this writer. In the format {cluster1: &lt;path&gt;, cluster2: &lt;path&gt;}</span>
<span class="sd">    bam_header : str</span>
<span class="sd">        pysam.AlignmentHeader used as template for output bam.</span>
<span class="sd">    progress_queue : multiprocessing.queues.Queue</span>
<span class="sd">        Queue to send progress updates to.</span>
<span class="sd">    pysam_threads : int, default 4</span>
<span class="sd">        Number of threads for pysam to use for writing. This is different from the threads used for the individual writers.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    int</span>
<span class="sd">        Returns 0 on success.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    Exception</span>
<span class="sd">        If buffered reader failes.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Check parameter that are not covered by beartype</span>
    <span class="n">utils</span><span class="o">.</span><span class="n">check_type</span><span class="p">(</span><span class="n">read_queue</span><span class="p">,</span> <span class="s2">&quot;read_queue&quot;</span><span class="p">,</span> <span class="n">BaseProxy</span><span class="p">)</span>
    <span class="n">utils</span><span class="o">.</span><span class="n">check_type</span><span class="p">(</span><span class="n">progress_queue</span><span class="p">,</span> <span class="s2">&quot;progress_queue&quot;</span><span class="p">,</span> <span class="n">BaseProxy</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="kn">import</span> <span class="nn">pysam</span>  <span class="c1"># install of pysam was checked in parent function</span>

        <span class="n">cluster_names</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">out_paths</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

        <span class="c1"># open bam files for writing cluster reads</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">handles</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">cluster_names</span><span class="p">:</span>
                <span class="n">handles</span><span class="p">[</span><span class="n">cluster</span><span class="p">]</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">out_paths</span><span class="p">[</span><span class="n">cluster</span><span class="p">],</span> <span class="s2">&quot;wb&quot;</span><span class="p">,</span> <span class="n">text</span><span class="o">=</span><span class="n">bam_header</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="n">pysam_threads</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">message</span><span class="p">)</span>
            <span class="k">raise</span> <span class="n">e</span>

        <span class="c1"># fetch reads from queue and write to bam</span>
        <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
            <span class="n">cluster</span><span class="p">,</span> <span class="n">read_lst</span> <span class="o">=</span> <span class="n">read_queue</span><span class="o">.</span><span class="n">get</span><span class="p">()</span>

            <span class="c1"># stop writing</span>
            <span class="k">if</span> <span class="n">cluster</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">break</span>

            <span class="c1"># write reads to file</span>
            <span class="n">handle</span> <span class="o">=</span> <span class="n">handles</span><span class="p">[</span><span class="n">cluster</span><span class="p">]</span>  <span class="c1"># the handle for this cluster</span>
            <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">read_lst</span><span class="p">:</span>

                <span class="c1"># create read object and write to handle</span>
                <span class="n">read</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignedSegment</span><span class="o">.</span><span class="n">fromstring</span><span class="p">(</span><span class="n">read</span><span class="p">,</span> <span class="n">handle</span><span class="o">.</span><span class="n">header</span><span class="p">)</span>
                <span class="n">handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">read</span><span class="p">)</span>

            <span class="c1"># Send written to progress</span>
            <span class="n">progress_queue</span><span class="o">.</span><span class="n">put</span><span class="p">((</span><span class="s2">&quot;written&quot;</span><span class="p">,</span> <span class="n">cluster</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">read_lst</span><span class="p">)))</span>

        <span class="c1"># Close files after use</span>
        <span class="k">for</span> <span class="n">handle</span> <span class="ow">in</span> <span class="n">handles</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="c1"># Tell monitoring queue that this writing job is done</span>
        <span class="n">progress_queue</span><span class="o">.</span><span class="n">put</span><span class="p">((</span><span class="s2">&quot;done&quot;</span><span class="p">,</span> <span class="n">cluster_names</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="kc">None</span><span class="p">))</span>  <span class="c1"># cluster is not used here, but is needed for the progress queue</span>

        <span class="k">return</span> <span class="mi">0</span>  <span class="c1"># success</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">message</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">e</span>


<div class="viewcode-block" id="bam_to_bigwig">
<a class="viewcode-back" href="../../../API/tools.html#sctoolbox.tools.bam.bam_to_bigwig">[docs]</a>
<span class="nd">@beartype</span>
<span class="k">def</span> <span class="nf">bam_to_bigwig</span><span class="p">(</span><span class="n">bam</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                  <span class="n">output</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                  <span class="n">scale</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
                  <span class="n">overwrite</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
                  <span class="n">tempdir</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;.&quot;</span><span class="p">,</span>
                  <span class="n">remove_temp</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
                  <span class="n">bedtools_path</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                  <span class="n">bgtobw_path</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Convert reads in a bam-file to bigwig format.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    bam : str</span>
<span class="sd">        Path to bam file.</span>
<span class="sd">    output : Optional[str], default None</span>
<span class="sd">        Path to output file. If None, output is written to same directory as bam file with same name and .bw extension.</span>
<span class="sd">    scale : bool, default True</span>
<span class="sd">        Scale output depth to reads per million mapped reads.</span>
<span class="sd">    overwrite : bool, default True</span>
<span class="sd">        Overwrite output file if it already exists.</span>
<span class="sd">    tempdir : str, default &quot;.&quot;</span>
<span class="sd">        Path to directory where temporary files are written.</span>
<span class="sd">    remove_temp : bool, default True</span>
<span class="sd">        Remove temporary files after conversion.</span>
<span class="sd">    bedtools_path : Optional[str], default None</span>
<span class="sd">        Path to bedtools binary. If None, the function will search for the binary in the path.</span>
<span class="sd">    bgtobw_path : Optional[str], default None</span>
<span class="sd">        Path to bedGraphToBigWig binary. If None, the function will search for the binary in the path.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>
<span class="sd">        Path to output file.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Set output name and check if bigwig already exists</span>
    <span class="k">if</span> <span class="n">output</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">output</span> <span class="o">=</span> <span class="n">bam</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.bam&quot;</span><span class="p">,</span> <span class="s2">&quot;.bw&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">output</span><span class="p">)</span> <span class="ow">and</span> <span class="n">overwrite</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Output file already exists. Set overwrite=True to overwrite.&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">output</span>

    <span class="c1"># Check required modules</span>
    <span class="n">utils</span><span class="o">.</span><span class="n">check_module</span><span class="p">(</span><span class="s2">&quot;pysam&quot;</span><span class="p">)</span>
    <span class="kn">import</span> <span class="nn">pysam</span>

    <span class="k">if</span> <span class="n">bedtools_path</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">bedtools_path</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">get_binary_path</span><span class="p">(</span><span class="s2">&quot;bedtools&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">bgtobw_path</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">bgtobw_path</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">get_binary_path</span><span class="p">(</span><span class="s2">&quot;bedGraphToBigWig&quot;</span><span class="p">)</span>

    <span class="c1"># Get size of genome and write a chromsizes file</span>
    <span class="n">bamobj</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">bam</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>
    <span class="n">chromsizes</span> <span class="o">=</span> <span class="p">{</span><span class="n">chrom</span><span class="p">:</span> <span class="n">bamobj</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">bamobj</span><span class="o">.</span><span class="n">references</span><span class="p">)}</span>
    <span class="n">chromsizes_file</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">get_temporary_filename</span><span class="p">(</span><span class="n">tempdir</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">chromsizes_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">size</span> <span class="ow">in</span> <span class="n">chromsizes</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">size</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Get number of mapped reads in file for normalization</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Getting scaling factor&quot;</span><span class="p">)</span>
    <span class="n">scaling_factor</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">if</span> <span class="n">scale</span><span class="p">:</span>
        <span class="n">n_reads</span> <span class="o">=</span> <span class="n">get_bam_reads</span><span class="p">(</span><span class="n">bamobj</span><span class="p">)</span>
        <span class="n">scaling_factor</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="n">n_reads</span> <span class="o">/</span> <span class="mf">1e6</span><span class="p">)</span>
        <span class="n">scaling_factor</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">scaling_factor</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>
    <span class="n">bamobj</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Convert bam to bedgraph</span>
    <span class="n">bedgraph_out</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">get_temporary_filename</span><span class="p">(</span><span class="n">tempdir</span><span class="p">)</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">bedtools_path</span><span class="si">}</span><span class="s2"> genomecov -bg -ibam </span><span class="si">{</span><span class="n">bam</span><span class="si">}</span><span class="s2"> &gt; </span><span class="si">{</span><span class="n">bedgraph_out</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running: &quot;</span> <span class="o">+</span> <span class="n">cmd</span><span class="p">)</span>
    <span class="n">utils</span><span class="o">.</span><span class="n">run_cmd</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>

    <span class="c1"># Sort and scale input</span>
    <span class="n">bedgraph_out_sorted</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">get_temporary_filename</span><span class="p">(</span><span class="n">tempdir</span><span class="p">)</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;sort -k1,1 -k2,2n -T </span><span class="si">{</span><span class="n">tempdir</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">bedgraph_out</span><span class="si">}</span><span class="s2"> |  awk &#39;</span><span class="se">{{</span><span class="s2">$4=$4*</span><span class="si">{</span><span class="n">scaling_factor</span><span class="si">}</span><span class="s2">; print $0</span><span class="se">}}</span><span class="s2">&#39; &gt; </span><span class="si">{</span><span class="n">bedgraph_out_sorted</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running: &quot;</span> <span class="o">+</span> <span class="n">cmd</span><span class="p">)</span>
    <span class="n">utils</span><span class="o">.</span><span class="n">run_cmd</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>

    <span class="c1"># Convert bedgraph to bigwig</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">bgtobw_path</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">bedgraph_out_sorted</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">chromsizes_file</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">output</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running: &quot;</span> <span class="o">+</span> <span class="n">cmd</span><span class="p">)</span>
    <span class="n">utils</span><span class="o">.</span><span class="n">run_cmd</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>

    <span class="c1"># Remove all temp files</span>
    <span class="k">if</span> <span class="n">remove_temp</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">utils</span><span class="o">.</span><span class="n">remove_files</span><span class="p">([</span><span class="n">chromsizes_file</span><span class="p">,</span> <span class="n">bedgraph_out</span><span class="p">,</span> <span class="n">bedgraph_out_sorted</span><span class="p">])</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Finished converting bam to bigwig! Output bigwig is found in: </span><span class="si">{</span><span class="n">output</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">output</span></div>



<span class="c1"># ---------------------------------------------------------------------------------- #</span>
<span class="c1"># ------------------------------ Bam to fragments file ----------------------------- #</span>
<span class="c1"># ---------------------------------------------------------------------------------- #</span>

<div class="viewcode-block" id="create_fragment_file">
<a class="viewcode-back" href="../../../API/tools.html#sctoolbox.tools.bam.create_fragment_file">[docs]</a>
<span class="nd">@beartype</span>
<span class="k">def</span> <span class="nf">create_fragment_file</span><span class="p">(</span><span class="n">bam</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                         <span class="n">barcode_tag</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;CB&#39;</span><span class="p">,</span>
                         <span class="n">barcode_regex</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                         <span class="n">outdir</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                         <span class="n">nproc</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
                         <span class="n">index</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
                         <span class="n">min_dist</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
                         <span class="n">max_dist</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5000</span><span class="p">,</span>
                         <span class="n">include_clipped</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
                         <span class="n">shift_plus</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
                         <span class="n">shift_minus</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="o">-</span><span class="mi">4</span><span class="p">,</span>
                         <span class="n">keep_temp</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create fragments file out of a BAM file.</span>

<span class="sd">    This is an alternative to using the sinto package,</span>
<span class="sd">    which is slow and has issues with large bam-files.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    bam : str</span>
<span class="sd">        Path to .bam file.</span>
<span class="sd">    barcode_tag : Optional[str], default &#39;CB&#39;</span>
<span class="sd">        The tag where cell barcodes are saved in the bam file. Set to None if the barcodes are in read names.</span>
<span class="sd">    barcode_regex : Optional[str], default None</span>
<span class="sd">        Regex to extract barcodes from read names. Set to None if barcodes are stored in a tag.</span>
<span class="sd">    outdir : Optional[str], default None</span>
<span class="sd">        Path to save fragments file. If None, the file will be saved in the same folder as the .bam file. Temporary intermediate files are also written to this directory.</span>
<span class="sd">    nproc : int, default 1</span>
<span class="sd">        Number of threads for parallelization.</span>
<span class="sd">    index : bool, default False</span>
<span class="sd">        If True, index fragments file. Requires bgzip and tabix.</span>
<span class="sd">    min_dist : int, default 10</span>
<span class="sd">        Minimum fragment length to consider.</span>
<span class="sd">    max_dist : int, default 5000</span>
<span class="sd">        Maximum fragment length to consider.</span>
<span class="sd">    include_clipped : bool, default True</span>
<span class="sd">        Whether to include soft-clipped bases in the fragment length. If True, the full length between reads is used. If False, the fragment length will be calculated from the aligned parts of the reads.</span>
<span class="sd">    shift_plus : int, default 5</span>
<span class="sd">        Shift the start position of the forward read by this value (standard for ATAC).</span>
<span class="sd">    shift_minus : int, default -4</span>
<span class="sd">        Shift the end position of the reverse read by this value (standard for ATAC).</span>
<span class="sd">    keep_temp : bool, default False</span>
<span class="sd">        If True, keep temporary files.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>
<span class="sd">        Path to fragments file.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    ValueError</span>
<span class="sd">        If both barcode_tag and barcode_regex (or neither) are set.</span>
<span class="sd">    FileNotFoundError</span>
<span class="sd">        If the input bam file does not exist.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">utils</span><span class="o">.</span><span class="n">check_module</span><span class="p">(</span><span class="s2">&quot;pysam&quot;</span><span class="p">)</span>
    <span class="kn">import</span> <span class="nn">pysam</span>

    <span class="c1"># Establish output filename</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">outdir</span><span class="p">:</span>
        <span class="n">outdir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">bam</span><span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>  <span class="c1"># create output directory if it does not exist</span>
    <span class="n">out_prefix</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">bam</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">outfile</span> <span class="o">=</span> <span class="n">out_prefix</span> <span class="o">+</span> <span class="s1">&#39;_fragments.tsv&#39;</span>  <span class="c1"># final output file</span>

    <span class="c1"># Check input tag/regex</span>
    <span class="k">if</span> <span class="n">barcode_tag</span> <span class="ow">and</span> <span class="n">barcode_regex</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Only one of &#39;barcode_tag&#39; and &#39;barcode_regex&#39; can be provided! Please set one of the parameters to None.&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="ow">not</span> <span class="n">barcode_tag</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">barcode_regex</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Either &#39;barcode_tag&#39; or &#39;barcode_regex&#39; must be provided!&quot;</span><span class="p">)</span>

    <span class="c1"># Check if bam file exists</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">bam</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">bam</span><span class="si">}</span><span class="s2"> does not exist!&quot;</span><span class="p">)</span>

    <span class="c1"># Ensure that bam-file is indexed</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">bam</span> <span class="o">+</span> <span class="s2">&quot;.bai&quot;</span><span class="p">):</span>

        <span class="n">index_success</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">sort_success</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">while</span> <span class="ow">not</span> <span class="n">index_success</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">sort_success</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;.bam-file has no index - trying to index .bam-file with pysam...&quot;</span><span class="p">)</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="n">pysam</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">bam</span><span class="p">)</span>
                <span class="n">index_success</span> <span class="o">=</span> <span class="kc">True</span>

            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>  <span class="c1"># try to sort the bam file before indexing again</span>

                <span class="k">if</span> <span class="n">sort_success</span><span class="p">:</span>  <span class="c1"># if we already sorted the file</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Unknown error when indexing bam file after sort - please see the error message.&quot;</span><span class="p">)</span>
                    <span class="k">raise</span> <span class="n">e</span>

                <span class="k">try</span><span class="p">:</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;.bam-file could not be indexed with pysam - bam might be unsorted. Trying to sort .bam-file using pysam...&quot;</span><span class="p">)</span>
                    <span class="n">bam_sorted</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">bam</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;_sorted.bam&quot;</span><span class="p">)</span>
                    <span class="n">pysam</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="s2">&quot;-o&quot;</span><span class="p">,</span> <span class="n">bam_sorted</span><span class="p">,</span> <span class="n">bam</span><span class="p">)</span>
                    <span class="n">bam</span> <span class="o">=</span> <span class="n">bam_sorted</span>   <span class="c1"># try to index again with this bam</span>
                    <span class="n">sort_success</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Unknown error when sorting .bam-file - please see the error message.&quot;</span><span class="p">)</span>
                    <span class="k">raise</span> <span class="n">e</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;.bam-file was successfully indexed.&quot;</span><span class="p">)</span>

    <span class="c1"># Chunk chromosomes</span>
    <span class="n">bamfile</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">bam</span><span class="p">)</span>
    <span class="n">chromosomes</span> <span class="o">=</span> <span class="n">bamfile</span><span class="o">.</span><span class="n">references</span>
    <span class="n">bamfile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">chromosome_chunks</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">split_list</span><span class="p">(</span><span class="n">chromosomes</span><span class="p">,</span> <span class="n">nproc</span><span class="p">)</span>

    <span class="c1"># Create fragments from bam</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Creating fragments from chromosomes...&quot;</span><span class="p">)</span>
    <span class="n">temp_files</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">nproc</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">chunk</span> <span class="o">=</span> <span class="n">chromosome_chunks</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>   <span class="c1"># only one chunk</span>
        <span class="n">temp_file</span> <span class="o">=</span> <span class="n">out_prefix</span> <span class="o">+</span> <span class="sa">f</span><span class="s2">&quot;_fragments_</span><span class="si">{</span><span class="n">chunk</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">chunk</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2">.tmp&quot;</span>
        <span class="n">temp_files</span> <span class="o">=</span> <span class="p">[</span><span class="n">temp_file</span><span class="p">]</span>
        <span class="n">n_written</span> <span class="o">=</span> <span class="n">_write_fragments</span><span class="p">(</span><span class="n">bam</span><span class="p">,</span> <span class="n">chromosomes</span><span class="p">,</span> <span class="n">temp_file</span><span class="p">,</span> <span class="n">barcode_tag</span><span class="o">=</span><span class="n">barcode_tag</span><span class="p">,</span> <span class="n">barcode_regex</span><span class="o">=</span><span class="n">barcode_regex</span><span class="p">,</span>
                                     <span class="n">min_dist</span><span class="o">=</span><span class="n">min_dist</span><span class="p">,</span> <span class="n">max_dist</span><span class="o">=</span><span class="n">max_dist</span><span class="p">,</span> <span class="n">include_clipped</span><span class="o">=</span><span class="n">include_clipped</span><span class="p">,</span> <span class="n">shift_plus</span><span class="o">=</span><span class="n">shift_plus</span><span class="p">,</span> <span class="n">shift_minus</span><span class="o">=</span><span class="n">shift_minus</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>

        <span class="n">pool</span> <span class="o">=</span> <span class="n">mp</span><span class="o">.</span><span class="n">Pool</span><span class="p">(</span><span class="n">nproc</span><span class="p">)</span>
        <span class="n">jobs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">chromosome_chunks</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Starting job for chromosomes: &quot;</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">chunk</span><span class="p">))</span>
            <span class="n">temp_file</span> <span class="o">=</span> <span class="n">out_prefix</span> <span class="o">+</span> <span class="sa">f</span><span class="s2">&quot;_fragments_</span><span class="si">{</span><span class="n">chunk</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">chunk</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2">.tmp&quot;</span>
            <span class="n">temp_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">temp_file</span><span class="p">)</span>
            <span class="n">job</span> <span class="o">=</span> <span class="n">pool</span><span class="o">.</span><span class="n">apply_async</span><span class="p">(</span><span class="n">_write_fragments</span><span class="p">,</span> <span class="n">args</span><span class="o">=</span><span class="p">(</span><span class="n">bam</span><span class="p">,</span> <span class="n">chunk</span><span class="p">,</span> <span class="n">temp_file</span><span class="p">,</span> <span class="n">barcode_tag</span><span class="p">,</span> <span class="n">barcode_regex</span><span class="p">,</span> <span class="n">min_dist</span><span class="p">,</span> <span class="n">max_dist</span><span class="p">,</span> <span class="n">include_clipped</span><span class="p">,</span> <span class="n">shift_plus</span><span class="p">,</span> <span class="n">shift_minus</span><span class="p">))</span>
            <span class="n">jobs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">job</span><span class="p">)</span>

        <span class="n">pool</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="c1"># monitor progress</span>
        <span class="n">utils</span><span class="o">.</span><span class="n">monitor_jobs</span><span class="p">(</span><span class="n">jobs</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Progress&quot;</span><span class="p">)</span>  <span class="c1"># waits for all jobs to finish</span>
        <span class="n">pool</span><span class="o">.</span><span class="n">join</span><span class="p">()</span>

        <span class="c1"># get number of written fragments</span>
        <span class="n">n_written</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="n">job</span><span class="o">.</span><span class="n">get</span><span class="p">()</span> <span class="k">for</span> <span class="n">job</span> <span class="ow">in</span> <span class="n">jobs</span><span class="p">])</span>

    <span class="c1"># check if any fragments were written</span>
    <span class="k">if</span> <span class="n">n_written</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;No fragments were written to file - please check your input files and parameters.&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found a total of </span><span class="si">{</span><span class="n">n_written</span><span class="si">}</span><span class="s2"> valid fragments in the .bam file.&quot;</span><span class="p">)</span>

    <span class="c1"># Merge temp files</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Merging identical fragments...&quot;</span><span class="p">)</span>
    <span class="n">tempdir</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">outdir</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">tempdir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>  <span class="c1"># create temp directory if it does not exist</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;cat </span><span class="si">{</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">temp_files</span><span class="p">)</span><span class="si">}</span><span class="s2"> | sort -T </span><span class="si">{</span><span class="n">tempdir</span><span class="si">}</span><span class="s2"> -k1,1 -k2,2n -k3,3n -k4 &quot;</span>
    <span class="n">cmd</span> <span class="o">+=</span> <span class="s2">&quot;| uniq -c | awk -v OFS=&#39;</span><span class="se">\\</span><span class="s2">t&#39; &#39;{$(NF+1)=$1;$1=</span><span class="se">\&quot;\&quot;</span><span class="s2">}1&#39; &quot;</span>  <span class="c1"># count number of fragments at position, move count to last column</span>
    <span class="n">cmd</span> <span class="o">+=</span> <span class="s2">&quot;| awk &#39;{gsub(/^[ </span><span class="se">\\</span><span class="s2">t]+/, </span><span class="se">\&quot;\&quot;</span><span class="s2">);}1&#39; &quot;</span>  <span class="c1"># remove leading whitespace</span>
    <span class="n">cmd</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;&gt; </span><span class="si">{</span><span class="n">outfile</span><span class="si">}</span><span class="s2">&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Running command: </span><span class="se">\&quot;</span><span class="si">{</span><span class="n">cmd</span><span class="si">}</span><span class="se">\&quot;</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Successfully created fragments file: </span><span class="si">{</span><span class="n">outfile</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Remove temp files</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">keep_temp</span><span class="p">:</span>
        <span class="n">utils</span><span class="o">.</span><span class="n">rm_tmp</span><span class="p">(</span><span class="n">temp_files</span><span class="o">=</span><span class="n">temp_files</span><span class="p">)</span>

    <span class="c1"># Index fragments file</span>
    <span class="k">if</span> <span class="n">index</span><span class="p">:</span>
        <span class="n">cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;bgzip --stdout </span><span class="si">{</span><span class="n">outfile</span><span class="si">}</span><span class="s2"> &gt; </span><span class="si">{</span><span class="n">outfile</span><span class="si">}</span><span class="s2">.gz&quot;</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Running command: </span><span class="se">\&quot;</span><span class="si">{</span><span class="n">cmd</span><span class="si">}</span><span class="se">\&quot;</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>

        <span class="n">cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;tabix -p bed </span><span class="si">{</span><span class="n">outfile</span><span class="si">}</span><span class="s2">.gz&quot;</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Running command: </span><span class="se">\&quot;</span><span class="si">{</span><span class="n">cmd</span><span class="si">}</span><span class="se">\&quot;</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>

    <span class="c1"># return path to sorted fragments file</span>
    <span class="k">return</span> <span class="n">outfile</span></div>



<span class="nd">@beartype</span>
<span class="k">def</span> <span class="nf">_get_barcode_from_readname</span><span class="p">(</span><span class="n">read</span><span class="p">:</span> <span class="s2">&quot;pysam.AlignedSegment&quot;</span><span class="p">,</span> <span class="n">regex</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extract barcode from read name.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    read : pysam.AlignedSegment</span>
<span class="sd">        Read from BAM file.</span>
<span class="sd">    regex : str</span>
<span class="sd">        Regex to extract barcode from read name.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>
<span class="sd">        Barcode.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">regex</span><span class="p">,</span> <span class="n">read</span><span class="o">.</span><span class="n">query_name</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>


<span class="nd">@beartype</span>
<span class="k">def</span> <span class="nf">_get_barcode_from_tag</span><span class="p">(</span><span class="n">read</span><span class="p">:</span> <span class="s2">&quot;pysam.AlignedSegment&quot;</span><span class="p">,</span> <span class="n">tag</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extract barcode from read tag.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    read : pysam.AlignedSegment</span>
<span class="sd">        Read from BAM file.</span>
<span class="sd">    tag : str</span>
<span class="sd">        Tag where barcode is stored.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>
<span class="sd">        Barcode.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">has_tag</span><span class="p">(</span><span class="n">tag</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">read</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="n">tag</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>


<span class="nd">@beartype</span>
<span class="k">def</span> <span class="nf">_write_fragments</span><span class="p">(</span><span class="n">bam</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                     <span class="n">chromosomes</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
                     <span class="n">outfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                     <span class="n">barcode_tag</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;CB&quot;</span><span class="p">,</span>
                     <span class="n">barcode_regex</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                     <span class="n">min_dist</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
                     <span class="n">max_dist</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5000</span><span class="p">,</span>
                     <span class="n">include_clipped</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
                     <span class="n">shift_plus</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
                     <span class="n">shift_minus</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="o">-</span><span class="mi">4</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Write fragments from a bam-file within a list of chromosomes to a text file.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    bam : str</span>
<span class="sd">        Path to .bam file.</span>
<span class="sd">    chromosomes : Sequence[str]</span>
<span class="sd">        List of chromosomes to fetch from bam file.</span>
<span class="sd">    outfile : str</span>
<span class="sd">        Path to output file.</span>
<span class="sd">    barcode_tag : Optional[str], default &#39;CB&#39;</span>
<span class="sd">        The tag where cell barcodes are saved in the bam file. Set to None if the barcodes are in read names.</span>
<span class="sd">    barcode_regex : Optional[str], default None</span>
<span class="sd">        Regex to extract barcodes from read names. Set to None if barcodes are stored in a tag.</span>
<span class="sd">    min_dist : int, default 10</span>
<span class="sd">        Minimum fragment length to consider.</span>
<span class="sd">    max_dist : int, default 5000</span>
<span class="sd">        Maximum fragment length to consider.</span>
<span class="sd">    include_clipped : boolean, default True</span>
<span class="sd">        Whether to include soft-clipped bases in the fragment length. If True, the full length between reads is used. If False, the fragment length will be calculated from the aligned parts of the reads.</span>
<span class="sd">    shift_plus : int, default 5</span>
<span class="sd">        Shift the start position of the forward read by this value (standard for ATAC).</span>
<span class="sd">    shift_minus : int, default -4</span>
<span class="sd">        Shift the end position of the reverse read by this value (standard for ATAC).</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    int</span>
<span class="sd">        Number of fragments written to file.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    ValueError</span>
<span class="sd">        If None of barcode_tag and barcode_regex are set.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">utils</span><span class="o">.</span><span class="n">check_module</span><span class="p">(</span><span class="s2">&quot;pysam&quot;</span><span class="p">)</span>
    <span class="kn">import</span> <span class="nn">pysam</span>

    <span class="c1"># Establish function to use for getting barcode</span>
    <span class="k">if</span> <span class="n">barcode_tag</span><span class="p">:</span>
        <span class="n">get_barcode</span> <span class="o">=</span> <span class="n">partial</span><span class="p">(</span><span class="n">_get_barcode_from_tag</span><span class="p">,</span> <span class="n">tag</span><span class="o">=</span><span class="n">barcode_tag</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">barcode_regex</span><span class="p">:</span>
        <span class="n">get_barcode</span> <span class="o">=</span> <span class="n">partial</span><span class="p">(</span><span class="n">_get_barcode_from_readname</span><span class="p">,</span> <span class="n">regex</span><span class="o">=</span><span class="n">barcode_regex</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Either barcode_tag or barcode_regex must be provided!&quot;</span><span class="p">)</span>

    <span class="n">fragments</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">n_written</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_txt</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">bam</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">bamfile</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">chromosome</span> <span class="ow">in</span> <span class="n">chromosomes</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">bamfile</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="n">chromosome</span><span class="p">):</span>  <span class="c1"># fetch all reads on this chromosome</span>

                    <span class="n">fragment_name</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">query_name</span>
                    <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">is_mapped</span> <span class="ow">and</span> <span class="n">read</span><span class="o">.</span><span class="n">is_paired</span><span class="p">:</span>  <span class="c1"># reads must be mapped and paired</span>
                        <span class="k">if</span> <span class="n">fragment_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">fragments</span><span class="p">:</span>
                            <span class="n">fragments</span><span class="p">[</span><span class="n">fragment_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">read</span>   <span class="c1"># first mate found; store it in the dictionary</span>

                        <span class="k">else</span><span class="p">:</span>  <span class="c1"># The mate was already found, so we can write the fragment</span>

                            <span class="c1"># Estimate fragment length</span>
                            <span class="n">reads</span> <span class="o">=</span> <span class="p">[</span><span class="n">fragments</span><span class="p">[</span><span class="n">fragment_name</span><span class="p">],</span> <span class="n">read</span><span class="p">]</span> <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">is_reverse</span> <span class="k">else</span> <span class="p">[</span><span class="n">read</span><span class="p">,</span> <span class="n">fragments</span><span class="p">[</span><span class="n">fragment_name</span><span class="p">]]</span>  <span class="c1"># put forward read first</span>
                            <span class="n">n_reverse</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="n">read</span><span class="o">.</span><span class="n">is_reverse</span> <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">reads</span><span class="p">])</span>  <span class="c1"># one of the reads must be reverse for it to be a proper forward-reverse pair</span>
                            <span class="n">fragment_start</span> <span class="o">=</span> <span class="n">reads</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">reference_start</span> <span class="o">+</span> <span class="n">shift_plus</span>  <span class="c1"># forward read; count whole read without soft-clipping</span>
                            <span class="n">fragment_end</span> <span class="o">=</span> <span class="n">reads</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">reference_end</span> <span class="o">+</span> <span class="n">shift_minus</span>  <span class="c1"># reverse read; count whole read without soft-clipping; shift_minus is negative</span>

                            <span class="k">if</span> <span class="n">include_clipped</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
                                <span class="n">fragment_start</span> <span class="o">-=</span> <span class="n">reads</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">query_alignment_start</span>
                                <span class="n">fragment_end</span> <span class="o">+=</span> <span class="n">reads</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">query_length</span> <span class="o">-</span> <span class="n">reads</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">query_alignment_end</span>  <span class="c1"># the difference between length and end position gives the number of clipped bases for reverse read</span>
                            <span class="n">fragment_length</span> <span class="o">=</span> <span class="n">fragment_end</span> <span class="o">-</span> <span class="n">fragment_start</span>

                            <span class="c1"># If the fragment is valid; we can save it</span>
                            <span class="k">if</span> <span class="n">n_reverse</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">fragment_length</span> <span class="o">&gt;=</span> <span class="n">min_dist</span> <span class="ow">and</span> <span class="n">fragment_length</span> <span class="o">&lt;=</span> <span class="n">max_dist</span><span class="p">:</span>

                                <span class="c1"># Get barcode</span>
                                <span class="n">barcode</span> <span class="o">=</span> <span class="n">get_barcode</span><span class="p">(</span><span class="n">read</span><span class="p">)</span>

                                <span class="k">if</span> <span class="n">barcode</span><span class="p">:</span>
                                    <span class="c1"># Write fragment to file</span>
                                    <span class="n">out_txt</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">read</span><span class="o">.</span><span class="n">reference_name</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">fragment_start</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">fragment_end</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">barcode</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
                                    <span class="n">n_written</span> <span class="o">+=</span> <span class="mi">1</span>

                            <span class="n">fragments</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">fragment_name</span><span class="p">)</span>  <span class="c1"># Remove the written fragment from the dictionary</span>

                <span class="c1"># Remove all remaining fragments from the dictionary; new chromosome or end of file</span>
                <span class="n">fragments</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">return</span> <span class="n">n_written</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Loosolab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>