<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Tools &mdash; SC FRAMEWORK  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../_static/plot_directive.css" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../_static/doctools.js?v=9a2dae69"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Plotting" href="plotting.html" />
    <link rel="prev" title="sctoolbox API reference" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            SC FRAMEWORK
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../CHANGES.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../rna-notebooks/index.html">RNA notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../atac-notebooks/index.html">ATAC notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general-notebooks/index.html">General notebooks</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">sctoolbox API reference</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">Tools</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#preprocessing">Preprocessing</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#module-sctoolbox.tools.qc_filter">qc_filter</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-sctoolbox.tools.highly_variable">highly_variable</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-sctoolbox.tools.dim_reduction">dim_reduction</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-sctoolbox.tools.norm_correct">norm_correct</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#analysis">Analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#module-sctoolbox.tools.embedding">embedding</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-sctoolbox.tools.clustering">clustering</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-sctoolbox.tools.marker_genes">marker_genes</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-sctoolbox.tools.celltype_annotation">celltype_annotation</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#additional">Additional</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#module-sctoolbox.tools.calc_overlap_fc">calc_overlap_fc</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-sctoolbox.tools.multiomics">multiomics</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-sctoolbox.tools.bam">bam</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-sctoolbox.tools.frip">frip</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-sctoolbox.tools.tsse">tsse</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-sctoolbox.tools.insertsize">insertsize</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-sctoolbox.tools.peak_annotation">peak_annotation</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-sctoolbox.tools.receptor_ligand">receptor_ligand</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-sctoolbox.tools.gene_correlation">gene_correlation</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-sctoolbox.tools.tobias">tobias</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="plotting.html">Plotting</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html">Utils</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">SC FRAMEWORK</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">sctoolbox API reference</a></li>
      <li class="breadcrumb-item active">Tools</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/API/tools.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="tools">
<h1>Tools<a class="headerlink" href="#tools" title="Link to this heading"></a></h1>
<section id="preprocessing">
<h2>Preprocessing<a class="headerlink" href="#preprocessing" title="Link to this heading"></a></h2>
<section id="module-sctoolbox.tools.qc_filter">
<span id="qc-filter"></span><h3>qc_filter<a class="headerlink" href="#module-sctoolbox.tools.qc_filter" title="Link to this heading"></a></h3>
<p>Tools for quality control.</p>
<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.qc_filter.calculate_qc_metrics">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.qc_filter.</span></span><span class="sig-name descname"><span class="pre">calculate_qc_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">percent_top</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/qc_filter.html#calculate_qc_metrics"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.qc_filter.calculate_qc_metrics" title="Link to this definition"></a></dt>
<dd><p>Calculate the qc metrics using <cite>scanpy.pp.calculate_qc_metrics</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Anndata object the quality metrics are added to.</p></li>
<li><p><strong>percent_top</strong> (<em>Optional</em><em>[</em><em>list</em><em>[</em><em>int</em><em>]</em><em>]</em><em>, </em><em>default None</em>) – Which proportions of top genes to cover.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default False</em>) – If the anndata object should be modified in place.</p></li>
<li><p><strong>**kwargs</strong> (<em>Any</em>) – Additional parameters forwarded to scanpy.pp.calculate_qc_metrics.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Returns anndata object with added quality metrics to .obs and .var. Returns None if <cite>inplace=True</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Optional[sc.AnnData]</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">scanpy.pp.calculate_qc_metrics</span></code></p>
<dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">https</span></code></dt><dd><p>//scanpy.readthedocs.io/en/stable/generated/scanpy.pp.calculate_qc_metrics.html</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">sctoolbox</span> <span class="k">as</span> <span class="nn">sct</span>

<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">pbmc3k</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Columns in .obs before &#39;calculate_qc_metrics&#39;:&quot;</span><span class="p">,</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
<span class="n">sct</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">calculate_qc_metrics</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Columns in .obs after &#39;calculate_qc_metrics&#39;:&quot;</span><span class="p">,</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Columns in .obs before &#39;calculate_qc_metrics&#39;: []
Columns in .obs after &#39;calculate_qc_metrics&#39;: [&#39;n_genes&#39;, &#39;log1p_n_genes&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;]
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.qc_filter.predict_cell_cycle">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.qc_filter.</span></span><span class="sig-name descname"><span class="pre">predict_cell_cycle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">g2m_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/qc_filter.html#predict_cell_cycle"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.qc_filter.predict_cell_cycle" title="Link to this definition"></a></dt>
<dd><p>Assign a score and a phase to each cell depending on the expression of cell cycle genes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Anndata object to predict cell cycle on.</p></li>
<li><p><strong>species</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – The species of data. Available species are: human, mouse, rat and zebrafish.
If both s_genes and g2m_genes are given, set species=None,
otherwise species is ignored.</p></li>
<li><p><strong>s_genes</strong> (<em>Optional</em><em>[</em><em>str</em><em> | </em><em>list</em><em>[</em><em>str</em><em>]</em><em>]</em><em>, </em><em>default None</em>) – If no species is given or desired species is not supported, you can provide
a list of genes for the S-phase or a txt file containing one gene in each row.
If only s_genes is provided and species is a supported input, the default
g2m_genes list will be used, otherwise the function will not run.</p></li>
<li><p><strong>g2m_genes</strong> (<em>Optional</em><em>[</em><em>str</em><em> | </em><em>list</em><em>[</em><em>str</em><em>]</em><em>]</em><em>, </em><em>default None</em>) – If no species is given or desired species is not supported, you can provide
a list of genes for the G2M-phase or a txt file containing one gene per row.
If only g2m_genes is provided and species is a supported input, the default
s_genes list will be used, otherwise the function will not run.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default True</em>) – if True, add new columns to the original anndata object.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>If inplace is False, return a copy of anndata object with the new column in the obs table.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Optional[sc.AnnData]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError:</strong> – 1: If s_genes or g2m_genes is not None and not of type list.
    2: If no cellcycle genes available for the given species.
    3. If given species is not supported and s_genes or g2m_genes are not given.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.qc_filter.estimate_doublets">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.qc_filter.</span></span><span class="sig-name descname"><span class="pre">estimate_doublets</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fill_na</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/qc_filter.html#estimate_doublets"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.qc_filter.estimate_doublets" title="Link to this definition"></a></dt>
<dd><p>Estimate doublet cells using scrublet.</p>
<p>Adds additional columns “doublet_score” and “predicted_doublet” in adata.obs,
as well as a “scrublet” key in adata.uns.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Anndata object to estimate doublets for.</p></li>
<li><p><strong>threshold</strong> (<em>float</em><em>, </em><em>default 0.25</em>) – Threshold for doublet detection.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default True</em>) – Whether to estimate doublets inplace or not.</p></li>
<li><p><strong>plot</strong> (<em>bool</em><em>, </em><em>default True</em>) – Whether to plot the doublet score distribution.</p></li>
<li><p><strong>groupby</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Key in adata.obs to use for batching during doublet estimation. If threads &gt; 1,
the adata is split into separate runs across threads. Otherwise each batch is run separately.</p></li>
<li><p><strong>threads</strong> (<em>int</em><em>, </em><em>default 4</em>) – Number of threads to use.</p></li>
<li><p><strong>fill_na</strong> (<em>bool</em><em>, </em><em>default True</em>) – If True, replaces NA values returned by scrublet with 0 and False. Scrublet returns NA if it cannot calculate
a doublet score. Keep in mind that this does not mean that it is no doublet.
By setting this parameter true it is assmuned that it is no doublet.</p></li>
<li><p><strong>**kwargs</strong> (<em>Any</em>) – Additional arguments are passed to scanpy.external.pp.scrublet.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Groupby should be set if the adata consists of multiple samples, as this improves the doublet estimation.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>If inplace is False, the function returns a copy of the adata object.
If inplace is True, the function returns None.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>Optional[sc.AnnData]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.qc_filter.predict_sex">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.qc_filter.</span></span><span class="sig-name descname"><span class="pre">predict_sex</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Xist'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_column</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/qc_filter.html#predict_sex"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.qc_filter.predict_sex" title="Link to this definition"></a></dt>
<dd><p>Predict sex based on expression of Xist (or another gene).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – An anndata object to predict sex for.</p></li>
<li><p><strong>groupby</strong> (<em>str</em>) – Column in adata.obs to group by.</p></li>
<li><p><strong>gene</strong> (<em>str</em><em>, </em><em>default &quot;Xist&quot;</em>) – Name of a female-specific gene to use for estimating Male/Female split.</p></li>
<li><p><strong>gene_column</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Name of the column in adata.var that contains the gene names. If not provided, adata.var.index is used.</p></li>
<li><p><strong>threshold</strong> (<em>float</em><em>, </em><em>default 0.3</em>) – Threshold for the minimum fraction of cells expressing the gene for the group to be considered “Female”.</p></li>
<li><p><strong>plot</strong> (<em>bool</em><em>, </em><em>default True</em>) – Whether to plot the distribution of gene expression per group.</p></li>
<li><p><strong>save</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – If provided, the plot will be saved to this path.</p></li>
<li><p><strong>**kwargs</strong> (<em>Any</em>) – Additional arguments are passed to scanpy.pl.violin.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>adata.X will be converted to numpy.ndarray if it is of type numpy.matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.qc_filter.automatic_thresholds">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.qc_filter.</span></span><span class="sig-name descname"><span class="pre">automatic_thresholds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">which</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'obs'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'var'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'obs'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">columns</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/qc_filter.html#automatic_thresholds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.qc_filter.automatic_thresholds" title="Link to this definition"></a></dt>
<dd><p>Get automatic thresholds for multiple data columns in adata.obs or adata.var.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Anndata object to find thresholds for.</p></li>
<li><p><strong>which</strong> (<em>Literal</em><em>[</em><em>&quot;obs&quot;</em><em>, </em><em>&quot;var&quot;</em><em>]</em><em>, </em><em>default &quot;obs&quot;</em>) – Which data to find thresholds for. Either “obs” or “var”.</p></li>
<li><p><strong>groupby</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Group rows by the column given in ‘groupby’ to find thresholds independently per group</p></li>
<li><p><strong>columns</strong> (<em>Optional</em><em>[</em><em>list</em><em>[</em><em>str</em><em>]</em><em>]</em><em>, </em><em>default None</em>) – Columns to calculate automatic thresholds for. If None, will take all numeric columns.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dict containing thresholds for each data column,
either grouped by groupby or directly containing “min” and “max” per column.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, dict[str, Union[float, dict[str, float]]]]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError:</strong> – If which is not set to ‘obs’ or ‘var’</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.qc_filter.thresholds_as_table">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.qc_filter.</span></span><span class="sig-name descname"><span class="pre">thresholds_as_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">threshold_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/qc_filter.html#thresholds_as_table"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.qc_filter.thresholds_as_table" title="Link to this definition"></a></dt>
<dd><p>Show the threshold dictionary as a table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>threshold_dict</strong> (<em>dict</em><em>[</em><em>str</em><em>, </em><em>dict</em><em>[</em><em>str</em><em>, </em><em>float</em><em> | </em><em>int</em><em> | </em><em>dict</em><em>[</em><em>str</em><em>, </em><em>int</em><em> | </em><em>float</em><em>]</em><em>]</em><em>]</em>) – Dictionary with thresholds.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.qc_filter.validate_threshold_dict">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.qc_filter.</span></span><span class="sig-name descname"><span class="pre">validate_threshold_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">thresholds</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/qc_filter.html#validate_threshold_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.qc_filter.validate_threshold_dict" title="Link to this definition"></a></dt>
<dd><p>Validate threshold dictionary.</p>
<p>Thresholds can be in the format:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">thresholds</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;chrM_percent&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;min&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;max&quot;</span><span class="p">:</span> <span class="mi">10</span><span class="p">},</span>
              <span class="s2">&quot;total_reads&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;min&quot;</span><span class="p">:</span> <span class="mi">1000</span><span class="p">}}</span>
</pre></div>
</div>
<p>Or per group in ‘groupby’:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">thresholds</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;chrM_percent&quot;</span><span class="p">:</span> <span class="p">{</span>
                        <span class="s2">&quot;Sample1&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;min&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;max&quot;</span><span class="p">:</span> <span class="mi">10</span><span class="p">},</span>
                        <span class="s2">&quot;Sample2&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;max&quot;</span><span class="p">:</span> <span class="mi">5</span><span class="p">}</span>
                        <span class="p">},</span>
              <span class="s2">&quot;total_reads&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;min&quot;</span><span class="p">:</span> <span class="mi">1000</span><span class="p">}}</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>table</strong> (<em>pd.DataFrame</em>) – Table to validate thresholds for.</p></li>
<li><p><strong>thresholds</strong> (<em>dict</em><em>[</em><em>str</em><em>, </em><em>dict</em><em>[</em><em>str</em><em>, </em><em>int</em><em> | </em><em>float</em><em>] </em><em>| </em><em>dict</em><em>[</em><em>str</em><em>, </em><em>dict</em><em>[</em><em>str</em><em>, </em><em>int</em><em> | </em><em>float</em><em>]</em><em>]</em><em>]</em>) – Dictionary of thresholds to validate.</p></li>
<li><p><strong>groupby</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Column for grouping thresholds.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the threshold dict is not valid.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.qc_filter.get_thresholds_wrapper">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.qc_filter.</span></span><span class="sig-name descname"><span class="pre">get_thresholds_wrapper</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">manual_thresholds</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">only_automatic_thresholds</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/qc_filter.html#get_thresholds_wrapper"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.qc_filter.get_thresholds_wrapper" title="Link to this definition"></a></dt>
<dd><p>Get the thresholds for the filtering.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Anndata object to find QC thresholds for.</p></li>
<li><p><strong>manual_thresholds</strong> (<em>dict</em><em>[</em><em>str</em><em>, </em><em>dict</em><em>[</em><em>str</em><em>, </em><em>Union</em><em>[</em><em>float</em><em>, </em><em>dict</em><em>[</em><em>str</em><em>, </em><em>float</em><em>]</em><em>]</em><em>]</em><em>]</em>) – Dictionary containing manually set thresholds</p></li>
<li><p><strong>only_automatic_thresholds</strong> (<em>bool</em><em>, </em><em>default True</em>) – If True, only set automatic thresholds.</p></li>
<li><p><strong>groupby</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Group cells by column in adata.obs.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary containing the thresholds</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, dict[str, Union[float | int, dict[str, float | int]]]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.qc_filter.get_keys">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.qc_filter.</span></span><span class="sig-name descname"><span class="pre">get_keys</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">manual_thresholds</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/qc_filter.html#get_keys"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.qc_filter.get_keys" title="Link to this definition"></a></dt>
<dd><p>Get threshold dictionary with keys that overlap with adata.obs.columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Anndata object</p></li>
<li><p><strong>manual_thresholds</strong> (<em>dict</em><em>[</em><em>str</em><em>, </em><em>Any</em><em>]</em>) – Dictionary with adata.obs colums as keys.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary with key - adata.obs.column overlap</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, dict[str, Union[float | int, dict[str, float | int]]]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.qc_filter.get_mean_thresholds">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.qc_filter.</span></span><span class="sig-name descname"><span class="pre">get_mean_thresholds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">thresholds</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/qc_filter.html#get_mean_thresholds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.qc_filter.get_mean_thresholds" title="Link to this definition"></a></dt>
<dd><p>Convert grouped thresholds to global thresholds by taking the mean across groups.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.qc_filter.apply_qc_thresholds">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.qc_filter.</span></span><span class="sig-name descname"><span class="pre">apply_qc_thresholds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">thresholds</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">which</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'obs'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'var'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'obs'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/qc_filter.html#apply_qc_thresholds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.qc_filter.apply_qc_thresholds" title="Link to this definition"></a></dt>
<dd><p>Apply QC thresholds to anndata object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Anndata object to filter.</p></li>
<li><p><strong>thresholds</strong> (<em>dict</em><em>[</em><em>str</em><em>, </em><em>Any</em><em>]</em>) – Dictionary of thresholds to apply.</p></li>
<li><p><strong>which</strong> (<em>Literal</em><em>[</em><em>&quot;obs&quot;</em><em>, </em><em>&quot;var&quot;</em><em>]</em><em>, </em><em>default 'obs'</em>) – Which table to filter on. Must be one of “obs” / “var”.</p></li>
<li><p><strong>groupby</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Column in table to group by.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default True</em>) – Change adata inplace or return a changed copy.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Anndata object with QC thresholds applied.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Optional[sc.AnnData]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError:</strong> – 1: If the keys in thresholds do not match with the columns in adata.[which].
    2: If grouped thesholds are not found. For example do not contain min and max values.
    3: If thresholds do not contain min and max values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.qc_filter.filter_cells">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.qc_filter.</span></span><span class="sig-name descname"><span class="pre">filter_cells</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cells</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_bool</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/qc_filter.html#filter_cells"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.qc_filter.filter_cells" title="Link to this definition"></a></dt>
<dd><p>Remove cells from anndata object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Anndata object to filter.</p></li>
<li><p><strong>cells</strong> (<em>str</em><em> | </em><em>list</em><em>[</em><em>str</em><em>]</em>) – A column in .obs containing boolean indicators or a list of cells to remove from object .obs table.</p></li>
<li><p><strong>remove_bool</strong> (<em>bool</em><em>, </em><em>default True</em>) – Is used if genes is a column in .obs table. If True, remove cells that are True. If False, remove cells that are False.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default True</em>) – If True, filter inplace. If False, return filtered adata object.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>If inplace is False, returns the filtered Anndata object. If inplace is True, returns None.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Optional[sc.AnnData]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.qc_filter.filter_genes">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.qc_filter.</span></span><span class="sig-name descname"><span class="pre">filter_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_bool</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/qc_filter.html#filter_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.qc_filter.filter_genes" title="Link to this definition"></a></dt>
<dd><p>Remove genes from adata object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Annotated data matrix object to filter</p></li>
<li><p><strong>genes</strong> (<em>str</em><em> | </em><em>list</em><em>[</em><em>str</em><em>]</em>) – A column containing boolean indicators or a list of genes to remove from object .var table.</p></li>
<li><p><strong>remove_bool</strong> (<em>bool</em><em>, </em><em>default True</em>) – Is used if genes is a column in .var table. If True, remove genes that are True. If False, remove genes that are False.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default True</em>) – If True, filter inplace. If False, return filtered adata object.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>If inplace is False, returns the filtered Anndata object. If inplace is True, returns None.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Optional[sc.AnnData]</p>
</dd>
</dl>
</dd></dl>

</section>
<hr class="docutils" />
<section id="module-sctoolbox.tools.highly_variable">
<span id="highly-variable"></span><h3>highly_variable<a class="headerlink" href="#module-sctoolbox.tools.highly_variable" title="Link to this heading"></a></h3>
<p>Tools to calculate and annotate highly variable genes.</p>
<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.highly_variable.annot_HVG">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.highly_variable.</span></span><span class="sig-name descname"><span class="pre">annot_HVG</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">anndata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_mean</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.0125</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_iterations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hvg_range</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">(1000,</span> <span class="pre">5000)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">step</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/highly_variable.html#annot_HVG"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.highly_variable.annot_HVG" title="Link to this definition"></a></dt>
<dd><p>Annotate highly variable genes (HVG). Tries to annotate in given range of HVGs, by gradually in-/ decreasing min_mean of scanpy.pp.highly_variable_genes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>anndata</strong> (<em>sc.AnnData</em>) – Anndata object to annotate.</p></li>
<li><p><strong>min_mean</strong> (<em>float</em><em>, </em><em>default 0.0125</em>) – Starting min_mean parameter for finding HVGs.</p></li>
<li><p><strong>max_iterations</strong> (<em>int</em><em>, </em><em>default 10</em>) – Maximum number of min_mean adjustments.</p></li>
<li><p><strong>hvg_range</strong> (<em>Tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em><em>, </em><em>default</em><em> (</em><em>1000</em><em>, </em><em>5000</em><em>)</em>) – Number of HVGs should be in the given range. Will issue a warning if result is not in range.
Default limits are chosen as proposed by <a class="reference external" href="https://doi.org/10.15252/msb.20188746">https://doi.org/10.15252/msb.20188746</a>.</p></li>
<li><p><strong>step</strong> (<em>int</em><em> | </em><em>float</em><em>, </em><em>default 10</em>) – Value min_mean is adjusted by in each iteration. Will divide min_value (below range) or multiply (above range) by this value.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether the anndata object is modified inplace.</p></li>
<li><p><strong>save</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Path to save the plot to. If None, the plot is not saved.</p></li>
<li><p><strong>**kwargs</strong> (<em>Any</em>) – Additional arguments forwarded to scanpy.pp.highly_variable_genes().</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Logarithmized data is expected.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Adds annotation of HVG to anndata object. Information is added to Anndata.var[“highly_variable”].
If inplace is False, the function returns None
Else returns a chagned copy of the input anndata object.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>Optional[sc.AnnData]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.highly_variable.get_variable_features">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.highly_variable.</span></span><span class="sig-name descname"><span class="pre">get_variable_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_cells</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_cells</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/highly_variable.html#get_variable_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.highly_variable.get_variable_features" title="Link to this definition"></a></dt>
<dd><p>Get the highly variable features of anndata object. Adds the column “highly_variable” to adata.var. If show is True, the plot is shown.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – The anndata object containing counts for variables.</p></li>
<li><p><strong>max_cells</strong> (<em>Optional</em><em>[</em><em>float</em><em> | </em><em>int</em><em>]</em><em>, </em><em>default None</em>) – The maximum variability score to set as threshold. Defaults to knee estimated threshold.</p></li>
<li><p><strong>min_cells</strong> (<em>Optional</em><em>[</em><em>float</em><em> | </em><em>int</em><em>]</em><em>, </em><em>default 0</em>) – The minimum variability score to set as threshold.</p></li>
<li><p><strong>show</strong> (<em>bool</em><em>, </em><em>default True</em>) – Show plot of variability scores and thresholds.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default True</em>) – If True, the anndata object is modified. Otherwise, a new anndata object is returned.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Designed for scATAC-seq data</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>KeyError</strong> – If adata.var[‘n_cells_by_counts’] is not available.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>If inplace is False, the function returns None
If inplace is True, the function returns an anndata object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Optional[sc.AnnData]</p>
</dd>
</dl>
</dd></dl>

</section>
<hr class="docutils" />
<section id="module-sctoolbox.tools.dim_reduction">
<span id="dim-reduction"></span><h3>dim_reduction<a class="headerlink" href="#module-sctoolbox.tools.dim_reduction" title="Link to this heading"></a></h3>
<p>Tools for dimensionality reduction with PCA/SVD.</p>
<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.dim_reduction.compute_PCA">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.dim_reduction.</span></span><span class="sig-name descname"><span class="pre">compute_PCA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">anndata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_highly_variable</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/dim_reduction.html#compute_PCA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.dim_reduction.compute_PCA" title="Link to this definition"></a></dt>
<dd><p>Compute a principal component analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>anndata</strong> (<em>sc.AnnData</em>) – Anndata object to add the PCA to.</p></li>
<li><p><strong>use_highly_variable</strong> (<em>bool</em><em>, </em><em>default True</em>) – If true, use highly variable genes to compute PCA.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether the anndata object is modified inplace.</p></li>
<li><p><strong>**kwargs</strong> (<em>Any</em>) – Additional parameters forwarded to scanpy.pp.pca().</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Returns anndata object with PCA components. Or None if inplace = True.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Optional[sc.AnnData]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.dim_reduction.lsi">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.dim_reduction.</span></span><span class="sig-name descname"><span class="pre">lsi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale_embeddings</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_comps</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_highly_variable</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/dim_reduction.html#lsi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.dim_reduction.lsi" title="Link to this definition"></a></dt>
<dd><p>Run Latent Semantic Indexing for dimensionality reduction.</p>
<p>Values represent the similarity of cells in the original space.
doi: 10.3389/fphys.2013.00008</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>sc.AnnData</em>) – AnnData object with peak counts.</p></li>
<li><p><strong>scale_embeddings</strong> (<em>bool</em><em>, </em><em>default True</em>) – Scale embeddings to zero mean and unit variance.</p></li>
<li><p><strong>n_comps</strong> (<em>int</em><em>, </em><em>default 50</em>) – Number of components to calculate with SVD.</p></li>
<li><p><strong>use_highly_variable</strong> (<em>bool</em><em>, </em><em>default True</em>) – If true, use highly variable genes to compute LSI.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError:</strong> – If highly variable genes are used and not found in adata.var[‘highly_variable’].</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Function is taken from muon package.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.dim_reduction.apply_svd">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.dim_reduction.</span></span><span class="sig-name descname"><span class="pre">apply_svd</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/dim_reduction.html#apply_svd"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.dim_reduction.apply_svd" title="Link to this definition"></a></dt>
<dd><p>Singular value decomposition of anndata object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – The anndata object to be decomposed.</p></li>
<li><p><strong>layer</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – The layer to be decomposed. If None, the layer is set to “X”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The decomposed anndata object containing .obsm, .varm and .uns information.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>sc.AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.dim_reduction.define_PC">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.dim_reduction.</span></span><span class="sig-name descname"><span class="pre">define_PC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">anndata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/dim_reduction.html#define_PC"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.dim_reduction.define_PC" title="Link to this definition"></a></dt>
<dd><p>Define threshold for most variable PCA components.</p>
<p>Note: Function expects PCA to be computed beforehand.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>anndata</strong> (<em>sc.AnnData</em>) – Anndata object with PCA to get significant PCs threshold from.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An int representing the number of PCs until elbow, defining PCs with significant variance.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError:</strong> – If PCA is not found in anndata.</p>
</dd>
</dl>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 0.5: </span>This will be removed in 0.7. Use the ‘sctoolbox.tools.propose_pcs’ function instead.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.dim_reduction.propose_pcs">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.dim_reduction.</span></span><span class="sig-name descname"><span class="pre">propose_pcs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">anndata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">how</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'variance'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'cumulative</span> <span class="pre">variance'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'correlation'</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['variance',</span> <span class="pre">'correlation']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'knee'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'percent'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'percent'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">perc_thresh</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">corr_thresh</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">corr_kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/dim_reduction.html#propose_pcs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.dim_reduction.propose_pcs" title="Link to this definition"></a></dt>
<dd><p>Propose a selection of PCs that can be used for further analysis.</p>
<p>Note: Function expects PCA to be computed beforehand.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>anndata</strong> (<em>sc.AnnData</em>) – Anndata object with PCA to get PCs from.</p></li>
<li><p><strong>how</strong> (<em>List</em><em>[</em><em>Literal</em><em>[</em><em>&quot;variance&quot;</em><em>, </em><em>&quot;cumulative variance&quot;</em><em>, </em><em>&quot;correlation&quot;</em><em>]</em><em>]</em><em>, </em><em>default</em><em> [</em><em>&quot;variance&quot;</em><em>, </em><em>&quot;correlation&quot;</em><em>]</em>) – Values to use for PC proposal. Will independently apply filters to all selected methods and return the union of PCs.</p></li>
<li><p><strong>var_method</strong> (<em>Literal</em><em>[</em><em>&quot;knee&quot;</em><em>, </em><em>&quot;percent&quot;</em><em>]</em><em>, </em><em>default &quot;percent&quot;</em>) – Either define a threshold based on a knee algorithm or use the percentile.</p></li>
<li><p><strong>perc_thresh</strong> (<em>Union</em><em>[</em><em>int</em><em>, </em><em>float</em><em>]</em><em>, </em><em>default 30</em>) – Percentile threshold of the PCs that should be included. Only for var_method=”percent” and expects a value from 0-100.</p></li>
<li><p><strong>corr_thresh</strong> (<em>float</em><em>, </em><em>default 0.3</em>) – Filter PCs with a correlation greater than the given value.</p></li>
<li><p><strong>corr_kwargs</strong> (<em>Optional</em><em>(</em><em>dict</em><em>)</em><em>, </em><em>default None</em>) – Parameters forwarded to <cite>sctoolbox.tools.correlation_matrix</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of PCs proposed for further use.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>List[int]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError:</strong> – If PCA is not found in anndata.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.dim_reduction.subset_PCA">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.dim_reduction.</span></span><span class="sig-name descname"><span class="pre">subset_PCA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_pcs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">select</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/dim_reduction.html#subset_PCA"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.dim_reduction.subset_PCA" title="Link to this definition"></a></dt>
<dd><p>Subset the PCA coordinates in adata.obsm[“X_pca”] to the given number of pcs.</p>
<p>Additionally, subset the PCs in adata.varm[“PCs”] and the variance ratio in adata.uns[“pca”][“variance_ratio”].</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Anndata object containing the PCA coordinates.</p></li>
<li><p><strong>n_pcs</strong> (<em>Optional</em><em>[</em><em>int</em><em>]</em><em>, </em><em>default None</em>) – Number of PCs to keep.</p></li>
<li><p><strong>start</strong> (<em>int</em><em>, </em><em>default 0</em>) – Index (0-based) of the first PC to keep. E.g. if start = 1 and n_pcs = 10, you will exclude the first PC to keep 9 PCs.</p></li>
<li><p><strong>select</strong> (<em>Optional</em><em>[</em><em>List</em><em>[</em><em>int</em><em>]</em><em>]</em><em>, </em><em>default None</em>) – Provide a list of PC numbers to keep. E.g. [2, 3, 5] will select the second, third and fifth PC. Will overwrite the n_pcs and start parameter.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default True</em>) – Whether to work inplace on the anndata object.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Anndata object with the subsetted PCA coordinates. Or None if inplace = True.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Optional[sc.AnnData]</p>
</dd>
</dl>
</dd></dl>

</section>
<hr class="docutils" />
<section id="module-sctoolbox.tools.norm_correct">
<span id="norm-correct"></span><h3>norm_correct<a class="headerlink" href="#module-sctoolbox.tools.norm_correct" title="Link to this heading"></a></h3>
<p>Normalization and correction tools.</p>
<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.norm_correct.atac_norm">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.norm_correct.</span></span><span class="sig-name descname"><span class="pre">atac_norm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/sctoolbox/tools/norm_correct.html#atac_norm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.norm_correct.atac_norm" title="Link to this definition"></a></dt>
<dd><p>Normalize ATAC data - deprecated functionality. Use normalize_adata instead.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.norm_correct.normalize_adata">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.norm_correct.</span></span><span class="sig-name descname"><span class="pre">normalize_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_highly_expressed</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_highly_variable</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_sum</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AnnData</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/norm_correct.html#normalize_adata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.norm_correct.normalize_adata" title="Link to this definition"></a></dt>
<dd><p>Normalize the count matrix and calculate dimension reduction using different methods.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>method</strong> (<em>str</em><em> | </em><em>list</em><em>[</em><em>str</em><em>]</em>) – Normalization method. Either ‘total’ and/or ‘tfidf’.
- ‘total’: Performs normalization for total counts, log1p and PCA.
- ‘tfidf’: Performs TFIDF normalization and LSI (corresponds to PCA). This method is often used for scATAC-seq data.</p></li>
<li><p><strong>exclude_highly_expressed</strong> (<em>bool</em><em>, </em><em>default True</em>) – Parameter for sc.pp.normalize_total. Decision to exclude highly expressed genes (HEG) from total normalization.</p></li>
<li><p><strong>use_highly_variable</strong> (<em>bool</em><em>, </em><em>default False</em>) – Parameter for sc.pp.pca and lsi. Decision to use highly variable genes for PCA/LSI.</p></li>
<li><p><strong>target_sum</strong> (<em>Optional</em><em>[</em><em>int</em><em>]</em><em>, </em><em>default None</em>) – Parameter for sc.pp.normalize_total. Decide the target sum of each cell after normalization.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary containing method name as key, and anndata as values.
Each anndata is the annotated data matrix with normalized count matrix and PCA/LSI calculated.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, sc.AnnData]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If method is not valid. Needs to be either ‘total’ or ‘tfidf’.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.norm_correct.tfidf">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.norm_correct.</span></span><span class="sig-name descname"><span class="pre">tfidf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_tf</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_idf</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_tfidf</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale_factor</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10000</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/norm_correct.html#tfidf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.norm_correct.tfidf" title="Link to this definition"></a></dt>
<dd><p>Transform peak counts with TF-IDF (Term Frequency - Inverse Document Frequency).</p>
<p>TF: peak counts are normalised by total number of counts per cell.
DF: total number of counts for each peak.
IDF: number of cells divided by DF.
By default, log(TF) * log(IDF) is returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>sc.AnnData</em>) – AnnData object with peak counts.</p></li>
<li><p><strong>log_tf</strong> (<em>bool</em><em>, </em><em>default True</em>) – Log-transform TF term if True.</p></li>
<li><p><strong>log_idf</strong> (<em>bool</em><em>, </em><em>default True</em>) – Log-transform IDF term if True.</p></li>
<li><p><strong>log_tfidf</strong> (<em>bool</em><em>, </em><em>default Frue</em>) – Log-transform TF*IDF term if True. Can only be used when log_tf and log_idf are False.</p></li>
<li><p><strong>scale_factor</strong> (<em>int</em><em>, </em><em>default 1e4</em>) – Scale factor to multiply the TF-IDF matrix by.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Function is from the muon package.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>AttributeError:</strong> – log(TF*IDF) requires log(TF) and log(IDF) to be False.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.norm_correct.tfidf_normalization">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.norm_correct.</span></span><span class="sig-name descname"><span class="pre">tfidf_normalization</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matrix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">spmatrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tf_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'raw'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'term_frequency'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'log'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'term_frequency'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">idf_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'unary'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'inverse_freq'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'inverse_freq_smooth'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'inverse_freq'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">csr_matrix</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/norm_correct.html#tfidf_normalization"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.norm_correct.tfidf_normalization" title="Link to this definition"></a></dt>
<dd><p>Perform TF-IDF normalization on a sparse matrix.</p>
<p>The different variants of the term frequency and inverse document frequency are obtained from <a class="reference external" href="https://en.wikipedia.org/wiki/Tf-idf">https://en.wikipedia.org/wiki/Tf-idf</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>matrix</strong> (<em>scipy.sparse matrix</em>) – The matrix to be normalized.</p></li>
<li><p><strong>tf_type</strong> (<em>Literal</em><em>[</em><em>&quot;term_frequency&quot;</em><em>, </em><em>&quot;log&quot;</em><em>, </em><em>&quot;raw&quot;</em><em>]</em><em>, </em><em>default &quot;term_frequency&quot;</em>) – The type of term frequency to use. Can be either “raw”, “term_frequency” or “log”.</p></li>
<li><p><strong>idf_type</strong> (<em>Literal</em><em>[</em><em>&quot;inverse_freq&quot;</em><em>, </em><em>&quot;unary&quot;</em><em>, </em><em>&quot;inverse_freq_smooth&quot;</em><em>]</em><em>, </em><em>default &quot;inverse_freq&quot;</em>) – The type of inverse document frequency to use. Can be either “unary”, “inverse_freq” or “inverse_freq_smooth”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>tfidf normalized sparse matrix.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>sparse.csr_matrix</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This function requires a lot of memory. Another option is to use the ac.pp.tfidf of the muon package.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.norm_correct.wrap_corrections">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.norm_correct.</span></span><span class="sig-name descname"><span class="pre">wrap_corrections</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">methods</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'bbknn'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'combat'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'mnn'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'harmony'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'scanorama'</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'bbknn'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'combat'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'mnn'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'harmony'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'scanorama'</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Callable</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['bbknn',</span> <span class="pre">'mnn']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method_kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AnnData</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/norm_correct.html#wrap_corrections"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.norm_correct.wrap_corrections" title="Link to this definition"></a></dt>
<dd><p>Calculate multiple batch corrections for adata using the ‘batch_correction’ function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – An annotated data matrix object to apply corrections to.</p></li>
<li><p><strong>batch_key</strong> (<em>str</em>) – The column in adata.obs containing batch information.</p></li>
<li><p><strong>methods</strong> (<em>list</em><em>[</em><em>batch_methods</em><em>] </em><em>| </em><em>Callable</em><em> | </em><em>batch_methods</em><em>, </em><em>default</em><em> [</em><em>&quot;bbknn&quot;</em><em>, </em><em>&quot;mnn&quot;</em><em>]</em>) – The method(s) to use for batch correction. Options are:
- bbknn
- mnn
- harmony
- scanorama
- combat
Or provide a custom batch correction function. See <cite>batch_correction(method)</cite> for more information.</p></li>
<li><p><strong>method_kwargs</strong> (<em>dict</em><em>, </em><em>default {}</em>) – Dict with methods as keys. Values are dicts of additional parameters forwarded to method. See batch_correction(<a href="#id1"><span class="problematic" id="id2">**</span></a>kwargs).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictonary of batch corrected anndata objects. Where the key is the correction method and the value is the corrected anndata.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, sc.AnnData]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If not all methods in methods are valid.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.norm_correct.batch_correction">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.norm_correct.</span></span><span class="sig-name descname"><span class="pre">batch_correction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'bbknn'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'combat'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'mnn'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'harmony'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'scanorama'</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'bbknn'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'combat'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'mnn'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'harmony'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'scanorama'</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Callable</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['bbknn',</span> <span class="pre">'mnn']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">highly_variable</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/norm_correct.html#batch_correction"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.norm_correct.batch_correction" title="Link to this definition"></a></dt>
<dd><p>Perform batch correction on the adata object using the ‘method’ given.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – An annotated data matrix object to apply corrections to.</p></li>
<li><p><strong>batch_key</strong> (<em>str</em>) – The column in adata.obs containing batch information.</p></li>
<li><p><strong>method</strong> (<em>str</em><em> or </em><em>function</em>) – <p>Either one of the predefined methods or a custom function for batch correction.
Note: The custom function is expected to accept an anndata object as the first parameter and return the batch corrected anndata.</p>
<dl class="simple">
<dt>Available methods:</dt><dd><ul>
<li><p>bbknn</p></li>
<li><p>mnn</p></li>
<li><p>harmony</p></li>
<li><p>scanorama</p></li>
<li><p>combat</p></li>
</ul>
</dd>
</dl>
</p></li>
<li><p><strong>highly_variable</strong> (<em>bool</em><em>, </em><em>default True</em>) – Only for method ‘mnn’. If True, only the highly variable genes (column ‘highly_variable’ in .var) will be used for batch correction.</p></li>
<li><p><strong>**kwargs</strong> (<em>Any</em>) – Additional arguments will be forwarded to the method function.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A copy of the anndata with applied batch correction.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>sc.AnnData</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError:</strong> – <ol class="arabic simple">
<li><p>If batch_key column is not in adata.obs
    2. If batch correction method is invalid.</p></li>
</ol>
</p></li>
<li><p><strong>KeyError:</strong> – If PCA has not been calculated before running bbknn.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.norm_correct.evaluate_batch_effect">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.norm_correct.</span></span><span class="sig-name descname"><span class="pre">evaluate_batch_effect</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obsm_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'X_umap'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'LISI_score'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_dims</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">perplexity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/norm_correct.html#evaluate_batch_effect"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.norm_correct.evaluate_batch_effect" title="Link to this definition"></a></dt>
<dd><p>Evaluate batch effect methods using LISI.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Anndata object with PCA and umap/tsne for batch evaluation.</p></li>
<li><p><strong>batch_key</strong> (<em>str</em>) – The column in adata.obs containing batch information.</p></li>
<li><p><strong>obsm_key</strong> (<em>str</em><em>, </em><em>default 'X_umap'</em>) – The column in adata.obsm containing coordinates.</p></li>
<li><p><strong>col_name</strong> (<em>str</em><em>, </em><em>default 'LISI_score'</em>) – Column name for storing the LISI score in .obs.</p></li>
<li><p><strong>max_dims</strong> (<em>int</em><em>, </em><em>default 5</em>) – Maximum number of dimensions of adata.obsm matrix to use for LISI (to speed up computation).</p></li>
<li><p><strong>perplexity</strong> (<em>int</em><em>, </em><em>default 30</em>) – Perplexity for the LISI score calculation.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to work inplace on the anndata object.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>if inplace is True, LISI_score is added to adata.obs inplace (returns None), otherwise a copy of the adata is returned.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Optional[sc.AnnData]</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>LISI score is calculated for each cell and it is between 1-n for a data-frame with n categorical variables.</p></li>
<li><p>indicates the effective number of different categories represented in the local neighborhood of each cell.</p></li>
<li><p>If the cells are well-mixed, then we expect the LISI score to be near n for a data with n batches.</p></li>
<li><p>The higher the LISI score is, the better batch correction method worked to normalize the batch effect and mix the cells from different batches.</p></li>
<li><p>For further information on LISI: <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1850-9">https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1850-9</a></p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>KeyError:</strong> – <ol class="arabic simple">
<li><p>If obsm_key is not in adata.obsm.
    2. If batch_key is no column in adata.obs.</p></li>
</ol>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.norm_correct.wrap_batch_evaluation">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.norm_correct.</span></span><span class="sig-name descname"><span class="pre">wrap_batch_evaluation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adatas</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AnnData</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obsm_keys</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['X_pca',</span> <span class="pre">'X_umap']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_dims</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AnnData</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/norm_correct.html#wrap_batch_evaluation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.norm_correct.wrap_batch_evaluation" title="Link to this definition"></a></dt>
<dd><p>Evaluate batch correction methods for a dict of anndata objects (using LISI score calculation).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adatas</strong> (<em>dict</em><em>[</em><em>str</em><em>, </em><em>sc.AnnData</em><em>]</em>) – Dict containing an anndata object for each batch correction method as values. Keys are the name of the respective method.
E.g.: {“bbknn”: anndata}</p></li>
<li><p><strong>batch_key</strong> (<em>str</em>) – The column in adata.obs containing batch information.</p></li>
<li><p><strong>obsm_keys</strong> (<em>str</em><em> | </em><em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default</em><em> [</em><em>'X_pca'</em><em>, </em><em>'X_umap'</em><em>]</em>) – Key(s) to coordinates on which the score is calculated.</p></li>
<li><p><strong>threads</strong> (<em>int</em><em>, </em><em>default 1</em>) – Number of threads to use for parallelization.</p></li>
<li><p><strong>max_dims</strong> (<em>int</em><em>, </em><em>default 5</em>) – Maximum number of dimensions of adata.obsm matrix to use for LISI (to speed up computation).</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to work inplace on the anndata dict.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dict containing an anndata object for each batch correction method as values of LISI scores added to .obs.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Optional[dict[str, sc.AnnData]]</p>
</dd>
</dl>
</dd></dl>

</section>
</section>
<section id="analysis">
<h2>Analysis<a class="headerlink" href="#analysis" title="Link to this heading"></a></h2>
<section id="module-sctoolbox.tools.embedding">
<span id="embedding"></span><h3>embedding<a class="headerlink" href="#module-sctoolbox.tools.embedding" title="Link to this heading"></a></h3>
<p>Embedding tools.</p>
<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.embedding.wrap_umap">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.embedding.</span></span><span class="sig-name descname"><span class="pre">wrap_umap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adatas</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Iterable</span><span class="p"><span class="pre">[</span></span><span class="pre">AnnData</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/embedding.html#wrap_umap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.embedding.wrap_umap" title="Link to this definition"></a></dt>
<dd><p>Compute umap for a list of adatas in parallel.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adatas</strong> (<em>Iterable</em><em>[</em><em>sc.AnnData</em><em>]</em>) – List of anndata objects to compute umap on.</p></li>
<li><p><strong>threads</strong> (<em>int</em><em>, </em><em>default 4</em>) – Number of threads to use.</p></li>
<li><p><strong>**kwargs</strong> (<em>Any</em>) – Additional arguments to be passed to sc.tl.umap.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.embedding.correlation_matrix">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.embedding.</span></span><span class="sig-name descname"><span class="pre">correlation_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">which</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'obs'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'var'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'obs'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">basis</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'pca'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_components</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">columns</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'spearmanr'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'pearsonr'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'spearmanr'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">DataFrame</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">DataFrame</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/embedding.html#correlation_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.embedding.correlation_matrix" title="Link to this definition"></a></dt>
<dd><p>Compute a matrix of correlation values between an embedding and given columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Annotated data matrix object.</p></li>
<li><p><strong>which</strong> (<em>Literal</em><em>[</em><em>&quot;obs&quot;</em><em>, </em><em>&quot;var&quot;</em><em>]</em><em>, </em><em>default &quot;obs&quot;</em>) – Whether to use the observations (“obs”) or variables (“var”) for the correlation.</p></li>
<li><p><strong>basis</strong> (<em>str</em><em>, </em><em>default &quot;pca&quot;</em>) – Dimensionality reduction to calculate correlation with. Must be a key in adata.obsm, or a basis available as “X_&lt;basis&gt;” such as “umap”, “tsne” or “pca”.</p></li>
<li><p><strong>n_components</strong> (<em>int</em><em>, </em><em>default None</em>) – Number of components to use for the correlation.</p></li>
<li><p><strong>columns</strong> (<em>Optional</em><em>[</em><em>list</em><em>[</em><em>str</em><em>]</em><em>]</em><em>, </em><em>default None</em>) – List of columns to use for the correlation. If None, all numeric columns are used.</p></li>
<li><p><strong>method</strong> (<em>Literal</em><em>[</em><em>&quot;spearmanr&quot;</em><em>, </em><em>&quot;pearson&quot;</em><em>]</em><em>, </em><em>default &quot;spearmanr&quot;</em>) – Method to use for correlation. Must be either “pearsonr” or “spearmanr”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>correlation coefficient, p-values</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple[pd.DataFrame, pd.DataFrame]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – If “basis” is not found in data, if “which” is not “obs” or “var”, if “method” is not “pearsonr” or “spearmanr”, or if “which” is “var” and “basis” not “pca”.</p></li>
<li><p><strong>KeyError</strong> – If any of the given columns is not found in the respective table.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<hr class="docutils" />
<section id="module-sctoolbox.tools.clustering">
<span id="clustering"></span><h3>clustering<a class="headerlink" href="#module-sctoolbox.tools.clustering" title="Link to this heading"></a></h3>
<p>Module for cell clustering.</p>
<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.clustering.recluster">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.clustering.</span></span><span class="sig-name descname"><span class="pre">recluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clusters</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">task</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'join'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'split'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'join'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'leiden'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'louvain'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'leiden'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolution</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">embedding</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'X_umap'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/clustering.html#recluster"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.clustering.recluster" title="Link to this definition"></a></dt>
<dd><p>Recluster an anndata object based on an existing clustering column in .obs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>column</strong> (<em>str</em><em> | </em><em>list</em><em>[</em><em>str</em><em>]</em>) – Column in adata.obs to use for re-clustering.</p></li>
<li><p><strong>clusters</strong> (<em>str</em><em> | </em><em>list</em><em>[</em><em>str</em><em>]</em>) – Clusters in <cite>column</cite> to re-cluster.</p></li>
<li><p><strong>task</strong> (<em>Literal</em><em>[</em><em>&quot;join&quot;</em><em>, </em><em>&quot;split&quot;</em><em>]</em><em>, </em><em>default &quot;join&quot;</em>) – Task to perform. Options are:
- “join”: Join clusters in <cite>clusters</cite> into one cluster.
- “split”: Split clusters in <cite>clusters</cite> are merged and then reclustered using <cite>method</cite> and <cite>resolution</cite>.</p></li>
<li><p><strong>method</strong> (<em>Literal</em><em>[</em><em>&quot;leiden&quot;</em><em>, </em><em>&quot;louvain&quot;</em><em>]</em><em>, </em><em>default &quot;leiden&quot;</em>) – Clustering method to use. Must be one of “leiden” or “louvain”.</p></li>
<li><p><strong>resolution</strong> (<em>float</em><em>, </em><em>default 1</em>) – Resolution parameter for clustering.</p></li>
<li><p><strong>key_added</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Name of the new column in adata.obs. If None, the column name is set to <cite>&lt;column&gt;_recluster</cite>.</p></li>
<li><p><strong>plot</strong> (<em>bool</em><em>, </em><em>default True</em>) – If a plot should be generated of the re-clustering.</p></li>
<li><p><strong>embedding</strong> (<em>str</em><em>, </em><em>default 'X_umap'</em>) – Select which embeding should be used.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>KeyError:</strong> – If the given embeding is not in the data.
    1. If the given embedding is not in the data.
    2. If given column is not found in adata.obs</p></li>
<li><p><strong>ValueError:</strong> – If a given cluster is not found in the adata.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">sctoolbox</span>

<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">pbmc68k_reduced</span><span class="p">()</span>
<span class="n">sctoolbox</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">recluster</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">column</span><span class="o">=</span><span class="s2">&quot;louvain&quot;</span><span class="p">,</span> <span class="n">clusters</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;5&quot;</span><span class="p">],</span> <span class="n">task</span><span class="o">=</span><span class="s2">&quot;join&quot;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;leiden&quot;</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<figure class="align-default">
<img alt="../_images/tools-1.png" class="plot-directive" src="../_images/tools-1.png" />
</figure>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">sctoolbox</span>

<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">pbmc68k_reduced</span><span class="p">()</span>
<span class="n">sctoolbox</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">recluster</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">column</span><span class="o">=</span><span class="s2">&quot;louvain&quot;</span><span class="p">,</span> <span class="n">clusters</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;6&quot;</span><span class="p">],</span> <span class="n">task</span><span class="o">=</span><span class="s2">&quot;split&quot;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;leiden&quot;</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<figure class="align-default">
<img alt="../_images/tools-2.png" class="plot-directive" src="../_images/tools-2.png" />
</figure>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.clustering.gini">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.clustering.</span></span><span class="sig-name descname"><span class="pre">gini</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">_SupportsArray</span><span class="p"><span class="pre">[</span></span><span class="pre">dtype</span><span class="p"><span class="pre">[</span></span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_NestedSequence</span><span class="p"><span class="pre">[</span></span><span class="pre">_SupportsArray</span><span class="p"><span class="pre">[</span></span><span class="pre">dtype</span><span class="p"><span class="pre">[</span></span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">complex</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bytes</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_NestedSequence</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">complex</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bytes</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/clustering.html#gini"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.clustering.gini" title="Link to this definition"></a></dt>
<dd><p>Calculate the Gini coefficient of a numpy array.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>x</strong> (<em>npt.ArrayLike</em>) – Array to calculate Gini coefficient for.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Gini coefficient.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.clustering.calc_ragi">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.clustering.</span></span><span class="sig-name descname"><span class="pre">calc_ragi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condition_column</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'clustering'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">binary_layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">AnnData</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float64</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/clustering.html#calc_ragi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.clustering.calc_ragi" title="Link to this definition"></a></dt>
<dd><p>Calculate the RAGI score over all clusters in the adata.</p>
<p>The RAGI score is a measure of how well a cluster is defined by a set of genes.
The score is the mean of the Gini coefficients of the gene enrichments across the clusters.
The functions uses binary sparse matrices ONLY. If the data is not binary, use <cite>sctoolbox.utils.binarize</cite>.
Binary layers can be selected using the <cite>binary_layer</cite> parameter.
The adata.var table also needs the total counts for each gene.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>condition_column</strong> (<em>str</em>) – Column in <cite>adata.obs</cite> to use for clustering.</p></li>
<li><p><strong>binary_layer</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Layer in <cite>adata.layers</cite> to use for calculating gene enrichment. If None, the raw layer is used.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Annotated data matrix with the Gini coefficients score in <cite>adata.var</cite> and RAGI score.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple[sc.AnnData, np.float64]</p>
</dd>
</dl>
</dd></dl>

</section>
<hr class="docutils" />
<section id="module-sctoolbox.tools.marker_genes">
<span id="marker-genes"></span><h3>marker_genes<a class="headerlink" href="#module-sctoolbox.tools.marker_genes" title="Link to this heading"></a></h3>
<p>Tools for marker gene analyis.</p>
<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.marker_genes.get_chromosome_genes">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.marker_genes.</span></span><span class="sig-name descname"><span class="pre">get_chromosome_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gtf</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosomes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/marker_genes.html#get_chromosome_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.marker_genes.get_chromosome_genes" title="Link to this definition"></a></dt>
<dd><p>Get a list of all genes in the gtf for certain chromosome(s).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gtf</strong> (<em>str</em>) – Path to the gtf file.</p></li>
<li><p><strong>chromosomes</strong> (<em>str</em><em> | </em><em>list</em><em>[</em><em>str</em><em>]</em>) – A chromosome or a list of chromosome names to search for genes in.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of all genes in the gtf for the given chromosome(s).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[str]</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This function is not directly used by the framework, but is used to create the marker gene lists for ‘label_genes’.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError:</strong> – If not all given chromosomes are found in the GTF-file.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.marker_genes.label_genes">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.marker_genes.</span></span><span class="sig-name descname"><span class="pre">label_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_column</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/marker_genes.html#label_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.marker_genes.label_genes" title="Link to this definition"></a></dt>
<dd><p>Label genes as ribosomal, mitochrondrial and gender genes.</p>
<p>Gene labels are added inplace.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – The anndata object.</p></li>
<li><p><strong>species</strong> (<em>str</em>) – Name of the species.</p></li>
<li><p><strong>gene_column</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Name of the column in adata.var that contains the gene names. Uses adata.var.index as default.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List containing the column names added to adata.var.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[str]</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#sctoolbox.tools.qc_filter.predict_cell_cycle" title="sctoolbox.tools.qc_filter.predict_cell_cycle"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sctoolbox.tools.qc_filter.predict_cell_cycle</span></code></a></dt><dd><p>for cell cycle prediction.</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.marker_genes.add_gene_expression">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.marker_genes.</span></span><span class="sig-name descname"><span class="pre">add_gene_expression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/marker_genes.html#add_gene_expression"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.marker_genes.add_gene_expression" title="Link to this definition"></a></dt>
<dd><p>Add values of gene/feature per cell to the adata.obs dataframe.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Anndata object containing gene expression/counts.</p></li>
<li><p><strong>gene</strong> (<em>str</em>) – Name of the gene/feature from the adata.var index to be added to adata.obs.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError:</strong> – If gene is not in adata.var.index.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.marker_genes.run_rank_genes">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.marker_genes.</span></span><span class="sig-name descname"><span class="pre">run_rank_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_in_group_fraction</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_fold_change</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_out_group_fraction</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/marker_genes.html#run_rank_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.marker_genes.run_rank_genes" title="Link to this definition"></a></dt>
<dd><p>Run scanpy rank_genes_groups and filter_rank_genes_groups.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Anndata object containing gene expression/counts.</p></li>
<li><p><strong>groupby</strong> (<em>str</em>) – Column by which the cells in adata should be grouped.</p></li>
<li><p><strong>min_in_group_fraction</strong> (<em>float</em><em>, </em><em>default 0.25</em>) – Minimum fraction of cells in a group that must express a gene to be considered as a marker gene.
Parameter forwarded to scanpy.tl.filter_rank_genes_groups.</p></li>
<li><p><strong>min_fold_change</strong> (<em>float</em><em>, </em><em>default 0.5</em>) – Minimum foldchange (+/-) to be considered as a marker gene.
Parameter forwarded to scanpy.tl.filter_rank_genes_groups.</p></li>
<li><p><strong>max_out_group_fraction</strong> (<em>float</em><em>, </em><em>default 0.8</em>) – Maximum fraction of cells in other groups that must express a gene to be considered as a marker gene.
Parameter forwarded to scanpy.tl.filter_rank_genes_groups.</p></li>
<li><p><strong>**kwargs</strong> (<em>Any</em>) – Additional arguments forwarded to scanpy.tl.rank_genes_groups.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError:</strong> – If number of groups defined by the groupby parameter is &lt; 2.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.marker_genes.pairwise_rank_genes">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.marker_genes.</span></span><span class="sig-name descname"><span class="pre">pairwise_rank_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">foldchange_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_in_group_fraction</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_out_group_fraction</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/marker_genes.html#pairwise_rank_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.marker_genes.pairwise_rank_genes" title="Link to this definition"></a></dt>
<dd><p>Rank genes pairwise between groups in ‘groupby’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Anndata object containing expression data.</p></li>
<li><p><strong>groupby</strong> (<em>str</em>) – Key in adata.obs containing groups to be compared.</p></li>
<li><p><strong>foldchange_threshold</strong> (<em>int</em><em> | </em><em>float</em><em>, </em><em>default 1</em>) – Minimum foldchange (+/-) to be considered as a marker gene.</p></li>
<li><p><strong>min_in_group_fraction</strong> (<em>float</em><em>, </em><em>default 0.25</em>) – Minimum fraction of cells in a group that must express a gene to be considered as a marker gene.</p></li>
<li><p><strong>max_out_group_fraction</strong> (<em>float</em><em>, </em><em>default 0.5</em>) – Maximum fraction of cells in other groups that must express a gene to be considered as a marker gene.</p></li>
<li><p><strong>**kwargs</strong> (<em>Any</em>) – Additional arguments to be passed to scanpy.tl.rank_genes_groups.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dataframe containinge the pairwise ranked gened between the groups.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.marker_genes.get_rank_genes_tables">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.marker_genes.</span></span><span class="sig-name descname"><span class="pre">get_rank_genes_tables</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'rank_genes_groups'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_group_fractions</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_columns</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_excel</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">DataFrame</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/marker_genes.html#get_rank_genes_tables"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.marker_genes.get_rank_genes_tables" title="Link to this definition"></a></dt>
<dd><p>Get gene tables containing “rank_genes_groups” genes and information per group (from previously chosen <cite>groupby</cite>).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Anndata object containing ranked genes.</p></li>
<li><p><strong>key</strong> (<em>str</em><em>, </em><em>default &quot;rank_genes_groups&quot;</em>) – The key in adata.uns to be used for fetching ranked genes.</p></li>
<li><p><strong>n_genes</strong> (<em>Optional</em><em>[</em><em>int</em><em>]</em><em>, </em><em>default 200</em>) – Number of genes to be included in the tables. If None, all genes are included.</p></li>
<li><p><strong>out_group_fractions</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True, the output tables will contain additional columns giving the fraction of genes per group.</p></li>
<li><p><strong>var_columns</strong> (<em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default</em><em> [</em><em>]</em>) – List of adata.var columns, which will be added to pandas.DataFrame.</p></li>
<li><p><strong>save_excel</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – The path to a file for writing the marker gene tables as an excel file (with one sheet per group).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary with group names as keys, and marker gene tables (pandas DataFrames) per group as values.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, pd.DataFrame]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError:</strong> – <ol class="arabic simple">
<li><p>If not all columns given in var_columns are in adata.var.
    2. If key cannot be found in adata.uns.</p></li>
</ol>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.marker_genes.mask_rank_genes">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.marker_genes.</span></span><span class="sig-name descname"><span class="pre">mask_rank_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'rank_genes_groups'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/marker_genes.html#mask_rank_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.marker_genes.mask_rank_genes" title="Link to this definition"></a></dt>
<dd><p>Mask names with “nan” in .uns[key][“names”] if they are found in given ‘genes’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Anndata object containing ranked genes.</p></li>
<li><p><strong>genes</strong> (<em>list</em><em>[</em><em>str</em><em>]</em>) – List of genes to be masked.</p></li>
<li><p><strong>key</strong> (<em>str</em><em>, </em><em>default &quot;rank_genes_groups&quot;</em>) – The key in adata.uns to be used for fetching ranked genes.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default True</em>) – If True, modifies adata.uns[key][“names”] in place. Otherwise, returns a copy of adata.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>If inplace = True, modifies adata.uns[key][“names”] in place and returns None.
Otherwise, returns a copy of adata.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Optional[sc.AnnData]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError:</strong> – If genes is not of type list.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.marker_genes.run_deseq2">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.marker_genes.</span></span><span class="sig-name descname"><span class="pre">run_deseq2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condition_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">confounders</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">percentile_range</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">(0,</span> <span class="pre">100)</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/marker_genes.html#run_deseq2"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.marker_genes.run_deseq2" title="Link to this definition"></a></dt>
<dd><p>Run DESeq2 on counts within adata. Must be run on the raw counts per sample. If the adata contains normalized counts in .X, ‘layer’ can be used to specify raw counts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Anndata object containing raw counts.</p></li>
<li><p><strong>sample_col</strong> (<em>str</em>) – Column name in adata.obs containing sample names.</p></li>
<li><p><strong>condition_col</strong> (<em>str</em>) – Column name in adata.obs containing condition names to be compared.</p></li>
<li><p><strong>confounders</strong> (<em>Optional</em><em>[</em><em>list</em><em>[</em><em>str</em><em>]</em><em>]</em><em>, </em><em>default None</em>) – List of additional column names in adata.obs containing confounders to be included in the model.</p></li>
<li><p><strong>layer</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Name of layer containing raw counts to be used for DESeq2. Default is None (use .X for counts)</p></li>
<li><p><strong>percentile_range</strong> (<em>Tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em><em>, </em><em>default</em><em> (</em><em>0</em><em>, </em><em>100</em><em>)</em>) – Percentile range of cells to be used for calculating pseudobulks. Setting (0,95) will restrict calculation
to the cells in the 0-95% percentile ranges. Default is (0, 100), which means all cells are used.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dataframe containing the results of the DESeq2 analysis.
Also adds the dataframe to adata.uns[“deseq_result”]</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError:</strong> – <ol class="arabic simple">
<li><p>If any given column name is not found in adata.obs.
    2. If any given column name contains characters not compatible with R.</p></li>
</ol>
</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Needs the package ‘diffexpr’ to be installed along with ‘bioconductor-deseq2’ and ‘rpy2’.
These can be obtained by installing the sctoolbox [deseq2] extra with pip using: <cite>pip install . .[deseq2]</cite>.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">sctoolbox.utils.pseudobulk_table</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.marker_genes.score_genes">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.marker_genes.</span></span><span class="sig-name descname"><span class="pre">score_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_set</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'score'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/marker_genes.html#score_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.marker_genes.score_genes" title="Link to this definition"></a></dt>
<dd><p>Assign a score to each cell depending on the expression of a set of genes. This is a wrapper for scanpy.tl.score_genes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Anndata object to score.</p></li>
<li><p><strong>gene_set</strong> (<em>str</em><em> | </em><em>list</em><em>[</em><em>str</em><em>]</em>) – A list of genes or path to a file containing a list of genes.
The txt file should have one gene per row.</p></li>
<li><p><strong>score_name</strong> (<em>str</em><em>, </em><em>default &quot;score&quot;</em>) – Name of the column in obs table where the score will be added.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default True</em>) – Adds the new column to the original anndata object.</p></li>
<li><p><strong>**kwargs</strong> (<em>Any</em>) – Additional arguments to be passed to scanpy.tl.score_genes.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>If inplace is False, return a copy of anndata object with the new column in the obs table.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Optional[sc.AnnData]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>FileNotFoundError:</strong> – If path given in gene_set does not lead to a file.</p>
</dd>
</dl>
</dd></dl>

</section>
<hr class="docutils" />
<section id="module-sctoolbox.tools.celltype_annotation">
<span id="celltype-annotation"></span><h3>celltype_annotation<a class="headerlink" href="#module-sctoolbox.tools.celltype_annotation" title="Link to this heading"></a></h3>
<p>Module for general celltype annotation.</p>
<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.celltype_annotation.add_cellxgene_annotation">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.celltype_annotation.</span></span><span class="sig-name descname"><span class="pre">add_cellxgene_annotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">csv</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/celltype_annotation.html#add_cellxgene_annotation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.celltype_annotation.add_cellxgene_annotation" title="Link to this definition"></a></dt>
<dd><p>Add columns from cellxgene annotation to the adata .obs table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Adata object to add annotations to.</p></li>
<li><p><strong>csv</strong> (<em>str</em>) – Path to the annotation file from cellxgene containing cell annotation.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.celltype_annotation.get_celltype_assignment">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.celltype_annotation.</span></span><span class="sig-name descname"><span class="pre">get_celltype_assignment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clustering</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_genes_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'celltype'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/celltype_annotation.html#get_celltype_assignment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.celltype_annotation.get_celltype_assignment" title="Link to this definition"></a></dt>
<dd><p>Get cell type assignment based on marker genes.</p>
<p>TODO make this more robust</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Anndata object containing raw counts.</p></li>
<li><p><strong>clustering</strong> (<em>str</em>) – Name of clustering column to use for cell type assignment.</p></li>
<li><p><strong>marker_genes_dict</strong> (<em>dict</em><em>[</em><em>str</em><em>, </em><em>list</em><em>[</em><em>str</em><em>]</em><em>]</em>) – Dictionary containing cell type names as keys and lists of marker genes as values.</p></li>
<li><p><strong>column_name</strong> (<em>str</em><em>, </em><em>default: &quot;celltype&quot;</em>) – Name of column to add to adata.obs containing cell type assignment.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Returns a dictionary with cluster-to-celltype mapping (key: cluster name, value: cell type)
Also adds the cell type assignment to adata.obs[&lt;column_name&gt;] in place.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.celltype_annotation.run_scsa">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.celltype_annotation.</span></span><span class="sig-name descname"><span class="pre">run_scsa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_column</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_symbol</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'auto'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'symbol'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'id'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'auto'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'rank_genes_groups'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column_added</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'SCSA_pred_celltype'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">python_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'Human'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'Mouse'</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Human'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pvalue</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tissue</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'All'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">user_db</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">celltype_column</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'cell_name'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/celltype_annotation.html#run_scsa"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.celltype_annotation.run_scsa" title="Link to this definition"></a></dt>
<dd><p>Run SCSA cell type annotation and assign cell types to cluster in an adata object.</p>
<p>This is a wrapper function that extracts ranked genes generated by scanpy.tl.rank_genes_groups
function and generates input matrix for SCSA, then runs SCSA and assigns cell types to clusters
in adata.obs.</p>
<p>Also adds adata.uns[‘SCSA’] as a dictionary with the following keys:
- ‘results’: SCSA result table
- ‘stderr’: SCSA stderr
- ‘stdout’: SCSA stdout
- ‘cmd’: SCSA command</p>
<p class="rubric">Notes</p>
<p>SCSA sometimes gives ValueError: MultiIndex (as covered in <a class="reference external" href="https://github.com/bioinfo-ibms-pumc/SCSA/issues/19">https://github.com/bioinfo-ibms-pumc/SCSA/issues/19</a>).
This can be solved by downgrading pandas to 1.2.4.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Adata object to be annotated, must contain ranked genes in adata.uns</p></li>
<li><p><strong>gene_column</strong> (<em>str</em><em>, </em><em>default None</em>) – Name of the column in adata.var that contains the gene names.</p></li>
<li><p><strong>gene_symbol</strong> (<em>str</em><em>, </em><em>default 'auto'</em>) – TODO Implement
The type of gene symbol. One of “auto”, “symbol” (gene name) or “id” (ensembl id).</p></li>
<li><p><strong>key</strong> (<em>str</em><em>, </em><em>default 'rank_genes_groups'</em>) – The key in adata.uns where ranked genes are stored.</p></li>
<li><p><strong>column_added</strong> (<em>str</em><em>, </em><em>default 'SCSA_pred_celltype'</em>) – The column name in adata.obs where the cell types will be added.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default True</em>) – If True, cell types will be added to adata.obs.</p></li>
<li><p><strong>python_path</strong> (<em>str</em><em>, </em><em>default None</em>) – SCSA parameter: Path to python. If not given, will be inferred from sys.executable.</p></li>
<li><p><strong>species</strong> (<em>Optional</em><em>[</em><em>Literal</em><em>[</em><em>'Human'</em><em>, </em><em>'Mouse'</em><em>]</em><em>]</em><em>, </em><em>default 'Human'</em>) – SCSA parameter: Supports only Human or Mouse. Set to None to use the user defined database given in user_db.</p></li>
<li><p><strong>fc</strong> (<em>float</em><em>, </em><em>default 1.5</em>) – SCSA parameter: Fold change threshold to filter genes.</p></li>
<li><p><strong>pvalue</strong> (<em>float</em><em>, </em><em>default 0.05</em>) – SCSA parameter: P-value threshold to filter genes.</p></li>
<li><p><strong>tissue</strong> (<em>str</em><em>, </em><em>default 'All'</em>) – TODO Implement
SCSA parameter: A specific tissue can be defined.</p></li>
<li><p><strong>user_db</strong> (<em>str</em><em>, </em><em>default None</em>) – SCSA parameter: Path to the user defined marker database.
Must contain at least two columns, one named “cell_name” (or set via celltype_column) for the cell type annotation,
and at least one more column with gene names or ids (selected automatically from best gene overlap).</p></li>
<li><p><strong>celltype_column</strong> (<em>str</em><em>, </em><em>default 'cell_name'</em>) – SCSA parameter: The column name in the user_db that contains the cell type annotation.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>If inplace==False, returns adata with cell types in adata.obs
Else return None</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Optional[sc.AnnData]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>KeyError:</strong> – <ol class="arabic simple">
<li><p>If key is not in adata.uns.
    2. If ‘params’ is not in adata.uns[key] or if ‘groupby’ is not in adata.uns[key][‘params’].
    3. If gene column is not in adata.var</p></li>
</ol>
</p></li>
<li><p><strong>ValueError:</strong> – <ol class="arabic simple">
<li><p>If species parameter is not Human, Mouse or None.
    2. If no species and no user database is provided.
    3. If SCSA run failes</p></li>
</ol>
</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
</section>
<section id="additional">
<h2>Additional<a class="headerlink" href="#additional" title="Link to this heading"></a></h2>
<section id="module-sctoolbox.tools.calc_overlap_fc">
<span id="calc-overlap-fc"></span><h3>calc_overlap_fc<a class="headerlink" href="#module-sctoolbox.tools.calc_overlap_fc" title="Link to this heading"></a></h3>
<p>Module to calculate fold change of reads or fragments from a BAM or fragments file that overlap specified regions.</p>
<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.calc_overlap_fc.pct_fragments_in_promoters">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.calc_overlap_fc.</span></span><span class="sig-name descname"><span class="pre">pct_fragments_in_promoters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gtf_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bam_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fragments_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cb_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cb_tag</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'CB'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'bos_taurus'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'caenorhabditis_elegans'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'canis_lupus_familiaris'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'danio_rerio'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'drosophila_melanogaster'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'gallus_gallus'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'homo_sapiens'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'mus_musculus'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'oryzias_latipes'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'rattus_norvegicus'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'sus_scrofa'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'xenopus_tropicalis'</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/calc_overlap_fc.html#pct_fragments_in_promoters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.calc_overlap_fc.pct_fragments_in_promoters" title="Link to this definition"></a></dt>
<dd><p>Calculate the fold change of fragments in promoters.</p>
<p>This function calculates for each cell, the fold change of fragments in a BAM alignment file
that overlap with a promoter region specified in a GTF file. The results are added to the anndata object
as new column.
This is a wrapper function for fc_fragments_in_regions to use promoters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – The anndata object containig cell barcodes in adata.obs.</p></li>
<li><p><strong>gtf_file</strong> (<em>str</em><em>, </em><em>default None</em>) – Path to GTF file for promoters regions. if None, the GTF file in flatfiles directory will be used.</p></li>
<li><p><strong>bam_file</strong> (<em>str</em><em>, </em><em>default None</em>) – Path to BAM file. If None, a fragments file must be provided in the parameter ‘fragments_file’.</p></li>
<li><p><strong>fragments_file</strong> (<em>str</em><em>, </em><em>default None</em>) – Path to fragments file. If None, a BAM file must be provided in the parameter ‘bam_file’. The
BAM file will be converted into fragments file.</p></li>
<li><p><strong>cb_col</strong> (<em>str</em><em>, </em><em>default None</em>) – The column in adata.obs containing cell barcodes. If None, adata.obs.index will be used.</p></li>
<li><p><strong>cb_tag</strong> (<em>str</em><em>, </em><em>default 'CB'</em>) – The tag where cell barcodes are saved in the bam file. Set to None if the barcodes are in read names.</p></li>
<li><p><strong>species</strong> (<em>str</em><em>, </em><em>default None</em>) – Name of the species, will only be used if gtf_file is None to use internal GTF files.
Species are {bos_taurus, caenorhabditis_elegans, canis_lupus_familiaris, danio_rerio, drosophila_melanogaster,
gallus_gallus, homo_sapiens, mus_musculus, oryzias_latipes, rattus_norvegicus, sus_scrofa, xenopus_tropicalis}</p></li>
<li><p><strong>threads</strong> (<em>int</em><em>, </em><em>default 1</em>) – Number of threads for parallelization. Will be used to convert BAM to fragments file.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If no species and no gtf_file is given.</p>
</dd>
</dl>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 0.4b: </span>This will be removed in 0.5. Use the overlap_fragments_in_regions function instead.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.calc_overlap_fc.fc_fragments_in_regions">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.calc_overlap_fc.</span></span><span class="sig-name descname"><span class="pre">fc_fragments_in_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regions_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bam_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fragments_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cb_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cb_tag</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'CB'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regions_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'list'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/calc_overlap_fc.html#fc_fragments_in_regions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.calc_overlap_fc.fc_fragments_in_regions" title="Link to this definition"></a></dt>
<dd><p>Calculate the fold change of fragments in a region against the background.</p>
<p>This function calculates the fold change of fragments overlapping a region against the background for each cell.
The regions are specified in a BED or GTF file and the fragments should be provided by a fragments or BAM file.
The results are added to the anndata object as new column.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – The anndata object containig cell barcodes in adata.obs.</p></li>
<li><p><strong>regions_file</strong> (<em>str</em>) – Path to BED or GTF file containing regions of interest.</p></li>
<li><p><strong>bam_file</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Path to BAM file. If None, a fragments file must be provided in the parameter ‘fragments_file’.</p></li>
<li><p><strong>fragments_file</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Path to fragments file. If None, a BAM file must be provided in the parameter ‘bam_file’. The
BAM file will be converted into fragments file.</p></li>
<li><p><strong>cb_col</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – The column in adata.obs containing cell barcodes. If None, adata.obs.index will be used.</p></li>
<li><p><strong>cb_tag</strong> (<em>str</em><em>, </em><em>default 'CB'</em>) – The tag where cell barcodes are saved in the bam file. Set to None if the barcodes are in read names.</p></li>
<li><p><strong>regions_name</strong> (<em>str</em><em>, </em><em>default 'list'</em>) – The name of the regions in the BED or GTF file (e.g. Exons). The name will be used as columns’ name
added to the anndata object (e.g. pct_fragments_in_{regions_name}).</p></li>
<li><p><strong>threads</strong> (<em>int</em><em>, </em><em>default 1</em>) – Number of threads for parallelization. Will be used to convert BAM to fragments file.</p></li>
<li><p><strong>temp_dir</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Path to temporary directory. Will use the current working directory by default.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If bam_file and fragment file is not provided.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.calc_overlap_fc.count_fragments_per_cell">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.calc_overlap_fc.</span></span><span class="sig-name descname"><span class="pre">count_fragments_per_cell</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">barcode_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'barcode'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frag_count</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'count'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/calc_overlap_fc.html#count_fragments_per_cell"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.calc_overlap_fc.count_fragments_per_cell" title="Link to this definition"></a></dt>
<dd><p>Get counts per cell from a dataframe containing fragments.</p>
<p>The dataframe must have a column with cell barcodes and a column with fragment counts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – The dataframe of a bedfile containing the fragments.</p></li>
<li><p><strong>barcode_col</strong> (<em>str</em><em>, </em><em>default 'barcode'</em>) – The column name containing the cell barcodes.</p></li>
<li><p><strong>frag_count</strong> (<em>str</em><em>, </em><em>default 'count'</em>) – The column name containing the fragment counts.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The dataframe containing the number of fragments per cell.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

</section>
<hr class="docutils" />
<section id="module-sctoolbox.tools.multiomics">
<span id="multiomics"></span><h3>multiomics<a class="headerlink" href="#module-sctoolbox.tools.multiomics" title="Link to this heading"></a></h3>
<p>Tools for multiomics analysis.</p>
<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.multiomics.merge_anndata">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.multiomics.</span></span><span class="sig-name descname"><span class="pre">merge_anndata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">anndata_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AnnData</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">join</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'inner'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'outer'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'inner'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/multiomics.html#merge_anndata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.multiomics.merge_anndata" title="Link to this definition"></a></dt>
<dd><p>Merge two h5ad files for dual cellxgene deplyoment.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>anndata_dict</strong> (<em>dict</em><em>[</em><em>str</em><em>, </em><em>sc.AnnData</em><em>]</em>) – Dictionary with labels as keys and anndata objects as values.</p></li>
<li><p><strong>join</strong> (<em>Literal</em><em>[</em><em>'inner'</em><em>, </em><em>'outer'</em><em>]</em><em>, </em><em>default 'inner'</em>) – <p>Set how to join cells of the adata objects: [‘inner’, ‘outer’].
This only affects the cells since the var/gene section is simply added.</p>
<ul>
<li><p>’inner’: only keep overlapping cells.</p></li>
<li><p>’outer’: keep all cells. This will add placeholder cells/dots to plots currently disabled.</p></li>
</ul>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Merged anndata object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>sc.AnnData</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Important: Depending on the size of the anndata objects the function takes
around 60 to 300 GB of RAM!
To save RAM and runtime the function generates a minimal anndata object.
Only .X, .var, .obs and .obsm are kept. Layers, .varm, etc is removed.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError:</strong> – If no indices of both adata.obs tables are overlapping.</p>
</dd>
</dl>
</dd></dl>

</section>
<hr class="docutils" />
<section id="module-sctoolbox.tools.bam">
<span id="bam"></span><h3>bam<a class="headerlink" href="#module-sctoolbox.tools.bam" title="Link to this heading"></a></h3>
<p>Functionality to split bam files into smaller bam files based on clustering in adata.obs.</p>
<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.bam.bam_adata_ov">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.bam.</span></span><span class="sig-name descname"><span class="pre">bam_adata_ov</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bamfile</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cb_tag</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'CB'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/bam.html#bam_adata_ov"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.bam.bam_adata_ov" title="Link to this definition"></a></dt>
<dd><p>Check if adata.obs barcodes existing in a column of a bamfile.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – adata object where adata.obs is stored</p></li>
<li><p><strong>bamfile</strong> (<em>str</em>) – path of the bamfile to investigate</p></li>
<li><p><strong>cb_tag</strong> (<em>str</em><em>, </em><em>default 'CB'</em>) – bamfile column to extract the barcodes from</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>hitrate of the barcodes in the bamfile</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.bam.check_barcode_tag">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.bam.</span></span><span class="sig-name descname"><span class="pre">check_barcode_tag</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bamfile</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cb_tag</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'CB'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/bam.html#check_barcode_tag"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.bam.check_barcode_tag" title="Link to this definition"></a></dt>
<dd><p>Check for the possibilty that the wrong barcode is used.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – adata object where adata.obs is stored</p></li>
<li><p><strong>bamfile</strong> (<em>str</em>) – path of the bamfile to investigate</p></li>
<li><p><strong>cb_tag</strong> (<em>str</em><em>, </em><em>default 'CB'</em>) – bamfile column to extract the barcodes from</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If barcode hit rate could not be identified</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.bam.subset_bam">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.bam.</span></span><span class="sig-name descname"><span class="pre">subset_bam</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bam_in</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bam_out</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">barcodes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Iterable</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">read_tag</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'CB'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pysam_threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/bam.html#subset_bam"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.bam.subset_bam" title="Link to this definition"></a></dt>
<dd><p>Subset a bam file based on a list of barcodes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bam_in</strong> (<em>str</em>) – Path to input bam file.</p></li>
<li><p><strong>bam_out</strong> (<em>str</em>) – Path to output bam file.</p></li>
<li><p><strong>barcodes</strong> (<em>Iterable</em><em>[</em><em>str</em><em>]</em>) – List of barcodes to keep.</p></li>
<li><p><strong>read_tag</strong> (<em>str</em><em>, </em><em>default &quot;CB&quot;</em>) – Tag in bam file to use for barcode.</p></li>
<li><p><strong>pysam_threads</strong> (<em>int</em><em>, </em><em>default 4</em>) – Number of threads to use for pysam.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>default False</em>) – Overwrite output file if it exists.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.bam.split_bam_clusters">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.bam.</span></span><span class="sig-name descname"><span class="pre">split_bam_clusters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bams</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Iterable</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">barcode_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">read_tag</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'CB'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'split_'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reader_threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">writer_threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parallel</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pysam_threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">buffer_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_queue_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individual_pbars</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sort_bams</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index_bams</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/bam.html#split_bam_clusters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.bam.split_bam_clusters" title="Link to this definition"></a></dt>
<dd><p>Split BAM files into clusters based on ‘groupby’ from the anndata.obs table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.Anndata</em>) – Annotated data matrix containing clustered cells in .obs.</p></li>
<li><p><strong>bams</strong> (<em>str</em><em> | </em><em>Iterable</em><em>[</em><em>str</em><em>]</em>) – One or more BAM files to split into clusters</p></li>
<li><p><strong>groupby</strong> (<em>str</em>) – Name of a column in adata.obs to cluster the pseudobulks by.</p></li>
<li><p><strong>barcode_col</strong> (<em>str</em><em>, </em><em>default None</em>) – Name of a column in adata.obs to use as barcodes. If None, use the index of .obs.</p></li>
<li><p><strong>read_tag</strong> (<em>str</em><em>, </em><em>default &quot;CB&quot;</em>) – Tag to use to identify the reads to split. Must match the barcodes of the barcode_col.</p></li>
<li><p><strong>output_prefix</strong> (str, default <cite>split_</cite>) – Prefix to use for the output files.</p></li>
<li><p><strong>reader_threads</strong> (<em>int</em><em>, </em><em>default 1</em>) – Number of threads to use for reading.</p></li>
<li><p><strong>writer_threads</strong> (<em>int</em><em>, </em><em>default 1</em>) – Number of threads to use for writing.</p></li>
<li><p><strong>parallel</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to enable parallel processsing.</p></li>
<li><p><strong>pysam_threads</strong> (<em>int</em><em>, </em><em>default 4</em>) – Number of threads for pysam.</p></li>
<li><p><strong>buffer_size</strong> (<em>int</em><em>, </em><em>default 10000</em>) – The size of the buffer between readers and writers.</p></li>
<li><p><strong>max_queue_size</strong> (<em>int</em><em>, </em><em>default 1000</em>) – The maximum size of the queue between readers and writers.</p></li>
<li><p><strong>individual_pbars</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to show a progress bar for each individual BAM file and output clusters. Default: False (overall progress bars).</p></li>
<li><p><strong>sort_bams</strong> (<em>bool</em><em>, </em><em>default False</em>) – Sort reads in each output bam</p></li>
<li><p><strong>index_bams</strong> (<em>bool</em><em>, </em><em>default False</em>) – Create an index file for each output bam. Will throw an error if <cite>sort_bams</cite> is False.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError:</strong> – <ol class="arabic simple">
<li><p>If groupby column is not in adata.obs
    2. If barcode column is not in adata.obs
    3. If index_bams is set and sort_bams is False</p></li>
</ol>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.bam.open_bam">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.bam.</span></span><span class="sig-name descname"><span class="pre">open_bam</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="m"><span class="pre">0</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="m"><span class="pre">1</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="m"><span class="pre">2</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="m"><span class="pre">3</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">pysam.AlignmentFile</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/bam.html#open_bam"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.bam.open_bam" title="Link to this definition"></a></dt>
<dd><p>Open bam file with pysam.AlignmentFile. On a specific verbosity level.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file</strong> (<em>str</em>) – Path to bam file.</p></li>
<li><p><strong>mode</strong> (<em>str</em>) – Mode to open the file in. See pysam.AlignmentFile</p></li>
<li><p><strong>verbosity</strong> (<em>Literal</em><em>[</em><em>0</em><em>, </em><em>1</em><em>, </em><em>2</em><em>, </em><em>3</em><em>]</em><em>, </em><em>default 3</em>) – Set verbosity level. Verbosity level 0 for no messages.</p></li>
<li><p><strong>**kwargs</strong> (<em>Any</em>) – Forwarded to pysam.AlignmentFile</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Object to work on SAM/BAM files.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pysam.AlignmentFile</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.bam.get_bam_reads">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.bam.</span></span><span class="sig-name descname"><span class="pre">get_bam_reads</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bam_obj</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">pysam.AlignmentFile</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/bam.html#get_bam_reads"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.bam.get_bam_reads" title="Link to this definition"></a></dt>
<dd><p>Get the number of reads from an open pysam.AlignmentFile.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>bam_obj</strong> (<em>pysam.AlignmentFile</em>) – An open pysam.AlignmentFile object to get the number of reads from.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>number of reads in the bam file</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.bam.bam_to_bigwig">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.bam.</span></span><span class="sig-name descname"><span class="pre">bam_to_bigwig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bam</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tempdir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'.'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_temp</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bedtools_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bgtobw_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/bam.html#bam_to_bigwig"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.bam.bam_to_bigwig" title="Link to this definition"></a></dt>
<dd><p>Convert reads in a bam-file to bigwig format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bam</strong> (<em>str</em>) – Path to bam file.</p></li>
<li><p><strong>output</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Path to output file. If None, output is written to same directory as bam file with same name and .bw extension.</p></li>
<li><p><strong>scale</strong> (<em>bool</em><em>, </em><em>default True</em>) – Scale output depth to reads per million mapped reads.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>default True</em>) – Overwrite output file if it already exists.</p></li>
<li><p><strong>tempdir</strong> (<em>str</em><em>, </em><em>default &quot;.&quot;</em>) – Path to directory where temporary files are written.</p></li>
<li><p><strong>remove_temp</strong> (<em>bool</em><em>, </em><em>default True</em>) – Remove temporary files after conversion.</p></li>
<li><p><strong>bedtools_path</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Path to bedtools binary. If None, the function will search for the binary in the path.</p></li>
<li><p><strong>bgtobw_path</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Path to bedGraphToBigWig binary. If None, the function will search for the binary in the path.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Path to output file.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.bam.create_fragment_file">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.bam.</span></span><span class="sig-name descname"><span class="pre">create_fragment_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bam</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">barcode_tag</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'CB'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">barcode_regex</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outdir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nproc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_dist</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_dist</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_clipped</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shift_plus</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shift_minus</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">-4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_temp</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/bam.html#create_fragment_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.bam.create_fragment_file" title="Link to this definition"></a></dt>
<dd><p>Create fragments file out of a BAM file.</p>
<p>This is an alternative to using the sinto package,
which is slow and has issues with large bam-files.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bam</strong> (<em>str</em>) – Path to .bam file.</p></li>
<li><p><strong>barcode_tag</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default 'CB'</em>) – The tag where cell barcodes are saved in the bam file. Set to None if the barcodes are in read names.</p></li>
<li><p><strong>barcode_regex</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Regex to extract barcodes from read names. Set to None if barcodes are stored in a tag.</p></li>
<li><p><strong>outdir</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Path to save fragments file. If None, the file will be saved in the same folder as the .bam file. Temporary intermediate files are also written to this directory.</p></li>
<li><p><strong>nproc</strong> (<em>int</em><em>, </em><em>default 1</em>) – Number of threads for parallelization.</p></li>
<li><p><strong>index</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True, index fragments file. Requires bgzip and tabix.</p></li>
<li><p><strong>min_dist</strong> (<em>int</em><em>, </em><em>default 10</em>) – Minimum fragment length to consider.</p></li>
<li><p><strong>max_dist</strong> (<em>int</em><em>, </em><em>default 5000</em>) – Maximum fragment length to consider.</p></li>
<li><p><strong>include_clipped</strong> (<em>bool</em><em>, </em><em>default True</em>) – Whether to include soft-clipped bases in the fragment length. If True, the full length between reads is used. If False, the fragment length will be calculated from the aligned parts of the reads.</p></li>
<li><p><strong>shift_plus</strong> (<em>int</em><em>, </em><em>default 5</em>) – Shift the start position of the forward read by this value (standard for ATAC).</p></li>
<li><p><strong>shift_minus</strong> (<em>int</em><em>, </em><em>default -4</em>) – Shift the end position of the reverse read by this value (standard for ATAC).</p></li>
<li><p><strong>keep_temp</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True, keep temporary files.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Path to fragments file.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – If both barcode_tag and barcode_regex (or neither) are set.</p></li>
<li><p><strong>FileNotFoundError</strong> – If the input bam file does not exist.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<hr class="docutils" />
<section id="module-sctoolbox.tools.frip">
<span id="frip"></span><h3>frip<a class="headerlink" href="#module-sctoolbox.tools.frip" title="Link to this heading"></a></h3>
<p>Module containing functions for calculating the FRiP score.</p>
<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.frip.calc_frip_scores">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.frip.</span></span><span class="sig-name descname"><span class="pre">calc_frip_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fragments</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">AnnData</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/frip.html#calc_frip_scores"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.frip.calc_frip_scores" title="Link to this definition"></a></dt>
<dd><p>Calculate the FRiP score for each barcode and adds it to adata.obs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – AnnData object containing the fragments</p></li>
<li><p><strong>fragments</strong> (<em>str</em>) – path to fragments bedfile</p></li>
<li><p><strong>temp_dir</strong> (<em>str</em><em>, </em><em>default &quot;&quot;</em>) – path to temp directory</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>AnnData object containing the fragments
total FRiP score</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple[sc.AnnData, float]</p>
</dd>
</dl>
</dd></dl>

</section>
<hr class="docutils" />
<section id="module-sctoolbox.tools.tsse">
<span id="tsse"></span><h3>tsse<a class="headerlink" href="#module-sctoolbox.tools.tsse" title="Link to this heading"></a></h3>
<p>Module to calculate TSS enrichment scores.</p>
<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.tsse.write_TSS_bed">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.tsse.</span></span><span class="sig-name descname"><span class="pre">write_TSS_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gtf</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">custom_TSS</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">negativ_shift</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">positiv_shift</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">list</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/tsse.html#write_TSS_bed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.tsse.write_TSS_bed" title="Link to this definition"></a></dt>
<dd><p>Write a custom TSS file from a gene gtf file.</p>
<p>negativ_shift and positiv_shift are the number of bases of the flanks.
The output is a bed file with the following columns:
1: chr, 2: start, 3:stop
and a list of lists with the following columns:
1: chr, 2: start, 3:stop</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gtf</strong> (<em>str</em>) – path to gtf file</p></li>
<li><p><strong>custom_TSS</strong> (<em>str</em>) – path to output file</p></li>
<li><p><strong>negativ_shift</strong> (<em>int</em><em>, </em><em>default 2000</em>) – number of bases to shift upstream</p></li>
<li><p><strong>positiv_shift</strong> (<em>int</em><em>, </em><em>default 2000</em>) – number of bases to shift downstream</p></li>
<li><p><strong>temp_dir</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – path to temporary directory</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>list of lists with the following columns:
1: chr, 2: start, 3:stop</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple[list, list]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.tsse.overlap_and_aggregate">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.tsse.</span></span><span class="sig-name descname"><span class="pre">overlap_and_aggregate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fragments</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">custom_TSS</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tss_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">negativ_shift</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">positiv_shift</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2000</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/tsse.html#overlap_and_aggregate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.tsse.overlap_and_aggregate" title="Link to this definition"></a></dt>
<dd><p>Overlap the fragments with the custom TSS file and aggregates the fragments in a dictionary.</p>
<p>The dictionary has the following structure:
{barcode: [tss_agg, n_fragments]}
tss_agg is a numpy array with the aggregated fragments around the TSS with flanks of negativ_shift and positiv_shift.
n_fragments is the number of fragments overlapping the TSS.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fragments</strong> (<em>str</em>) – path to fragments file</p></li>
<li><p><strong>custom_TSS</strong> (<em>str</em>) – path to custom TSS file</p></li>
<li><p><strong>overlap</strong> (<em>str</em>) – path to output file</p></li>
<li><p><strong>tss_list</strong> (<em>list</em><em>[</em><em>list</em><em>[</em><em>str</em><em> | </em><em>int</em><em>]</em><em>]</em>) – list of lists with the following columns:
1: chr, 2: start, 3:stop</p></li>
<li><p><strong>negativ_shift</strong> (<em>int</em><em>, </em><em>default 2000</em>) – number of bases to shift upstream</p></li>
<li><p><strong>positiv_shift</strong> (<em>int</em><em>, </em><em>default 2000</em>) – number of bases to shift downstream</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>dictionary with the following structure:
{barcode: [tss_agg, n_fragments]}
tss_agg is a numpy array with the aggregated fragments around the TSS with flanks of negativ_shift and positiv_shift.
n_fragments is the number of fragments overlapping the TSS.
list[str] contains a list of temporary files.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple[dict[str, list], list[str]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.tsse.calc_per_base_tsse">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.tsse.</span></span><span class="sig-name descname"><span class="pre">calc_per_base_tsse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tSSe_df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_bias</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">100</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">_SupportsArray</span><span class="p"><span class="pre">[</span></span><span class="pre">dtype</span><span class="p"><span class="pre">[</span></span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_NestedSequence</span><span class="p"><span class="pre">[</span></span><span class="pre">_SupportsArray</span><span class="p"><span class="pre">[</span></span><span class="pre">dtype</span><span class="p"><span class="pre">[</span></span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">complex</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bytes</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_NestedSequence</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">complex</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bytes</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/tsse.html#calc_per_base_tsse"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.tsse.calc_per_base_tsse" title="Link to this definition"></a></dt>
<dd><p>Calculate per base tSSe by dividing the tSSe by the bias. The bias is calculated by averaging the edges of the tSSe.</p>
<p>The edges are defined by the edge_size.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tSSe_df</strong> (<em>pd.DataFrame</em>) – dataframe with the following columns:
1: barcode, 2: tSSe, 3: n_fragments</p></li>
<li><p><strong>min_bias</strong> (<em>float</em><em>, </em><em>default 0.01</em>) – minimum bias to avoid division by zero</p></li>
<li><p><strong>edge_size</strong> (<em>int</em><em>, </em><em>default 100</em>) – number of bases to use for the edges</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>numpy array with the per base tSSe</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>npt.ArrayLike</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.tsse.global_tsse_score">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.tsse.</span></span><span class="sig-name descname"><span class="pre">global_tsse_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">per_base_tsse</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">_SupportsArray</span><span class="p"><span class="pre">[</span></span><span class="pre">dtype</span><span class="p"><span class="pre">[</span></span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_NestedSequence</span><span class="p"><span class="pre">[</span></span><span class="pre">_SupportsArray</span><span class="p"><span class="pre">[</span></span><span class="pre">dtype</span><span class="p"><span class="pre">[</span></span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">complex</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bytes</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_NestedSequence</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">complex</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bytes</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">negativ_shift</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">_SupportsArray</span><span class="p"><span class="pre">[</span></span><span class="pre">dtype</span><span class="p"><span class="pre">[</span></span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_NestedSequence</span><span class="p"><span class="pre">[</span></span><span class="pre">_SupportsArray</span><span class="p"><span class="pre">[</span></span><span class="pre">dtype</span><span class="p"><span class="pre">[</span></span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">complex</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bytes</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_NestedSequence</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">complex</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bytes</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/tsse.html#global_tsse_score"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.tsse.global_tsse_score" title="Link to this definition"></a></dt>
<dd><p>Calculate the global tSSe score by averaging the per base tSSe around the TSS.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>per_base_tsse</strong> (<em>npt.ArrayLike</em>) – numpy array with the per base tSSe</p></li>
<li><p><strong>negativ_shift</strong> (<em>int</em>) – number of bases to shift upstream</p></li>
<li><p><strong>edge_size</strong> (<em>int</em><em>, </em><em>default 50</em>) – number of bases to use for the edges</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>numpy array with the global tSSe score</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>npt.ArrayLike</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.tsse.tsse_scoring">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.tsse.</span></span><span class="sig-name descname"><span class="pre">tsse_scoring</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fragments</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gtf</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">negativ_shift</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">positiv_shift</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_size_total</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_size_per_base</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_bias</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_tmp</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/tsse.html#tsse_scoring"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.tsse.tsse_scoring" title="Link to this definition"></a></dt>
<dd><p>Calculate the tSSe score for each cell.</p>
<p>Calculating the TSSe score is done like described in: “Chromatin accessibility profiling by ATAC-seq” Fiorella et al. 2022</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fragments</strong> (<em>str</em>) – path to fragments file</p></li>
<li><p><strong>gtf</strong> (<em>str</em>) – path to gtf file</p></li>
<li><p><strong>negativ_shift</strong> (<em>int</em><em>, </em><em>default 2000</em>) – number of bases to shift upstream</p></li>
<li><p><strong>positiv_shift</strong> (<em>int</em><em>, </em><em>default 2000</em>) – number of bases to shift downstream</p></li>
<li><p><strong>edge_size_total</strong> (<em>int</em><em>, </em><em>default 100</em>) – number of bases to use for the edges for the global tSSe score</p></li>
<li><p><strong>edge_size_per_base</strong> (<em>int</em><em>, </em><em>default 50</em>) – number of bases to use for the edges for the per base tSSe score</p></li>
<li><p><strong>min_bias</strong> (<em>float</em><em>, </em><em>default 0.01</em>) – minimum bias to avoid division by zero</p></li>
<li><p><strong>keep_tmp</strong> (<em>bool</em><em>, </em><em>default False</em>) – keep temporary files</p></li>
<li><p><strong>temp_dir</strong> (<em>str</em><em>, </em><em>default &quot;&quot;</em>) – path to temporary directory</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>dataframe with the following columns:
1: barcode, 2: tSSe, 3: n_fragments</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.tsse.add_tsse_score">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.tsse.</span></span><span class="sig-name descname"><span class="pre">add_tsse_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fragments</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gtf</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">negativ_shift</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">positiv_shift</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_size_total</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_size_per_base</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_bias</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_tmp</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/tsse.html#add_tsse_score"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.tsse.add_tsse_score" title="Link to this definition"></a></dt>
<dd><p>Add the tSSe score to the adata object.</p>
<p>Calculating the TSSe score is done like described in: “Chromatin accessibility profiling by ATAC-seq” Fiorella et al. 2022</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – AnnData object</p></li>
<li><p><strong>fragments</strong> (<em>str</em>) – path to fragments.bed</p></li>
<li><p><strong>gtf</strong> (<em>str</em>) – path to gtf file</p></li>
<li><p><strong>negativ_shift</strong> (<em>int</em><em>, </em><em>default 2000</em>) – number of bases to shift upstream for the flanking regions</p></li>
<li><p><strong>positiv_shift</strong> (<em>int</em><em>, </em><em>default 2000</em>) – number of bases to shift downstream for the flanking regions</p></li>
<li><p><strong>edge_size_total</strong> (<em>int</em><em>, </em><em>default 100</em>) – number of bases to use for the edges for the global tSSe score</p></li>
<li><p><strong>edge_size_per_base</strong> (<em>int</em><em>, </em><em>default 50</em>) – number of bases to use for the edges for the per base tSSe score</p></li>
<li><p><strong>min_bias</strong> (<em>float</em><em>, </em><em>default 0.01</em>) – minimum bias to avoid division by zero</p></li>
<li><p><strong>keep_tmp</strong> (<em>bool</em><em>, </em><em>default False</em>) – keep temporary files</p></li>
<li><p><strong>temp_dir</strong> (<em>str</em><em>, </em><em>default &quot;&quot;</em>) – path to temporary directory</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>AnnData object with added tSSe score</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>sc.AnnData</p>
</dd>
</dl>
</dd></dl>

</section>
<hr class="docutils" />
<section id="module-sctoolbox.tools.insertsize">
<span id="insertsize"></span><h3>insertsize<a class="headerlink" href="#module-sctoolbox.tools.insertsize" title="Link to this heading"></a></h3>
<p>Tools to calculate fragemnt- and insertsize for scATAC.</p>
<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.insertsize.add_insertsize">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.insertsize.</span></span><span class="sig-name descname"><span class="pre">add_insertsize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bam</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fragments</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">barcode_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">barcode_tag</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'CB'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regions</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/insertsize.html#add_insertsize"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.insertsize.add_insertsize" title="Link to this definition"></a></dt>
<dd><p>Add information on insertsize to the adata object using either a .bam-file or a fragments file.</p>
<p>Adds columns “insertsize_count” and “mean_insertsize” to adata.obs and a key “insertsize_distribution” to adata.uns containing the
insertsize distribution as a pandas dataframe.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – AnnData object to add insertsize information to.</p></li>
<li><p><strong>bam</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Path to bam file containing paired-end reads. If None, the fragments file is used instead.</p></li>
<li><p><strong>fragments</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Path to fragments file containing fragments information. If None, the bam file is used instead.</p></li>
<li><p><strong>barcode_col</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Column in adata.obs containing the name of the cell barcode. If barcode_col is None, it is assumed that the index of adata.obs contains the barcode.</p></li>
<li><p><strong>barcode_tag</strong> (<em>str</em><em>, </em><em>default 'CB'</em>) – Only for bamfiles: Tag used for the cell barcode for each read.</p></li>
<li><p><strong>regions</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Only for bamfiles: A list of regions to obtain reads from, e.g. [‘chr1:1-2000000’]. If None, all reads in the .bam-file are used.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError:</strong> – <ol class="arabic simple">
<li><p>If bam and fragments is given.
    2. If bam and fragments is not given.
    3. If no barcodes between bam- or fragment-file and adata overlap</p></li>
</ol>
</p>
</dd>
</dl>
</dd></dl>

</section>
<hr class="docutils" />
<section id="module-sctoolbox.tools.peak_annotation">
<span id="peak-annotation"></span><h3>peak_annotation<a class="headerlink" href="#module-sctoolbox.tools.peak_annotation" title="Link to this heading"></a></h3>
<p>Tools for peak annotation required by scATAC-seq.</p>
<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.peak_annotation.annotate_adata">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.peak_annotation.</span></span><span class="sig-name descname"><span class="pre">annotate_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gtf</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">best</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coordinate_cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_temp</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/peak_annotation.html#annotate_adata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.peak_annotation.annotate_adata" title="Link to this definition"></a></dt>
<dd><p>Annotate adata .var features with genes from .gtf using UROPA <a class="footnote-reference brackets" href="#id4" id="id3" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a>.</p>
<p>The function assumes that the adata.var contains genomic coordinates in the first
three columns, e.g. “chromosome”, “start”, “end”.
If specific columns should be used, please adjust the names via ‘coordinate_cols’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – The anndata object containing features to annotate.</p></li>
<li><p><strong>gtf</strong> (<em>str</em>) – Path to .gtf file containing genomic elements for annotation.</p></li>
<li><p><strong>config</strong> (<em>Optional</em><em>[</em><em>dict</em><em>[</em><em>str</em><em>, </em><em>Any</em><em>]</em><em>]</em><em>, </em><em>default None</em>) – A dictionary indicating how regions should be annotated.
Default (None) is to annotate feature ‘gene’ within -10000;1000bp of the gene start.
See ‘Examples’ of how to set up a custom configuration dictionary.</p></li>
<li><p><strong>best</strong> (<em>bool</em><em>, </em><em>default True</em>) – Whether to return the best annotation or all valid annotations.</p></li>
<li><p><strong>threads</strong> (<em>int</em><em>, </em><em>default 1</em>) – Number of threads to use for multiprocessing.</p></li>
<li><p><strong>coordinate_cols</strong> (<em>Optional</em><em>[</em><em>list</em><em>[</em><em>str</em><em>]</em><em>]</em><em>, </em><em>default None</em>) – A list of column names in the regions DataFrame that contain the chromosome, start and end coordinates.
If None the first three columns are taken.</p></li>
<li><p><strong>temp_dir</strong> (<em>str</em><em>, </em><em>default ''</em>) – Path to a directory to store files. Is only used if input .gtf-file needs sorting.</p></li>
<li><p><strong>remove_temp</strong> (<em>boolean</em><em>, </em><em>default True</em>) – If True remove temporary directory after execution.</p></li>
<li><p><strong>inplace</strong> (<em>boolean</em><em>, </em><em>default True</em>) – Whether to add the annotations to the adata object in place.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>If inplace == True, the annotation is added to adata.var in place.
Else, a copy of the adata object is returned with the annotations added.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Optional[sc.AnnData]</p>
</dd>
</dl>
<p class="rubric">References</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id4" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id3">1</a><span class="fn-bracket">]</span></span>
<p>Kondili M, Fust A, Preussner J, Kuenne C, Braun T, and Looso M. UROPA: a tool for Universal RObust Peak Annotation. Scientific Reports 7 (2017), doi: 10.1038/s41598-017-02464-y</p>
</aside>
</aside>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">custom_config</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;queries&quot;</span><span class="p">:</span> <span class="p">[{</span><span class="s2">&quot;distance&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">10000</span><span class="p">,</span> <span class="mi">1000</span><span class="p">],</span>
<span class="go">                                  &quot;feature_anchor&quot;: &quot;start&quot;,</span>
<span class="go">                                  &quot;feature&quot;: &quot;gene&quot;}],</span>
<span class="go">                             &quot;priority&quot;: True,</span>
<span class="go">                             &quot;show_attributes&quot;: &quot;all&quot;}</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">annotate_regions</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">gtf</span><span class="o">=</span><span class="s2">&quot;genes.gtf&quot;</span><span class="p">,</span>
<span class="go">                            config=custom_config)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.peak_annotation.annotate_narrowPeak">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.peak_annotation.</span></span><span class="sig-name descname"><span class="pre">annotate_narrowPeak</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gtf</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">best</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_temp</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/peak_annotation.html#annotate_narrowPeak"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.peak_annotation.annotate_narrowPeak" title="Link to this definition"></a></dt>
<dd><p>Annotate narrowPeak files with genes from .gtf using UROPA.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filepath</strong> (<em>str</em>) – Path to the narrowPeak file to be annotated.</p></li>
<li><p><strong>gtf</strong> (<em>str</em>) – Path to the .gtf file containing the genes to be annotated.</p></li>
<li><p><strong>config</strong> (<em>Optional</em><em>[</em><em>dict</em><em>[</em><em>str</em><em>, </em><em>Any</em><em>]</em><em>]</em><em>, </em><em>default None</em>) – A dictionary indicating how regions should be annotated.
Default (None) is to annotate feature ‘gene’ within -10000;1000bp of the gene start.
See ‘Examples’ of how to set up a custom configuration dictionary.</p></li>
<li><p><strong>best</strong> (<em>bool</em><em>, </em><em>default True</em>) – Whether to return the best annotation or all valid annotations.</p></li>
<li><p><strong>threads</strong> (<em>int</em><em>, </em><em>default 1</em>) – Number of threads to perform the annotation.</p></li>
<li><p><strong>temp_dir</strong> (<em>str</em><em>, </em><em>default ''</em>) – Path to the directory where the temporary files should be written.</p></li>
<li><p><strong>remove_temp</strong> (<em>bool</em><em>, </em><em>default True</em>) – If True remove temporary directory after execution.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dataframe containing the annotations.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

</section>
<hr class="docutils" />
<section id="module-sctoolbox.tools.receptor_ligand">
<span id="receptor-ligand"></span><h3>receptor_ligand<a class="headerlink" href="#module-sctoolbox.tools.receptor_ligand" title="Link to this heading"></a></h3>
<p>Tools for a receptor-ligand analysis.</p>
<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.receptor_ligand.download_db">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.receptor_ligand.</span></span><span class="sig-name descname"><span class="pre">download_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ligand_column</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">receptor_column</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'\t'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/receptor_ligand.html#download_db"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.receptor_ligand.download_db" title="Link to this definition"></a></dt>
<dd><p>Download table of receptor-ligand interactions and store in adata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Analysis object the database will be added to.</p></li>
<li><p><strong>db_path</strong> (<em>str</em>) – Path to database table. A valid database needs a column with receptor gene ids/symbols and ligand gene ids/symbols.
Human: <a class="reference external" href="http://tcm.zju.edu.cn/celltalkdb/download/processed_data/human_lr_pair.txt">http://tcm.zju.edu.cn/celltalkdb/download/processed_data/human_lr_pair.txt</a>
Mouse: <a class="reference external" href="http://tcm.zju.edu.cn/celltalkdb/download/processed_data/mouse_lr_pair.txt">http://tcm.zju.edu.cn/celltalkdb/download/processed_data/mouse_lr_pair.txt</a></p></li>
<li><p><strong>ligand_column</strong> (<em>str</em>) – Name of the column with ligand gene names.
Use ‘ligand_gene_symbol’ for the urls provided above.</p></li>
<li><p><strong>receptor_column</strong> (<em>str</em>) – Name of column with receptor gene names.
Use ‘receptor_gene_symbol’ for the urls provided above.</p></li>
<li><p><strong>sep</strong> (<em>str</em><em>, </em><em>default 't'</em>) – Separator of database table.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to copy <cite>adata</cite> or modify it inplace.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True will overwrite existing database.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This will remove all information stored in adata.uns[‘receptor-ligand’]</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>If not inplace, return copy of adata with added database path and
database table to adata.uns[‘receptor-ligand’]</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>Optional[sc.AnnData]</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError:</strong> – 1: If ligand_column is not in database.
    2: If receptor_column is not in database.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.receptor_ligand.calculate_interaction_table">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.receptor_ligand.</span></span><span class="sig-name descname"><span class="pre">calculate_interaction_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_column</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_index</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/receptor_ligand.html#calculate_interaction_table"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.receptor_ligand.calculate_interaction_table" title="Link to this definition"></a></dt>
<dd><p>Calculate an interaction table of the clusters defined in adata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – AnnData object that holds the expression values and clustering</p></li>
<li><p><strong>cluster_column</strong> (<em>str</em>) – Name of the cluster column in adata.obs.</p></li>
<li><p><strong>gene_index</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Column in adata.var that holds gene symbols/ ids.
Corresponds to <cite>download_db(ligand_column, receptor_column)</cite>.
Uses index when None.</p></li>
<li><p><strong>normalize</strong> (<em>int</em><em>, </em><em>default 1000</em>) – Correct clusters to given size.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to copy <cite>adata</cite> or modify it inplace.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True will overwrite existing interaction table.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>If not inpalce, return copy of adata with added interactions table to adata.uns[‘receptor-ligand’][‘interactions’]</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Optional[sc.AnnData]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError:</strong> – 1: If receptor-ligand database cannot be found.
    2: Id database genes do not match adata genes.</p></li>
<li><p><strong>Exception:</strong> – If not interactions were found.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.receptor_ligand.interaction_violin_plot">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.receptor_ligand.</span></span><span class="sig-name descname"><span class="pre">interaction_violin_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_perc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">(5,</span> <span class="pre">20)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">100</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">_SupportsArray</span><span class="p"><span class="pre">[</span></span><span class="pre">dtype</span><span class="p"><span class="pre">[</span></span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_NestedSequence</span><span class="p"><span class="pre">[</span></span><span class="pre">_SupportsArray</span><span class="p"><span class="pre">[</span></span><span class="pre">dtype</span><span class="p"><span class="pre">[</span></span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">complex</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bytes</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_NestedSequence</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">complex</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bytes</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/receptor_ligand.html#interaction_violin_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.receptor_ligand.interaction_violin_plot" title="Link to this definition"></a></dt>
<dd><p>Generate violin plot of pairwise cluster interactions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – AnnData object</p></li>
<li><p><strong>min_perc</strong> (<em>int</em><em> | </em><em>float</em>) – Minimum percentage of cells in a cluster that express the respective gene. A value from 0-100.</p></li>
<li><p><strong>save</strong> (<em>str</em><em>, </em><em>default None</em>) – Output filename. Uses the internal ‘sctoolbox.settings.figure_dir’.</p></li>
<li><p><strong>figsize</strong> (<em>int tuple</em><em>, </em><em>default</em><em> (</em><em>5</em><em>, </em><em>20</em><em>)</em>) – Figure size</p></li>
<li><p><strong>dpi</strong> (<em>float</em><em>, </em><em>default 100</em>) – The resolution of the figure in dots-per-inch.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Object containing all plots. As returned by matplotlib.pyplot.subplots</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>npt.ArrayLike</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.receptor_ligand.hairball">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.receptor_ligand.</span></span><span class="sig-name descname"><span class="pre">hairball</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_perc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interaction_score</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interaction_perc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Network'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_min</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_max</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cbar_label</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Interaction</span> <span class="pre">count'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_count</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">restrict_to</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">additional_nodes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hide_edges</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">_SupportsArray</span><span class="p"><span class="pre">[</span></span><span class="pre">dtype</span><span class="p"><span class="pre">[</span></span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_NestedSequence</span><span class="p"><span class="pre">[</span></span><span class="pre">_SupportsArray</span><span class="p"><span class="pre">[</span></span><span class="pre">dtype</span><span class="p"><span class="pre">[</span></span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">complex</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bytes</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_NestedSequence</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">complex</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bytes</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/receptor_ligand.html#hairball"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.receptor_ligand.hairball" title="Link to this definition"></a></dt>
<dd><p>Generate network graph of interactions between clusters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – AnnData object</p></li>
<li><p><strong>min_perc</strong> (<em>int</em><em> | </em><em>float</em>) – Minimum percentage of cells in a cluster that express the respective gene. A value from 0-100.</p></li>
<li><p><strong>interaction_score</strong> (<em>float</em><em> | </em><em>int</em><em>, </em><em>default 0</em>) – Interaction score must be above this threshold for the interaction to be counted in the graph.</p></li>
<li><p><strong>interaction_perc</strong> (<em>Optional</em><em>[</em><em>int</em><em> | </em><em>float</em><em>]</em><em>, </em><em>default None</em>) – Select interaction scores above or equal to the given percentile. Will overwrite parameter interaction_score. A value from 0-100.</p></li>
<li><p><strong>save</strong> (<em>str</em><em>, </em><em>default None</em>) – Output filename. Uses the internal ‘sctoolbox.settings.figure_dir’.</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>default 'Network'</em>) – The plots title.</p></li>
<li><p><strong>color_min</strong> (<em>float</em><em>, </em><em>default 0</em>) – Min value for color range.</p></li>
<li><p><strong>color_max</strong> (<em>Optional</em><em>[</em><em>float</em><em> | </em><em>int</em><em>]</em><em>, </em><em>default None</em>) – Max value for color range.</p></li>
<li><p><strong>cbar_label</strong> (<em>str</em><em>, </em><em>default 'Interaction count'</em>) – Label above the colorbar.</p></li>
<li><p><strong>show_count</strong> (<em>bool</em><em>, </em><em>default False</em>) – Show the interaction count in the hairball.</p></li>
<li><p><strong>restrict_to</strong> (<em>Optional</em><em>[</em><em>list</em><em>[</em><em>str</em><em>]</em><em>]</em><em>, </em><em>default None</em>) – Only show given clusters provided in list.</p></li>
<li><p><strong>additional_nodes</strong> (<em>Optional</em><em>[</em><em>list</em><em>[</em><em>str</em><em>]</em><em>]</em><em>, </em><em>default None</em>) – List of additional node names displayed in the hairball.</p></li>
<li><p><strong>hide_edges</strong> (<em>Optional</em><em>[</em><em>list</em><em>[</em><em>Tuple</em><em>[</em><em>str</em><em>, </em><em>str</em><em>]</em><em>]</em><em>]</em><em>, </em><em>default None</em>) – List of tuples with node names that should not have an edge shown. Order doesn’t matter. E.g. <cite>[(“a”, “b”)]</cite> to omit the edge between node a and b.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Object containing all plots. As returned by matplotlib.pyplot.subplots</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>npt.ArrayLike</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError:</strong> – If restrict_to contains invalid clusters.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.receptor_ligand.progress_violins">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.receptor_ligand.</span></span><span class="sig-name descname"><span class="pre">progress_violins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">datalist</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">DataFrame</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datalabel</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_a</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_b</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_perc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">(12,</span> <span class="pre">6)</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/receptor_ligand.html#progress_violins"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.receptor_ligand.progress_violins" title="Link to this definition"></a></dt>
<dd><p>Show cluster interactions over timepoints.</p>
<p>CURRENTLY NOT FUNCTIONAL!</p>
<p>TODO Implement function</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>datalist</strong> (<em>list</em><em>[</em><em>pd.DataFrame</em><em>]</em>) – List of interaction DataFrames. Each DataFrame represents a timepoint.</p></li>
<li><p><strong>datalabel</strong> (<em>list</em><em>[</em><em>str</em><em>]</em>) – List of strings. Used to label the violins.</p></li>
<li><p><strong>cluster_a</strong> (<em>str</em>) – Name of the first interacting cluster.</p></li>
<li><p><strong>cluster_b</strong> (<em>str</em>) – Name of the second interacting cluster.</p></li>
<li><p><strong>min_perc</strong> (<em>float</em><em> | </em><em>int</em>) – Minimum percentage of cells in a cluster each gene must be expressed in.</p></li>
<li><p><strong>save</strong> (<em>str</em>) – Path to output file.</p></li>
<li><p><strong>figsize</strong> (<em>Tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em><em>, </em><em>default</em><em> (</em><em>12</em><em>, </em><em>6</em><em>)</em>) – Tuple of plot (width, height).</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.receptor_ligand.interaction_progress">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.receptor_ligand.</span></span><span class="sig-name descname"><span class="pre">interaction_progress</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">datalist</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">AnnData</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datalabel</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">receptor</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ligand</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">receptor_cluster</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ligand_cluster</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">(4,</span> <span class="pre">4)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Axes</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/receptor_ligand.html#interaction_progress"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.receptor_ligand.interaction_progress" title="Link to this definition"></a></dt>
<dd><p>Barplot that shows the interaction score of a single interaction between two given clusters over multiple datasets.</p>
<p>TODO add checks &amp; error messages</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>datalist</strong> (<em>list</em><em>[</em><em>sc.AnnData</em><em>]</em>) – List of anndata objects.</p></li>
<li><p><strong>datalabel</strong> (<em>list</em><em>[</em><em>str</em><em>]</em>) – List of labels for the given datalist.</p></li>
<li><p><strong>receptor</strong> (<em>str</em>) – Name of the receptor gene.</p></li>
<li><p><strong>ligand</strong> (<em>str</em>) – Name of the ligand gene.</p></li>
<li><p><strong>receptor_cluster</strong> (<em>str</em>) – Name of the receptor cluster.</p></li>
<li><p><strong>ligand_cluster</strong> (<em>str</em>) – Name of the ligand cluster.</p></li>
<li><p><strong>figsize</strong> (<em>Tuple</em><em>[</em><em>int</em><em> | </em><em>float</em><em>, </em><em>int</em><em> | </em><em>float</em><em>]</em><em>, </em><em>default</em><em> (</em><em>4</em><em>, </em><em>4</em><em>)</em>) – Figure size in inch.</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>default 100</em>) – Dots per inch.</p></li>
<li><p><strong>save</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Output filename. Uses the internal ‘sctoolbox.settings.figure_dir’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The plotting object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>matplotlib.axes.Axes</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.receptor_ligand.connectionPlot">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.receptor_ligand.</span></span><span class="sig-name descname"><span class="pre">connectionPlot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">restrict_to</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">(10,</span> <span class="pre">15)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">connection_alpha</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'interaction_score'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">receptor_cluster_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'receptor_cluster'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">receptor_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'receptor_gene'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">receptor_hue</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'receptor_score'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">receptor_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'receptor_percent'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">receptor_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ligand_cluster_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'ligand_cluster'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ligand_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'ligand_gene'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ligand_hue</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'ligand_score'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ligand_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'ligand_percent'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ligand_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lw_multiplier</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wspace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">line_colors</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'rainbow'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">_SupportsArray</span><span class="p"><span class="pre">[</span></span><span class="pre">dtype</span><span class="p"><span class="pre">[</span></span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_NestedSequence</span><span class="p"><span class="pre">[</span></span><span class="pre">_SupportsArray</span><span class="p"><span class="pre">[</span></span><span class="pre">dtype</span><span class="p"><span class="pre">[</span></span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">complex</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bytes</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">_NestedSequence</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">complex</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bytes</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/receptor_ligand.html#connectionPlot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.receptor_ligand.connectionPlot" title="Link to this definition"></a></dt>
<dd><p>Show specific receptor-ligand connections between clusters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – AnnData object</p></li>
<li><p><strong>restrict_to</strong> (<em>Optional</em><em>[</em><em>list</em><em>[</em><em>str</em><em>]</em><em>]</em><em>, </em><em>default None</em>) – Restrict plot to given cluster names.</p></li>
<li><p><strong>figsize</strong> (<em>Tuple</em><em>[</em><em>int</em><em> | </em><em>float</em><em>, </em><em>int</em><em> | </em><em>float</em><em>]</em><em>, </em><em>default</em><em> (</em><em>10</em><em>, </em><em>15</em><em>)</em>) – Figure size</p></li>
<li><p><strong>dpi</strong> (<em>float</em><em>, </em><em>default 100</em>) – The resolution of the figure in dots-per-inch.</p></li>
<li><p><strong>connection_alpha</strong> (<em>str</em><em>, </em><em>default 'interaction_score'</em>) – Name of column that sets alpha value of lines between plots. None to disable.</p></li>
<li><p><strong>save</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Output filename. Uses the internal ‘sctoolbox.settings.figure_dir’.</p></li>
<li><p><strong>title</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Title of the plot</p></li>
<li><p><strong>receptor_cluster_col</strong> (<em>str</em><em>, </em><em>default 'receptor_cluster'</em>) – Name of column containing cluster names of receptors. Shown on x-axis.</p></li>
<li><p><strong>receptor_col</strong> (<em>str</em><em>, </em><em>default 'receptor_gene'</em>) – Name of column containing gene names of receptors. Shown on y-axis.</p></li>
<li><p><strong>receptor_hue</strong> (<em>str</em><em>, </em><em>default 'receptor_score'</em>) – Name of column containing receptor scores. Shown as point color.</p></li>
<li><p><strong>receptor_size</strong> (<em>str</em><em>, </em><em>default 'receptor_percent'</em>) – Name of column containing receptor expression percentage. Shown as point size.</p></li>
<li><p><strong>receptor_genes</strong> (<em>Optional</em><em>[</em><em>list</em><em>[</em><em>str</em><em>]</em><em>]</em><em>, </em><em>default None</em>) – Restrict receptors to given genes.</p></li>
<li><p><strong>ligand_cluster_col</strong> (<em>str</em><em>, </em><em>default 'ligand_cluster'</em>) – Name of column containing cluster names of ligands. Shown on x-axis.</p></li>
<li><p><strong>ligand_col</strong> (<em>str</em><em>, </em><em>default 'ligand_gene'</em>) – Name of column containing gene names of ligands. Shown on y-axis.</p></li>
<li><p><strong>ligand_hue</strong> (<em>str</em><em>, </em><em>default 'ligand_score'</em>) – Name of column containing ligand scores. Shown as point color.</p></li>
<li><p><strong>ligand_size</strong> (<em>str</em><em>, </em><em>default 'ligand_percent'</em>) – Name of column containing ligand expression percentage. Shown as point size.</p></li>
<li><p><strong>ligand_genes</strong> (<em>Optional</em><em>[</em><em>list</em><em>[</em><em>str</em><em>]</em><em>]</em><em>, </em><em>default None</em>) – Restrict ligands to given genes.</p></li>
<li><p><strong>filter</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Conditions to filter the interaction table on. E.g. ‘column_name &gt; 5 &amp; other_column &lt; 2’. Forwarded to pandas.DataFrame.query.</p></li>
<li><p><strong>lw_multiplier</strong> (<em>int</em><em> | </em><em>float</em><em>, </em><em>default 2</em>) – Linewidth multiplier.</p></li>
<li><p><strong>wspace</strong> (<em>float</em><em>, </em><em>default 0.4</em>) – Width between plots. Fraction of total width.</p></li>
<li><p><strong>line_colors</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default 'rainbow'</em>) – Name of colormap used to color lines. All lines are black if None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Object containing all plots. As returned by matplotlib.pyplot.subplots</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>npt.ArrayLike</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>Exception:</strong> – If no onteractions between clsuters are found.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.receptor_ligand.get_interactions">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.receptor_ligand.</span></span><span class="sig-name descname"><span class="pre">get_interactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">anndata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_perc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interaction_score</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interaction_perc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_a</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_b</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/receptor_ligand.html#get_interactions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.receptor_ligand.get_interactions" title="Link to this definition"></a></dt>
<dd><p>Get interaction table from anndata and apply filters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>anndata</strong> (<em>sc.AnnData</em>) – Anndata object to pull interaction table from.</p></li>
<li><p><strong>min_perc</strong> (<em>Optional</em><em>[</em><em>float</em><em> | </em><em>int</em><em>]</em><em>, </em><em>default None</em>) – Minimum percent of cells in a cluster that express the ligand/ receptor gene. Value from 0-100.</p></li>
<li><p><strong>interaction_score</strong> (<em>Optional</em><em>[</em><em>float</em><em> | </em><em>int</em><em>]</em><em>, </em><em>default None</em>) – Filter receptor-ligand interactions below given score. Ignored if <cite>interaction_perc</cite> is set.</p></li>
<li><p><strong>interaction_perc</strong> (<em>Optional</em><em>[</em><em>float</em><em> | </em><em>int</em><em>]</em><em>, </em><em>default None</em>) – Filter receptor-ligand interactions below the given percentile. Overwrite <cite>interaction_score</cite>. Value from 0-100.</p></li>
<li><p><strong>group_a</strong> (<em>Optional</em><em>[</em><em>list</em><em>[</em><em>str</em><em>]</em><em>]</em><em>, </em><em>default None</em>) – List of cluster names that must be present in any given receptor-ligand interaction.</p></li>
<li><p><strong>group_b</strong> (<em>Optional</em><em>[</em><em>list</em><em>[</em><em>str</em><em>]</em><em>]</em><em>, </em><em>default None</em>) – List of cluster names that must be present in any given receptor-ligand interaction.</p></li>
<li><p><strong>save</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Output filename. Uses the internal ‘sctoolbox.settings.table_dir’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>Table that contains interactions. Columns:</p>
<blockquote>
<div><ul class="simple">
<li><p>receptor_cluster      = name of the receptor cluster</p></li>
<li><p>ligand_cluster        = name of the ligand cluster</p></li>
<li><p>receptor_gene         = name of the receptor gene</p></li>
<li><p>ligand_gene           = name of the ligand gene</p></li>
<li><p>receptor_score        = zscore of receptor gene cluster mean expression (scaled by cluster size)</p></li>
<li><p>ligand_score          = zscore of ligand gene cluster mean expression (scaled by cluster size)</p></li>
<li><p>receptor_percent      = percent of cells in cluster expressing receptor gene</p></li>
<li><p>ligand_percent        = percent of cells in cluster expressing ligand gene</p></li>
<li><p>receptor_cluster_size = number of cells in receptor cluster</p></li>
<li><p>ligand_cluster_size   = number of cells in ligand cluster</p></li>
<li><p>interaction_score     = sum of receptor_score and ligand_score</p></li>
</ul>
</div></blockquote>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

</section>
<hr class="docutils" />
<section id="module-sctoolbox.tools.gene_correlation">
<span id="gene-correlation"></span><h3>gene_correlation<a class="headerlink" href="#module-sctoolbox.tools.gene_correlation" title="Link to this heading"></a></h3>
<p>Tools for gene-gene correlation.</p>
<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.gene_correlation.correlate_conditions">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.gene_correlation.</span></span><span class="sig-name descname"><span class="pre">correlate_conditions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condition_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condition_A</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condition_B</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/gene_correlation.html#correlate_conditions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.gene_correlation.correlate_conditions" title="Link to this definition"></a></dt>
<dd><p>Calculate the correlation of a gene expression over two conditions and compares the two conditions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Annotated adata object.</p></li>
<li><p><strong>gene</strong> (<em>str</em>) – Gene of interest.</p></li>
<li><p><strong>condition_col</strong> (<em>str</em>) – Column in adata.obs containing conditions.</p></li>
<li><p><strong>condition_A</strong> (<em>str</em>) – Name of the first condition.</p></li>
<li><p><strong>condition_B</strong> (<em>str</em>) – Name of the second condition.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dataframe containing the correlation of a gene expression over two conditions.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If one or both condition columns are not in adata.obs.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.gene_correlation.correlate_ref_vs_all">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.gene_correlation.</span></span><span class="sig-name descname"><span class="pre">correlate_ref_vs_all</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correlation_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/gene_correlation.html#correlate_ref_vs_all"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.gene_correlation.correlate_ref_vs_all" title="Link to this definition"></a></dt>
<dd><p>Calculate the correlation of the reference gene vs all other genes.</p>
<p>Additionally, plots umap highlighting correlating gene expression.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>ref_gene</strong> (<em>str</em>) – Reference gene.</p></li>
<li><p><strong>correlation_threshold</strong> (<em>float</em><em>, </em><em>default 0.4</em>) – Threshold for correlation value. Correlating genes with a correlation &gt;= threshold will be plotted
if plot parameter is set to True.</p></li>
<li><p><strong>save</strong> (<em>str</em><em>, </em><em>default None</em>) – Path to save the figure to.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dataframe containing correlation of refrence gene to other genes.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.gene_correlation.compare_two_conditons">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.gene_correlation.</span></span><span class="sig-name descname"><span class="pre">compare_two_conditons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_cond_A</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_cond_B</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cells_A</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cells_B</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/gene_correlation.html#compare_two_conditons"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.gene_correlation.compare_two_conditons" title="Link to this definition"></a></dt>
<dd><p>Compare two conditions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_cond_A</strong> (<em>pd.DataFrame</em>) – Dataframe containing correlation analysis from correlate_ref_vs_all</p></li>
<li><p><strong>df_cond_B</strong> (<em>pd.DataFrame</em>) – Dataframe containing correlation analysis from correlate_ref_vs_all</p></li>
<li><p><strong>n_cells_A</strong> (<em>int</em>) – Number of cells within condition A</p></li>
<li><p><strong>n_cells_B</strong> (<em>int</em>) – Number of cells within condition B</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dataframe containing single correlation and Fischer Z transformation</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

</section>
<hr class="docutils" />
<section id="module-sctoolbox.tools.tobias">
<span id="tobias"></span><h3>tobias<a class="headerlink" href="#module-sctoolbox.tools.tobias" title="Link to this heading"></a></h3>
<p>Tools for TOBIAS usage.</p>
<dl class="py function">
<dt class="sig sig-object py" id="sctoolbox.tools.tobias.write_TOBIAS_config">
<span class="sig-prename descclassname"><span class="pre">sctoolbox.tools.tobias.</span></span><span class="sig-name descname"><span class="pre">write_TOBIAS_config</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">out_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bams</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fasta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">blacklist</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gtf</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motifs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">organism</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'human'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'mouse'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'zebrafish'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'human'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'TOBIAS_output'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_comparison</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_correction</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_venn</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coverage</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wilson</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/sctoolbox/tools/tobias.html#write_TOBIAS_config"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sctoolbox.tools.tobias.write_TOBIAS_config" title="Link to this definition"></a></dt>
<dd><p>Write a TOBIAS config file from input bams/fasta/blacklist etc.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>out_path</strong> (<em>str</em>) – Path to output yaml file.</p></li>
<li><p><strong>bams</strong> (<em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default</em><em> [</em><em>]</em>) – List of paths to bam files.</p></li>
<li><p><strong>names</strong> (<em>Optional</em><em>[</em><em>list</em><em>[</em><em>str</em><em>]</em><em>]</em><em>, </em><em>default None</em>) – List of names for the bams. If None, the names are set to the bam file names with common prefix and suffix removed.</p></li>
<li><p><strong>fasta</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Path to fasta file.</p></li>
<li><p><strong>blacklist</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Path to blacklist file.</p></li>
<li><p><strong>gtf</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Path to gtf file.</p></li>
<li><p><strong>motifs</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>default None</em>) – Path to motif file.</p></li>
<li><p><strong>organism</strong> (<em>Literal</em><em>[</em><em>&quot;human&quot;</em><em>, </em><em>&quot;mouse&quot;</em><em>, </em><em>&quot;zebrafish&quot;</em><em>]</em><em>, </em><em>default 'human'</em>) – Organism name. TOBIAS supports ‘human’, ‘mouse’ or ‘zebrafish’.</p></li>
<li><p><strong>output</strong> (<em>str</em><em>, </em><em>default 'Tobias_output'</em>) – Output directory of the TOBIAS run.</p></li>
<li><p><strong>plot_comparison</strong> (<em>bool</em><em>, </em><em>default True</em>) – Flag for the step of plotting comparison of the TOBIAS run.</p></li>
<li><p><strong>plot_correction</strong> (<em>bool</em><em>, </em><em>default True</em>) – Flag for the step of plotting correction of the TOBIAS run.</p></li>
<li><p><strong>plot_venn</strong> (<em>bool</em><em>, </em><em>default True</em>) – Flag for the step of plotting venn diagramms of the TOBIAS run.</p></li>
<li><p><strong>coverage</strong> (<em>bool</em><em>, </em><em>default True</em>) – Flag for coverage step of the TOBIAS run.</p></li>
<li><p><strong>wilson</strong> (<em>bool</em><em>, </em><em>default True</em>) – Flag for wilson step of the TOBIAS run.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="sctoolbox API reference" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="plotting.html" class="btn btn-neutral float-right" title="Plotting" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Loosolab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>