<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>sctoolbox.plotting.genometracks &mdash; SC FRAMEWORK  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../../../_static/plot_directive.css" />

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../../../_static/doctools.js?v=9a2dae69"></script>
        <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            SC FRAMEWORK
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../CHANGES.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../rna-notebooks/index.html">RNA notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../atac-notebooks/index.html">ATAC notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../general-notebooks/index.html">General notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../API/index.html">sctoolbox API reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">SC FRAMEWORK</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">sctoolbox.plotting.genometracks</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for sctoolbox.plotting.genometracks</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;Class to create a genome track plot via pyGenomeTracks.&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">tempfile</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="kn">from</span> <span class="nn">beartype</span> <span class="kn">import</span> <span class="n">beartype</span>
<span class="kn">from</span> <span class="nn">beartype.typing</span> <span class="kn">import</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Any</span><span class="p">,</span> <span class="n">Literal</span><span class="p">,</span> <span class="n">Iterable</span>

<span class="kn">import</span> <span class="nn">sctoolbox.utils</span> <span class="k">as</span> <span class="nn">utils</span>
<span class="kn">from</span> <span class="nn">sctoolbox._settings</span> <span class="kn">import</span> <span class="n">settings</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">settings</span><span class="o">.</span><span class="n">logger</span>


<div class="viewcode-block" id="GenomeTracks">
<a class="viewcode-back" href="../../../API/plotting.html#sctoolbox.plotting.genometracks.GenomeTracks">[docs]</a>
<span class="nd">@beartype</span>
<span class="k">class</span> <span class="nc">GenomeTracks</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Class for creating a genome track plot via pyGenomeTracks by collecting different tracks and writing the .ini file.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>
<span class="sd">        :nofigs:</span>

<span class="sd">        import sctoolbox.plotting as pl</span>

<span class="sd">        G = pl.GenomeTracks()</span>

<span class="sd">        #Add bigwig tracks</span>
<span class="sd">        G.add_track(&quot;data/tracks/bigwig1.bw&quot;, color=&quot;red&quot;)</span>
<span class="sd">        G.add_track(&quot;data/tracks/bigwig2.bw&quot;, color=&quot;blue&quot;, orientation=&quot;inverted&quot;)</span>

<span class="sd">        #Add hlines to previous bigwig track</span>
<span class="sd">        G.add_hlines([100, 200], color=&quot;red&quot;)</span>
<span class="sd">        G.add_hlines([250], color=&quot;blue&quot;, line_style=&quot;dashed&quot;)</span>

<span class="sd">        #Add links</span>
<span class="sd">        G.add_track(&quot;data/tracks/links.arcs&quot;, orientation=&quot;inverted&quot;)</span>

<span class="sd">        #Add one line between tracks</span>
<span class="sd">        G.add_hline()</span>

<span class="sd">        #Add .bed-file regions</span>
<span class="sd">        G.add_track(&quot;data/tracks/tad_classification.bed&quot;, title=&quot;bed&quot;)</span>
<span class="sd">        G.add_track(&quot;data/tracks/tad_classification.bed&quot;, color=&quot;Reds&quot;, title=&quot;bed colored by score column&quot;)</span>

<span class="sd">        #Add vlines and highlight</span>
<span class="sd">        G.add_track(&quot;data/tracks/vlines.bed&quot;, file_type=&quot;vlines&quot;)</span>
<span class="sd">        G.add_track(&quot;data/tracks/vhighlight.bed&quot;, file_type=&quot;vhighlight&quot;)</span>

<span class="sd">        #Add a spacer</span>
<span class="sd">        G.add_spacer()</span>

<span class="sd">        #Add genes</span>
<span class="sd">        G.add_track(&quot;data/tracks/genes.gtf&quot;, gene_rows=5)</span>

<span class="sd">        #Add x-axis</span>
<span class="sd">        G.add_spacer()</span>
<span class="sd">        G.add_xaxis()</span>

<span class="sd">        # Plot</span>
<span class="sd">        G.plot(region=&quot;X:3000000-3500000&quot;, output=&quot;genometrack_X.png&quot;, trackLabelFraction=0.2)</span>

<span class="sd">    .. image:: genometrack_X.png</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Initialize the GenomeTracks object.&quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">tracks</span> <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># dictionary of tracks</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">type_count</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">global_defaults</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;height&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">}</span>  <span class="c1"># dictionary of default values</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">type_defaults</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;gtf&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;merge_transcripts&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span> <span class="s2">&quot;fontsize&quot;</span><span class="p">:</span> <span class="mi">12</span><span class="p">},</span>
                              <span class="s2">&quot;hlines&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;overlay_previous&quot;</span><span class="p">:</span> <span class="s2">&quot;share-y&quot;</span><span class="p">}}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">available_types</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;bed&quot;</span><span class="p">,</span> <span class="s2">&quot;bedgraph&quot;</span><span class="p">,</span> <span class="s2">&quot;bedgraph_matrix&quot;</span><span class="p">,</span> <span class="s2">&quot;bigwig&quot;</span><span class="p">,</span> <span class="s2">&quot;domains&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;epilogos&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;gtf&quot;</span><span class="p">,</span> <span class="s2">&quot;hic_matrix&quot;</span><span class="p">,</span> <span class="s2">&quot;hic_matrix_square&quot;</span><span class="p">,</span> <span class="s2">&quot;links&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;maf&quot;</span><span class="p">,</span> <span class="s2">&quot;narrow_peak&quot;</span><span class="p">,</span> <span class="s2">&quot;scalebar&quot;</span><span class="p">,</span> <span class="s2">&quot;vhighlight&quot;</span><span class="p">,</span> <span class="s2">&quot;vlines&quot;</span><span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">output</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># path to the output file if written</span>

    <span class="k">def</span> <span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return a string representation of the GenomeTracks object.&quot;&quot;&quot;</span>
        <span class="n">n_tracks</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tracks</span><span class="p">)</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;GenomeTracks object with </span><span class="si">{</span><span class="n">n_tracks</span><span class="si">}</span><span class="s2"> track(s). See &lt;obj&gt;.tracks for details.&quot;</span>

<div class="viewcode-block" id="GenomeTracks.add_track">
<a class="viewcode-back" href="../../../API/plotting.html#sctoolbox.plotting.genometracks.GenomeTracks.add_track">[docs]</a>
    <span class="k">def</span> <span class="nf">add_track</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                  <span class="n">file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                  <span class="n">file_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                  <span class="n">name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                  <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="n">Any</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add a track to the GenomeTracks object.</span>

<span class="sd">        The track will be added to the configuration file as one element, e.g. .add_track(&quot;file1.bed&quot;, file_type=&quot;bed&quot;, name=&quot;my_bed&quot;) will add the following to the configuration file:</span>
<span class="sd">        ```</span>
<span class="sd">        [my_bed]</span>
<span class="sd">        file = file1.bed</span>
<span class="sd">        file_type = bed</span>
<span class="sd">        ```</span>

<span class="sd">        Additional parameters are decided by &lt;obj&gt;.global_defaults and &lt;obj&gt;.type_defaults, or can be given by kwargs. All options and parameters are available at:</span>
<span class="sd">        https://pygenometracks.readthedocs.io/en/latest/content/all_tracks.html</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        file : str</span>
<span class="sd">            Path to the file containing information to be plotted. Can be .bed, .bw, .gtf etc.</span>
<span class="sd">        file_type : str, default None</span>
<span class="sd">            Specify the &#39;file_type&#39; argument for pyGenomeTracks. If None, the type will be predicted from the file ending.</span>
<span class="sd">        name : str, default None</span>
<span class="sd">            Name of the track. If None, the name will be estimated from the file_type e.g. &#39;bigwig 1&#39;. or &#39;bed 2&#39;. If the file_type is not available, the name will be the file path.</span>
<span class="sd">        **kwargs : arguments</span>
<span class="sd">            Additional arguments to be passed to pyGenomeTracks track configuration, for example `height=5` or `title=&quot;My track&quot;`.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        ValueError</span>
<span class="sd">            If the file_type is not valid.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Setup</span>
        <span class="n">track_dict</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">global_defaults</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">track_dict</span><span class="p">[</span><span class="s2">&quot;file&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">file</span>

        <span class="c1"># Predict file type</span>
        <span class="k">if</span> <span class="n">file_type</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">file_type</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_predict_type</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Check if file_type is valid</span>
            <span class="k">if</span> <span class="n">file_type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">available_types</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">file_type</span> <span class="o">==</span> <span class="s2">&quot;spacer&quot;</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;file_type &#39;spacer&#39; is not valid. Use GenomeTracks.add_spacer() instead.&quot;</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">file_type</span> <span class="o">==</span> <span class="s2">&quot;x-axis&quot;</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;file_type &#39;x-axis&#39; is not valid. Use GenomeTracks.add_xaxis() instead.&quot;</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">file_type</span> <span class="o">==</span> <span class="s2">&quot;hlines&quot;</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;file_type &#39;hlines&#39; is not valid. Use GenomeTracks.add_hlines() instead.&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;file_type &#39;</span><span class="si">{</span><span class="n">file_type</span><span class="si">}</span><span class="s2">&#39; not valid. Choose from </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">available_types</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># If filetype was predicted or given; add to track dict</span>
        <span class="k">if</span> <span class="n">file_type</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">file_type</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;vlines&quot;</span><span class="p">,</span> <span class="s2">&quot;vhighlight&quot;</span><span class="p">]:</span>
                <span class="n">track_dict</span><span class="p">[</span><span class="s2">&quot;type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_type</span>   <span class="c1"># file_type = type for some options</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">track_dict</span><span class="p">[</span><span class="s2">&quot;file_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_type</span>
        <span class="n">track_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">type_defaults</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">file_type</span><span class="p">,</span> <span class="p">{}))</span>  <span class="c1"># add type defaults</span>

        <span class="c1"># Set title depending on file_type</span>
        <span class="k">if</span> <span class="n">file_type</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;vlines&quot;</span><span class="p">,</span> <span class="s2">&quot;vhighlight&quot;</span><span class="p">]:</span>
            <span class="k">del</span> <span class="n">track_dict</span><span class="p">[</span><span class="s2">&quot;height&quot;</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">track_dict</span><span class="p">[</span><span class="s2">&quot;title&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>  <span class="c1"># per default, can be overwritten with kwargs</span>

        <span class="c1"># Final overwrite with kwargs</span>
        <span class="n">track_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="c1"># Count file-types</span>
        <span class="k">if</span> <span class="n">file_type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">type_count</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">type_count</span><span class="p">[</span><span class="n">file_type</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">type_count</span><span class="p">[</span><span class="n">file_type</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="c1"># Add track to tracks dictionary</span>
        <span class="k">if</span> <span class="n">name</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">file_type</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">name</span> <span class="o">=</span> <span class="n">file</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">name</span> <span class="o">=</span> <span class="n">file_type</span> <span class="o">+</span> <span class="s2">&quot; &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">type_count</span><span class="p">[</span><span class="n">file_type</span><span class="p">])</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">tracks</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="n">name</span><span class="p">:</span> <span class="n">track_dict</span><span class="p">})</span></div>


<div class="viewcode-block" id="GenomeTracks.add_hlines">
<a class="viewcode-back" href="../../../API/plotting.html#sctoolbox.plotting.genometracks.GenomeTracks.add_hlines">[docs]</a>
    <span class="k">def</span> <span class="nf">add_hlines</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                   <span class="n">y_values</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">int</span> <span class="o">|</span> <span class="nb">float</span><span class="p">],</span>
                   <span class="n">overlay_previous</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;share-y&quot;</span><span class="p">,</span> <span class="s2">&quot;no&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;share-y&quot;</span><span class="p">,</span>
                   <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="n">Any</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add horizontal lines to the previous plot.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        y_values : list of int or float</span>
<span class="sd">            List of y values to plot horizontal lines at.</span>
<span class="sd">        overlay_previous : str, default &quot;share-y&quot;</span>
<span class="sd">            Whether to plot the lines on the same y-axis as the previous plot (&quot;share-y&quot;) or on a new y-axis (&quot;no&quot;).</span>
<span class="sd">        **kwargs : arguments</span>
<span class="sd">            Additional arguments to be passed to pyGenomeTracks track configuration, for example `title=&quot;My lines&quot;`.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">y_values</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">y</span><span class="p">)</span> <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">y_values</span><span class="p">]</span>

        <span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;hlines&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;y_values&quot;</span><span class="p">:</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">y_values</span><span class="p">),</span> <span class="s2">&quot;overlay_previous&quot;</span><span class="p">:</span> <span class="n">overlay_previous</span><span class="p">,</span> <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="p">}}</span>
        <span class="n">d</span><span class="p">[</span><span class="s2">&quot;hlines&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="n">d</span><span class="p">[</span><span class="s2">&quot;hlines&quot;</span><span class="p">][</span><span class="s2">&quot;file_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;hlines&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">tracks</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">d</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomeTracks.add_hline">
<a class="viewcode-back" href="../../../API/plotting.html#sctoolbox.plotting.genometracks.GenomeTracks.add_hline">[docs]</a>
    <span class="k">def</span> <span class="nf">add_hline</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                  <span class="n">height</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
                  <span class="n">line_width</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
                  <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="n">Any</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add a horizontal line between tracks, not within a track.</span>

<span class="sd">        Can be used to visually separate tracks.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        height : int, default 1</span>
<span class="sd">            Height of the track with the line in the middle.</span>
<span class="sd">        line_width : int, default 2</span>
<span class="sd">            Width of the line.</span>
<span class="sd">        **kwargs : arguments</span>
<span class="sd">            Additional arguments to be passed to pyGenomeTracks track configuration, for example `title=&quot;A line&quot;`.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">d</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">d</span><span class="p">[</span><span class="s2">&quot;height&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">height</span>
        <span class="n">d</span><span class="p">[</span><span class="s2">&quot;line_width&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">line_width</span>
        <span class="n">d</span><span class="p">[</span><span class="s2">&quot;show_data_range&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;show_data_range&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>  <span class="c1"># default is False</span>
        <span class="n">d</span><span class="p">[</span><span class="s2">&quot;overlay_previous&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;no&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">add_hlines</span><span class="p">([</span><span class="mi">1</span><span class="p">],</span> <span class="n">min_value</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_value</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="o">**</span><span class="n">d</span><span class="p">)</span>  <span class="c1"># line is in the middle of the track</span></div>


<div class="viewcode-block" id="GenomeTracks.add_spacer">
<a class="viewcode-back" href="../../../API/plotting.html#sctoolbox.plotting.genometracks.GenomeTracks.add_spacer">[docs]</a>
    <span class="k">def</span> <span class="nf">add_spacer</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                   <span class="n">height</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">1</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add a spacer between tracks.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        height : int, default 1</span>
<span class="sd">            Height of the spacer track.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;spacer&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;height&quot;</span><span class="p">:</span> <span class="n">height</span><span class="p">}}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tracks</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">d</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomeTracks.add_xaxis">
<a class="viewcode-back" href="../../../API/plotting.html#sctoolbox.plotting.genometracks.GenomeTracks.add_xaxis">[docs]</a>
    <span class="k">def</span> <span class="nf">add_xaxis</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                  <span class="n">height</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
                  <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="n">Any</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add the x-axis to the plot.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        height : int, default 1</span>
<span class="sd">            Height of the x-axis track.</span>
<span class="sd">        **kwargs : arguments</span>
<span class="sd">            Additional arguments to be passed to pyGenomeTracks track configuration.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;height&quot;</span><span class="p">:</span> <span class="n">height</span><span class="p">}</span>
        <span class="n">d</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tracks</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s2">&quot;x-axis&quot;</span><span class="p">:</span> <span class="n">d</span><span class="p">})</span></div>


    <span class="k">def</span> <span class="nf">_predict_type</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Predict the file type from the file ending or the contents of the file.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        file : str</span>
<span class="sd">            Path of the file to be plotted.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        str | None</span>
<span class="sd">            String of predicted file type or None if the file type could not be predicted.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.bed&quot;</span><span class="p">):</span>
            <span class="k">return</span> <span class="s2">&quot;bed&quot;</span>

        <span class="k">elif</span> <span class="n">file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.bw&quot;</span><span class="p">):</span>
            <span class="k">return</span> <span class="s2">&quot;bigwig&quot;</span>

        <span class="k">elif</span> <span class="n">file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.gtf&quot;</span><span class="p">):</span>
            <span class="k">return</span> <span class="s2">&quot;gtf&quot;</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Could not predict file type for &#39;</span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2">&#39;. pyGenometracks will try to predict the file type. For more control, please specify &#39;file_type&#39; manually in &#39;.add_track&#39;.&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_create_config_str</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Create configuration string based on tracks list.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        config_str : str</span>
<span class="sd">            String containing the configuration file content</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">config_str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tracks</span><span class="p">:</span>
            <span class="n">track_name</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">track_dict</span> <span class="o">=</span> <span class="n">d</span><span class="p">[</span><span class="n">track_name</span><span class="p">]</span>

            <span class="n">config_str</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;[</span><span class="si">{</span><span class="n">track_name</span><span class="si">}</span><span class="s2">]</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">track_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">config_str</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2"> = </span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="n">config_str</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="k">return</span> <span class="n">config_str</span>

    <span class="k">def</span> <span class="nf">_write_config</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">config_file</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Write the configuration file to disk.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        config_file : str, default None</span>
<span class="sd">            Path to the configuration file to create. If None, a temporary file will be created in the system&#39;s temp directory.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        config_file : str</span>
<span class="sd">            Path to the configuration file.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">config_file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">config_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">tempfile</span><span class="o">.</span><span class="n">gettempdir</span><span class="p">(),</span> <span class="nb">next</span><span class="p">(</span><span class="n">tempfile</span><span class="o">.</span><span class="n">_get_candidate_names</span><span class="p">()))</span>

        <span class="n">config_str</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_create_config_str</span><span class="p">()</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">config_str</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">config_file</span>

<div class="viewcode-block" id="GenomeTracks.show_plot">
<a class="viewcode-back" href="../../../API/plotting.html#sctoolbox.plotting.genometracks.GenomeTracks.show_plot">[docs]</a>
    <span class="k">def</span> <span class="nf">show_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Display the plot.&quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">output</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No output file was created. Run GenomeTracks.plot() first.&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">utils</span><span class="o">.</span><span class="n">_is_notebook</span><span class="p">():</span>
            <span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">Image</span><span class="p">,</span> <span class="n">IFrame</span><span class="p">,</span> <span class="n">display</span>

            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">output</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.png&quot;</span><span class="p">):</span>
                <span class="n">display</span><span class="p">(</span><span class="n">Image</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">output</span><span class="p">))</span>
            <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">output</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.pdf&quot;</span><span class="p">):</span>
                <span class="n">display</span><span class="p">(</span><span class="n">IFrame</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="kn">import</span> <span class="nn">matplotlib.image</span> <span class="k">as</span> <span class="nn">mpimg</span>

            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">output</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.png&quot;</span><span class="p">):</span>
                <span class="n">img</span> <span class="o">=</span> <span class="n">mpimg</span><span class="o">.</span><span class="n">imread</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output</span><span class="p">)</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">img</span><span class="p">)</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Only .png files can be shown in the console.&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomeTracks.show_config">
<a class="viewcode-back" href="../../../API/plotting.html#sctoolbox.plotting.genometracks.GenomeTracks.show_config">[docs]</a>
    <span class="k">def</span> <span class="nf">show_config</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Show the current configuration file as a string.&quot;&quot;&quot;</span>

        <span class="n">config_str</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_create_config_str</span><span class="p">()</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">config_str</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomeTracks.plot">
<a class="viewcode-back" href="../../../API/plotting.html#sctoolbox.plotting.genometracks.GenomeTracks.plot">[docs]</a>
    <span class="k">def</span> <span class="nf">plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
             <span class="n">region</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
             <span class="n">output</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;genometracks.png&quot;</span><span class="p">,</span>
             <span class="n">config_file</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
             <span class="n">title</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
             <span class="n">show</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
             <span class="n">dpi</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">300</span><span class="p">,</span>
             <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="n">Any</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plot the final GenomeTracks plot based on the collected tracks.</span>

<span class="sd">        Runs pyGenomeTracks with the configuration file and the given parameters, and saves the output to the given file.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        region : str</span>
<span class="sd">            Region to plot, e.g. &quot;chr1:1000000-2000000&quot;.</span>
<span class="sd">        output : str, default &quot;genometracks.png&quot;</span>
<span class="sd">            Path to the output file.</span>
<span class="sd">        config_file : str, default None</span>
<span class="sd">            Path to the configuration file to create. If None, a temporary file will be created in the system&#39;s temp directory.</span>
<span class="sd">        title : str, default None</span>
<span class="sd">            Title of the plot. If None, no title will be shown.</span>
<span class="sd">        show : bool, default True</span>
<span class="sd">            If the function is run in a jupyter notebook, &#39;show&#39; controls whether to show the plot at the end of the function run.</span>
<span class="sd">        dpi : int, default 300</span>
<span class="sd">            DPI of the plot.</span>
<span class="sd">        **kwargs : arguments</span>
<span class="sd">            Additional arguments to be passed to pyGenomeTracks, for example `trackLabelFraction=0.2`.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        ValueError</span>
<span class="sd">            If the pyGenomeTracks command fails.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;title&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">title</span><span class="si">}</span><span class="s2">&#39;&quot;</span> <span class="k">if</span> <span class="n">title</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="kc">None</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;dpi&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dpi</span>

        <span class="c1"># create the .ini file</span>
        <span class="n">ini_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_write_config</span><span class="p">(</span><span class="n">config_file</span><span class="o">=</span><span class="n">config_file</span><span class="p">)</span>

        <span class="c1"># Build command</span>
        <span class="n">pgtracks_path</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">get_binary_path</span><span class="p">(</span><span class="s2">&quot;pyGenomeTracks&quot;</span><span class="p">)</span>
        <span class="n">cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pgtracks_path</span><span class="si">}</span><span class="s2"> --tracks </span><span class="si">{</span><span class="n">ini_file</span><span class="si">}</span><span class="s2"> --region </span><span class="si">{</span><span class="n">region</span><span class="si">}</span><span class="s2"> --outFileName </span><span class="si">{</span><span class="n">output</span><span class="si">}</span><span class="s2"> &quot;</span>

        <span class="c1"># Add additional kwargs</span>
        <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">value</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>  <span class="c1"># title might be None</span>
                <span class="n">cmd</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot; --</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="s2"> &quot;</span>

        <span class="c1"># Run pygenometracks</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Running command: &#39;</span><span class="si">{</span><span class="n">cmd</span><span class="si">}</span><span class="s2">&#39;&quot;</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">except</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">CalledProcessError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">output</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># reset output in case it was previously plotted</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error while running pyGenomeTracks: </span><span class="si">{</span><span class="n">e</span><span class="o">.</span><span class="n">output</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Remove config file</span>
        <span class="k">if</span> <span class="n">config_file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>  <span class="c1"># config_file was created by _write_config</span>
            <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">ini_file</span><span class="p">)</span>

        <span class="c1"># Show in notebook</span>
        <span class="k">if</span> <span class="n">show</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">output</span> <span class="o">=</span> <span class="n">output</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">show_plot</span><span class="p">()</span></div>
</div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Loosolab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>