<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>sctoolbox.plotting.marker_genes &mdash; SC FRAMEWORK  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../../../_static/plot_directive.css" />

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../../../_static/doctools.js?v=9a2dae69"></script>
        <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            SC FRAMEWORK
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../CHANGES.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../rna-notebooks/index.html">RNA notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../atac-notebooks/index.html">ATAC notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../general-notebooks/index.html">General notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../API/index.html">sctoolbox API reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">SC FRAMEWORK</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">sctoolbox.plotting.marker_genes</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for sctoolbox.plotting.marker_genes</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;Plots for marker genes e.g. as results of sc.tl.rank_genes_groups.&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">qnorm</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">sklearn.preprocessing</span> <span class="kn">import</span> <span class="n">MinMaxScaler</span>
<span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">zscore</span>
<span class="kn">import</span> <span class="nn">re</span>

<span class="c1"># for plotting</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">matplotlib.patches</span> <span class="kn">import</span> <span class="n">Patch</span>

<span class="kn">from</span> <span class="nn">beartype.typing</span> <span class="kn">import</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">,</span> <span class="n">Literal</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">,</span> <span class="n">Any</span>
<span class="kn">from</span> <span class="nn">beartype</span> <span class="kn">import</span> <span class="n">beartype</span>

<span class="c1"># sctoolbox functions</span>
<span class="kn">import</span> <span class="nn">sctoolbox.utils</span> <span class="k">as</span> <span class="nn">utils</span>
<span class="kn">from</span> <span class="nn">sctoolbox.plotting.general</span> <span class="kn">import</span> <span class="n">bidirectional_barplot</span><span class="p">,</span> <span class="n">_save_figure</span><span class="p">,</span> <span class="n">_make_square</span>
<span class="kn">import</span> <span class="nn">sctoolbox.utils.decorator</span> <span class="k">as</span> <span class="nn">deco</span>


<div class="viewcode-block" id="rank_genes_plot">
<a class="viewcode-back" href="../../../API/plotting.html#sctoolbox.plotting.marker_genes.rank_genes_plot">[docs]</a>
<span class="nd">@deco</span><span class="o">.</span><span class="n">log_anndata</span>
<span class="nd">@beartype</span>
<span class="k">def</span> <span class="nf">rank_genes_plot</span><span class="p">(</span><span class="n">adata</span><span class="p">:</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span>
                    <span class="n">key</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;rank_genes_groups&quot;</span><span class="p">,</span>
                    <span class="n">genes</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                    <span class="n">n_genes</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">15</span><span class="p">,</span>
                    <span class="n">dendrogram</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
                    <span class="n">groupby</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                    <span class="n">title</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                    <span class="n">style</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;dots&quot;</span><span class="p">,</span> <span class="s2">&quot;heatmap&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;dots&quot;</span><span class="p">,</span>
                    <span class="n">measure</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;expression&quot;</span><span class="p">,</span>
                    <span class="n">save</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                    <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="n">Any</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plot expression of genes from rank_genes_groups or from a gene list/dict.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    adata : sc.AnnData</span>
<span class="sd">        Annotated data matrix.</span>
<span class="sd">    key : Optional[str], default &quot;rank_genes_groups&quot;</span>
<span class="sd">        Key from `adata.uns` to plot. For example, `rank_genes_groups` or `rank_genes_groups_filtered`.</span>
<span class="sd">    genes : Optional[list[str] | dict[str, list[str]]], default None</span>
<span class="sd">        List of genes to plot across groups in &#39;groupby&#39;. If a dict is passed, the keys are the group names and the values are lists of genes. Setting &#39;genes&#39; overrides the &#39;key&#39; parameter.</span>
<span class="sd">    n_genes : int, default 15</span>
<span class="sd">        Number of genes to plot if `key` is specified.</span>
<span class="sd">    dendrogram : bool, default False</span>
<span class="sd">        Whether to show the dendrogram for groups.</span>
<span class="sd">    groupby : Optional[str], default None</span>
<span class="sd">        Key from `adata.obs` to group cells by.</span>
<span class="sd">    title : Optional[str], default None</span>
<span class="sd">        Title for the plot.</span>
<span class="sd">    style : Literal[&quot;dots&quot;, &quot;heatmap&quot;], default &quot;dots&quot;</span>
<span class="sd">        Style of the plot. Either `dots` or `heatmap`.</span>
<span class="sd">    measure : str, default &quot;expression&quot;</span>
<span class="sd">        Measure to write in colorbar label. For example, `expression` or `accessibility`.</span>
<span class="sd">    save : Optional[str], default None</span>
<span class="sd">        If given, save the figure to this path.</span>
<span class="sd">    **kwargs : Any</span>
<span class="sd">        Additional arguments passed to `sc.pl.rank_genes_groups_dotplot` or `sc.pl.rank_genes_groups_matrixplot`.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    ValueError</span>
<span class="sd">        1. If `style` is not one of `dots` or `heatmap`</span>
<span class="sd">        2. If `groupby` is not specified when `genes` is specified.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    g : dict</span>
<span class="sd">        Dictionary containing the matplotlib axes objects for the plot.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>

<span class="sd">        pl.rank_genes_plot(adata, n_genes=5)</span>

<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>

<span class="sd">        pl.rank_genes_plot(adata, genes={&quot;group1&quot;: adata.var.index[:10], &quot;group2&quot;: adata.var.index[10:20]}, groupby=&quot;bulk_labels&quot;)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Key is not needed if genes are specified</span>
    <span class="k">if</span> <span class="n">genes</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">key</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="c1"># Plot genes from rank_genes_groups or from gene list</span>
    <span class="n">parameters</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;swap_axes&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">}</span>  <span class="c1"># default parameters</span>
    <span class="n">parameters</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">key</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>  <span class="c1"># from rank_genes_groups output</span>

        <span class="k">if</span> <span class="n">style</span> <span class="o">==</span> <span class="s2">&quot;dots&quot;</span><span class="p">:</span>
            <span class="n">g</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_genes_groups_dotplot</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span>
                                                <span class="n">key</span><span class="o">=</span><span class="n">key</span><span class="p">,</span>
                                                <span class="n">n_genes</span><span class="o">=</span><span class="n">n_genes</span><span class="p">,</span>
                                                <span class="n">dendrogram</span><span class="o">=</span><span class="n">dendrogram</span><span class="p">,</span>
                                                <span class="n">groupby</span><span class="o">=</span><span class="n">groupby</span><span class="p">,</span>
                                                <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                                <span class="o">**</span><span class="n">parameters</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">style</span> <span class="o">==</span> <span class="s2">&quot;heatmap&quot;</span><span class="p">:</span>
            <span class="n">g</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_genes_groups_matrixplot</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span>
                                                   <span class="n">key</span><span class="o">=</span><span class="n">key</span><span class="p">,</span>
                                                   <span class="n">n_genes</span><span class="o">=</span><span class="n">n_genes</span><span class="p">,</span>
                                                   <span class="n">dendrogram</span><span class="o">=</span><span class="n">dendrogram</span><span class="p">,</span>
                                                   <span class="n">groupby</span><span class="o">=</span><span class="n">groupby</span><span class="p">,</span>
                                                   <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                                   <span class="o">**</span><span class="n">parameters</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>  <span class="c1"># from a gene list</span>

        <span class="k">if</span> <span class="n">groupby</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;The parameter &#39;groupby&#39; is needed if &#39;genes&#39; is given.&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">style</span> <span class="o">==</span> <span class="s2">&quot;dots&quot;</span><span class="p">:</span>
            <span class="n">g</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dotplot</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span>
                              <span class="n">dendrogram</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                              <span class="n">groupby</span><span class="o">=</span><span class="n">groupby</span><span class="p">,</span>
                              <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                              <span class="o">**</span><span class="n">parameters</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">style</span> <span class="o">==</span> <span class="s2">&quot;heatmap&quot;</span><span class="p">:</span>
            <span class="n">g</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">matrixplot</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span>
                                 <span class="n">dendrogram</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                 <span class="n">groupby</span><span class="o">=</span><span class="n">groupby</span><span class="p">,</span>
                                 <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                 <span class="o">**</span><span class="n">parameters</span><span class="p">)</span>

    <span class="n">g</span><span class="p">[</span><span class="s2">&quot;mainplot_ax&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">g</span><span class="p">[</span><span class="s2">&quot;mainplot_ax&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get_xticklabels</span><span class="p">(),</span> <span class="n">ha</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>

    <span class="c1"># Rotate gene group names</span>
    <span class="k">if</span> <span class="s2">&quot;gene_group_ax&quot;</span> <span class="ow">in</span> <span class="n">g</span> <span class="ow">and</span> <span class="n">parameters</span><span class="p">[</span><span class="s2">&quot;swap_axes&quot;</span><span class="p">]</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">:</span>  <span class="c1"># only rotate if axes is not swapped</span>
        <span class="k">for</span> <span class="n">text</span> <span class="ow">in</span> <span class="n">g</span><span class="p">[</span><span class="s2">&quot;gene_group_ax&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">_children</span><span class="p">:</span>
            <span class="n">text</span><span class="o">.</span><span class="n">_rotation</span> <span class="o">=</span> <span class="mi">45</span>
            <span class="n">text</span><span class="o">.</span><span class="n">_horizontalalignment</span> <span class="o">=</span> <span class="s2">&quot;left&quot;</span>

    <span class="c1"># Change title of colorbar (for example expression -&gt; accessibility)</span>
    <span class="n">default_title</span> <span class="o">=</span> <span class="n">g</span><span class="p">[</span><span class="s2">&quot;color_legend_ax&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get_title</span><span class="p">()</span>
    <span class="n">updated_title</span> <span class="o">=</span> <span class="n">default_title</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;expression&quot;</span><span class="p">,</span> <span class="n">measure</span><span class="p">)</span>
    <span class="n">g</span><span class="p">[</span><span class="s2">&quot;color_legend_ax&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">updated_title</span><span class="p">)</span>

    <span class="c1"># Add title to plot above groups</span>
    <span class="k">if</span> <span class="n">title</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>

        <span class="k">if</span> <span class="s2">&quot;gene_group_ax&quot;</span> <span class="ow">in</span> <span class="n">g</span><span class="p">:</span>

            <span class="k">if</span> <span class="n">parameters</span><span class="p">[</span><span class="s2">&quot;swap_axes&quot;</span><span class="p">]:</span>
                <span class="n">g</span><span class="p">[</span><span class="s2">&quot;mainplot_ax&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">title</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">*</span> <span class="mi">2</span><span class="p">)</span>  <span class="c1"># \n to ensure a little space between title and plot</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">gcf</span><span class="p">()</span>
                <span class="n">fig</span><span class="o">.</span><span class="n">canvas</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
                <span class="n">renderer</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">canvas</span><span class="o">.</span><span class="n">get_renderer</span><span class="p">()</span>

                <span class="n">highest_y</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="k">for</span> <span class="n">text</span> <span class="ow">in</span> <span class="n">g</span><span class="p">[</span><span class="s2">&quot;gene_group_ax&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">_children</span><span class="p">:</span>

                    <span class="c1"># Find highest y value for all labels</span>
                    <span class="n">ax</span> <span class="o">=</span> <span class="n">g</span><span class="p">[</span><span class="s2">&quot;gene_group_ax&quot;</span><span class="p">]</span>
                    <span class="n">transf</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">transData</span><span class="o">.</span><span class="n">inverted</span><span class="p">()</span>
                    <span class="n">bb</span> <span class="o">=</span> <span class="n">text</span><span class="o">.</span><span class="n">get_window_extent</span><span class="p">(</span><span class="n">renderer</span><span class="o">=</span><span class="n">renderer</span><span class="p">)</span>
                    <span class="n">bb_datacoords</span> <span class="o">=</span> <span class="n">bb</span><span class="o">.</span><span class="n">transformed</span><span class="p">(</span><span class="n">transf</span><span class="p">)</span>

                    <span class="n">highest_y</span> <span class="o">=</span> <span class="n">bb_datacoords</span><span class="o">.</span><span class="n">y1</span> <span class="k">if</span> <span class="n">bb_datacoords</span><span class="o">.</span><span class="n">y1</span> <span class="o">&gt;</span> <span class="n">highest_y</span> <span class="k">else</span> <span class="n">highest_y</span>

                <span class="n">x_mid</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">g</span><span class="p">[</span><span class="s2">&quot;gene_group_ax&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get_xlim</span><span class="p">())</span>
                <span class="n">g</span><span class="p">[</span><span class="s2">&quot;gene_group_ax&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">x_mid</span><span class="p">,</span> <span class="n">highest_y</span> <span class="o">+</span> <span class="mf">0.2</span><span class="p">,</span> <span class="n">title</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s2">&quot;bottom&quot;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s2">&quot;center&quot;</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">g</span><span class="p">[</span><span class="s2">&quot;mainplot_ax&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>

    <span class="c1"># Save figure</span>
    <span class="n">_save_figure</span><span class="p">(</span><span class="n">save</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">g</span></div>



<span class="c1">#####################################################################</span>
<span class="c1">#          Violin / boxplot / bar for genes between groups          #</span>
<span class="c1">#####################################################################</span>

<div class="viewcode-block" id="grouped_violin">
<a class="viewcode-back" href="../../../API/plotting.html#sctoolbox.plotting.marker_genes.grouped_violin">[docs]</a>
<span class="nd">@deco</span><span class="o">.</span><span class="n">log_anndata</span>
<span class="nd">@beartype</span>
<span class="k">def</span> <span class="nf">grouped_violin</span><span class="p">(</span><span class="n">adata</span><span class="p">:</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span>
                   <span class="n">x</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
                   <span class="n">y</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                   <span class="n">groupby</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                   <span class="n">figsize</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span> <span class="o">|</span> <span class="nb">float</span><span class="p">,</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">float</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                   <span class="n">title</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                   <span class="n">style</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;violin&quot;</span><span class="p">,</span> <span class="s2">&quot;boxplot&quot;</span><span class="p">,</span> <span class="s2">&quot;bar&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;violin&quot;</span><span class="p">,</span>
                   <span class="n">normalize</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
                   <span class="n">ax</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                   <span class="n">save</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                   <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="n">Any</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create violinplot of values across cells in an adata object grouped by x and &#39;groupby&#39;.</span>

<span class="sd">    Can for example show the expression of one gene across groups (x = obs_group, y = gene),</span>
<span class="sd">    expression of multiple genes grouped by cell type (x = gene_list, groupby = obs_cell_type),</span>
<span class="sd">    or values from adata.obs across cells (x = obs_group, y = obs_column).</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    adata : sc.AnnData</span>
<span class="sd">        Annotated data matrix.</span>
<span class="sd">    x : str | list[str]</span>
<span class="sd">        Column name in adata.obs or gene name(s) in adata.var.index to group by on the x-axis. Multiple gene names can be given in a list.</span>
<span class="sd">    y : Optional[str], default None</span>
<span class="sd">        A column name in adata.obs or a gene in adata.var.index to plot values for. Only needed if x is a column in adata.obs.</span>
<span class="sd">    groupby : Optional[str], default None</span>
<span class="sd">        Column name in adata.obs to create grouped violins. If None, a single violin is plotted per group in &#39;x&#39;.</span>
<span class="sd">    figsize : Optional[Tuple[int | float, int | float]], default None</span>
<span class="sd">        Figure size.</span>
<span class="sd">    title : Optional[str], default None</span>
<span class="sd">        Title of the plot. If None, no title is shown.</span>
<span class="sd">    style : Literal[&quot;violin&quot;, &quot;boxplot&quot;, &quot;bar&quot;], default &quot;violin&quot;</span>
<span class="sd">        Plot style. Either &quot;violin&quot; or &quot;boxplot&quot; or &quot;bar&quot;.</span>
<span class="sd">    normalize : bool, default False</span>
<span class="sd">        If True, normalize the values in &#39;y&#39; to the range [0, 1] per group in &#39;x&#39;.</span>
<span class="sd">    ax : Optional[matplotlib.axes.Axes], default None</span>
<span class="sd">        A matplotlib axes object to plot violinplots in. If None, a new figure and axes is created.</span>
<span class="sd">    save : Optional[str], default None</span>
<span class="sd">        Path to save the figure to. If None, the figure is not saved.</span>
<span class="sd">    **kwargs : Any</span>
<span class="sd">        Additional arguments passed to seaborn.violinplot or seaborn.boxplot.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    matplotlib.axes.Axes</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    ValueError</span>
<span class="sd">        If x or y are not columns in adata.obs or a genes in adata.var.index.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>

<span class="sd">        pl.grouped_violin(adata, &#39;phase&#39;, y=&#39;G2M_score&#39;)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">x</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">]</span>
    <span class="n">x</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>  <span class="c1"># convert to list incase x was a numpy array or other iterable</span>

    <span class="c1"># Establish if x is a column in adata.obs or a gene in adata.var.index</span>
    <span class="n">x_assignment</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">element</span> <span class="ow">in</span> <span class="n">x</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">element</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="n">element</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">element</span><span class="si">}</span><span class="s2"> is not a column in adata.obs or a gene in adata.var.index&quot;</span><span class="p">)</span>
        <span class="n">x_assignment</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;obs&quot;</span> <span class="k">if</span> <span class="n">element</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span> <span class="k">else</span> <span class="s2">&quot;var&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">x_assignment</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;x must be either a column in adata.obs or all genes in adata.var.index&quot;</span><span class="p">)</span>

    <span class="n">x_assignment</span> <span class="o">=</span> <span class="n">x_assignment</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># Establish if y is a column in adata.obs or a gene in adata.var.index</span>
    <span class="k">if</span> <span class="n">x_assignment</span> <span class="o">==</span> <span class="s2">&quot;obs&quot;</span> <span class="ow">and</span> <span class="n">y</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Because &#39;x&#39; is a column in obs, &#39;y&#39; must be given as parameter&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">y</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">y_assignment</span> <span class="o">=</span> <span class="s2">&quot;obs&quot;</span>
        <span class="k">elif</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
            <span class="n">y_assignment</span> <span class="o">=</span> <span class="s2">&quot;var&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;y&#39; (</span><span class="si">{</span><span class="n">y</span><span class="si">}</span><span class="s2">) was not found in either adata.obs or adata.var.index&quot;</span><span class="p">)</span>

    <span class="c1"># Create obs table with column</span>
    <span class="n">obs_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">y</span><span class="p">,</span> <span class="n">groupby</span><span class="p">]</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
    <span class="n">obs_table</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">copy</span><span class="p">()[</span><span class="n">obs_cols</span><span class="p">]</span>  <span class="c1"># creates a copy</span>

    <span class="k">for</span> <span class="n">element</span> <span class="ow">in</span> <span class="n">x</span> <span class="o">+</span> <span class="p">[</span><span class="n">y</span><span class="p">]:</span>
        <span class="k">if</span> <span class="n">element</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
            <span class="n">gene_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argwhere</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span> <span class="o">==</span> <span class="n">element</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">vals</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">[:,</span> <span class="n">gene_idx</span><span class="p">]</span><span class="o">.</span><span class="n">todense</span><span class="p">()</span><span class="o">.</span><span class="n">A1</span>
            <span class="n">obs_table</span><span class="p">[</span><span class="n">element</span><span class="p">]</span> <span class="o">=</span> <span class="n">vals</span>

    <span class="c1"># Convert table to long format if the x-axis contains gene expressions</span>
    <span class="k">if</span> <span class="n">x_assignment</span> <span class="o">==</span> <span class="s2">&quot;var&quot;</span><span class="p">:</span>
        <span class="n">index_name</span> <span class="o">=</span> <span class="n">obs_table</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span>
        <span class="n">id_vars</span> <span class="o">=</span> <span class="p">[</span><span class="n">index_name</span><span class="p">,</span> <span class="n">groupby</span><span class="p">]</span>
        <span class="n">id_vars</span> <span class="o">=</span> <span class="n">id_vars</span> <span class="o">+</span> <span class="n">x</span> <span class="k">if</span> <span class="n">x_assignment</span> <span class="o">==</span> <span class="s2">&quot;obs&quot;</span> <span class="k">else</span> <span class="n">id_vars</span>
        <span class="n">id_vars</span> <span class="o">=</span> <span class="p">[</span><span class="n">v</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">id_vars</span> <span class="k">if</span> <span class="n">v</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">]</span>

        <span class="n">value_name</span> <span class="o">=</span> <span class="s2">&quot;Expression&quot;</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">normalize</span> <span class="k">else</span> <span class="s2">&quot;Normalized expression&quot;</span>
        <span class="n">obs_table</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">obs_table</span> <span class="o">=</span> <span class="n">obs_table</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="n">id_vars</span><span class="p">,</span> <span class="n">value_vars</span><span class="o">=</span><span class="n">x</span><span class="p">,</span>
                                   <span class="n">var_name</span><span class="o">=</span><span class="s2">&quot;Gene&quot;</span><span class="p">,</span> <span class="n">value_name</span><span class="o">=</span><span class="n">value_name</span><span class="p">)</span>
        <span class="n">x_var</span> <span class="o">=</span> <span class="s2">&quot;Gene&quot;</span>
        <span class="n">y_var</span> <span class="o">=</span> <span class="n">value_name</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">x_var</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">y_var</span> <span class="o">=</span> <span class="n">y</span>

    <span class="c1"># Normalize values to 0-1 per group in x_var</span>
    <span class="k">if</span> <span class="n">normalize</span><span class="p">:</span>
        <span class="n">obs_table</span><span class="p">[</span><span class="n">y_var</span><span class="p">]</span> <span class="o">=</span> <span class="n">obs_table</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">x_var</span><span class="p">,</span> <span class="n">group_keys</span><span class="o">=</span><span class="kc">False</span><span class="p">)[</span><span class="n">y_var</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span> <span class="o">-</span> <span class="n">x</span><span class="o">.</span><span class="n">min</span><span class="p">())</span> <span class="o">/</span> <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">x</span><span class="o">.</span><span class="n">min</span><span class="p">()))</span>

    <span class="c1"># Plot expression from obs table</span>
    <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">_</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">style</span> <span class="o">==</span> <span class="s2">&quot;violin&quot;</span><span class="p">:</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;scale&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;width&quot;</span> <span class="k">if</span> <span class="s2">&quot;scale&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">kwargs</span> <span class="k">else</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;scale&quot;</span><span class="p">]</span>  <span class="c1"># set defaults</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;cut&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span> <span class="k">if</span> <span class="s2">&quot;cut&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">kwargs</span> <span class="k">else</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;cut&quot;</span><span class="p">]</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">violinplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">obs_table</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="n">x_var</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">y_var</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="n">groupby</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">style</span> <span class="o">==</span> <span class="s2">&quot;boxplot&quot;</span><span class="p">:</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">obs_table</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="n">x_var</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">y_var</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="n">groupby</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">style</span> <span class="o">==</span> <span class="s2">&quot;bar&quot;</span><span class="p">:</span>

        <span class="k">pass</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Style &#39;</span><span class="si">{</span><span class="n">style</span><span class="si">}</span><span class="s2">&#39; is not valid for this function. Style must be one of &#39;violin&#39; or &#39;boxplot&#39;&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">groupby</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="n">groupby</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;center left&#39;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">),</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>  <span class="c1"># Set location of legend</span>

    <span class="c1"># Final adjustments of labels</span>
    <span class="k">if</span> <span class="n">x_assignment</span> <span class="o">==</span> <span class="s2">&quot;obs&quot;</span> <span class="ow">and</span> <span class="n">y_assignment</span> <span class="o">==</span> <span class="s2">&quot;var&quot;</span><span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="n">ax</span><span class="o">.</span><span class="n">get_ylabel</span><span class="p">()</span> <span class="o">+</span> <span class="s2">&quot; expression&quot;</span><span class="p">)</span>

    <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">ax</span><span class="o">.</span><span class="n">get_xticklabels</span><span class="p">(),</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;right&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;top&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># save figure if output is given</span>
    <span class="n">_save_figure</span><span class="p">(</span><span class="n">save</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">ax</span></div>



<div class="viewcode-block" id="group_expression_boxplot">
<a class="viewcode-back" href="../../../API/plotting.html#sctoolbox.plotting.marker_genes.group_expression_boxplot">[docs]</a>
<span class="nd">@deco</span><span class="o">.</span><span class="n">log_anndata</span>
<span class="nd">@beartype</span>
<span class="k">def</span> <span class="nf">group_expression_boxplot</span><span class="p">(</span><span class="n">adata</span><span class="p">:</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span>
                             <span class="n">gene_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
                             <span class="n">groupby</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                             <span class="n">figsize</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span> <span class="o">|</span> <span class="nb">float</span><span class="p">,</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">float</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                             <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="n">Any</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plot a boxplot showing summarized gene expression of genes in `gene_list` across the groups in `groupby`.</span>

<span class="sd">    The total gene expression is quantile normalized per group, and are subsequently normalized to 0-1 per gene across groups.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    adata : sc.AnnData</span>
<span class="sd">        An annotated data matrix object containing counts in .X.</span>
<span class="sd">    gene_list : list[str]</span>
<span class="sd">        A list of genes to show expression for.</span>
<span class="sd">    groupby : str</span>
<span class="sd">        A column in .obs for grouping cells into groups on the x-axis</span>
<span class="sd">    figsize : Optional[Tuple[int | float, int | float]], default None (matplotlib default)</span>
<span class="sd">        Control the size of the output figure, e.g. (6,10).</span>
<span class="sd">    **kwargs : Any</span>
<span class="sd">        Additional arguments passed to seaborn.boxplot.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    matplotlib.axes.Axes</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>

<span class="sd">        gene_list=(&quot;HES4&quot;, &quot;PRMT2&quot;, &quot;ITGB2&quot;)</span>
<span class="sd">        pl.group_expression_boxplot(adata, gene_list, groupby=&quot;bulk_labels&quot;)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Obtain pseudobulk</span>
    <span class="n">gene_table</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">pseudobulk_table</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groupby</span><span class="p">)</span>

    <span class="c1"># Normalize across clusters</span>
    <span class="n">gene_table</span> <span class="o">=</span> <span class="n">qnorm</span><span class="o">.</span><span class="n">quantile_normalize</span><span class="p">(</span><span class="n">gene_table</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Normalize to 0-1 across groups</span>
    <span class="n">scaler</span> <span class="o">=</span> <span class="n">MinMaxScaler</span><span class="p">()</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">gene_table</span><span class="o">.</span><span class="n">T</span>
    <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span> <span class="o">=</span> <span class="n">scaler</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">])</span>
    <span class="n">gene_table</span> <span class="o">=</span> <span class="n">df</span>

    <span class="c1"># Melt to long format</span>
    <span class="n">gene_table_melted</span> <span class="o">=</span> <span class="n">gene_table</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="s2">&quot;index&quot;</span><span class="p">,</span> <span class="n">var_name</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">)</span>
    <span class="n">gene_table_melted</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="n">groupby</span><span class="p">},</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Subset to input gene list</span>
    <span class="n">gene_table_melted</span> <span class="o">=</span> <span class="n">gene_table_melted</span><span class="p">[</span><span class="n">gene_table_melted</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)]</span>

    <span class="c1"># Sort by median value</span>
    <span class="n">medians</span> <span class="o">=</span> <span class="n">gene_table_melted</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">groupby</span><span class="p">)[</span><span class="s2">&quot;value&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">median</span><span class="p">()</span><span class="o">.</span><span class="n">to_frame</span><span class="p">()</span>
    <span class="n">medians</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;medians&quot;</span><span class="p">]</span>
    <span class="n">gene_table_melted_sorted</span> <span class="o">=</span> <span class="n">gene_table_melted</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">medians</span><span class="p">,</span> <span class="n">left_on</span><span class="o">=</span><span class="n">groupby</span><span class="p">,</span> <span class="n">right_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;medians&quot;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># Joined figure with all</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
    <span class="n">g</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">gene_table_melted_sorted</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="n">groupby</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;value&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;darkgrey&quot;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Normalized expression&quot;</span><span class="p">)</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">ax</span><span class="o">.</span><span class="n">get_xticklabels</span><span class="p">(),</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">g</span></div>



<div class="viewcode-block" id="gene_expression_heatmap">
<a class="viewcode-back" href="../../../API/plotting.html#sctoolbox.plotting.marker_genes.gene_expression_heatmap">[docs]</a>
<span class="nd">@deco</span><span class="o">.</span><span class="n">log_anndata</span>
<span class="nd">@beartype</span>
<span class="k">def</span> <span class="nf">gene_expression_heatmap</span><span class="p">(</span><span class="n">adata</span><span class="p">:</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span>
                            <span class="n">genes</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
                            <span class="n">cluster_column</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                            <span class="n">gene_name_column</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                            <span class="n">title</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                            <span class="n">groupby</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                            <span class="n">row_cluster</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
                            <span class="n">col_cluster</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
                            <span class="n">show_row_dendrogram</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
                            <span class="n">show_col_dendrogram</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
                            <span class="n">figsize</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span> <span class="o">|</span> <span class="nb">float</span><span class="p">,</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">float</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                            <span class="n">save</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="n">Any</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Any</span><span class="p">:</span>  <span class="c1"># Any since beartype cannot handle sns datatypes</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plot a heatmap of z-score normalized gene expression across clusters/groups.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    adata : sc.AnnData</span>
<span class="sd">        Annotated data matrix.</span>
<span class="sd">    genes : list[str]</span>
<span class="sd">        List of genes to plot. Must match names in `adata.var.index`.</span>
<span class="sd">    cluster_column : str</span>
<span class="sd">        Key in `adata.obs` for which to cluster the x-axis.</span>
<span class="sd">    gene_name_column : Optional[str], default None</span>
<span class="sd">        Column in `adata.var` for which to use for gene row names. Default is to use the .var index.</span>
<span class="sd">    title : Optional[str], default None</span>
<span class="sd">        Title of the plot.</span>
<span class="sd">    groupby : Optional[str], default None</span>
<span class="sd">        Key in `adata.obs` for which to plot a colorbar per cluster.</span>
<span class="sd">    row_cluster : bool, default True</span>
<span class="sd">        Whether to cluster the rows.</span>
<span class="sd">    col_cluster : bool, default False</span>
<span class="sd">        Whether to cluster the columns.</span>
<span class="sd">    show_row_dendrogram : bool, default False</span>
<span class="sd">        Whether to show the dendrogram for the rows.</span>
<span class="sd">    show_col_dendrogram : bool, default False</span>
<span class="sd">        Whether to show the dendrogram for the columns.</span>
<span class="sd">    figsize : Optional[Tuple[int | float, int | float]], default None</span>
<span class="sd">        Size of the figure. If `None`, use default size.</span>
<span class="sd">    save : Optional[str], default None</span>
<span class="sd">        If given, save the figure to this path.</span>
<span class="sd">    **kwargs : Any</span>
<span class="sd">        Additional arguments passed to `seaborn.clustermap`.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    g : Any</span>
<span class="sd">        sns.matrix.ClusterGrid: The seaborn ClusterGrid object containing the heatmap.</span>
<span class="sd">        Note: Any since sns.matrix.ClusterGrid cannot be checked by beartype.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    KeyError</span>
<span class="sd">        If `gene_name_column` is not a column in `adata.var`.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>

<span class="sd">        adata.obs[&quot;samples&quot;] = np.random.choice([&quot;CTRL1&quot;, &quot;CTRL2&quot;, &quot;CTRL3&quot;, &quot;CTRL4&quot;, &quot;TREAT1&quot;, &quot;TREAT2&quot;, &quot;TREAT3&quot;, &quot;TREAT4&quot;], size=adata.shape[0])</span>
<span class="sd">        adata.obs[&quot;condition&quot;] = adata.obs[&quot;samples&quot;].str.extract(&quot;([A-Z]+)&quot;)</span>

<span class="sd">        genes = adata.var.index[:15]</span>
<span class="sd">        pl.gene_expression_heatmap(adata, genes, cluster_column=&quot;samples&quot;,</span>
<span class="sd">                                   groupby=&quot;condition&quot;,</span>
<span class="sd">                                   title=&quot;Gene expression&quot;,</span>
<span class="sd">                                   col_cluster=True,</span>
<span class="sd">                                   show_col_dendrogram=True,</span>
<span class="sd">                                   colors_ratio=0.03)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">genes</span><span class="p">]</span>  <span class="c1"># Subset to genes</span>

    <span class="c1"># Decide which combination to cluster by</span>
    <span class="n">groupby_col</span> <span class="o">=</span> <span class="s2">&quot;_cluster_by&quot;</span>
    <span class="k">if</span> <span class="n">groupby</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">groupby_col</span><span class="p">]</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">cluster_column</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">groupby</span><span class="p">]))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">groupby_col</span><span class="p">]</span> <span class="o">=</span> <span class="p">[(</span><span class="n">s</span><span class="p">,</span> <span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">cluster_column</span><span class="p">]]</span>

    <span class="c1"># Collect counts for each gene per sample</span>
    <span class="n">counts</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">pseudobulk_table</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="n">groupby_col</span><span class="p">)</span>
    <span class="n">counts_z</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">zscore</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>

    <span class="c1"># Color dict for groupby</span>
    <span class="k">if</span> <span class="n">groupby</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">groups</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">groupby</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="n">color_list</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">()[:</span><span class="nb">len</span><span class="p">(</span><span class="n">groups</span><span class="p">)]</span>
        <span class="n">color_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">groups</span><span class="p">,</span> <span class="n">color_list</span><span class="p">))</span>

        <span class="c1"># Get color per column</span>
        <span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="n">color_dict</span><span class="p">[</span><span class="n">col</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]]</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">counts_z</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
        <span class="n">col_colors</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">counts_z</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">colors</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">col_colors</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="c1"># Translation dict for row names</span>
    <span class="k">if</span> <span class="n">gene_name_column</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">row_name_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">gene_name_column</span><span class="p">]))</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Column &#39;</span><span class="si">{</span><span class="n">gene_name_column</span><span class="si">}</span><span class="s2">&#39; not found in adata.var&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">row_name_dict</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="n">nrows</span><span class="p">,</span> <span class="n">ncols</span> <span class="o">=</span> <span class="n">counts_z</span><span class="o">.</span><span class="n">shape</span>
    <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="n">ncols</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="n">nrows</span> <span class="o">/</span> <span class="mi">3</span><span class="p">)</span> <span class="k">if</span> <span class="n">figsize</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">figsize</span>  <span class="c1"># (width, height)</span>

    <span class="c1"># Plot heatmap</span>
    <span class="n">parameters</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;cmap&quot;</span><span class="p">:</span> <span class="s2">&quot;bwr&quot;</span><span class="p">,</span> <span class="s2">&quot;center&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">}</span>
    <span class="n">parameters</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="n">g</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">clustermap</span><span class="p">(</span><span class="n">counts_z</span><span class="p">,</span>
                       <span class="n">xticklabels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                       <span class="n">yticklabels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="c1"># show all genes</span>
                       <span class="n">row_cluster</span><span class="o">=</span><span class="n">row_cluster</span><span class="p">,</span>
                       <span class="n">col_cluster</span><span class="o">=</span><span class="n">col_cluster</span><span class="p">,</span>
                       <span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span>
                       <span class="n">col_colors</span><span class="o">=</span><span class="n">col_colors</span><span class="p">,</span>
                       <span class="o">**</span><span class="n">parameters</span><span class="p">)</span>

    <span class="n">yticklabes</span> <span class="o">=</span> <span class="p">[</span><span class="n">row_name_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">g</span><span class="o">.</span><span class="n">data2d</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
    <span class="n">g</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">yticklabes</span><span class="p">)</span>

    <span class="n">xticklabels</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">g</span><span class="o">.</span><span class="n">data2d</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>  <span class="c1"># first index contains the cluster_column</span>
    <span class="n">g</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">xticklabels</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">)</span>
    <span class="n">g</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">labelleft</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelright</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">g</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">heatmap_bbox</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">get_position</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">show_row_dendrogram</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">:</span>
        <span class="n">g</span><span class="o">.</span><span class="n">ax_row_dendrogram</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">show_col_dendrogram</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">:</span>
        <span class="n">g</span><span class="o">.</span><span class="n">ax_col_dendrogram</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># Invert order of x-axis</span>
    <span class="c1"># g.ax_col_colors.invert_xaxis()</span>
    <span class="c1"># g.ax_col_dendrogram.invert_xaxis()</span>
    <span class="c1"># g.ax_heatmap.invert_xaxis()</span>

    <span class="c1"># Move colorbar</span>
    <span class="n">cbar_ax</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">ax_cbar</span>
    <span class="n">cbar_width</span> <span class="o">=</span> <span class="n">heatmap_bbox</span><span class="o">.</span><span class="n">width</span> <span class="o">/</span> <span class="n">ncols</span>  <span class="c1"># width of 1 column</span>
    <span class="n">cbar_height</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">heatmap_bbox</span><span class="o">.</span><span class="n">height</span><span class="p">,</span> <span class="n">heatmap_bbox</span><span class="o">.</span><span class="n">height</span> <span class="o">/</span> <span class="n">nrows</span> <span class="o">*</span> <span class="mi">5</span><span class="p">)</span>  <span class="c1"># 5 rows high, but ensure colorbar is not taller than heatmap</span>
    <span class="n">cbar_ax</span><span class="o">.</span><span class="n">set_position</span><span class="p">([</span><span class="n">heatmap_bbox</span><span class="o">.</span><span class="n">x1</span> <span class="o">+</span> <span class="n">cbar_width</span><span class="p">,</span> <span class="n">heatmap_bbox</span><span class="o">.</span><span class="n">y0</span><span class="p">,</span> <span class="n">cbar_width</span><span class="p">,</span> <span class="n">cbar_height</span><span class="p">])</span>
    <span class="n">cbar_ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Mean expr.</span><span class="se">\n</span><span class="s2">z-score&quot;</span><span class="p">)</span>

    <span class="c1"># Add color legend for groupby above cbar</span>
    <span class="k">if</span> <span class="n">groupby</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>

        <span class="n">g</span><span class="o">.</span><span class="n">ax_col_colors</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">right</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelright</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="n">handles</span> <span class="o">=</span> <span class="p">[</span><span class="n">Patch</span><span class="p">(</span><span class="n">facecolor</span><span class="o">=</span><span class="n">color_dict</span><span class="p">[</span><span class="n">name</span><span class="p">])</span> <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">color_dict</span><span class="p">]</span>
        <span class="n">legend</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="p">,</span> <span class="n">color_dict</span><span class="p">,</span>
                            <span class="n">title</span><span class="o">=</span><span class="n">groupby</span><span class="p">,</span>
                            <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">1.25</span><span class="p">),</span>  <span class="c1"># 1.25 to make sure there is space for cbar label</span>
                            <span class="n">bbox_transform</span><span class="o">=</span><span class="n">cbar_ax</span><span class="o">.</span><span class="n">transAxes</span><span class="p">,</span>
                            <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;lower left&#39;</span><span class="p">,</span>
                            <span class="n">handlelength</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">handleheight</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                            <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                            <span class="n">borderpad</span><span class="o">=</span><span class="mi">0</span>
                            <span class="p">)</span>
        <span class="n">legend</span><span class="o">.</span><span class="n">_legend_box</span><span class="o">.</span><span class="n">align</span> <span class="o">=</span> <span class="s2">&quot;left&quot;</span>

    <span class="c1"># Set title on top of heatmap</span>
    <span class="k">if</span> <span class="n">title</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">g</span><span class="o">.</span><span class="n">ax_col_dendrogram</span><span class="o">.</span><span class="n">get_visible</span><span class="p">():</span>
            <span class="n">g</span><span class="o">.</span><span class="n">ax_col_dendrogram</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">title</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">13</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">g</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">counts_z</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mi">2</span><span class="p">,</span> <span class="n">title</span><span class="p">,</span>  <span class="c1"># ensures space beetween title and heatmap</span>
                              <span class="n">transform</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">transData</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s2">&quot;center&quot;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s2">&quot;bottom&quot;</span><span class="p">,</span>
                              <span class="n">fontsize</span><span class="o">=</span><span class="mi">13</span><span class="p">)</span>

    <span class="n">_save_figure</span><span class="p">(</span><span class="n">save</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">g</span></div>



<div class="viewcode-block" id="plot_differential_genes">
<a class="viewcode-back" href="../../../API/plotting.html#sctoolbox.plotting.marker_genes.plot_differential_genes">[docs]</a>
<span class="nd">@beartype</span>
<span class="k">def</span> <span class="nf">plot_differential_genes</span><span class="p">(</span><span class="n">rank_table</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
                            <span class="n">title</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;Differentially expressed genes&quot;</span><span class="p">,</span>
                            <span class="n">save</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="n">Any</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plot number of differentially expressed genes per contrast in a barplot.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    rank_table : pd.DataFrame</span>
<span class="sd">        Output of sctoolbox.tools.marker_genes.pairwise_rank_genes.</span>
<span class="sd">    title : str, default &quot;Differentially expressed genes&quot;</span>
<span class="sd">        Title of the plot.</span>
<span class="sd">    save : Optional[str], default None</span>
<span class="sd">        If given, save the figure to this path.</span>
<span class="sd">    **kwargs : Any</span>
<span class="sd">        Keyword arguments passed to pl.bidirectional_barplot.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    ValueError</span>
<span class="sd">        If no significant differentially expressed genes are found in the data.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    matplotlib.axes.Axes</span>
<span class="sd">        Axes object.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>

<span class="sd">        import sctoolbox.tools as tl</span>
<span class="sd">        adata.obs[&quot;groups&quot;] = np.random.choice([&quot;G1&quot;, &quot;G2&quot;, &quot;G3&quot;], size=adata.shape[0])</span>
<span class="sd">        pairwise_table = tl.marker_genes.pairwise_rank_genes(adata, foldchange_threshold=0.2, groupby=&quot;groups&quot;)</span>

<span class="sd">        pl.plot_differential_genes(pairwise_table)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">group_columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">rank_table</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="s2">&quot;_group&quot;</span> <span class="ow">in</span> <span class="n">col</span><span class="p">]</span>

    <span class="n">info</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">group_columns</span><span class="p">:</span>
        <span class="n">m</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s2">&quot;(.+)/(.+)_group&quot;</span><span class="p">,</span> <span class="n">col</span><span class="p">)</span>  <span class="c1"># tuple(col.split(&quot;_&quot;)[0].split(&quot;/&quot;))</span>
        <span class="n">contrast</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">([</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">2</span><span class="p">)])</span>

        <span class="n">counts</span> <span class="o">=</span> <span class="n">rank_table</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">all</span><span class="p">(</span><span class="n">x</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">counts</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;C1&#39;</span><span class="p">,</span> <span class="s1">&#39;C2&#39;</span><span class="p">]):</span>
            <span class="n">info</span><span class="p">[</span><span class="n">contrast</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;left_value&quot;</span><span class="p">:</span> <span class="n">counts</span><span class="p">[</span><span class="s2">&quot;C1&quot;</span><span class="p">],</span> <span class="s2">&quot;right_value&quot;</span><span class="p">:</span> <span class="n">counts</span><span class="p">[</span><span class="s2">&quot;C2&quot;</span><span class="p">]}</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">info</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No significant differentially expressed genes in the data. Abort.&quot;</span><span class="p">)</span>

    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">info</span><span class="p">,</span> <span class="n">orient</span><span class="o">=</span><span class="s2">&quot;index&quot;</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;left_label&quot;</span><span class="p">,</span> <span class="s2">&quot;right_label&quot;</span><span class="p">])</span>

    <span class="n">ax</span> <span class="o">=</span> <span class="n">bidirectional_barplot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="n">save</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Number of genes&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">ax</span></div>



<div class="viewcode-block" id="plot_gene_correlation">
<a class="viewcode-back" href="../../../API/plotting.html#sctoolbox.plotting.marker_genes.plot_gene_correlation">[docs]</a>
<span class="nd">@beartype</span>
<span class="k">def</span> <span class="nf">plot_gene_correlation</span><span class="p">(</span><span class="n">adata</span><span class="p">:</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span>
                          <span class="n">ref_gene</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                          <span class="n">gene_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="nb">str</span><span class="p">,</span>
                          <span class="n">ncols</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span>
                          <span class="n">figsize</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span> <span class="o">|</span> <span class="nb">float</span><span class="p">,</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">float</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                          <span class="n">save</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                          <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="n">Any</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plot the gene expression of one reference gene against the expression of a set of genes.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    adata : sc.AnnData</span>
<span class="sd">        An annotated data matrix object containing counts in .X.</span>
<span class="sd">    ref_gene : str</span>
<span class="sd">        Reference gene to which other genes are comapred to.</span>
<span class="sd">    gene_list : list[str] | str</span>
<span class="sd">        A list of genes to show expression for.</span>
<span class="sd">    ncols : int, default 3</span>
<span class="sd">        Number of columns in plot grid.</span>
<span class="sd">    figsize : Optional[Tuple[int | float, int | float]], default None</span>
<span class="sd">        Control the size of the output figure, e.g. (6,10).</span>
<span class="sd">    save : Optional[str], default None</span>
<span class="sd">        Save the figure to a file.</span>
<span class="sd">    **kwargs : Any</span>
<span class="sd">        Additional arguments passed to seaborn.regplot.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    Iterable[matplotlib.axes.Axes]</span>
<span class="sd">        List containing all axis objects.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    .. plot::</span>
<span class="sd">        :context: close-figs</span>

<span class="sd">        gene_list=(&quot;HES4&quot;, &quot;PRMT2&quot;, &quot;ITGB2&quot;)</span>
<span class="sd">        pl.plot_gene_correlation(adata, &quot;SUMO3&quot;, gene_list)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gene_list</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">gene_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">gene_list</span><span class="p">]</span>

    <span class="c1"># Find out how many rows we need</span>
    <span class="n">nrows</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)</span> <span class="o">/</span> <span class="n">ncols</span><span class="p">))</span>

    <span class="k">if</span> <span class="n">figsize</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="n">ncols</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span> <span class="n">nrows</span> <span class="o">*</span> <span class="mi">3</span><span class="p">)</span>

    <span class="n">fig</span><span class="p">,</span> <span class="n">axarr</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">ncols</span><span class="o">=</span><span class="n">ncols</span><span class="p">,</span> <span class="n">nrows</span><span class="o">=</span><span class="n">nrows</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
    <span class="n">axes_list</span> <span class="o">=</span> <span class="n">axarr</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>

    <span class="c1"># Get expression values of reference gene</span>
    <span class="n">ref</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">ref_gene</span><span class="p">]</span><span class="o">.</span><span class="n">to_df</span><span class="p">()[</span><span class="n">ref_gene</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_list</span><span class="p">):</span>
        <span class="n">ax</span> <span class="o">=</span> <span class="n">axes_list</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">gene_expr</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">to_df</span><span class="p">()[</span><span class="n">gene</span><span class="p">]</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">regplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">ref</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">gene_expr</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

    <span class="c1"># Hide axes not used</span>
    <span class="k">for</span> <span class="n">ax</span> <span class="ow">in</span> <span class="n">axes_list</span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_list</span><span class="p">):]:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">axes_list</span> <span class="o">=</span> <span class="n">axes_list</span><span class="p">[:</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)]</span>

    <span class="c1"># Make plots square</span>
    <span class="k">for</span> <span class="n">ax</span> <span class="ow">in</span> <span class="n">axes_list</span><span class="p">:</span>
        <span class="n">_make_square</span><span class="p">(</span><span class="n">ax</span><span class="p">)</span>

    <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>

    <span class="c1"># Save figure if chosen</span>
    <span class="k">if</span> <span class="n">save</span><span class="p">:</span>
        <span class="n">_save_figure</span><span class="p">(</span><span class="n">save</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">axes_list</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Loosolab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>