{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "728e1d46",
   "metadata": {
    "editable": false,
    "hide_input": true,
    "init_cell": true
   },
   "outputs": [],
   "source": [
    "from sctoolbox.utils.jupyter import bgcolor, _compare_version\n",
    "\n",
    "# change the background of input cells\n",
    "bgcolor(\"PowderBlue\", select=[2, 4, 5])\n",
    "\n",
    "nb_name = \"99-report.ipynb\"\n",
    "_compare_version(nb_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4cec3c5",
   "metadata": {},
   "source": [
    "# 01 - Analysis report\n",
    "<hr style=\"border:2px solid black\"> </hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c0c3a6d",
   "metadata": {},
   "source": [
    "## 1 - Description\n",
    "This notebook creates an analysis report PowerPoint by collecting key information, e.g. plots, versions and thresholds, created during the individual analysis steps and combines them into a comprehensive report detailing the whole analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "613ede07",
   "metadata": {},
   "source": [
    "______"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5370836",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-07-24T06:21:23.347393Z",
     "start_time": "2025-07-24T06:21:23.344391Z"
    }
   },
   "source": [
    "## 2 - Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "810050e9",
   "metadata": {
    "editable": false
   },
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import sctoolbox.tools.report as sc_report"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e75188d9",
   "metadata": {},
   "source": [
    "______"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d488738d",
   "metadata": {},
   "source": [
    "## 3 - Report options and content selection\n",
    "<hr style=\"border:2px solid black\"> </hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "950d391d",
   "metadata": {},
   "source": [
    "### 3.1 Global report settings\n",
    "The following box provides options concerning the whole report.\n",
    "\n",
    "|     Name     | Description | Default |\n",
    "|--------------|-------------|---------|\n",
    "|`dataset_name`|The name of the dataset. Will be displayed on the title slide.|-|\n",
    "| `report_dir` |The directory that contains all of the report information.|`../report/`|\n",
    "|  `template`  |A PowerPoint providing template slides. Used for the slide design.|`scRNA-template.pptx`|\n",
    "| `slide_size` |Set the slide format. Can be “standard”, “widescreen”, “a4-portait”, “a4-landscape” or a tuple of numbers indicating (height, width) in cm.|`widescreen`|\n",
    "| `max_pixels` |The maximum number of pixels an image is allowed to have. Images exceeding the size are automatically scaled down.|`5e7`|\n",
    "|  `file_ext`  | A list of the file extensions that will be used to fill the slides with content.|`[\"png\", \"md\", \"txt\"]`|\n",
    "\n",
    "<h1><center>⬐ Fill in input data here ⬎</center></h1>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af2a76c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_name = \"ext12345\"\n",
    "\n",
    "# advanced options\n",
    "report_dir = \"../report/\"\n",
    "template = \"scRNA-template.pptx\"\n",
    "slide_size = \"widescreen\"\n",
    "max_pixels = 5e7\n",
    "file_ext = [\"png\", \"md\", \"txt\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0098f38f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-07-24T06:47:44.706545Z",
     "start_time": "2025-07-24T06:47:44.703593Z"
    }
   },
   "source": [
    "### 3.2 Sections\n",
    "Select the sections that should be displayed in the final PowerPoint report.\n",
    "\n",
    "**No changes required unless the section order or the displayed section names need changes or a new section should be included.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "172275e7",
   "metadata": {
    "editable": false
   },
   "outputs": [],
   "source": [
    "print(\"Found the following sections to report:\")\n",
    "for sec in sorted(list(Path(report_dir).glob(\"*\"))):\n",
    "    if not Path(sec).is_dir():\n",
    "        continue\n",
    "\n",
    "    print(f\"    - {sec.name}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "266a4691",
   "metadata": {},
   "source": [
    "**Select** and **order** the sections of the final report PowerPoint by editing the dictionary below.  \n",
    "\n",
    "`\"<section_id>\": \"<title displayed in the slide>\"`  \n",
    "The *section_id* has to be in the list above. Sections that are specified below but not available are ignored."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1c894e9",
   "metadata": {},
   "source": [
    "<h1><center>⬐ Fill in input data here ⬎</center></h1>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a80dc827",
   "metadata": {},
   "outputs": [],
   "source": [
    "# a dict to define section order and the displayed titles\n",
    "section_titles = {\n",
    "    \"01_assembly\": \"Dataset Assembly\",\n",
    "    \"02_QC\": \"Quality Control\",\n",
    "    \"03_batch_correction\": \"Normalization and Noise Reduction\",\n",
    "    \"04_clustering\": \"Embedding and Clustering\",\n",
    "    \"group_markers\": \"Group Marker prediction\",\n",
    "    \"annotation\": \"Cell Type Annotation\",\n",
    "    \"GSEA\": \"Gene Set Enrichment Analysis\",\n",
    "    \"0A1_receptor_ligand\": \"Receptor-Ligand Analysis\",\n",
    "    \"0A2_receptor_ligand_differences\": \"Receptor-Ligand Difference Analysis\",\n",
    "    \"0B_velocity\": \"RNA Velocity Analysis\",\n",
    "    \"proportion_analysis\": \"Proportion Analysis\",\n",
    "    \"pseudotime_analysis\": \"Pseudotime Trajectory Analysis\"\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d5e031f",
   "metadata": {},
   "source": [
    "### 3.3 Slide options\n",
    "<hr style=\"border:1px solid black\"> </hr>\n",
    "The dictionary below holds all the information needed to format and populate the slides of the report.\n",
    "\n",
    "**No changes needed unless slide order or slide content needs to be changed or new slides are required.**\n",
    "\n",
    "#### Structure\n",
    "The top-level keys are the *section ids*, the corresponding value can either be a dictionary or a list containing dictionaries. The bottom-level dictionary provides the slide options. Available slide options can be found [here (PowerPointReport.add_slide)](https://loosolab.pages.gwdg.de/software/pptreport/API/index.html#pptreport.powerpointreport.PowerPointReport.add_slide).\n",
    "\n",
    "<h1><center>⬐ Fill in input data here ⬎</center></h1>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75608139",
   "metadata": {},
   "outputs": [],
   "source": [
    "slide_section_args = {\n",
    "    \"01_assembly\": {\n",
    "        \"title\": \"Dataset Assembly\",\n",
    "        \"content\": [\n",
    "            \"../report/01_assembly/01_assembly.png\",\n",
    "            \"../report/01_assembly/01_dataset_overview.md\",\n",
    "            \"../report/01_assembly/01_obs.png\",\n",
    "            \"../report/01_assembly/01_var.png\"\n",
    "        ],\n",
    "        \"notes\": [\n",
    "            \"A dataset may be constructed from multiple parts (samples, sequenceing batches, etc.). The first table shows the individual parts and their location from which the dataset is constructed.\",\n",
    "            \"\\n\",\n",
    "            \"The text provides a short overview of the dataset dimensions and available information. In a typical scRNA experiment each observation translates to a cell and each variable to a gene. This means that observation information contains information about each cell, e.g. the timepoint a cell belongs to. Likewise, provides variable information details about individual genes, for example how many cells had expression of a certain gene.\",\n",
    "            \"Datasets can contain additional layers, i.e. data matrixes, such as raw expression or normalized expression values.\"\n",
    "            \"The last two tables show a detailed view on the observation (obs) and variable (var) of the dataset.\",\n",
    "        ],\n",
    "        \"content_alignment\": \"left\",\n",
    "        \"content_layout\": \"vertical\"\n",
    "    },\n",
    "    \"02_QC\": [\n",
    "        {\n",
    "            \"title\": \"Gender prediction\",\n",
    "            \"notes\": \"The gender of each sample is predicted based on the expression of a female gene. Female samples are shown with a red background.\",\n",
    "            \"content\": \"../report/02_QC/01*\"\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Cell filter I\",\n",
    "            \"content\": \"../report/02_QC/02*\",\n",
    "            \"notes\": [\n",
    "                \"Low quality cells are removed from the dataset by applying thresholds to quality metrics. This is either done using global thresholds or thresholds for each individual sample.\",\n",
    "                \"\\n\",\n",
    "                \"Common quality metrics:\",\n",
    "                \"n_genes:\t\t\tThe number of genes associated with a barcode (cell).\",\n",
    "                \"total_counts:\t\t\tThe total amount of reads detected for this barcode (cell).\",\n",
    "                \"log1p_total_counts:\t\tSame as above but on a logarithmic scale.\",\n",
    "                \"total_counts_is_mito:\t\tThe total amount of reads associated with mitochondrial genes.\",\n",
    "                \"log1p_total_counts_is_mito:\t\tSame as above but on a logarithmic scale.\",\n",
    "                \"pct_counts_is_mito:\t\tPercentage of mitochondrial reads per barcode (cell).\",\n",
    "                \"total_counts_is_ribo:\t\tThe total amount of reads associated with ribosomal genes.\",\n",
    "                \"log1p_total_counts_is_ribo:\t\tSame as above but on a logarithmic scale.\",\n",
    "                \"pct_counts_is_ribo:\t\tPercentage of ribosomal reads per barcode (cell).\",\n",
    "                \"total_counts_is_gender:\t\tThe total amount of reads associated with gender genes.\",\n",
    "                \"log1p_total_counts_is_gender:\tSame as above but on a logarithmic scale.\",\n",
    "                \"pct_counts_is_gender:\t\tPercentage of gender related reads per barcode (cell).\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Cell filter II\",\n",
    "            \"content\": \"../report/02_QC/03*\",\n",
    "            \"notes\": \"This plot provides a pairwise view of the quality metrics, their distributions, and the respective thresholds.\"\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Cell filter III\",\n",
    "            \"content\": [\n",
    "                \"../report/02_QC/04_cell_filter_impact.png\",\n",
    "                \"../report/02_QC/04_doublet_info.txt\",\n",
    "                \"../report/02_QC/04_cell_filter_info.txt\"\n",
    "            ],\n",
    "            \"notes\": [\n",
    "                \"The UpSetPlot provides the impact each threshold has on the dataset. This is shown for each metric alone and in combination with other metrics to estimate the overlap between the filters.\",\n",
    "                \"The final amount of filtered cells is given in the text.\",\n",
    "                \"The dataset may also be filtered for doublets, observations that contain more than one cell, to avoid skewing the following analysis.\"\n",
    "            ],\n",
    "            \"content_layout\": \"vertical\",\n",
    "            \"height_ratios\": [0.8, 0.1, 0.1]\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Gene filter I\",\n",
    "            \"content\": \"../report/02_QC/05*\",\n",
    "            \"notes\": [\n",
    "                \"Similar to cell filtering, genes may also be removed from the dataset with the aid of quality metrics.\",\n",
    "                \"\\n\",\n",
    "                \"Common quality metrics:\",\n",
    "                \"n_cells_by_counts:\tThe number of cells that contain reads associated with the gene.\",\n",
    "                \"mean_counts:\t\tThe mean amount of reads over all cells.\",\n",
    "                \"log1p_mean_counts:\tSame as above but on a logarithmic scale.\",\n",
    "                \"pct_dropout_by_counts:\tPercentage of cells this gene does not appear in.\",\n",
    "                \"total_counts:\t\tThe total amount of reads associated to this gene.\",\n",
    "                \"log1p_total_counts:\tSame as above but on a logarithmic scale.\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Gene filter II\",\n",
    "            \"notes\": \"Genes are categorized into mitochondrial, ribosomal or gender related. This is either based on a list of genes or a search string e.g. genes starting with 'mt' for mitochondrial genes. Genes related to mitochondrial, ribosomal or gender functions are optionally filtered from the dataset.\",\n",
    "            \"content\": [\n",
    "                \"../report/02_QC/06_gene_labels.png\",\n",
    "                \"../report/02_QC/06_*info.txt\"\n",
    "            ],\n",
    "            \"content_layout\": \"vertical\",\n",
    "            \"height_ratios\": [0.7, 0.1, 0.1, 0.1]\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Dataset denoising\",\n",
    "            \"content\": \"../report/02_QC/07*\",\n",
    "            \"notes\": \"We remove ambient RNA and technical noise from the count matrix using scAR. The tool estimates the ambient profile by averaging cell-free droplets. An autoencoder neural network later corrects the count matrix.\"\n",
    "        }\n",
    "    ],\n",
    "    \"03_batch_correction\": [\n",
    "        {\n",
    "            \"title\": \"Cell Cycle\",\n",
    "            \"notes\": \"The cell cycle phase of each cell is predicted based on the activity of known cell cycle genes. The plot shows the distribution of cell cycle phases across the samples.\",\n",
    "            \"content\": \"../report/03_batch_correction/01*\"\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"PCA\",\n",
    "            \"content\": \"../report/03_batch_correction/02*\",\n",
    "            \"notes\": [\n",
    "                \"A principle component analysis (PCA) is computed to reduce the dimensionality by reorganizing the data into principle components (PCs). The plot shows the first two PCs with each dot (cell) colored for quality metrics.\",\n",
    "                \"There should be no gradient or grouping visible to provide a clean biological signal.\"\n",
    "            ]\n",
    "            \n",
    "        },\n",
    "        {\n",
    "            \"title\": \"PCA - component subset\",\n",
    "            \"notes\": [\n",
    "                \"Each principle component (PC) is evaluated based on its correlation to quality metrics and variance. PCs exceeding one of the two thresholds are filtered from the dataset, indicated as a lightgrey bar.\",\n",
    "                \"Removing PCs helps to reduce technical or other unwanted noise to create a clean biological signal.\"\n",
    "            ],\n",
    "            \"content\": \"../report/03_batch_correction/03*\"\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Batch correction\",\n",
    "            \"content\": \"../report/03_batch_correction/04*\",\n",
    "            \"width_ratios\": [0.4, 0.6],\n",
    "            \"notes\": [\n",
    "                \"Batch correction aims to reduce technical differences in the dataset. For example, two separate sequencing runs may show as two distinct groups despite being biologically similar. There are multiple batch correction methods available with different concepts causing them to perform differently for each dataset. The plot compares multiple batch correction methods (columns) to select the optimal one for the dataset.\",\n",
    "                \"The plot shows PCAs and preliminary UMAPs to assess how well the batches are mixed after batch correction. The bottom plots show the LISI score, a measurement for the 'mixedness' of the batches (maximum = number of batches). A high LISI is not a sufficient indicator, the PCA and UMAP shapes should be included into the consideration as well.\"\n",
    "            ]\n",
    "        }\n",
    "        \n",
    "    ],\n",
    "    \"04_clustering\": [\n",
    "        {\n",
    "            \"title\": \"Embedding I\",\n",
    "            \"content\": \"../report/04_clustering/01*\",\n",
    "            \"width_ratios\": [0.7, 0.3],\n",
    "            \"notes\": [\n",
    "                \"The final dimension reduction step is to compute a 2D embedding of the dataset. This can either be a Uniform Manifold Approximation and Projection (UMAP) or distributed stochastic neighbor embedding (t-SNE).\",\n",
    "                \"Both methods rely on two main parameters to compute the respective embedding, which are shown in the table. The plot shows the embedding colored for various quality measurements. Ideally, there shouldn't be a gradient or grouping visible, unless expected through the experimental design.\",\n",
    "                \"Note: Dot = Cell\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Embedding II\",\n",
    "            \"content\": \"../report/04_clustering/02*\",\n",
    "            \"notes\": \"This plot shows how the cells for each sample/ condition are distributed across the dataset.\"\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Clustering I\",\n",
    "            \"content\": \"../report/04_clustering/03*\",\n",
    "            \"notes\": \"The embedding colored for sample/ condition (left) and cells assigned to clusters (right). Clustering is typically done using the leiden algorithm.\"\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Clustering II\",\n",
    "            \"content\": \"../report/04_clustering/04*\",\n",
    "            \"notes\": [\n",
    "                \"Here, the number of cells per clusters is shown. The color shows the proportion of cells assigned to samples/ conditions.\",\n",
    "                \"Small clusters or clusters predominantly from one sample/ condition may be outliers or part of a bigger cluster, as such, they may be removed or combined with another cluster. However, this is tied to the expectations of the individual dataset and must be decided on a case-by-case basis.\"\n",
    "            ]\n",
    "        }\n",
    "    ],\n",
    "    \"0A1_receptor_ligand\": [\n",
    "        {\n",
    "            \"title\": \"Database\",\n",
    "            \"content\": \"../report/0A1_receptor_ligand/01*\",\n",
    "            \"content_layout\": \"vertical\",\n",
    "            \"height_ratios\": [0.1, 0.9],\n",
    "            \"notes\": \"Receptor-ligand analysis requires a database providing the receptor-ligand gene pairs. The table shows an excerpt from the database selected for this analysis.\"\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Interaction overview\",\n",
    "            \"content\": \"../report/0A1_receptor_ligand/02*\",\n",
    "            \"notes\": [\n",
    "                \"The Cyclone plot shows the amount of interactions between groups, e.g. cell types. An interaction is counted if the receptor- and ligand-gene expression are enriched for the respective groups. The inner-grey shell provides the number of cells per group, while the outermost shell shows the top receptor and ligand genes for the respective group.\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Interaction detail\",\n",
    "            \"content\": \"../report/0A1_receptor_ligand/03*\",\n",
    "            \"notes\": [\n",
    "                \"The Connection plot provides a qualitative view on individual interactions. It shows the percentage of cell expressing the gene (dot size), the gene enrichment for the respective group (dot color) and interaction strength (line width).\",\n",
    "                \"The higher each value the better.\"\n",
    "            ]\n",
    "        }\n",
    "    ],\n",
    "    \"0A2_receptor_ligand_differences\": [\n",
    "        {\n",
    "            \"title\": \"Pairwise Differences\",\n",
    "            \"notes\": [\n",
    "                \"The network shows the top interaction differences, using quantile rank differences, between two conditions. Blue lines indicate increased interaction in one condition, while red lines indicate increased interaction in the other (the closer to 1 or -1 the bigger the difference).\",\n",
    "                \"The node color and shape define the cell type(s) involved in the interaction. Hubs are centered around a receptor or ligand gene that interacts with multiple partners/ cell types.\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Temporal Progression\",\n",
    "            \"notes\": [\n",
    "                \"These plots focus on the differences between continuous conditions, e.g. day 1 <-> day 2, day 2 <-> day 3. They are the same as the pairwise difference plots, apart from their focus on specific combinations.\",\n",
    "                \"The network shows the top interaction differences, using quantile rank differences, between two conditions. Blue lines indicate increased interaction in one condition, while red lines indicate increased interaction in the other (the closer to 1 or -1 the bigger the difference).\",\n",
    "                \"The node color and shape define the cell type(s) involved in the interaction. Hubs are centered around a receptor or ligand gene that interacts with multiple partners/ cell types.\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Interaction over Time\",\n",
    "            \"notes\": \"This plot shows how the expression of receptor and ligand genes changes over time in specific cell types.\"\n",
    "        }\n",
    "    ],\n",
    "    \"0B_velocity\": [\n",
    "        {\n",
    "            \"title\": \"Splicing Rate\",\n",
    "            \"notes\": [\n",
    "                \"The plot shows the rate between spliced and unspliced RNA.\",\n",
    "                \"Depending on the sequencing method the rate is typically between 10-25% unspliced.\",\n",
    "                \"https://scvelo.readthedocs.io/en/stable/VelocityBasics.html#Load-the-Data\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Velocity Confidence\",\n",
    "            \"content_layout\": \"vertical\",\n",
    "            \"notes\": [\n",
    "                \"The velocity length describes the differentiation speed of a cell, while the confidence describes the coherence of the velocity vectors with neighboring cells, i.e. if neighboring cells differentiate in the same direction.\",\n",
    "                \"https://scvelo.readthedocs.io/en/stable/VelocityBasics.html#Speed-and-coherence\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"RNA Velocity\",\n",
    "            \"notes\": \"The arrows on top of the embedding provide a view on the differential/ developmental path the different cell population may follow over time.\"\n",
    "        }\n",
    "    ],\n",
    "    \"group_markers\": [\n",
    "        {\n",
    "            \"title\": \"Top Marker Genes\",\n",
    "            \"content_layout\": \"vertical\",\n",
    "            \"notes\": [\n",
    "                \"The top marker genes for each group, e.g. a clustering or cell types, presented in heatmap and dotplot style.\",\n",
    "                \"Good quality marker genes are exclusively expressed in one group, meaning they show up in only one row of the plots.\",\n",
    "                \"A dataset with groups that are well separated by gene markers shows a 'stair-like' pattern in the plots.\",\n",
    "                \"Groups with shared markers might be combined to provide better separated groups.\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Top Marker Expression per Group I\",\n",
    "            \"notes\": [\n",
    "                \"The top marker for each group (row). The first embedding shows the location of the group and the following plots show the respective gene expression binned into hexagonal tiles.\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Top Marker Expression per Group II\",\n",
    "            \"notes\": [\n",
    "                \"The top marker for each group (row). The first embedding shows the location of the group and the following plots show which cells express the gene.\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Genes of Interest I\",\n",
    "            \"notes\": \"The mean expression of manually selected genes and a separate embedding showing the expression of each gene summarized into hexagonal tiles.\"\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Genes of Interest II\",\n",
    "            \"notes\": [\n",
    "                \"The expression of the manually selected genes next to each group (row).\",\n",
    "                \"The first embedding shows the location of the group and the following plots show the respective gene expression binned into hexagonal tiles.\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Genes of Interest III\",\n",
    "            \"notes\": [\n",
    "                \"The expression of the manually selected genes next to each group (row).\",\n",
    "                \"The first embedding shows the location of the group and the following plots show which cells express the gene.\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Condition Markers\",\n",
    "            \"notes\": [\n",
    "                \"This slide shows marker genes that separate the condition within a group. A dotplot is shown for each group where condition markers are found.\"\n",
    "            ]\n",
    "        },\n",
    "    ],\n",
    "    \"annotation\": [\n",
    "        {\n",
    "            \"title\": \"Cell Type Annotation\",\n",
    "            \"content\": [\n",
    "                \"../report/annotation/01_marker_genes_*\",\n",
    "                \"../report/annotation/01_Comparison_of_cell_type_annotations.png\"\n",
    "            ],\n",
    "            \"content_layout\": \"vertical\",\n",
    "            \"notes\": [\n",
    "                \"The dotplot shows the groups to assign cell types and their respective top gene markers.\",\n",
    "                \"The table shows different annotation versions (columns) based on different algorithms. The Annotation is either based on SCSA or on the SC-Framework internal marker repo (MR).\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Cell Type Annotation\",\n",
    "            \"notes\": [\n",
    "                \"The first embedding is colored for the initial groups, while the following embeddings show the different annotation versions from the previous slides.\"\n",
    "            ]\n",
    "        }\n",
    "    ],\n",
    "    \"GSEA\": [\n",
    "        {\n",
    "            \"title\": \"Top Pathways per Group\",\n",
    "            \"notes\": [\n",
    "                \"The dotplot shows the top enriched pathways per cluster. The size of the dot indicates the fraction of genes in the cluster that match the pathway and the color of the dot indicates statistical significance (higher is better).\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Pathway related Genes\",\n",
    "            \"notes\": \"The term dotplot focuses on a single term/pathway and thus shows individual genes instead of pathways on the y-axis. A Z-Score is applied to the mean gene expression per cluster to highlight differences in expression between the clusters (x-axis).\"\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Pathway Gene-sharing Network\",\n",
    "            \"notes\": \"The network plot shows connections between enriched pathways per cluster. The node size corresponds to the percentage of gene overlap in a certain term of interest. The colour of the node corresponds to the significance of the enriched terms and the edge size corresponds to the number of genes that overlap between two connected nodes.\"\n",
    "        }\n",
    "    ],\n",
    "    \"proportion_analysis\": [\n",
    "        {\n",
    "            \"title\": \"Proportion Analysis\",\n",
    "            \"notes\": [\n",
    "                \"The table shows the proportions for each group of the dataset (rows) split by the conditions.\",\n",
    "                \"\\n\",\n",
    "                \"Column descriptions:\",\n",
    "                \"baseline_props:\tThe proportion of cells belonging to the respective group.\",\n",
    "                \"mean_props_*:\t\tThe proportion of cells belonging to the respective group and condition.\",\n",
    "                \"t_/f_statistics:\t\tt-test or ANOVA statistical significance. Higher means bigger difference between conditions (higher is better).\",\n",
    "                \"p_values:\t\tSignificance of proportional changes between conditions.\",\n",
    "                \"adjusted_p_values:\tSignificance corrected for the number of groups.\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Proportion Analysis\",\n",
    "            \"notes\": [\n",
    "                \"The plots show the proportion (amount) of cells of each group (e.g. cell type) allocated to each of the conditions. The p-value on top of each plot describes whether there is a significant change in proportion between any of the conditions. In case there are no replicates Scanpro will create simulated replicates (similar to random subsamples) to improve statistical robustness. The replicates are either shown as separate entities (upper plot) or as a box-distribution (lower plot). E.g. for a dataset where clustering_col = 'celltype' and condition_col = 'injury' a plot with low p-value can be interpreted as 'Cell Type X shows a high change in the number of cells between injured and healthy' and a high p-value can be interpreted as 'Cell Type Y shows a low change in the number of cells between injured and healthy'.\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Proportion Analysis\",\n",
    "            \"notes\": [\n",
    "                \"The plots show the proportion (amount) of cells of each group (e.g. cell type) allocated to each of the conditions. The p-value on top of each plot describes whether there is a significant change in proportion between any of the conditions. In case there are no replicates Scanpro will create simulated replicates (similar to random subsamples) to improve statistical robustness. The replicates are either shown as separate entities (upper plot) or as a box-distribution (lower plot). E.g. for a dataset where clustering_col = 'celltype' and condition_col = 'injury' a plot with low p-value can be interpreted as 'Cell Type X shows a high change in the number of cells between injured and healthy' and a high p-value can be interpreted as 'Cell Type Y shows a low change in the number of cells between injured and healthy'.\"\n",
    "            ]\n",
    "        }\n",
    "    ],\n",
    "    \"pseudotime_analysis\": [\n",
    "        {\n",
    "            \"title\": \"Trajectory and Root Selection\",\n",
    "            \"notes\": [\n",
    "                \"A principal graph is inferred from the embedding. Cells are assigned to a node by a value between 0 and 1, which allows the use of probabilistic mappings to account for variability. In case of the clustering showing cell types this can be interpreted as potential differentiation paths a cell could take. However, they are not yet in a pseudotime context so there is no direction in the presented graph.\",\n",
    "                \"The pseudotime start point is either defined as one of the nodes or the expression strength and density of a gene.\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Pseudotime Trajectory I\",\n",
    "            \"notes\": \"The same graph as before but with the directions defined through the previously selected root.\"\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Pseudotime Trajectory II\",\n",
    "            \"notes\": [\n",
    "                \"1. Segments are edges (or branches) in the trajectory tree or graph. Each segment represents a developmental transition between two milestones (nodes), corresponding to cells that are transitioning between states.\",\n",
    "                \"2. The milestones are the nodes of the trajectory tree. Each milestone represents a key point or “state” during the differentiation process.\",\n",
    "                \"3. The pseudotime values defines potential cell trajectories based on the root as starting points. The grey area is the starting location. The time progresses from light to dark red.\",\n",
    "                \"4. The grouping is shown as a reference to the other plots to identify e.g. potential biological differentiation paths.\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"title\": \"Pseudotime Trajectory III\",\n",
    "            \"notes\": [\n",
    "                \"The dendrogram shows the branching of the segments as depicted in the previous slide.\",\n",
    "                \"The dendrogram is colored by the selected grouping to e.g. show the emergence of cell types over pseudotime.\"\n",
    "            ]\n",
    "        }\n",
    "    ]\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd38952f",
   "metadata": {},
   "source": [
    "______"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ad20092",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-07-24T06:21:23.347393Z",
     "start_time": "2025-07-24T06:21:23.344391Z"
    }
   },
   "source": [
    "## 4 - Generate the report\n",
    "This part combines all the information and settings from above into the final PowerPoint structure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6f6240b",
   "metadata": {
    "editable": false
   },
   "outputs": [],
   "source": [
    "report = sc_report.generate_report(\n",
    "    dataset_name=dataset_name,\n",
    "    section_titles=section_titles,\n",
    "    slide_sec_kwargs=slide_section_args,\n",
    "    file_ext=file_ext,\n",
    "    template=template,\n",
    "    slide_format=slide_size,\n",
    "    report_dir=report_dir,\n",
    "    max_pixels=max_pixels,\n",
    "    method_template=\"RNA\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af8f2391",
   "metadata": {},
   "source": [
    "______"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33ea5f69",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-07-24T06:21:23.347393Z",
     "start_time": "2025-07-24T06:21:23.344391Z"
    }
   },
   "source": [
    "## 5 - Save Report\n",
    "Render the PowerPoint report and save it to the disk."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b900fa0d",
   "metadata": {
    "editable": false
   },
   "outputs": [],
   "source": [
    "# Save presentation\n",
    "out_path = Path(report_dir) / \"report.pptx\"\n",
    "\n",
    "report.save(str(out_path))\n",
    "\n",
    "display(f\"Saved to: {out_path}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sctoolbox",
   "language": "python",
   "name": "sctoolbox"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  },
  "sc_framework": {
   "version": "0.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
