Search.setIndex({"alltitles": {"0.1 (22-05-2023)": [[4, "id15"]], "0.1.1 (24-05-2023)": [[4, "id14"]], "0.2 (30-08-2023)": [[4, "id13"]], "0.3 (30-11-2023)": [[4, "id9"]], "0.4 (31-1-24)": [[4, "id7"]], "0.5 (04-03-24)": [[4, "id5"]], "0.6 (27-03-24)": [[4, "id3"]], "0.6.1 (28-03-24)": [[4, "id2"]], "0.7.0 (23-04-24)": [[4, "id1"]], "01 - Assembling or loading anndata object": [[5, "01---Assembling-or-loading-anndata-object"], [16, "01---Assembling-or-loading-anndata-object"]], "02 - QC and filtering": [[6, "02---QC-and-filtering"], [17, "02---QC-and-filtering"]], "03 - Batch effect correction and comparisons": [[7, "03---Batch-effect-correction-and-comparisons"], [18, "03---Batch-effect-correction-and-comparisons"]], "04 - Embedding and clustering": [[8, "04---Embedding-and-clustering"], [19, "04---Embedding-and-clustering"]], "0A - Marker gene computation and analysis": [[20, "0A---Marker-gene-computation-and-analysis"]], "0B - Receptor-Ligand Analysis": [[21, "0B---Receptor-Ligand-Analysis"]], "0C - Velocity Analysis": [[22, "0C---Velocity-Analysis"]], "1 - Description": [[5, "1---Description"], [6, "1---Description"], [7, "1---Description"], [8, "1---Description"], [10, "1---Description"], [12, "1---Description"], [13, "1---Description"], [14, "1---Description"], [16, "1---Description"], [17, "1---Description"], [18, "1---Description"], [19, "1---Description"], [20, "1---Description"], [21, "1---Description"], [22, "1---Description"]], "1. .h5ad file": [[16, "1.-.h5ad-file"]], "1.1 - Parameter Overview": [[10, "1.1---Parameter-Overview"]], "1.1 - Preparing for cellxgene": [[12, "1.1---Preparing-for-cellxgene"]], "1.1 - Pseudotime Analysis": [[14, "1.1---Pseudotime-Analysis"]], "1.1 Embedding": [[8, "1.1-Embedding"], [19, "1.1-Embedding"]], "1.1 Sources to create the anndata object:": [[5, "1.1-Sources-to-create-the-anndata-object:"]], "1.1 Velocity Analysis": [[22, "1.1-Velocity-Analysis"]], "1.1.1 - Essential input data": [[10, "1.1.1---Essential-input-data"]], "1.1.2 - Marker List Assembly: wrap.create_multiple_marker_lists": [[10, "1.1.2---Marker-List-Assembly:-wrap.create_multiple_marker_lists"]], "1.1.3 - Annotation Parameters: wrap.run_annotation": [[10, "1.1.3---Annotation-Parameters:-wrap.run_annotation"]], "1.2 - MaMPlan creation": [[12, "1.2---MaMPlan-creation"]], "1.2 - Parameters": [[14, "1.2---Parameters"]], "1.2.1 - Parameters": [[12, "1.2.1---Parameters"]], "10 - Batch correction": [[18, "10---Batch-correction"]], "10 - PAGA": [[22, "10---PAGA"]], "10 - Saving adata for next notebook": [[8, "10---Saving-adata-for-next-notebook"], [19, "10---Saving-adata-for-next-notebook"]], "10.1 - Plot overview of batch corrections": [[18, "10.1---Plot-overview-of-batch-corrections"]], "10.2 - Select the final object": [[18, "10.2---Select-the-final-object"]], "11 - Saving adata for the next notebook": [[18, "11---Saving-adata-for-the-next-notebook"]], "11 - Velocity genes": [[22, "11---Velocity-genes"]], "11.1 - Gene of interest": [[22, "11.1---Gene-of-interest"]], "12 - Saving adata": [[22, "12---Saving-adata"]], "2 - Input/output settings": [[13, "2---Input/output-settings"], [14, "2---Input/output-settings"]], "2 - Loading packages": [[20, "2---Loading-packages"]], "2 - Setup": [[5, "2---Setup"], [6, "2---Setup"], [7, "2---Setup"], [8, "2---Setup"], [12, "2---Setup"], [16, "2---Setup"], [17, "2---Setup"], [18, "2---Setup"], [19, "2---Setup"], [21, "2---Setup"], [22, "2---Setup"]], "2- Setup": [[10, "2--Setup"]], "2. star solo quant folder": [[16, "2.-star-solo-quant-folder"]], "2.2 ATAC specific quality control metrics": [[5, "2.2-ATAC-specific-quality-control-metrics"]], "3 - General Input": [[12, "3---General-Input"]], "3 - Load adata": [[21, "3---Load-adata"]], "3 - Load anndata": [[6, "3---Load-anndata"], [7, "3---Load-anndata"], [8, "3---Load-anndata"], [17, "3---Load-anndata"], [18, "3---Load-anndata"], [19, "3---Load-anndata"], [22, "3---Load-anndata"]], "3 - Loading adata": [[10, "3---Loading-adata"], [20, "3---Loading-adata"]], "3 - Loading packages": [[13, "3---Loading-packages"], [14, "3---Loading-packages"]], "3 - Read in data": [[5, "3---Read-in-data"], [16, "3---Read-in-data"]], "3. .mtx, barcode.tsv, genes.tsv": [[16, "3.-.mtx,-barcode.tsv,-genes.tsv"]], "3.1 - Option 1: Read from h5ad": [[5, "3.1---Option-1:-Read-from-h5ad"], [16, "3.1---Option-1:-Read-from-h5ad"]], "3.2 - Option 2: Assemble from preprocessing pipeline \u2018quant\u2019 folder": [[16, "3.2---Option-2:-Assemble-from-preprocessing-pipeline-'quant'-folder"]], "3.2 - Option 2: Convert from Seurat to anndata object": [[5, "3.2---Option-2:-Convert-from-Seurat-to-anndata-object"]], "3.3 - Option 3: Create an anndata object from .mtx, barcodes.tsv and genes.tsv": [[16, "3.3---Option-3:-Create-an-anndata-object-from-.mtx,-barcodes.tsv-and-genes.tsv"]], "3.4 - Option 4: Convert from Seurat to anndata object": [[16, "3.4---Option-4:-Convert-from-Seurat-to-anndata-object"]], "4 - Automatic markers per cluster using rank_genes_groups": [[20, "4---Automatic-markers-per-cluster-using-rank_genes_groups"]], "4 - Essential Input": [[10, "4---Essential-Input"]], "4 - General input": [[7, "4---General-input"], [8, "4---General-input"], [18, "4---General-input"], [19, "4---General-input"], [21, "4---General-input"], [22, "4---General-input"]], "4 - Load anndata": [[12, "4---Load-anndata"], [13, "4---Load-anndata"], [14, "4---Load-anndata"]], "4 - Prepare anndata": [[5, "4---Prepare-anndata"], [16, "4---Prepare-anndata"]], "4 - QC and filtering": [[6, "4---QC-and-filtering"], [17, "4---QC-and-filtering"]], "4. convert from R object": [[16, "4.-convert-from-R-object"]], "4.1 - Rename and delete columns": [[5, "4.1---Rename-and-delete-columns"]], "4.1 - Show STARsolo quality (optional)": [[6, "4.1---Show-STARsolo-quality-(optional)"], [17, "4.1---Show-STARsolo-quality-(optional)"]], "4.2 - Format anndata indices": [[5, "4.2---Format-anndata-indices"]], "4.2 - Label genes": [[17, "4.2---Label-genes"]], "4.2 - Remove empty cells and features": [[6, "4.2---Remove-empty-cells-and-features"]], "4.3 - Binarize matrix": [[6, "4.3---Binarize-matrix"]], "4.3 - Calculate QC metrics": [[17, "4.3---Calculate-QC-metrics"]], "4.4 - Calculate doublet scores": [[17, "4.4---Calculate-doublet-scores"]], "4.4 Calculate and remove doublets": [[6, "4.4-Calculate-and-remove-doublets"]], "4.5 - Predict sex per sample": [[17, "4.5---Predict-sex-per-sample"]], "4.6 - Cell filtering": [[6, "4.6---Cell-filtering"], [17, "4.6---Cell-filtering"]], "4.6.1 - Estimate initial thresholds automatically": [[17, "4.6.1---Estimate-initial-thresholds-automatically"]], "4.6.1 Adapt thresholds": [[6, "4.6.1-Adapt-thresholds"]], "4.6.2 - Apply gene filtering": [[6, "4.6.2---Apply-gene-filtering"]], "4.6.2 - Customize thresholds via sliders": [[17, "4.6.2---Customize-thresholds-via-sliders"]], "4.6.3 - Apply final thresholds": [[17, "4.6.3---Apply-final-thresholds"]], "4.6.3 - Show data after filtering": [[6, "4.6.3---Show-data-after-filtering"]], "4.6.4 - Show data after filtering": [[17, "4.6.4---Show-data-after-filtering"]], "4.7 - Gene filtering": [[17, "4.7---Gene-filtering"]], "4.7.1 - Customize thresholds via sliders": [[17, "4.7.1---Customize-thresholds-via-sliders"]], "4.7.2 - Apply gene filtering": [[17, "4.7.2---Apply-gene-filtering"]], "4.7.3 - Show data after filtering": [[17, "4.7.3---Show-data-after-filtering"]], "4.7.4 - Filter additional marked genes": [[17, "4.7.4---Filter-additional-marked-genes"]], "5 - Add ATAC specific metrices": [[5, "5---Add-ATAC-specific-metrices"]], "5 - Assemble marker lists": [[10, "5---Assemble-marker-lists"]], "5 - Calculate UMAP/TSNE and find best setting": [[8, "5---Calculate-UMAP/TSNE-and-find-best-setting"], [19, "5---Calculate-UMAP/TSNE-and-find-best-setting"]], "5 - Check for spliced/unspliced/ambiguous layer": [[22, "5---Check-for-spliced/unspliced/ambiguous-layer"]], "5 - Download receptor-ligand database": [[21, "5---Download-receptor-ligand-database"]], "5 - Feature processing": [[6, "5---Feature-processing"]], "5 - General Input": [[13, "5---General-Input"]], "5 - Normalization": [[7, "5---Normalization"]], "5 - Plot expression": [[20, "5---Plot-expression"]], "5 - Plot highly expressed genes": [[18, "5---Plot-highly-expressed-genes"]], "5 - Prepare adata for cellxgene": [[12, "5---Prepare-adata-for-cellxgene"]], "5 - Save filtered adata": [[17, "5---Save-filtered-adata"]], "5 - Saving the loaded anndata object": [[16, "5---Saving-the-loaded-anndata-object"]], "5 - scFates": [[14, "5---scFates"]], "5.1 - Add leiden columns": [[12, "5.1---Add-leiden-columns"]], "5.1 - Calculate Graph": [[14, "5.1---Calculate-Graph"]], "5.1 - Check barcode tag": [[5, "5.1---Check-barcode-tag"]], "5.1 - Filter additional marked features": [[6, "5.1---Filter-additional-marked-features"]], "5.1 - Plot custom gene list": [[20, "5.1---Plot-custom-gene-list"]], "5.2 - Clean up adata": [[12, "5.2---Clean-up-adata"]], "5.2 - Plot cluster marker": [[20, "5.2---Plot-cluster-marker"]], "5.2 - Plot graph": [[14, "5.2---Plot-graph"]], "5.2 - Score fragment length distributions": [[5, "5.2---Score-fragment-length-distributions"]], "5.2 - Select highly variable features": [[6, "5.2---Select-highly-variable-features"]], "5.3 - Calculate an overlap": [[5, "5.3---Calculate-an-overlap"]], "5.3 - Pseudotime": [[14, "5.3---Pseudotime"]], "5.3 - Save adata": [[12, "5.3---Save-adata"]], "5.4 - Plot Trajectory graph": [[14, "5.4---Plot-Trajectory-graph"]], "5.5 - Plot pseudotime": [[14, "5.5---Plot-pseudotime"]], "5.6 - Plot dendogram": [[14, "5.6---Plot-dendogram"]], "6 - Annotate adata": [[10, "6---Annotate-adata"]], "6 - Cell clustering": [[8, "6---Cell-clustering"]], "6 - Compute interactions": [[21, "6---Compute-interactions"]], "6 - DEG between conditions per cluster using scanpy rank gene groups": [[20, "6---DEG-between-conditions-per-cluster-using-scanpy-rank-gene-groups"]], "6 - Normalization": [[18, "6---Normalization"]], "6 - PCA and neighbors for uncorrected data": [[7, "6---PCA-and-neighbors-for-uncorrected-data"]], "6 - Plot final embedding with quality measures": [[19, "6---Plot-final-embedding-with-quality-measures"]], "6 - Preprocessing": [[22, "6---Preprocessing"]], "6 - Proportion analysis with Scanpro": [[13, "6---Proportion-analysis-with-Scanpro"]], "6 - Save filtered adata": [[6, "6---Save-filtered-adata"]], "6 - Saving adata": [[14, "6---Saving-adata"]], "6 - Saving the anndata object": [[5, "6---Saving-the-anndata-object"]], "6 - Write MaMPlan": [[12, "6---Write-MaMPlan"]], "6.1 - Plot correlations": [[7, "6.1---Plot-correlations"]], "6.1 - Reclustering": [[8, "6.1---Reclustering"]], "6.1 - Show annotated .obs table": [[10, "6.1---Show-annotated-.obs-table"]], "6.2 - Choose a subset of PCs": [[7, "6.2---Choose-a-subset-of-PCs"]], "6.2 - Final clustering of cells": [[8, "6.2---Final-clustering-of-cells"]], "6.3 - Calculate neighbors": [[7, "6.3---Calculate-neighbors"]], "6.4 - Wrap UMAP": [[7, "6.4---Wrap-UMAP"]], "6.5 - Choose a normalization": [[7, "6.5---Choose-a-normalization"]], "7 - Batch correction": [[7, "7---Batch-correction"]], "7 - Calculate velocity": [[22, "7---Calculate-velocity"]], "7 - Cell clustering": [[19, "7---Cell-clustering"]], "7 - Plot distribution of cells across clusters": [[8, "7---Plot-distribution-of-cells-across-clusters"]], "7 - Plotting": [[21, "7---Plotting"]], "7 - Predict Cell Cycle": [[18, "7---Predict-Cell-Cycle"]], "7 - Save adata": [[10, "7---Save-adata"]], "7 - Saving adata": [[13, "7---Saving-adata"]], "7.1 - Plot overview of batch corrections": [[7, "7.1---Plot-overview-of-batch-corrections"]], "7.1 - Reclustering (optional)": [[19, "7.1---Reclustering-(optional)"]], "7.1 - Violin": [[21, "7.1---Violin"]], "7.2 - Final clustering of cells": [[19, "7.2---Final-clustering-of-cells"]], "7.2 - Network": [[21, "7.2---Network"]], "7.3 - Receptor-ligand connections": [[21, "7.3---Receptor-ligand-connections"]], "8 - Find highly variable genes": [[18, "8---Find-highly-variable-genes"]], "8 - Generating 3D Object with UMAP coordinates in HTML": [[8, "8---Generating-3D-Object-with-UMAP-coordinates-in-HTML"]], "8 - Get interaction table": [[21, "8---Get-interaction-table"]], "8 - Plot distribution of cells across clusters": [[19, "8---Plot-distribution-of-cells-across-clusters"]], "8 - Save adata": [[20, "8---Save-adata"]], "8 - Saving adata for the next notebook": [[7, "8---Saving-adata-for-the-next-notebook"]], "8 - Visualization": [[22, "8---Visualization"]], "8.1 - Proportion": [[22, "8.1---Proportion"]], "8.2 - Embedding": [[22, "8.2---Embedding"]], "8.3 - Velocity length and confidence": [[22, "8.3---Velocity-length-and-confidence"]], "8.4 - Velocity graph": [[22, "8.4---Velocity-graph"]], "9 - Annotate regions to genes": [[8, "9---Annotate-regions-to-genes"]], "9 - Generating 3D Object with UMAP coordinates in HTML": [[19, "9---Generating-3D-Object-with-UMAP-coordinates-in-HTML"]], "9 - PCA and neighbors for uncorrected data": [[18, "9---PCA-and-neighbors-for-uncorrected-data"]], "9 - Pseudotime - scVelo": [[22, "9---Pseudotime---scVelo"]], "9.1 - Choose a subset of PCs (optional)": [[18, "9.1---Choose-a-subset-of-PCs-(optional)"]], "9.2 - Calculate neighbors": [[18, "9.2---Calculate-neighbors"]], "ATAC notebooks": [[9, "atac-notebooks"]], "Additional": [[2, "additional"]], "Analysis": [[2, "analysis"]], "Cell type annotation and marker list assembly": [[10, "Cell-type-annotation-and-marker-list-assembly"]], "Changelog": [[4, "changelog"]], "Changes to ATAC notebooks": [[4, "changes-to-atac-notebooks"]], "Changes to RNA notebooks": [[4, "changes-to-rna-notebooks"]], "Changes to notebooks": [[4, "changes-to-notebooks"], [4, "id4"], [4, "id6"], [4, "id8"], [4, "id12"]], "Contents:": [[15, null]], "General notebooks": [[11, "general-notebooks"]], "Note!": [[14, "Note!"]], "Plotting": [[1, "plotting"]], "Preparing adata for cellxgene / MaMPlan creation": [[12, "Preparing-adata-for-cellxgene-/-MaMPlan-creation"]], "Preprocessing": [[2, "preprocessing"]], "Proportion Analysis": [[13, "Proportion-Analysis"]], "RNA notebooks": [[23, "rna-notebooks"]], "Tools": [[2, "tools"]], "Trajectory and Pseudotime Analysis": [[14, "Trajectory-and-Pseudotime-Analysis"]], "Utils": [[3, "utils"]], "Welcome to the documentation of the single cell framework!": [[15, "welcome-to-the-documentation-of-the-single-cell-framework"]], "adata": [[3, "module-sctoolbox.utils.adata"]], "assemblers": [[3, "module-sctoolbox.utils.assemblers"]], "bam": [[2, "module-sctoolbox.tools.bam"]], "bioutils": [[3, "module-sctoolbox.utils.bioutils"]], "calc_overlap_fc": [[2, "module-sctoolbox.tools.calc_overlap_fc"]], "celltype_annotation": [[2, "module-sctoolbox.tools.celltype_annotation"]], "checker": [[3, "module-sctoolbox.utils.checker"]], "clustering": [[1, "module-sctoolbox.plotting.clustering"], [2, "module-sctoolbox.tools.clustering"]], "creators": [[3, "module-sctoolbox.utils.creators"]], "decorator": [[3, "module-sctoolbox.utils.decorator"]], "dim_reduction": [[2, "module-sctoolbox.tools.dim_reduction"]], "embedding": [[1, "module-sctoolbox.plotting.embedding"], [2, "module-sctoolbox.tools.embedding"]], "frip": [[2, "module-sctoolbox.tools.frip"]], "gene_correlation": [[2, "module-sctoolbox.tools.gene_correlation"]], "general": [[1, "module-sctoolbox.plotting.general"], [3, "module-sctoolbox.utils.general"]], "genometracks": [[1, "module-sctoolbox.plotting.genometracks"]], "highly_variable": [[1, "module-sctoolbox.plotting.highly_variable"], [2, "module-sctoolbox.tools.highly_variable"]], "insertsize": [[2, "module-sctoolbox.tools.insertsize"]], "io": [[3, "module-sctoolbox.utils.io"]], "jupyter": [[3, "module-sctoolbox.utils.jupyter"]], "marker_genes": [[1, "module-sctoolbox.plotting.marker_genes"], [2, "module-sctoolbox.tools.marker_genes"]], "multiomics": [[2, "module-sctoolbox.tools.multiomics"]], "multiprocessing": [[3, "module-sctoolbox.utils.multiprocessing"]], "norm_correct": [[2, "module-sctoolbox.tools.norm_correct"]], "peak_annotation": [[2, "module-sctoolbox.tools.peak_annotation"]], "qc_filter": [[1, "module-sctoolbox.plotting.qc_filter"], [2, "module-sctoolbox.tools.qc_filter"]], "receptor_ligand": [[2, "module-sctoolbox.tools.receptor_ligand"]], "sctoolbox API reference": [[0, "sctoolbox-api-reference"]], "tables": [[3, "module-sctoolbox.utils.tables"]], "tobias": [[2, "module-sctoolbox.tools.tobias"]], "tsse": [[2, "module-sctoolbox.tools.tsse"]], "velocity": [[1, "module-sctoolbox.plotting.velocity"]]}, "docnames": ["API/index", "API/plotting", "API/tools", "API/utils", "CHANGES", "atac-notebooks/01_assembling", "atac-notebooks/02_QC", "atac-notebooks/03_normalization_batch_correction", "atac-notebooks/04_clustering", "atac-notebooks/index", "general-notebooks/annotation", "general-notebooks/index", "general-notebooks/prepare_for_cellxgene", "general-notebooks/proportion_analysis", "general-notebooks/pseudotime_analysis", "index", "rna-notebooks/01_assembling_anndata", "rna-notebooks/02_QC_filtering", "rna-notebooks/03_batch", "rna-notebooks/04_clustering", "rna-notebooks/0A_marker_genes", "rna-notebooks/0B_ligand_receptor", "rna-notebooks/0C_velocity_analysis", "rna-notebooks/index"], "envversion": {"nbsphinx": 4, "sphinx": 61, "sphinx.domains.c": 3, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 9, "sphinx.domains.index": 1, "sphinx.domains.javascript": 3, "sphinx.domains.math": 2, "sphinx.domains.python": 4, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.intersphinx": 1, "sphinx.ext.viewcode": 1}, "filenames": ["API/index.rst", "API/plotting.rst", "API/tools.rst", "API/utils.rst", "CHANGES.rst", "atac-notebooks/01_assembling.nblink", "atac-notebooks/02_QC.nblink", "atac-notebooks/03_normalization_batch_correction.nblink", "atac-notebooks/04_clustering.nblink", "atac-notebooks/index.rst", "general-notebooks/annotation.nblink", "general-notebooks/index.rst", "general-notebooks/prepare_for_cellxgene.nblink", "general-notebooks/proportion_analysis.nblink", "general-notebooks/pseudotime_analysis.nblink", "index.rst", "rna-notebooks/01_assembling_anndata.nblink", "rna-notebooks/02_QC_filtering.nblink", "rna-notebooks/03_batch.nblink", "rna-notebooks/04_clustering.nblink", "rna-notebooks/0A_marker_genes.nblink", "rna-notebooks/0B_ligand_receptor.nblink", "rna-notebooks/0C_velocity_analysis.nblink", "rna-notebooks/index.rst"], "indexentries": {"add_analysis() (in module sctoolbox.utils.creators)": [[3, "sctoolbox.utils.creators.add_analysis", false]], "add_cellxgene_annotation() (in module sctoolbox.tools.celltype_annotation)": [[2, "sctoolbox.tools.celltype_annotation.add_cellxgene_annotation", false]], "add_expr_to_obs() (in module sctoolbox.utils.adata)": [[3, "sctoolbox.utils.adata.add_expr_to_obs", false]], "add_gene_expression() (in module sctoolbox.tools.marker_genes)": [[2, "sctoolbox.tools.marker_genes.add_gene_expression", false]], "add_hline() (sctoolbox.plotting.genometracks.genometracks method)": [[1, "sctoolbox.plotting.genometracks.GenomeTracks.add_hline", false]], "add_hlines() (sctoolbox.plotting.genometracks.genometracks method)": [[1, "sctoolbox.plotting.genometracks.GenomeTracks.add_hlines", false]], "add_insertsize() (in module sctoolbox.tools.insertsize)": [[2, "sctoolbox.tools.insertsize.add_insertsize", false]], "add_spacer() (sctoolbox.plotting.genometracks.genometracks method)": [[1, "sctoolbox.plotting.genometracks.GenomeTracks.add_spacer", false]], "add_track() (sctoolbox.plotting.genometracks.genometracks method)": [[1, "sctoolbox.plotting.genometracks.GenomeTracks.add_track", false]], "add_tsse_score() (in module sctoolbox.tools.tsse)": [[2, "sctoolbox.tools.tsse.add_tsse_score", false]], "add_uns_info() (in module sctoolbox.utils.adata)": [[3, "sctoolbox.utils.adata.add_uns_info", false]], "add_xaxis() (sctoolbox.plotting.genometracks.genometracks method)": [[1, "sctoolbox.plotting.genometracks.GenomeTracks.add_xaxis", false]], "anndata_overview() (in module sctoolbox.plotting.embedding)": [[1, "sctoolbox.plotting.embedding.anndata_overview", false]], "annot_hvg() (in module sctoolbox.tools.highly_variable)": [[2, "sctoolbox.tools.highly_variable.annot_HVG", false]], "annotate_adata() (in module sctoolbox.tools.peak_annotation)": [[2, "sctoolbox.tools.peak_annotation.annotate_adata", false]], "annotate_narrowpeak() (in module sctoolbox.tools.peak_annotation)": [[2, "sctoolbox.tools.peak_annotation.annotate_narrowPeak", false]], "apply_qc_thresholds() (in module sctoolbox.tools.qc_filter)": [[2, "sctoolbox.tools.qc_filter.apply_qc_thresholds", false]], "apply_svd() (in module sctoolbox.tools.dim_reduction)": [[2, "sctoolbox.tools.dim_reduction.apply_svd", false]], "atac_norm() (in module sctoolbox.tools.norm_correct)": [[2, "sctoolbox.tools.norm_correct.atac_norm", false]], "automatic_thresholds() (in module sctoolbox.tools.qc_filter)": [[2, "sctoolbox.tools.qc_filter.automatic_thresholds", false]], "bam_adata_ov() (in module sctoolbox.tools.bam)": [[2, "sctoolbox.tools.bam.bam_adata_ov", false]], "bam_to_bigwig() (in module sctoolbox.tools.bam)": [[2, "sctoolbox.tools.bam.bam_to_bigwig", false]], "barcode_index() (in module sctoolbox.utils.bioutils)": [[3, "sctoolbox.utils.bioutils.barcode_index", false]], "batch_correction() (in module sctoolbox.tools.norm_correct)": [[2, "sctoolbox.tools.norm_correct.batch_correction", false]], "bidirectional_barplot() (in module sctoolbox.plotting.general)": [[1, "sctoolbox.plotting.general.bidirectional_barplot", false]], "boxplot() (in module sctoolbox.plotting.general)": [[1, "sctoolbox.plotting.general.boxplot", false]], "build_notebooks_regex() (in module sctoolbox.utils.creators)": [[3, "sctoolbox.utils.creators.build_notebooks_regex", false]], "calc_frip_scores() (in module sctoolbox.tools.frip)": [[2, "sctoolbox.tools.frip.calc_frip_scores", false]], "calc_per_base_tsse() (in module sctoolbox.tools.tsse)": [[2, "sctoolbox.tools.tsse.calc_per_base_tsse", false]], "calc_ragi() (in module sctoolbox.tools.clustering)": [[2, "sctoolbox.tools.clustering.calc_ragi", false]], "calculate_interaction_table() (in module sctoolbox.tools.receptor_ligand)": [[2, "sctoolbox.tools.receptor_ligand.calculate_interaction_table", false]], "calculate_qc_metrics() (in module sctoolbox.tools.qc_filter)": [[2, "sctoolbox.tools.qc_filter.calculate_qc_metrics", false]], "check_barcode_tag() (in module sctoolbox.tools.bam)": [[2, "sctoolbox.tools.bam.check_barcode_tag", false]], "check_columns() (in module sctoolbox.utils.checker)": [[3, "sctoolbox.utils.checker.check_columns", false]], "check_file_ending() (in module sctoolbox.utils.checker)": [[3, "sctoolbox.utils.checker.check_file_ending", false]], "check_marker_lists() (in module sctoolbox.utils.checker)": [[3, "sctoolbox.utils.checker.check_marker_lists", false]], "check_module() (in module sctoolbox.utils.checker)": [[3, "sctoolbox.utils.checker.check_module", false]], "check_type() (in module sctoolbox.utils.checker)": [[3, "sctoolbox.utils.checker.check_type", false]], "clean_flanking_strings() (in module sctoolbox.utils.general)": [[3, "sctoolbox.utils.general.clean_flanking_strings", false]], "clear() (in module sctoolbox.utils.jupyter)": [[3, "sctoolbox.utils.jupyter.clear", false]], "clustermap_dotplot() (in module sctoolbox.plotting.general)": [[1, "sctoolbox.plotting.general.clustermap_dotplot", false]], "compare_embeddings() (in module sctoolbox.plotting.embedding)": [[1, "sctoolbox.plotting.embedding.compare_embeddings", false]], "compare_two_conditons() (in module sctoolbox.tools.gene_correlation)": [[2, "sctoolbox.tools.gene_correlation.compare_two_conditons", false]], "compute_pca() (in module sctoolbox.tools.dim_reduction)": [[2, "sctoolbox.tools.dim_reduction.compute_PCA", false]], "connectionplot() (in module sctoolbox.tools.receptor_ligand)": [[2, "sctoolbox.tools.receptor_ligand.connectionPlot", false]], "convert_id() (in module sctoolbox.utils.bioutils)": [[3, "sctoolbox.utils.bioutils.convert_id", false]], "converttoadata() (in module sctoolbox.utils.assemblers)": [[3, "sctoolbox.utils.assemblers.convertToAdata", false]], "correlate_conditions() (in module sctoolbox.tools.gene_correlation)": [[2, "sctoolbox.tools.gene_correlation.correlate_conditions", false]], "correlate_ref_vs_all() (in module sctoolbox.tools.gene_correlation)": [[2, "sctoolbox.tools.gene_correlation.correlate_ref_vs_all", false]], "correlation_matrix() (in module sctoolbox.tools.embedding)": [[2, "sctoolbox.tools.embedding.correlation_matrix", false]], "count_fragments_per_cell() (in module sctoolbox.tools.calc_overlap_fc)": [[2, "sctoolbox.tools.calc_overlap_fc.count_fragments_per_cell", false]], "create_dir() (in module sctoolbox.utils.io)": [[3, "sctoolbox.utils.io.create_dir", false]], "create_fragment_file() (in module sctoolbox.tools.bam)": [[2, "sctoolbox.tools.bam.create_fragment_file", false]], "debug_func_log() (in module sctoolbox.utils.decorator)": [[3, "sctoolbox.utils.decorator.debug_func_log", false]], "define_pc() (in module sctoolbox.tools.dim_reduction)": [[2, "sctoolbox.tools.dim_reduction.define_PC", false]], "download_db() (in module sctoolbox.tools.receptor_ligand)": [[2, "sctoolbox.tools.receptor_ligand.download_db", false]], "estimate_doublets() (in module sctoolbox.tools.qc_filter)": [[2, "sctoolbox.tools.qc_filter.estimate_doublets", false]], "evaluate_batch_effect() (in module sctoolbox.tools.norm_correct)": [[2, "sctoolbox.tools.norm_correct.evaluate_batch_effect", false]], "fc_fragments_in_regions() (in module sctoolbox.tools.calc_overlap_fc)": [[2, "sctoolbox.tools.calc_overlap_fc.fc_fragments_in_regions", false]], "fill_na() (in module sctoolbox.utils.tables)": [[3, "sctoolbox.utils.tables.fill_na", false]], "filter_cells() (in module sctoolbox.tools.qc_filter)": [[2, "sctoolbox.tools.qc_filter.filter_cells", false]], "filter_genes() (in module sctoolbox.tools.qc_filter)": [[2, "sctoolbox.tools.qc_filter.filter_genes", false]], "flip_embedding() (in module sctoolbox.plotting.embedding)": [[1, "sctoolbox.plotting.embedding.flip_embedding", false]], "format_adata_var() (in module sctoolbox.utils.checker)": [[3, "sctoolbox.utils.checker.format_adata_var", false]], "from_mtx() (in module sctoolbox.utils.assemblers)": [[3, "sctoolbox.utils.assemblers.from_mtx", false]], "from_quant() (in module sctoolbox.utils.assemblers)": [[3, "sctoolbox.utils.assemblers.from_quant", false]], "from_single_mtx() (in module sctoolbox.utils.assemblers)": [[3, "sctoolbox.utils.assemblers.from_single_mtx", false]], "from_single_starsolo() (in module sctoolbox.utils.assemblers)": [[3, "sctoolbox.utils.assemblers.from_single_starsolo", false]], "gene_expression_heatmap() (in module sctoolbox.plotting.marker_genes)": [[1, "sctoolbox.plotting.marker_genes.gene_expression_heatmap", false]], "gene_id_to_name() (in module sctoolbox.utils.bioutils)": [[3, "sctoolbox.utils.bioutils.gene_id_to_name", false]], "genometracks (class in sctoolbox.plotting.genometracks)": [[1, "sctoolbox.plotting.genometracks.GenomeTracks", false]], "get_adata_subsets() (in module sctoolbox.utils.adata)": [[3, "sctoolbox.utils.adata.get_adata_subsets", false]], "get_bam_reads() (in module sctoolbox.tools.bam)": [[2, "sctoolbox.tools.bam.get_bam_reads", false]], "get_binary_path() (in module sctoolbox.utils.general)": [[3, "sctoolbox.utils.general.get_binary_path", false]], "get_cell_values() (in module sctoolbox.utils.adata)": [[3, "sctoolbox.utils.adata.get_cell_values", false]], "get_celltype_assignment() (in module sctoolbox.tools.celltype_annotation)": [[2, "sctoolbox.tools.celltype_annotation.get_celltype_assignment", false]], "get_chromosome_genes() (in module sctoolbox.tools.marker_genes)": [[2, "sctoolbox.tools.marker_genes.get_chromosome_genes", false]], "get_datetime() (in module sctoolbox.utils.general)": [[3, "sctoolbox.utils.general.get_datetime", false]], "get_index_type() (in module sctoolbox.utils.checker)": [[3, "sctoolbox.utils.checker.get_index_type", false]], "get_interactions() (in module sctoolbox.tools.receptor_ligand)": [[2, "sctoolbox.tools.receptor_ligand.get_interactions", false]], "get_keys() (in module sctoolbox.tools.qc_filter)": [[2, "sctoolbox.tools.qc_filter.get_keys", false]], "get_mean_thresholds() (in module sctoolbox.tools.qc_filter)": [[2, "sctoolbox.tools.qc_filter.get_mean_thresholds", false]], "get_minimal_adata() (in module sctoolbox.utils.adata)": [[3, "sctoolbox.utils.adata.get_minimal_adata", false]], "get_organism() (in module sctoolbox.utils.bioutils)": [[3, "sctoolbox.utils.bioutils.get_organism", false]], "get_package_versions() (in module sctoolbox.utils.general)": [[3, "sctoolbox.utils.general.get_package_versions", false]], "get_parameter_table() (in module sctoolbox.utils.decorator)": [[3, "sctoolbox.utils.decorator.get_parameter_table", false]], "get_pbar() (in module sctoolbox.utils.multiprocessing)": [[3, "sctoolbox.utils.multiprocessing.get_pbar", false]], "get_rank_genes_tables() (in module sctoolbox.tools.marker_genes)": [[2, "sctoolbox.tools.marker_genes.get_rank_genes_tables", false]], "get_slider_thresholds() (in module sctoolbox.plotting.qc_filter)": [[1, "sctoolbox.plotting.qc_filter.get_slider_thresholds", false]], "get_temporary_filename() (in module sctoolbox.utils.io)": [[3, "sctoolbox.utils.io.get_temporary_filename", false]], "get_thresholds_wrapper() (in module sctoolbox.tools.qc_filter)": [[2, "sctoolbox.tools.qc_filter.get_thresholds_wrapper", false]], "get_user() (in module sctoolbox.utils.general)": [[3, "sctoolbox.utils.general.get_user", false]], "get_variable_features() (in module sctoolbox.tools.highly_variable)": [[2, "sctoolbox.tools.highly_variable.get_variable_features", false]], "gini() (in module sctoolbox.tools.clustering)": [[2, "sctoolbox.tools.clustering.gini", false]], "gitlab_download() (in module sctoolbox.utils.creators)": [[3, "sctoolbox.utils.creators.gitlab_download", false]], "global_tsse_score() (in module sctoolbox.tools.tsse)": [[2, "sctoolbox.tools.tsse.global_tsse_score", false]], "grey_colormap() (in module sctoolbox.plotting.embedding)": [[1, "sctoolbox.plotting.embedding.grey_colormap", false]], "group_correlation() (in module sctoolbox.plotting.qc_filter)": [[1, "sctoolbox.plotting.qc_filter.group_correlation", false]], "group_expression_boxplot() (in module sctoolbox.plotting.marker_genes)": [[1, "sctoolbox.plotting.marker_genes.group_expression_boxplot", false]], "grouped_violin() (in module sctoolbox.plotting.marker_genes)": [[1, "sctoolbox.plotting.marker_genes.grouped_violin", false]], "gunzip_file() (in module sctoolbox.utils.checker)": [[3, "sctoolbox.utils.checker.gunzip_file", false]], "hairball() (in module sctoolbox.tools.receptor_ligand)": [[2, "sctoolbox.tools.receptor_ligand.hairball", false]], "identify_columns() (in module sctoolbox.utils.general)": [[3, "sctoolbox.utils.general.identify_columns", false]], "in_range() (in module sctoolbox.utils.checker)": [[3, "sctoolbox.utils.checker.in_range", false]], "interaction_progress() (in module sctoolbox.tools.receptor_ligand)": [[2, "sctoolbox.tools.receptor_ligand.interaction_progress", false]], "interaction_violin_plot() (in module sctoolbox.tools.receptor_ligand)": [[2, "sctoolbox.tools.receptor_ligand.interaction_violin_plot", false]], "is_integer_array() (in module sctoolbox.utils.checker)": [[3, "sctoolbox.utils.checker.is_integer_array", false]], "is_regex() (in module sctoolbox.utils.checker)": [[3, "sctoolbox.utils.checker.is_regex", false]], "is_str_numeric() (in module sctoolbox.utils.checker)": [[3, "sctoolbox.utils.checker.is_str_numeric", false]], "label_genes() (in module sctoolbox.tools.marker_genes)": [[2, "sctoolbox.tools.marker_genes.label_genes", false]], "load_h5ad() (in module sctoolbox.utils.adata)": [[3, "sctoolbox.utils.adata.load_h5ad", false]], "log_anndata() (in module sctoolbox.utils.decorator)": [[3, "sctoolbox.utils.decorator.log_anndata", false]], "longest_common_suffix() (in module sctoolbox.utils.general)": [[3, "sctoolbox.utils.general.longest_common_suffix", false]], "lsi() (in module sctoolbox.tools.dim_reduction)": [[2, "sctoolbox.tools.dim_reduction.lsi", false]], "marker_gene_clustering() (in module sctoolbox.plotting.clustering)": [[1, "sctoolbox.plotting.clustering.marker_gene_clustering", false]], "mask_rank_genes() (in module sctoolbox.tools.marker_genes)": [[2, "sctoolbox.tools.marker_genes.mask_rank_genes", false]], "merge_anndata() (in module sctoolbox.tools.multiomics)": [[2, "sctoolbox.tools.multiomics.merge_anndata", false]], "module": [[1, "module-sctoolbox.plotting.clustering", false], [1, "module-sctoolbox.plotting.embedding", false], [1, "module-sctoolbox.plotting.general", false], [1, "module-sctoolbox.plotting.genometracks", false], [1, "module-sctoolbox.plotting.highly_variable", false], [1, "module-sctoolbox.plotting.marker_genes", false], [1, "module-sctoolbox.plotting.qc_filter", false], [1, "module-sctoolbox.plotting.velocity", false], [2, "module-sctoolbox.tools.bam", false], [2, "module-sctoolbox.tools.calc_overlap_fc", false], [2, "module-sctoolbox.tools.celltype_annotation", false], [2, "module-sctoolbox.tools.clustering", false], [2, "module-sctoolbox.tools.dim_reduction", false], [2, "module-sctoolbox.tools.embedding", false], [2, "module-sctoolbox.tools.frip", false], [2, "module-sctoolbox.tools.gene_correlation", false], [2, "module-sctoolbox.tools.highly_variable", false], [2, "module-sctoolbox.tools.insertsize", false], [2, "module-sctoolbox.tools.marker_genes", false], [2, "module-sctoolbox.tools.multiomics", false], [2, "module-sctoolbox.tools.norm_correct", false], [2, "module-sctoolbox.tools.peak_annotation", false], [2, "module-sctoolbox.tools.qc_filter", false], [2, "module-sctoolbox.tools.receptor_ligand", false], [2, "module-sctoolbox.tools.tobias", false], [2, "module-sctoolbox.tools.tsse", false], [3, "module-sctoolbox.utils.adata", false], [3, "module-sctoolbox.utils.assemblers", false], [3, "module-sctoolbox.utils.bioutils", false], [3, "module-sctoolbox.utils.checker", false], [3, "module-sctoolbox.utils.creators", false], [3, "module-sctoolbox.utils.decorator", false], [3, "module-sctoolbox.utils.general", false], [3, "module-sctoolbox.utils.io", false], [3, "module-sctoolbox.utils.jupyter", false], [3, "module-sctoolbox.utils.multiprocessing", false], [3, "module-sctoolbox.utils.tables", false]], "monitor_jobs() (in module sctoolbox.utils.multiprocessing)": [[3, "sctoolbox.utils.multiprocessing.monitor_jobs", false]], "n_cells_barplot() (in module sctoolbox.plotting.qc_filter)": [[1, "sctoolbox.plotting.qc_filter.n_cells_barplot", false]], "normalize_adata() (in module sctoolbox.tools.norm_correct)": [[2, "sctoolbox.tools.norm_correct.normalize_adata", false]], "open_bam() (in module sctoolbox.tools.bam)": [[2, "sctoolbox.tools.bam.open_bam", false]], "overlap_and_aggregate() (in module sctoolbox.tools.tsse)": [[2, "sctoolbox.tools.tsse.overlap_and_aggregate", false]], "pairwise_rank_genes() (in module sctoolbox.tools.marker_genes)": [[2, "sctoolbox.tools.marker_genes.pairwise_rank_genes", false]], "pairwise_scatter() (in module sctoolbox.plotting.general)": [[1, "sctoolbox.plotting.general.pairwise_scatter", false]], "pct_fragments_in_promoters() (in module sctoolbox.tools.calc_overlap_fc)": [[2, "sctoolbox.tools.calc_overlap_fc.pct_fragments_in_promoters", false]], "plot() (sctoolbox.plotting.genometracks.genometracks method)": [[1, "sctoolbox.plotting.genometracks.GenomeTracks.plot", false]], "plot_3d_umap() (in module sctoolbox.plotting.embedding)": [[1, "sctoolbox.plotting.embedding.plot_3D_UMAP", false]], "plot_differential_genes() (in module sctoolbox.plotting.marker_genes)": [[1, "sctoolbox.plotting.marker_genes.plot_differential_genes", false]], "plot_embedding() (in module sctoolbox.plotting.embedding)": [[1, "sctoolbox.plotting.embedding.plot_embedding", false]], "plot_gene_correlation() (in module sctoolbox.plotting.marker_genes)": [[1, "sctoolbox.plotting.marker_genes.plot_gene_correlation", false]], "plot_group_embeddings() (in module sctoolbox.plotting.embedding)": [[1, "sctoolbox.plotting.embedding.plot_group_embeddings", false]], "plot_insertsize() (in module sctoolbox.plotting.qc_filter)": [[1, "sctoolbox.plotting.qc_filter.plot_insertsize", false]], "plot_pca_correlation() (in module sctoolbox.plotting.embedding)": [[1, "sctoolbox.plotting.embedding.plot_pca_correlation", false]], "plot_pca_variance() (in module sctoolbox.plotting.embedding)": [[1, "sctoolbox.plotting.embedding.plot_pca_variance", false]], "plot_starsolo_quality() (in module sctoolbox.plotting.qc_filter)": [[1, "sctoolbox.plotting.qc_filter.plot_starsolo_quality", false]], "plot_starsolo_umi() (in module sctoolbox.plotting.qc_filter)": [[1, "sctoolbox.plotting.qc_filter.plot_starsolo_UMI", false]], "plot_venn() (in module sctoolbox.plotting.general)": [[1, "sctoolbox.plotting.general.plot_venn", false]], "predict_cell_cycle() (in module sctoolbox.tools.qc_filter)": [[2, "sctoolbox.tools.qc_filter.predict_cell_cycle", false]], "predict_sex() (in module sctoolbox.tools.qc_filter)": [[2, "sctoolbox.tools.qc_filter.predict_sex", false]], "prepare_atac_anndata() (in module sctoolbox.utils.assemblers)": [[3, "sctoolbox.utils.assemblers.prepare_atac_anndata", false]], "prepare_for_cellxgene() (in module sctoolbox.utils.adata)": [[3, "sctoolbox.utils.adata.prepare_for_cellxgene", false]], "progress_violins() (in module sctoolbox.tools.receptor_ligand)": [[2, "sctoolbox.tools.receptor_ligand.progress_violins", false]], "propose_pcs() (in module sctoolbox.tools.dim_reduction)": [[2, "sctoolbox.tools.dim_reduction.propose_pcs", false]], "pseudobulk_table() (in module sctoolbox.utils.bioutils)": [[3, "sctoolbox.utils.bioutils.pseudobulk_table", false]], "pseudotime_heatmap() (in module sctoolbox.plotting.velocity)": [[1, "sctoolbox.plotting.velocity.pseudotime_heatmap", false]], "quality_violin() (in module sctoolbox.plotting.qc_filter)": [[1, "sctoolbox.plotting.qc_filter.quality_violin", false]], "rank_genes_plot() (in module sctoolbox.plotting.marker_genes)": [[1, "sctoolbox.plotting.marker_genes.rank_genes_plot", false]], "read_list_file() (in module sctoolbox.utils.general)": [[3, "sctoolbox.utils.general.read_list_file", false]], "recluster() (in module sctoolbox.tools.clustering)": [[2, "sctoolbox.tools.clustering.recluster", false]], "remove_files() (in module sctoolbox.utils.io)": [[3, "sctoolbox.utils.io.remove_files", false]], "remove_prefix() (in module sctoolbox.utils.general)": [[3, "sctoolbox.utils.general.remove_prefix", false]], "remove_suffix() (in module sctoolbox.utils.general)": [[3, "sctoolbox.utils.general.remove_suffix", false]], "rename_categories() (in module sctoolbox.utils.tables)": [[3, "sctoolbox.utils.tables.rename_categories", false]], "rm_tmp() (in module sctoolbox.utils.io)": [[3, "sctoolbox.utils.io.rm_tmp", false]], "run_cmd() (in module sctoolbox.utils.general)": [[3, "sctoolbox.utils.general.run_cmd", false]], "run_deseq2() (in module sctoolbox.tools.marker_genes)": [[2, "sctoolbox.tools.marker_genes.run_deseq2", false]], "run_rank_genes() (in module sctoolbox.tools.marker_genes)": [[2, "sctoolbox.tools.marker_genes.run_rank_genes", false]], "run_scsa() (in module sctoolbox.tools.celltype_annotation)": [[2, "sctoolbox.tools.celltype_annotation.run_scsa", false]], "sanitize_string() (in module sctoolbox.utils.general)": [[3, "sctoolbox.utils.general.sanitize_string", false]], "save_h5ad() (in module sctoolbox.utils.adata)": [[3, "sctoolbox.utils.adata.save_h5ad", false]], "sc_colormap() (in module sctoolbox.plotting.embedding)": [[1, "sctoolbox.plotting.embedding.sc_colormap", false]], "scale_values() (in module sctoolbox.utils.general)": [[3, "sctoolbox.utils.general.scale_values", false]], "scatter_hvf_distribution() (in module sctoolbox.plotting.highly_variable)": [[1, "sctoolbox.plotting.highly_variable.scatter_HVF_distribution", false]], "score_genes() (in module sctoolbox.tools.marker_genes)": [[2, "sctoolbox.tools.marker_genes.score_genes", false]], "sctoolbox.plotting.clustering": [[1, "module-sctoolbox.plotting.clustering", false]], "sctoolbox.plotting.embedding": [[1, "module-sctoolbox.plotting.embedding", false]], "sctoolbox.plotting.general": [[1, "module-sctoolbox.plotting.general", false]], "sctoolbox.plotting.genometracks": [[1, "module-sctoolbox.plotting.genometracks", false]], "sctoolbox.plotting.highly_variable": [[1, "module-sctoolbox.plotting.highly_variable", false]], "sctoolbox.plotting.marker_genes": [[1, "module-sctoolbox.plotting.marker_genes", false]], "sctoolbox.plotting.qc_filter": [[1, "module-sctoolbox.plotting.qc_filter", false]], "sctoolbox.plotting.velocity": [[1, "module-sctoolbox.plotting.velocity", false]], "sctoolbox.tools.bam": [[2, "module-sctoolbox.tools.bam", false]], "sctoolbox.tools.calc_overlap_fc": [[2, "module-sctoolbox.tools.calc_overlap_fc", false]], "sctoolbox.tools.celltype_annotation": [[2, "module-sctoolbox.tools.celltype_annotation", false]], "sctoolbox.tools.clustering": [[2, "module-sctoolbox.tools.clustering", false]], "sctoolbox.tools.dim_reduction": [[2, "module-sctoolbox.tools.dim_reduction", false]], "sctoolbox.tools.embedding": [[2, "module-sctoolbox.tools.embedding", false]], "sctoolbox.tools.frip": [[2, "module-sctoolbox.tools.frip", false]], "sctoolbox.tools.gene_correlation": [[2, "module-sctoolbox.tools.gene_correlation", false]], "sctoolbox.tools.highly_variable": [[2, "module-sctoolbox.tools.highly_variable", false]], "sctoolbox.tools.insertsize": [[2, "module-sctoolbox.tools.insertsize", false]], "sctoolbox.tools.marker_genes": [[2, "module-sctoolbox.tools.marker_genes", false]], "sctoolbox.tools.multiomics": [[2, "module-sctoolbox.tools.multiomics", false]], "sctoolbox.tools.norm_correct": [[2, "module-sctoolbox.tools.norm_correct", false]], "sctoolbox.tools.peak_annotation": [[2, "module-sctoolbox.tools.peak_annotation", false]], "sctoolbox.tools.qc_filter": [[2, "module-sctoolbox.tools.qc_filter", false]], "sctoolbox.tools.receptor_ligand": [[2, "module-sctoolbox.tools.receptor_ligand", false]], "sctoolbox.tools.tobias": [[2, "module-sctoolbox.tools.tobias", false]], "sctoolbox.tools.tsse": [[2, "module-sctoolbox.tools.tsse", false]], "sctoolbox.utils.adata": [[3, "module-sctoolbox.utils.adata", false]], "sctoolbox.utils.assemblers": [[3, "module-sctoolbox.utils.assemblers", false]], "sctoolbox.utils.bioutils": [[3, "module-sctoolbox.utils.bioutils", false]], "sctoolbox.utils.checker": [[3, "module-sctoolbox.utils.checker", false]], "sctoolbox.utils.creators": [[3, "module-sctoolbox.utils.creators", false]], "sctoolbox.utils.decorator": [[3, "module-sctoolbox.utils.decorator", false]], "sctoolbox.utils.general": [[3, "module-sctoolbox.utils.general", false]], "sctoolbox.utils.io": [[3, "module-sctoolbox.utils.io", false]], "sctoolbox.utils.jupyter": [[3, "module-sctoolbox.utils.jupyter", false]], "sctoolbox.utils.multiprocessing": [[3, "module-sctoolbox.utils.multiprocessing", false]], "sctoolbox.utils.tables": [[3, "module-sctoolbox.utils.tables", false]], "search_clustering_parameters() (in module sctoolbox.plotting.clustering)": [[1, "sctoolbox.plotting.clustering.search_clustering_parameters", false]], "search_tsne_parameters() (in module sctoolbox.plotting.embedding)": [[1, "sctoolbox.plotting.embedding.search_tsne_parameters", false]], "search_umap_parameters() (in module sctoolbox.plotting.embedding)": [[1, "sctoolbox.plotting.embedding.search_umap_parameters", false]], "setup_experiment() (in module sctoolbox.utils.creators)": [[3, "sctoolbox.utils.creators.setup_experiment", false]], "setup_r() (in module sctoolbox.utils.general)": [[3, "sctoolbox.utils.general.setup_R", false]], "show_config() (sctoolbox.plotting.genometracks.genometracks method)": [[1, "sctoolbox.plotting.genometracks.GenomeTracks.show_config", false]], "show_plot() (sctoolbox.plotting.genometracks.genometracks method)": [[1, "sctoolbox.plotting.genometracks.GenomeTracks.show_plot", false]], "shuffle_cells() (in module sctoolbox.utils.adata)": [[3, "sctoolbox.utils.adata.shuffle_cells", false]], "split_bam_clusters() (in module sctoolbox.tools.bam)": [[2, "sctoolbox.tools.bam.split_bam_clusters", false]], "split_list() (in module sctoolbox.utils.general)": [[3, "sctoolbox.utils.general.split_list", false]], "split_list_size() (in module sctoolbox.utils.general)": [[3, "sctoolbox.utils.general.split_list_size", false]], "subset_bam() (in module sctoolbox.tools.bam)": [[2, "sctoolbox.tools.bam.subset_bam", false]], "subset_pca() (in module sctoolbox.tools.dim_reduction)": [[2, "sctoolbox.tools.dim_reduction.subset_PCA", false]], "table_zscore() (in module sctoolbox.utils.tables)": [[3, "sctoolbox.utils.tables.table_zscore", false]], "tfidf() (in module sctoolbox.tools.norm_correct)": [[2, "sctoolbox.tools.norm_correct.tfidf", false]], "tfidf_normalization() (in module sctoolbox.tools.norm_correct)": [[2, "sctoolbox.tools.norm_correct.tfidf_normalization", false]], "thresholds_as_table() (in module sctoolbox.tools.qc_filter)": [[2, "sctoolbox.tools.qc_filter.thresholds_as_table", false]], "tsse_scoring() (in module sctoolbox.tools.tsse)": [[2, "sctoolbox.tools.tsse.tsse_scoring", false]], "umap_marker_overview() (in module sctoolbox.plotting.embedding)": [[1, "sctoolbox.plotting.embedding.umap_marker_overview", false]], "unify_genes_column() (in module sctoolbox.utils.bioutils)": [[3, "sctoolbox.utils.bioutils.unify_genes_column", false]], "validate_regions() (in module sctoolbox.utils.checker)": [[3, "sctoolbox.utils.checker.validate_regions", false]], "validate_threshold_dict() (in module sctoolbox.tools.qc_filter)": [[2, "sctoolbox.tools.qc_filter.validate_threshold_dict", false]], "var_index_from() (in module sctoolbox.utils.checker)": [[3, "sctoolbox.utils.checker.var_index_from", false]], "violin_hvf_distribution() (in module sctoolbox.plotting.highly_variable)": [[1, "sctoolbox.plotting.highly_variable.violin_HVF_distribution", false]], "violinplot() (in module sctoolbox.plotting.general)": [[1, "sctoolbox.plotting.general.violinplot", false]], "wrap_batch_evaluation() (in module sctoolbox.tools.norm_correct)": [[2, "sctoolbox.tools.norm_correct.wrap_batch_evaluation", false]], "wrap_corrections() (in module sctoolbox.tools.norm_correct)": [[2, "sctoolbox.tools.norm_correct.wrap_corrections", false]], "wrap_umap() (in module sctoolbox.tools.embedding)": [[2, "sctoolbox.tools.embedding.wrap_umap", false]], "write_excel() (in module sctoolbox.utils.tables)": [[3, "sctoolbox.utils.tables.write_excel", false]], "write_list_file() (in module sctoolbox.utils.general)": [[3, "sctoolbox.utils.general.write_list_file", false]], "write_tobias_config() (in module sctoolbox.tools.tobias)": [[2, "sctoolbox.tools.tobias.write_TOBIAS_config", false]], "write_tss_bed() (in module sctoolbox.tools.tsse)": [[2, "sctoolbox.tools.tsse.write_TSS_bed", false]]}, "objects": {"sctoolbox.plotting": [[1, 0, 0, "-", "clustering"], [1, 0, 0, "-", "embedding"], [1, 0, 0, "-", "general"], [1, 0, 0, "-", "genometracks"], [1, 0, 0, "-", "highly_variable"], [1, 0, 0, "-", "marker_genes"], [1, 0, 0, "-", "qc_filter"], [1, 0, 0, "-", "velocity"]], "sctoolbox.plotting.clustering": [[1, 1, 1, "", "marker_gene_clustering"], [1, 1, 1, "", "search_clustering_parameters"]], "sctoolbox.plotting.embedding": [[1, 1, 1, "", "anndata_overview"], [1, 1, 1, "", "compare_embeddings"], [1, 1, 1, "", "flip_embedding"], [1, 1, 1, "", "grey_colormap"], [1, 1, 1, "", "plot_3D_UMAP"], [1, 1, 1, "", "plot_embedding"], [1, 1, 1, "", "plot_group_embeddings"], [1, 1, 1, "", "plot_pca_correlation"], [1, 1, 1, "", "plot_pca_variance"], [1, 1, 1, "", "sc_colormap"], [1, 1, 1, "", "search_tsne_parameters"], [1, 1, 1, "", "search_umap_parameters"], [1, 1, 1, "", "umap_marker_overview"]], "sctoolbox.plotting.general": [[1, 1, 1, "", "bidirectional_barplot"], [1, 1, 1, "", "boxplot"], [1, 1, 1, "", "clustermap_dotplot"], [1, 1, 1, "", "pairwise_scatter"], [1, 1, 1, "", "plot_venn"], [1, 1, 1, "", "violinplot"]], "sctoolbox.plotting.genometracks": [[1, 2, 1, "", "GenomeTracks"]], "sctoolbox.plotting.genometracks.GenomeTracks": [[1, 3, 1, "", "add_hline"], [1, 3, 1, "", "add_hlines"], [1, 3, 1, "", "add_spacer"], [1, 3, 1, "", "add_track"], [1, 3, 1, "", "add_xaxis"], [1, 3, 1, "", "plot"], [1, 3, 1, "", "show_config"], [1, 3, 1, "", "show_plot"]], "sctoolbox.plotting.highly_variable": [[1, 1, 1, "", "scatter_HVF_distribution"], [1, 1, 1, "", "violin_HVF_distribution"]], "sctoolbox.plotting.marker_genes": [[1, 1, 1, "", "gene_expression_heatmap"], [1, 1, 1, "", "group_expression_boxplot"], [1, 1, 1, "", "grouped_violin"], [1, 1, 1, "", "plot_differential_genes"], [1, 1, 1, "", "plot_gene_correlation"], [1, 1, 1, "", "rank_genes_plot"]], "sctoolbox.plotting.qc_filter": [[1, 1, 1, "", "get_slider_thresholds"], [1, 1, 1, "", "group_correlation"], [1, 1, 1, "", "n_cells_barplot"], [1, 1, 1, "", "plot_insertsize"], [1, 1, 1, "", "plot_starsolo_UMI"], [1, 1, 1, "", "plot_starsolo_quality"], [1, 1, 1, "", "quality_violin"]], "sctoolbox.plotting.velocity": [[1, 1, 1, "", "pseudotime_heatmap"]], "sctoolbox.tools": [[2, 0, 0, "-", "bam"], [2, 0, 0, "-", "calc_overlap_fc"], [2, 0, 0, "-", "celltype_annotation"], [2, 0, 0, "-", "clustering"], [2, 0, 0, "-", "dim_reduction"], [2, 0, 0, "-", "embedding"], [2, 0, 0, "-", "frip"], [2, 0, 0, "-", "gene_correlation"], [2, 0, 0, "-", "highly_variable"], [2, 0, 0, "-", "insertsize"], [2, 0, 0, "-", "marker_genes"], [2, 0, 0, "-", "multiomics"], [2, 0, 0, "-", "norm_correct"], [2, 0, 0, "-", "peak_annotation"], [2, 0, 0, "-", "qc_filter"], [2, 0, 0, "-", "receptor_ligand"], [2, 0, 0, "-", "tobias"], [2, 0, 0, "-", "tsse"]], "sctoolbox.tools.bam": [[2, 1, 1, "", "bam_adata_ov"], [2, 1, 1, "", "bam_to_bigwig"], [2, 1, 1, "", "check_barcode_tag"], [2, 1, 1, "", "create_fragment_file"], [2, 1, 1, "", "get_bam_reads"], [2, 1, 1, "", "open_bam"], [2, 1, 1, "", "split_bam_clusters"], [2, 1, 1, "", "subset_bam"]], "sctoolbox.tools.calc_overlap_fc": [[2, 1, 1, "", "count_fragments_per_cell"], [2, 1, 1, "", "fc_fragments_in_regions"], [2, 1, 1, "", "pct_fragments_in_promoters"]], "sctoolbox.tools.celltype_annotation": [[2, 1, 1, "", "add_cellxgene_annotation"], [2, 1, 1, "", "get_celltype_assignment"], [2, 1, 1, "", "run_scsa"]], "sctoolbox.tools.clustering": [[2, 1, 1, "", "calc_ragi"], [2, 1, 1, "", "gini"], [2, 1, 1, "", "recluster"]], "sctoolbox.tools.dim_reduction": [[2, 1, 1, "", "apply_svd"], [2, 1, 1, "", "compute_PCA"], [2, 1, 1, "", "define_PC"], [2, 1, 1, "", "lsi"], [2, 1, 1, "", "propose_pcs"], [2, 1, 1, "", "subset_PCA"]], "sctoolbox.tools.embedding": [[2, 1, 1, "", "correlation_matrix"], [2, 1, 1, "", "wrap_umap"]], "sctoolbox.tools.frip": [[2, 1, 1, "", "calc_frip_scores"]], "sctoolbox.tools.gene_correlation": [[2, 1, 1, "", "compare_two_conditons"], [2, 1, 1, "", "correlate_conditions"], [2, 1, 1, "", "correlate_ref_vs_all"]], "sctoolbox.tools.highly_variable": [[2, 1, 1, "", "annot_HVG"], [2, 1, 1, "", "get_variable_features"]], "sctoolbox.tools.insertsize": [[2, 1, 1, "", "add_insertsize"]], "sctoolbox.tools.marker_genes": [[2, 1, 1, "", "add_gene_expression"], [2, 1, 1, "", "get_chromosome_genes"], [2, 1, 1, "", "get_rank_genes_tables"], [2, 1, 1, "", "label_genes"], [2, 1, 1, "", "mask_rank_genes"], [2, 1, 1, "", "pairwise_rank_genes"], [2, 1, 1, "", "run_deseq2"], [2, 1, 1, "", "run_rank_genes"], [2, 1, 1, "", "score_genes"]], "sctoolbox.tools.multiomics": [[2, 1, 1, "", "merge_anndata"]], "sctoolbox.tools.norm_correct": [[2, 1, 1, "", "atac_norm"], [2, 1, 1, "", "batch_correction"], [2, 1, 1, "", "evaluate_batch_effect"], [2, 1, 1, "", "normalize_adata"], [2, 1, 1, "", "tfidf"], [2, 1, 1, "", "tfidf_normalization"], [2, 1, 1, "", "wrap_batch_evaluation"], [2, 1, 1, "", "wrap_corrections"]], "sctoolbox.tools.peak_annotation": [[2, 1, 1, "", "annotate_adata"], [2, 1, 1, "", "annotate_narrowPeak"]], "sctoolbox.tools.qc_filter": [[2, 1, 1, "", "apply_qc_thresholds"], [2, 1, 1, "", "automatic_thresholds"], [2, 1, 1, "", "calculate_qc_metrics"], [2, 1, 1, "", "estimate_doublets"], [2, 1, 1, "", "filter_cells"], [2, 1, 1, "", "filter_genes"], [2, 1, 1, "", "get_keys"], [2, 1, 1, "", "get_mean_thresholds"], [2, 1, 1, "", "get_thresholds_wrapper"], [2, 1, 1, "", "predict_cell_cycle"], [2, 1, 1, "", "predict_sex"], [2, 1, 1, "", "thresholds_as_table"], [2, 1, 1, "", "validate_threshold_dict"]], "sctoolbox.tools.receptor_ligand": [[2, 1, 1, "", "calculate_interaction_table"], [2, 1, 1, "", "connectionPlot"], [2, 1, 1, "", "download_db"], [2, 1, 1, "", "get_interactions"], [2, 1, 1, "", "hairball"], [2, 1, 1, "", "interaction_progress"], [2, 1, 1, "", "interaction_violin_plot"], [2, 1, 1, "", "progress_violins"]], "sctoolbox.tools.tobias": [[2, 1, 1, "", "write_TOBIAS_config"]], "sctoolbox.tools.tsse": [[2, 1, 1, "", "add_tsse_score"], [2, 1, 1, "", "calc_per_base_tsse"], [2, 1, 1, "", "global_tsse_score"], [2, 1, 1, "", "overlap_and_aggregate"], [2, 1, 1, "", "tsse_scoring"], [2, 1, 1, "", "write_TSS_bed"]], "sctoolbox.utils": [[3, 0, 0, "-", "adata"], [3, 0, 0, "-", "assemblers"], [3, 0, 0, "-", "bioutils"], [3, 0, 0, "-", "checker"], [3, 0, 0, "-", "creators"], [3, 0, 0, "-", "decorator"], [3, 0, 0, "-", "general"], [3, 0, 0, "-", "io"], [3, 0, 0, "-", "jupyter"], [3, 0, 0, "-", "multiprocessing"], [3, 0, 0, "-", "tables"]], "sctoolbox.utils.adata": [[3, 1, 1, "", "add_expr_to_obs"], [3, 1, 1, "", "add_uns_info"], [3, 1, 1, "", "get_adata_subsets"], [3, 1, 1, "", "get_cell_values"], [3, 1, 1, "", "get_minimal_adata"], [3, 1, 1, "", "load_h5ad"], [3, 1, 1, "", "prepare_for_cellxgene"], [3, 1, 1, "", "save_h5ad"], [3, 1, 1, "", "shuffle_cells"]], "sctoolbox.utils.assemblers": [[3, 1, 1, "", "convertToAdata"], [3, 1, 1, "", "from_mtx"], [3, 1, 1, "", "from_quant"], [3, 1, 1, "", "from_single_mtx"], [3, 1, 1, "", "from_single_starsolo"], [3, 1, 1, "", "prepare_atac_anndata"]], "sctoolbox.utils.bioutils": [[3, 1, 1, "", "barcode_index"], [3, 1, 1, "", "convert_id"], [3, 1, 1, "", "gene_id_to_name"], [3, 1, 1, "", "get_organism"], [3, 1, 1, "", "pseudobulk_table"], [3, 1, 1, "", "unify_genes_column"]], "sctoolbox.utils.checker": [[3, 1, 1, "", "check_columns"], [3, 1, 1, "", "check_file_ending"], [3, 1, 1, "", "check_marker_lists"], [3, 1, 1, "", "check_module"], [3, 1, 1, "", "check_type"], [3, 1, 1, "", "format_adata_var"], [3, 1, 1, "", "get_index_type"], [3, 1, 1, "", "gunzip_file"], [3, 1, 1, "", "in_range"], [3, 1, 1, "", "is_integer_array"], [3, 1, 1, "", "is_regex"], [3, 1, 1, "", "is_str_numeric"], [3, 1, 1, "", "validate_regions"], [3, 1, 1, "", "var_index_from"]], "sctoolbox.utils.creators": [[3, 1, 1, "", "add_analysis"], [3, 1, 1, "", "build_notebooks_regex"], [3, 1, 1, "", "gitlab_download"], [3, 1, 1, "", "setup_experiment"]], "sctoolbox.utils.decorator": [[3, 1, 1, "", "debug_func_log"], [3, 1, 1, "", "get_parameter_table"], [3, 1, 1, "", "log_anndata"]], "sctoolbox.utils.general": [[3, 1, 1, "", "clean_flanking_strings"], [3, 1, 1, "", "get_binary_path"], [3, 1, 1, "", "get_datetime"], [3, 1, 1, "", "get_package_versions"], [3, 1, 1, "", "get_user"], [3, 1, 1, "", "identify_columns"], [3, 1, 1, "", "longest_common_suffix"], [3, 1, 1, "", "read_list_file"], [3, 1, 1, "", "remove_prefix"], [3, 1, 1, "", "remove_suffix"], [3, 1, 1, "", "run_cmd"], [3, 1, 1, "", "sanitize_string"], [3, 1, 1, "", "scale_values"], [3, 1, 1, "", "setup_R"], [3, 1, 1, "", "split_list"], [3, 1, 1, "", "split_list_size"], [3, 1, 1, "", "write_list_file"]], "sctoolbox.utils.io": [[3, 1, 1, "", "create_dir"], [3, 1, 1, "", "get_temporary_filename"], [3, 1, 1, "", "remove_files"], [3, 1, 1, "", "rm_tmp"]], "sctoolbox.utils.jupyter": [[3, 1, 1, "", "clear"]], "sctoolbox.utils.multiprocessing": [[3, 1, 1, "", "get_pbar"], [3, 1, 1, "", "monitor_jobs"]], "sctoolbox.utils.tables": [[3, 1, 1, "", "fill_na"], [3, 1, 1, "", "rename_categories"], [3, 1, 1, "", "table_zscore"], [3, 1, 1, "", "write_excel"]]}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "function", "Python function"], "2": ["py", "class", "Python class"], "3": ["py", "method", "Python method"]}, "objtypes": {"0": "py:module", "1": "py:function", "2": "py:class", "3": "py:method"}, "terms": {"": [1, 2, 3, 7, 14, 17], "0": [1, 2, 3, 5, 6, 7, 8, 12, 13, 14, 15, 17, 18, 19, 20, 21, 22], "00008": 2, "01": [1, 2, 4, 6, 9, 14, 15, 23], "0125": [2, 18], "017": 2, "018": 22, "019": [2, 7], "02": [4, 9, 15, 23], "020": 6, "021": 6, "02464": 2, "025": 14, "03": [1, 9, 15, 23], "03_batch": 4, "04": [9, 15, 23], "0414": 22, "04_cluster": [20, 21, 22], "05": [2, 13, 15], "05_marker_gen": 4, "0619": 7, "08": 15, "08bf36": 20, "0a": [15, 23], "0b": [15, 23], "0c": [15, 23], "0rc1": 4, "1": [1, 2, 3, 15], "10": [1, 2, 3, 5, 6, 7, 14, 17, 20, 21], "100": [1, 2, 3, 6, 7, 13, 14, 18, 21, 22], "1000": [1, 2, 3, 8, 14, 18, 19, 21], "10000": 2, "1000000": [1, 5], "1000bp": 2, "103": 4, "1038": [2, 6, 7, 22], "104": 5, "11": 15, "110": 4, "111": 4, "117": 4, "118": 4, "1186": [2, 6], "119": 4, "12": 2, "122": 4, "124": 4, "126": 4, "128": 4, "129": 4, "130": 4, "134": 4, "135": 4, "136": 4, "137": 4, "138": 4, "139": 4, "140": 4, "143": 4, "145": 4, "146": 4, "147": 4, "148": 4, "149": 4, "14_velocity_analysi": 4, "15": [1, 2, 7, 8, 18, 19, 20], "150": 5, "151": 4, "152": 4, "15252": [2, 17], "153": 4, "156": 4, "157": 4, "162": 4, "164": 4, "165": 4, "166": 4, "167": 4, "173": 4, "174": 4, "175": 4, "176": 4, "177": 4, "179": 4, "180": 4, "181": 4, "1850": 2, "187": 4, "188": 4, "19": 2, "190": 4, "191": 4, "192": 4, "1929": 6, "193": 4, "194": 4, "195": 4, "199": 4, "1e4": 2, "2": [1, 2, 3, 15], "20": [1, 2], "200": [1, 2, 4, 8, 19], "2000": [2, 14], "2000000": [1, 2], "201": 4, "2013": 2, "2017": 2, "2018": 22, "20188746": [2, 17], "202": 4, "2022": 2, "2023": 15, "205": 4, "207": 4, "209": 4, "211": 4, "213": 4, "214": 4, "21583": 6, "22": 15, "220": 4, "227": 4, "229": 4, "23": 15, "232": 4, "233": 4, "24": 15, "25": [2, 14, 20], "250": 1, "2500": 1, "27": 15, "28": 15, "2d": 1, "2g": 22, "3": [1, 2, 3, 15], "30": [1, 2, 8, 15, 19, 21], "300": [1, 2, 14], "3000000": 1, "31": 15, "33": 4, "3389": 2, "3500000": 1, "37071230": 3, "3d": [1, 4], "3ea053a67c1242ceb8f0e7616e1bd482": 13, "4": [1, 2, 3, 15], "400": [8, 19], "42": [1, 3], "43": 4, "4346": 7, "44": 4, "45": 1, "46": 4, "494": 22, "498": 22, "4b": [2, 3], "5": [1, 2, 3, 15], "50": [2, 7, 8, 14, 18, 19, 21, 22], "5000": [2, 6, 18], "54": 4, "560": 22, "57": 4, "6": [1, 2, 3, 15], "60": [1, 2, 3, 8, 19], "600": [1, 7, 8, 18, 19], "7": [1, 2, 15, 20], "78": 4, "8": [1, 2, 5, 13], "800": [8, 19], "8000": 1, "89": 3, "9": [1, 2, 4, 6, 12, 20], "90": [3, 4, 21], "95": [2, 3], "99": 3, "A": [1, 2, 3, 7, 10, 12, 14, 18, 22], "As": [2, 5, 7, 18], "But": 3, "By": [2, 3, 14], "For": [1, 2, 5, 6, 7, 10, 13, 14, 16, 17, 20], "If": [1, 2, 3, 5, 6, 7, 10, 12, 13, 14, 17, 18], "In": [3, 7, 12, 13, 14, 17, 18, 21], "It": [3, 10, 12], "NOT": 2, "No": 22, "On": [2, 3, 6, 22], "One": [2, 10], "Or": [1, 2], "The": [1, 2, 3, 4, 5, 6, 7, 8, 10, 12, 14, 16, 17, 18, 19, 20, 21, 22], "There": [7, 14, 18, 20], "These": [2, 17, 18], "To": [2, 5, 6, 7, 8, 12, 18, 19], "Will": [2, 3, 7, 18, 21], "With": 3, "_": [1, 3, 6, 7, 8, 12, 14, 17, 18, 19, 20], "__version__": 12, "_add_path": 4, "_barcod": [3, 16], "_boxplot": 13, "_cellxgen": 12, "_cluster_": 20, "_clustering_": 20, "_color": 4, "_correlation_ob": 7, "_correlation_var": 7, "_custom_gene_list_express": 20, "_express": 20, "_filter": 20, "_filter_object": 4, "_gene": [3, 16], "_mamplan": 12, "_matrix": [3, 16], "_nestedsequ": [1, 2, 3], "_overlap_two_bedfil": 4, "_pca_variance_select": 7, "_reclust": 2, "_save_figur": [8, 14, 18, 19, 20], "_stripplot": 13, "_supportsarrai": [1, 2, 3], "_version": 4, "_vs_": 20, "abcd": 1, "abil": 4, "abl": 12, "about": [7, 8, 10, 12, 18, 19], "abov": [2, 5, 7, 14, 16, 18, 21], "absolut": [1, 6, 7, 17, 18], "ac": 2, "accept": [2, 3, 7, 18], "access": [1, 2, 4, 12], "accord": 3, "account": [6, 13, 14], "across": [1, 2, 6, 17, 20], "act": 12, "action": 5, "activ": [7, 18, 19], "actual": 10, "ad": [1, 2, 3, 4, 5], "adapt": 4, "adata": [0, 1, 2, 4, 5, 11, 15, 16], "adata1": 1, "adata2": 1, "adata_1": 1, "adata_2": 1, "adata_correct": 7, "adata_input_dir": [3, 10, 12, 13, 14], "adata_input_prefix": 3, "adata_list": 1, "adata_nam": [1, 7], "adata_norm": 7, "adata_output": [5, 6, 7, 12, 16, 17, 18], "adata_output_dir": [3, 10, 12, 13, 14], "adata_output_prefix": 3, "adata_rna": 16, "adata_sub": 20, "add": [1, 2, 3, 4, 6, 8, 10, 13, 19], "add_analysi": [0, 3], "add_cellxgene_annot": [0, 2], "add_expr_to_ob": [0, 3], "add_fld_metr": 5, "add_gene_express": [0, 2], "add_hlin": [0, 1], "add_inserts": [0, 2], "add_insertsize_metr": 4, "add_label": 1, "add_spac": [0, 1], "add_track": [0, 1], "add_tsse_scor": [0, 2], "add_uns_info": [0, 3, 17, 18], "add_xaxi": [0, 1], "addit": [0, 1, 3, 4, 5, 7, 10, 12, 16, 18, 19], "addition": [2, 12, 20], "additional_nod": 2, "additon": 12, "address": 3, "adjust": [2, 3, 4, 5, 8, 16, 18, 19, 20], "adjusted_p_valu": 13, "advanc": 10, "affect": [2, 7, 18], "aforement": 16, "after": [2, 5, 7, 18, 22], "ag": [1, 12], "again": 3, "against": [1, 2], "aggreg": [2, 3], "aid": 18, "aim": [7, 13, 20], "al": [2, 22], "algorithm": [2, 7, 8, 14, 19], "align": 2, "alignmentfil": 2, "all": [1, 2, 3, 4, 5, 6, 7, 10, 12, 13, 17, 18, 20, 21], "all_track": 1, "allow": [4, 7, 8, 14, 18, 19], "along": [1, 2, 14], "alpha": [2, 14, 22], "alphabet": 1, "alreadi": [2, 3], "also": [2, 6, 13, 14, 20], "altern": 2, "ambigu": 3, "amount": [12, 20], "an": [1, 2, 3, 4, 7, 10, 12, 18, 20, 22], "analyi": 2, "analys": 17, "analysi": [0, 3, 4, 5, 7, 10, 11, 12, 15, 16, 18, 23], "analysis_nam": 3, "analyst": [4, 7, 12, 18], "ani": [1, 2, 3, 7, 18], "anndata": [1, 2, 3, 9, 10, 15, 23], "anndata_1": [5, 6, 16, 17], "anndata_14": 22, "anndata_2": [6, 7, 17, 18], "anndata_3": [7, 8, 18, 19], "anndata_4": [8, 10, 12, 14, 19, 20, 21, 22], "anndata_5": [13, 20], "anndata_annot": 10, "anndata_dict": 2, "anndata_fil": [20, 21], "anndata_overview": [0, 1, 4, 7, 18], "anndata_pseudotim": 14, "anndata_scanpro": 13, "annot": [1, 2, 3, 11, 13, 15], "annot_hvg": [0, 2, 18], "annotate_adata": [0, 2, 8], "annotate_gen": 4, "annotate_narrowpeak": [0, 2], "annotate_region": 2, "anoth": [2, 6, 7], "api": 15, "append": [3, 7, 17, 18, 20], "appli": [2, 4, 5, 7, 10, 16, 18], "applic": 1, "apply_qc_threshold": [0, 2, 6, 17], "apply_svd": [0, 2], "approach": 14, "approxim": [8, 19], "ar": [1, 2, 3, 4, 5, 6, 7, 8, 10, 13, 14, 16, 17, 18, 19, 20, 22], "arc": 1, "arcsin": 13, "arg": 2, "argument": [1, 2, 3], "around": [1, 2], "arpack": 18, "arr": 3, "arrai": [1, 2, 3, 20], "arraylik": [1, 2, 3], "arrow": [14, 22], "arrow_color": 22, "arrow_length": 22, "arrow_s": 22, "articl": 2, "artifact": [7, 18], "aspect": 6, "assembl": [0, 4, 9, 15, 23], "assembli": [4, 11, 15], "assign": [2, 3, 10, 14, 17, 20], "assmun": 2, "associ": 17, "assum": [2, 13], "astyp": [6, 7, 17, 18, 20], "atac": [2, 3, 6, 15], "atac_norm": [0, 2, 4], "atac_util": 5, "attribut": 1, "attributeerror": 2, "auto": [2, 3, 4], "automat": [1, 2, 4, 13, 14], "automatic_threshold": [0, 2, 17], "avail": [1, 2, 3, 5, 6, 7, 10, 13, 16, 18], "averag": 2, "avoid": [2, 3, 7, 18], "ax": [1, 2, 3, 7], "axarr": 1, "axi": [1, 2, 6, 22], "b": [1, 2, 3], "back": [7, 18], "backend": 20, "backend_pdf": 20, "background": 2, "background_gradi": 22, "bam": [0, 4, 5], "bam_adata_ov": [0, 2], "bam_fil": [2, 5], "bam_in": 2, "bam_obj": 2, "bam_out": 2, "bam_to_bigwig": [0, 2], "bamfil": [2, 5], "banana": 1, "bar": [1, 2, 3], "barcod": [1, 2, 3, 6, 17, 22], "barcode_col": [2, 5], "barcode_index": [0, 3], "barcode_regex": 2, "barcode_tag": [2, 5], "barplot": [1, 2, 4], "base": [1, 2, 3, 5, 6, 7, 8, 14, 16, 17, 18, 19, 22], "basi": [1, 2, 7, 8, 14, 18, 19, 20, 22], "basic": [7, 10, 14, 18], "batch": [1, 2, 4, 9, 12, 15, 23], "batch_column": [7, 18], "batch_correct": [0, 2, 7, 18], "batch_correction_overview": 18, "batch_kei": [2, 7, 18], "batch_method": [2, 7, 18], "bay": 13, "bbknn": [1, 2, 7, 18], "bcu": 12, "beartyp": [1, 3, 4], "becom": 10, "bed": [1, 2, 5], "bedfil": [2, 5], "bedgraphtobigwig": 2, "bedtool": 2, "bedtools_path": 2, "been": 2, "befor": [1, 2, 3, 6, 17, 18, 20], "beforehand": 2, "behavior": 21, "behind": 1, "being": [8, 19], "belong": 1, "below": [1, 2, 7, 14, 16, 18, 19], "bend": 14, "besid": [3, 7], "best": [2, 7, 18], "better": [2, 18], "between": [1, 2, 3, 4, 5, 7, 8, 12, 13, 14, 18, 19, 21, 22], "bgcolor": [5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22], "bgtobw_path": 2, "bgzip": 2, "bia": 2, "bidirect": 1, "bidirectional_barplot": [0, 1], "bigwig": [1, 2], "bigwig1": 1, "bigwig2": 1, "bin": [1, 3], "binar": 2, "binari": [2, 3, 6], "binarize_mtx": 6, "binary_lay": 2, "bio": 3, "bioconductor": 2, "bioinfo": 2, "biolog": [6, 7, 13, 17, 18, 21], "biomedcentr": 2, "bioutil": [0, 4], "black": 2, "blacklist": 2, "blood": 10, "blue": [1, 18, 22], "blueberri": 1, "bn": 12, "bn_public": [4, 12], "bool": [1, 2, 3, 6, 12], "boolean": [1, 2, 3, 4, 5, 10], "bootstrap": 13, "border": 1, "bos_tauru": 2, "both": [2, 7, 18], "box": 1, "boxplot": [0, 1, 13], "branch": [3, 14], "braun": 2, "broken": [3, 17], "browser": 12, "buffer": 2, "buffer_s": 2, "bug": [4, 12], "bugfix": 4, "build": [3, 4], "build_notebooks_regex": [0, 3], "bulk": 20, "bulk_label": 1, "bw": [1, 2], "byte": [1, 2, 3], "c": [1, 2, 3, 12, 22], "c0c2c4": 20, "c1": [1, 13], "c2": [1, 13], "c3": 1, "caenorhabditis_elegan": 2, "calc_frip_scor": [0, 2], "calc_overlap_fc": [0, 4], "calc_overlap_pct": 5, "calc_per_base_tss": [0, 2], "calc_ragi": [0, 2], "calcul": [1, 2, 3], "calculate_fld_scor": 5, "calculate_interaction_t": [0, 2, 21], "calculate_overlap": 5, "calculate_qc_metr": [0, 2, 6, 17], "call": [1, 3], "callabl": [1, 2, 3], "calledprocesserror": 3, "can": [1, 2, 3, 5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22], "canis_lupus_familiari": 2, "cannot": [1, 2, 3], "capit": 12, "care": [7, 18], "case": [3, 6, 7, 14, 16, 17, 18], "cast": 12, "categor": [2, 3], "categori": [2, 3, 4, 6, 7, 17, 18], "caus": [4, 14], "cb": [2, 5], "cb_col": [2, 5], "cb_tag": [2, 5], "cbar_label": [1, 2], "cd": 4, "ceil": 20, "cell": [1, 2, 3, 4, 5, 7, 11, 12, 14, 16, 20, 21, 22], "cell_count": 22, "cell_distribution_barplot": [8, 19], "cell_filt": [6, 17], "cell_filtering_fin": [6, 17], "cell_filtering_scatt": [6, 17], "cell_nam": 2, "cellcycl": 2, "celltalkdb": [2, 21], "celltyp": [2, 12, 13], "celltype_annot": 0, "celltype_column": 2, "celltype_column_nam": 10, "cellular": [6, 8, 17, 19, 21], "cellxgen": [2, 3, 4, 11, 15], "certain": [2, 3, 5, 12], "chagn": 2, "challeng": [7, 18], "chamber": [13, 20], "chang": [2, 3, 5, 6, 8, 16, 18, 19, 20, 22], "changelog": 15, "char_list": 3, "charact": 2, "character": [8, 19], "check": [1, 2, 3, 4, 12, 13, 20], "check_barcode_tag": [0, 2, 5], "check_chang": 4, "check_column": [0, 3], "check_ensembl": 10, "check_file_end": [0, 3], "check_marker_list": [0, 3], "check_modul": [0, 3], "check_onlin": 12, "check_typ": [0, 3], "checker": 0, "choic": [1, 3, 8, 14, 19], "choos": [3, 5, 6, 8, 14, 16, 17, 19], "chosen": [1, 2, 7, 18], "chr": [2, 3, 5], "chr1": [1, 2], "chrm": 6, "chrm_percent": 2, "chromatin": 2, "chromosom": [2, 6], "chrx": 6, "chry": 6, "chunk": 3, "chunk_siz": 3, "chunk_size_bam": 5, "ci": 4, "cicd": 4, "citat": 12, "class": 1, "clean": 4, "clean_flanking_str": [0, 3], "clear": [0, 3], "click": 10, "clip": 2, "close": [3, 6, 17, 20], "close_logfil": [5, 6, 7, 8, 13, 14, 16, 17, 18, 19, 22], "closer": 18, "clsuter": 2, "cluster": [0, 3, 4, 6, 7, 9, 10, 12, 13, 14, 15, 18, 21, 22, 23], "cluster_": 20, "cluster_a": 2, "cluster_b": 2, "cluster_col": [20, 21], "cluster_column": [1, 2, 21], "cluster_marker_genes_": 20, "cluster_on": 1, "clustered_adata": 10, "clustergrid": 1, "clustering_col": [13, 20], "clustering_column": [8, 10, 19], "clustering_method": [8, 19], "clustering_search": 19, "clustermap": 1, "clustermap_dotplot": [0, 1], "clusters_col": 13, "cmap": [1, 3, 20, 22], "cmd": [2, 3], "cml_paramet": 10, "cn": [2, 21], "coars": [8, 19], "code": 4, "coeffici": [1, 2, 14], "coher": 22, "col": [3, 12, 22], "col_clust": 1, "col_nam": 2, "collect": 1, "color": [1, 2, 3, 4, 7, 8, 14, 18, 19, 20, 21, 22], "color_bi": [1, 7, 18], "color_cel": 14, "color_list": 1, "color_max": [2, 21], "color_min": [2, 21], "colorbar": [1, 2, 4], "colormap": [1, 2], "colors_ratio": 1, "colum": 2, "column": [1, 2, 3, 4, 6, 7, 8, 10, 13, 14, 16, 17, 18, 19, 20, 21, 22], "column_3d": [8, 19], "column_ad": 2, "column_nam": 2, "column_specific_term": 10, "columns_ad": 3, "com": [2, 3], "comapr": 1, "combat": [2, 18], "combin": [1, 7, 8, 16, 18, 19, 21], "command": [1, 2, 3], "commandlin": 3, "commit": 3, "common": [2, 3], "commonli": 17, "commun": 21, "compar": [1, 2, 7, 13, 18], "compare_embed": [0, 1, 7], "compare_two_conditon": [0, 2], "comparison": [2, 4, 6, 9, 10, 15, 17, 23], "compat": 2, "complet": 4, "complex": [1, 2, 3, 6, 14], "compon": [1, 2, 4, 7, 8, 18, 19], "compos": 14, "composit": 13, "compress": 3, "comput": [2, 7, 12, 15, 18, 23], "compute_pca": [0, 2], "con_figs": 21, "con_filt": 21, "con_lig_whitelist": 21, "con_lw_multipli": 21, "con_out_suffix": 21, "con_rec_whitelist": 21, "con_restrict_to": 21, "con_titl": 21, "con_wspac": 21, "cond1": [6, 13, 17], "cond2": [6, 13, 17], "conda": 3, "condit": [1, 2, 3, 4, 6, 13, 16, 17, 21], "condition_a": 2, "condition_b": 2, "condition_col": [2, 13, 20], "condition_column": [2, 6, 8, 17, 19], "condition_doublet_remov": 6, "conds_col": 13, "conduc": 6, "config": [2, 4, 5, 6, 7, 8, 12, 16, 17, 18, 19, 20, 21, 22], "config_fil": 1, "configur": [1, 2, 3], "confirm": 10, "confound": 2, "connect": 2, "connection_alpha": [2, 21], "connectionerror": 3, "connectionplot": [0, 2, 4, 21], "conserv": 17, "consid": [2, 14, 18, 20], "consider": [7, 18], "consist": 2, "consol": [1, 3], "constraint": 4, "construct": 3, "conta": 3, "contain": [1, 2, 3, 4, 5, 6, 8, 10, 12, 14, 16, 17, 18, 22], "containig": 2, "containing": 2, "content": [1, 12], "context": 21, "contin": [1, 3], "continu": [4, 14, 20], "contour": 1, "contrast": [1, 17], "control": [1, 2, 3, 6, 8, 10, 17, 19], "convers": 2, "convert": [2, 3, 4, 6, 12], "convert_id": [0, 3], "converttoadata": [0, 3, 5, 16], "coolwarm": 22, "coordin": [1, 2, 3, 5], "coordinate_col": [2, 3, 5, 8], "coordinate_column": 3, "copi": [1, 2, 3, 4, 5, 6, 16, 18, 19, 22], "core": [12, 18], "corr": 1, "corr_kwarg": [2, 7, 18], "corr_on": [1, 7, 18], "corr_plot": [1, 7, 18], "corr_thresh": [1, 2, 7, 18], "corrcoef": 1, "correct": [1, 2, 3, 9, 15, 21, 23], "correl": [1, 2, 4, 18, 22], "correlate_condit": [0, 2], "correlate_ref_vs_al": [0, 2], "correlation_matrix": [0, 2, 4], "correlation_threshold": 2, "correspond": [1, 2, 3, 6, 17], "could": [2, 3, 5, 7, 14, 16, 18], "count": [1, 2, 3, 6, 13, 17, 18, 20, 21, 22], "count_fragments_per_cel": [0, 2], "covari": [13, 17], "cover": [2, 4], "coverag": [2, 4], "cpu": [12, 14], "cpu_limit": 12, "cpu_request": 12, "creat": [1, 2, 3, 4, 8, 10, 12, 17, 18, 19, 20, 21], "create_dir": [0, 3], "create_fragment_fil": [0, 2, 4], "creation": [11, 15], "creator": 0, "critic": 10, "crucial": 21, "csr_matrix": 2, "csv": [2, 3], "ctrl1": 1, "ctrl2": 1, "ctrl3": 1, "ctrl4": 1, "cumul": [1, 2, 7, 18], "current": [1, 2, 3, 10, 12], "curv": 14, "custom": [1, 2, 3, 4, 10, 18], "custom_config": 2, "custom_gene_list": 20, "custom_marker_annot": 4, "custom_tss": 2, "cwt": 4, "cycl": [2, 7], "cytoplasm": 17, "d": 20, "danio_rerio": 2, "dash": 1, "data": [1, 2, 3, 4, 8, 12, 13, 14, 19, 20, 21, 22], "databas": [2, 12], "datafam": 1, "datafram": [1, 2, 3, 4, 20, 22], "datalabel": 2, "datalist": 2, "dataset": [1, 2, 3, 4, 5, 7, 12, 16, 18, 21], "datatyp": 12, "date": [3, 10], "db_path": [2, 21], "dd": 3, "de": [3, 8, 19], "debug_func_log": [0, 3], "decid": [1, 2, 6, 17, 18], "decis": [2, 7, 18], "deco": 13, "decompos": 2, "decomposit": 2, "decompress": 3, "decor": [0, 4, 13], "decreas": [2, 4, 7, 12], "dedic": [5, 16], "defaul": 1, "default": [1, 2, 3, 4, 6, 7, 8, 12, 18, 19], "default_threshold": 6, "defin": [1, 2, 3, 5], "define_pc": [0, 2], "deg_": 20, "deg_max_out_group_fract": 20, "deg_min_fold_chang": 20, "deg_min_in_group_fract": 20, "deg_n_gen": 20, "deg_ranking_method": 20, "deg_styl": 20, "deg_tabl": 20, "degrad": [7, 18], "degre": 1, "del": [5, 16], "delet": [3, 4, 16], "delimit": 3, "dend_height": 1, "dend_width": 1, "dendrogram": [1, 14], "densiti": [1, 4], "depedn": 3, "depend": [1, 2, 3, 4, 7, 18], "deplet": [21, 22], "deploi": 12, "deploy": [3, 4], "deplyoment": [2, 12], "deprec": [2, 3, 4], "depth": [2, 17, 18], "deriv": [6, 14], "describ": [2, 22], "descript": 3, "deseq2": [2, 20], "deseq_": 20, "deseq_result": 2, "deseq_t": 20, "design": [2, 7, 10, 18], "desir": [2, 5], "dest": 3, "destin": 3, "detail": [10, 12, 21], "detect": [2, 3, 14, 17], "determin": [1, 4, 6, 7, 13], "df": [1, 2, 3, 22], "df_cond_a": 2, "df_cond_b": 2, "diagram": 1, "diagramm": 2, "dict": [1, 2, 3, 20], "dictionari": [1, 2, 3, 10, 13], "dictonari": 2, "differ": [1, 2, 3, 4, 7, 8, 12, 13, 14, 18, 19, 20], "differenti": [1, 13, 20, 22], "diffexpr": 2, "dim_reduct": 0, "dimens": [1, 2, 4, 6, 8, 14, 19], "dimension": [1, 2, 7, 8, 19], "dir": 3, "direct": 3, "directli": [2, 14, 18, 22], "directori": [1, 2, 3, 4, 16], "disabl": [2, 3, 4, 7, 18], "disintegr": 4, "displai": [1, 2, 3, 4, 5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22], "dist": [8, 19], "dist_rang": [8, 19], "distanc": [2, 8, 14, 19], "distinct": [8, 19], "distribut": [1, 2, 6, 17, 21], "divid": [2, 10], "divis": [2, 18], "dna": 6, "do": [2, 3, 7], "do_clust": [7, 18], "doc": 4, "docker": 12, "document": [2, 4, 13, 14, 22], "doe": [1, 2, 3], "doesn": 2, "doi": [2, 6, 7, 17, 22], "doku": 4, "done": [2, 3, 5, 16, 18], "dot": [1, 2, 20], "dotplot": [1, 4, 20], "doublet": 2, "doublet_scor": [2, 6], "doublet_threshold": [6, 17], "down": 1, "downgrad": 2, "download": [2, 3], "download_db": [0, 2, 21], "downstream": [2, 7, 18], "dpi": [1, 2, 14, 21], "dpi_sav": [7, 8, 14, 18, 19], "draw": 1, "drawn": 1, "driven": [7, 18], "drop": [5, 16], "drop_ob": [5, 16], "drop_var": [5, 16], "drosophila_melanogast": 2, "dst123": 12, "dt": 1, "dtype": [1, 2, 3, 14], "dual": 2, "due": [3, 4, 14, 18, 20], "duplic": [3, 4, 6, 17], "dure": [2, 18], "dy": 17, "e": [1, 2, 3, 4, 5, 6, 7, 10, 12, 14, 16, 17, 18, 20, 21, 22], "each": [1, 2, 3, 6, 7, 12, 13, 14, 18, 20], "earli": 16, "earliest": 14, "edg": 2, "edge_s": 2, "edge_size_per_bas": 2, "edge_size_tot": 2, "edu": [2, 21], "effect": [2, 3, 8, 9, 15, 19, 23], "either": [1, 2, 3, 5, 6, 7, 14, 16, 18, 19, 20], "elast": 14, "elbow": 2, "element": [1, 2, 3], "elif": [7, 8, 14, 18, 19, 20], "elpigraph": 14, "els": [2, 5, 6, 7, 12, 14, 17, 18, 20], "email": 10, "embed": [0, 3, 4, 6, 7, 9, 14, 15, 18, 20, 23], "embedding_clust": [8, 19], "embedding_nam": 3, "embedding_qu": 19, "embedding_sample_distribut": 19, "empir": 13, "empti": [8, 12, 19], "en": [1, 2, 14], "enabl": [1, 2, 10], "encod": 4, "end": [1, 2, 3, 4, 5], "endswith": 5, "enrich": [2, 6, 21], "ensembl": [2, 3, 10], "ensembl_id": 3, "ensg00000164690": 3, "ensur": [6, 7, 17, 18], "enter": 10, "entir": 3, "entri": 3, "env": 4, "environ": 4, "epg": 14, "epg_lambda": 14, "epg_mu": 14, "epi": [6, 7], "episcanpi": [6, 7], "equal": [2, 21], "error": [2, 3, 4, 5, 16, 18, 20], "errror": 3, "especi": 5, "estim": [1, 2, 3, 6], "estimate_doublet": [0, 2, 6, 17], "et": [2, 22], "etc": [1, 2, 3, 6, 7, 17, 18, 20], "euclidean": 7, "evalu": [2, 4, 6, 17], "evaluate_batch_effect": [0, 2], "even": 10, "everi": [1, 3, 12], "examin": 17, "exampl": [1, 2, 3, 4, 7, 9, 11, 14, 17, 18, 20, 23], "exce": 3, "excel": [2, 3], "except": [2, 3, 10, 12, 14], "exclud": [2, 3, 6, 7, 18], "exclude_highly_express": [2, 18], "exclus": 21, "execut": [2, 3], "exist": [2, 3, 5, 10, 16, 21], "exon": 2, "exp_id": 12, "expect": [1, 2, 3, 7, 18], "experi": [3, 7, 18], "experiment": [7, 18], "explain": [1, 7, 18], "explore_sigma": 14, "export": 10, "expr": 1, "express": [1, 2, 3, 4, 8, 14, 16, 17, 19, 21, 22], "exsit": 3, "ext123": 12, "extens": 2, "extern": 2, "extra": 2, "extract": [1, 2, 3], "f": [1, 7, 8, 12, 13, 14, 17, 18, 19, 20, 21, 22], "f_in": 3, "f_out": 3, "fa2": 4, "factor": [1, 2, 7, 18], "fail": [1, 2], "fall": [7, 18], "fals": [1, 2, 3, 5, 6, 7, 8, 14, 17, 18, 19, 20, 21, 22], "fasta": 2, "fc": 2, "fc_fragments_in_region": [0, 2, 5], "fcgr3a": 20, "featur": [1, 2, 3, 4, 5, 7], "feature_anchor": 2, "femal": [2, 17], "female_gen": 17, "female_predict": 17, "fetch": 2, "few": 17, "field": 22, "fifth": 2, "fig": 20, "figdir": 22, "figsiz": [1, 2, 21], "figur": [1, 2, 3, 6, 13, 14, 17, 18, 19, 21], "figure_dir": [2, 13, 14, 20, 22], "fike": 3, "file": [1, 2, 3, 4, 5, 8, 10, 12, 21, 22], "file1": 1, "file_convert": 16, "file_list": 3, "file_nam": 10, "file_regex": 3, "file_typ": 1, "filenam": [1, 2, 3, 22], "filenotfounderror": [2, 3], "filepath": 2, "fill": [3, 5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22], "fill_na": [0, 2, 3], "fillna": 1, "filter": [2, 3, 4, 7, 9, 15, 18, 20, 21, 23], "filter_and_norm": 22, "filter_cel": [0, 2, 6, 17], "filter_chrm": 6, "filter_doublet": [6, 17], "filter_gen": [0, 2, 17], "filter_gend": 17, "filter_method": [7, 18], "filter_mito": 17, "filter_rank_genes_group": 2, "filter_ribo": 17, "filter_xi": 6, "filterwarn": 14, "final": [1, 5, 6, 7, 10, 12, 13, 14, 16, 22], "final_pc_select": 18, "final_threshold": [6, 17], "find": [2, 3], "fine": 10, "fiorella": 2, "first": [1, 2, 4, 7, 10, 18, 19, 20], "fischer": 2, "fit": [4, 12], "five": 10, "fix": [1, 4, 12, 20], "flag": [2, 4], "flank": 2, "flat": 1, "flatfil": 2, "fld": [4, 5], "fld_score": 5, "fld_score_cwt": 6, "flip": 1, "flip_embed": [0, 1], "float": [1, 2, 3, 6, 14, 18], "float32": 14, "float64": [2, 14], "floatrangeslid": 1, "fold": [2, 20], "foldchang": [2, 20], "foldchange_threshold": [1, 2], "folder": [1, 2, 3, 6, 10, 17], "follow": [1, 2, 3, 7, 10, 16, 18], "force_homologi": 10, "fork": 14, "form": [1, 3], "format": [1, 2, 3, 16, 22], "format_adata_var": [0, 3], "forward": [1, 2, 3], "found": [1, 2, 3, 4, 14, 20, 22], "four": 1, "fphy": 2, "fraction": [1, 2, 6, 17, 20], "frag_count": 2, "fragemnt": 2, "fragment": [2, 4, 6], "fragments_fil": [2, 5], "frame": 2, "framework": [2, 13], "frequenc": 2, "frip": [0, 4, 6], "from": [1, 2, 3, 4, 6, 7, 8, 10, 12, 13, 14, 17, 18, 19, 20, 21, 22], "from_column": 3, "from_mtx": [0, 3, 16], "from_quant": [0, 3, 16], "from_single_mtx": [0, 3], "from_single_starsolo": [0, 3], "frue": 2, "fruit": 1, "full": [1, 2, 3], "full_figure_prefix": [8, 19], "fulli": 4, "func": 3, "function": [1, 2, 3, 4, 5, 6, 8, 19, 21], "function_nam": 21, "funtion": 1, "further": [2, 12], "fust": 2, "futur": [1, 22], "g": [1, 2, 3, 4, 5, 6, 7, 10, 12, 14, 16, 17, 18, 20, 21, 22], "g1": 1, "g2": 1, "g2m": [1, 2], "g2m_gene": [2, 18], "g2m_score": 1, "g3": 1, "gallus_gallu": 2, "gb": [2, 12], "gender": [2, 6, 10, 17], "gene": [1, 2, 3, 4, 7, 10, 12, 14, 15, 19, 21, 23], "gene_col": 21, "gene_column": [2, 17], "gene_correl": 0, "gene_expression_heatmap": [0, 1, 4], "gene_filt": 17, "gene_filtering_fin": 17, "gene_id_to_nam": [0, 3], "gene_index": [2, 21], "gene_list": 1, "gene_name_column": 1, "gene_row": 1, "gene_set": 2, "gene_symbol": 2, "gener": [0, 2, 4, 5, 15, 17], "genes_index": 3, "genom": [1, 2], "genomebiologi": 2, "genometrack": [0, 4], "genometrack_x": 1, "genotyp": 10, "gerner": 4, "get": [1, 2, 3, 4, 6, 8, 12, 14, 17, 19], "get_adata_subset": [0, 3], "get_atac_thresholds_wrapp": 4, "get_bam_read": [0, 2], "get_binary_path": [0, 3], "get_cell_valu": [0, 3], "get_celltype_assign": [0, 2], "get_chromosome_gen": [0, 2], "get_datetim": [0, 3], "get_df": 22, "get_index_typ": [0, 3], "get_interact": [0, 2, 21], "get_kei": [0, 2], "get_mean_threshold": [0, 2, 6, 17], "get_minimal_adata": [0, 3], "get_organ": [0, 3], "get_package_vers": [0, 3], "get_parameter_t": [0, 3], "get_pbar": [0, 3], "get_rank_genes_t": [0, 2, 4, 20], "get_slider_threshold": [0, 1, 6, 17], "get_temporary_filenam": [0, 3], "get_thresholds_wrapp": [0, 2, 4, 6], "get_us": [0, 3, 12], "get_variable_featur": [0, 2, 6], "gi": 12, "gini": [0, 2], "github": 2, "gitlab": [3, 4], "gitlab_download": [0, 3], "gitlab_token": 3, "give": [1, 2], "given": [1, 2, 3, 6, 7, 14, 17, 18, 20, 21, 22], "glob": 3, "global": [1, 2, 6, 17], "global_default": 1, "global_threshold": [1, 6, 17], "global_tsse_scor": [0, 2], "gnuplot": 22, "go": 1, "goal": 18, "goi": 22, "gradual": 2, "grapefruit": 1, "graph": [2, 18, 21], "graph_object": 1, "greater": 2, "grei": 1, "grey_colormap": [0, 1], "grid": 1, "group": [1, 2, 3, 4, 6, 8, 17, 19, 21, 22], "group1": [1, 3], "group2": [1, 3], "group_a": [2, 21], "group_b": [2, 21], "group_correl": [0, 1], "group_expression_boxplot": [0, 1], "group_heatmap": 4, "groupbi": [1, 2, 3, 4, 6, 8, 17, 19, 22], "grouped_violin": [0, 1], "groups_dict": 1, "gtf": [1, 2, 3, 5, 8], "gtf_file": 2, "gtf_path": 8, "gunzip_fil": [0, 3], "gwdg": [3, 12], "gwdgmanagt": 12, "h": 3, "h5ad": [2, 3, 6, 7, 8, 10, 12, 13, 14, 17, 18, 19, 20, 21, 22], "h5ad_path": 3, "ha": [1, 2, 3, 7, 17], "had": 17, "hairbal": [0, 2, 21], "hand": 22, "handl": 4, "harmoni": [2, 7, 18], "have": [1, 2, 3, 5, 6, 8, 16, 18, 19], "hbox": 1, "head": [20, 22], "header": [1, 3], "heart_mark": 10, "heatmap": [1, 7, 18, 20], "heg": 2, "height": [1, 2, 8, 19], "help": [7, 18, 21], "here": [5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22], "hes4": 1, "heterogen": [7, 8, 19], "hexbin": [1, 4], "hexbin_grids": 1, "hg38_gene": 8, "hidden": 3, "hide_edg": 2, "high": [3, 6, 7, 8, 17, 19], "higher": [1, 2, 8, 18, 19, 22], "highest": [1, 7, 14, 18], "highest_expr_gen": 18, "highli": [1, 2, 4, 7], "highlight": [1, 2, 14, 20], "highly_express": 18, "highly_vari": [0, 6, 18], "highly_variable_gen": 2, "hint": 4, "hit": 2, "hitrat": 2, "hline": 1, "hmgb2": 1, "hold": [2, 5, 12, 21], "home": 3, "homo_sapien": [2, 5], "homologi": 10, "horizont": 1, "host": 3, "how": [1, 2, 3, 7, 17, 18, 22], "howev": [7, 10, 18, 20], "html": [1, 2, 4], "html_file": [8, 19], "http": [1, 2, 3, 7, 14, 21, 22], "hue": 1, "human": [2, 10, 17, 18], "human_lr_pair": [2, 21], "hvf": [1, 6], "hvg": 2, "hvg_rang": [2, 18], "i": [1, 2, 3, 4, 5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22], "ibm": 2, "id": [2, 3, 10, 12, 14, 21], "id_col_nam": 3, "identif": [18, 20], "identifi": [2, 3, 7, 8, 13, 18, 19, 20], "identify_column": [0, 3], "idf": 2, "idf_typ": 2, "ifram": [8, 19], "ignor": [2, 6, 14], "ignore_overwrit": 10, "ii": 6, "imag": [3, 14, 22], "impact": 21, "implement": [2, 4, 6], "import": [1, 2, 3, 4, 5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22], "importerror": 3, "improv": [2, 3], "in_rang": [0, 3, 4], "inacces": 3, "inch": [1, 2], "includ": [2, 3, 4, 10, 12, 13, 18], "include_clip": 2, "include_limit": 3, "incorrect": 3, "increas": [4, 22], "independ": 2, "index": [1, 2, 3, 5, 7, 17, 18, 20, 21, 22], "index_bam": 2, "index_from": [3, 5], "indic": [2, 17, 18, 22], "individu": [2, 3, 5, 6, 8, 17, 19, 22], "individual_pbar": 2, "inf": 18, "infer": [2, 3, 7, 14], "inform": [1, 2, 3, 4, 5, 10, 12, 13, 16, 18, 21], "inher": [8, 19], "ini": 1, "init": 18, "initi": [1, 3], "initial_threshold": [6, 17], "inlin": [6, 17], "inner": 2, "inpalc": 2, "inplac": [2, 3, 5, 7, 8, 12, 16, 18, 20, 21], "input": [2, 3, 5, 6, 16, 17, 20], "inserts": [0, 1], "insertsize_count": 2, "insertsize_distribut": [1, 2], "insid": [1, 3], "insight": [8, 19, 21], "instal": [2, 3, 10, 12], "instead": [1, 2, 3, 10], "insuffic": 20, "insufficient_s": 20, "int": [1, 2, 3, 12, 14, 20, 22], "int32": 14, "int64": 14, "integ": [1, 3], "integr": [4, 17], "intend": [7, 16, 18, 20], "interact": [1, 2, 10], "interaction_perc": [2, 21], "interaction_progress": [0, 2], "interaction_scor": [2, 21], "interaction_t": 21, "interaction_violin_plot": [0, 2, 21], "interest": [2, 14], "interfer": [7, 14, 18], "intermedi": 2, "intern": [2, 3, 4], "internal_path": 3, "interpret": 21, "intersect": [7, 18], "introduc": [7, 18], "invalid": [2, 3, 12, 14, 20], "invers": [2, 7], "inverse_freq": 2, "inverse_freq_smooth": 2, "invert": 1, "investig": [2, 21], "io": [0, 1, 2], "ipython": [8, 19], "ipywidget": [1, 4], "is_gend": 17, "is_integer_arrai": [0, 3], "is_mito": 17, "is_regex": [0, 3], "is_ribo": 17, "is_str_numer": [0, 3], "isin": 20, "issu": [2, 3, 4], "issubset": 22, "item": [1, 3, 7, 20], "iter": [1, 2, 3, 4, 8, 19], "itgb2": 1, "its": [7, 22], "j": [2, 20], "job": [3, 4], "join": [2, 8, 19], "jupyt": [0, 1, 10], "just": 18, "k": 22, "keep": [2, 3, 4, 10, 12], "keep_al": 10, "keep_ob": [3, 12], "keep_temp": 2, "keep_tmp": 2, "keep_var": [3, 12], "kei": [1, 2, 3, 5, 6, 7, 8, 16, 17, 18, 19, 20, 21, 22], "kendal": 1, "kept": [2, 3, 7, 18], "kernel": 4, "key_ad": [2, 8, 19], "keyerror": [1, 2, 3, 7, 18], "keyword": [1, 3], "kind": 13, "knee": 2, "known": [6, 20], "kondili": 2, "kubernet": 12, "kuenn": 2, "kwarg": [1, 2, 3, 4], "la": 22, "label": [1, 2, 3, 12], "label_gen": [0, 2, 17], "lager": 1, "lambda": [5, 16], "larg": [2, 17], "larger": 1, "last": [7, 18], "last_notebook_adata": [12, 13, 14, 22], "late": 16, "latent": 2, "later": [8, 19], "latest": [1, 10, 12], "latter": 6, "layer": [1, 2, 3, 6, 18, 20], "ldap": 12, "lead": [2, 8, 19], "leak": 17, "learn": [4, 8, 19], "learning_r": [1, 8, 19], "learning_rate_rang": [1, 8, 19], "least": [1, 2, 3], "leav": 1, "left": 1, "left_label": 1, "left_valu": 1, "legend": 1, "legend_loc": 1, "leiden": [1, 2, 7, 8, 10, 18, 19], "leiden_0": [8, 10, 19, 20], "leiden_col": 12, "lemon": 1, "len": [6, 17, 20, 22], "length": [1, 2, 3, 6, 14], "less": [1, 8, 17, 19], "level": [2, 18], "lf": 4, "lib": 3, "librari": 6, "life": 10, "ligand": [2, 4, 15, 23], "ligand_clust": [2, 21], "ligand_cluster_col": [2, 21], "ligand_cluster_s": 2, "ligand_col": [2, 21], "ligand_column": [2, 21], "ligand_gen": [2, 4, 21], "ligand_gene_symbol": [2, 21], "ligand_hu": [2, 21], "ligand_perc": [2, 21], "ligand_s": [2, 21], "ligand_scor": [2, 21], "lightgrei": 1, "like": [2, 7, 22], "likelihood": 14, "limit": [2, 3, 10, 12, 13], "line": [1, 2, 3, 4], "line_color": [2, 21], "line_styl": 1, "line_width": 1, "linear": [13, 14], "lineplot": 1, "linewidth": [2, 21], "link": [1, 3], "lisi": [1, 2, 7, 18], "lisi_scor": [1, 2], "list": [1, 2, 3, 4, 5, 6, 7, 9, 11, 12, 15, 18, 19, 21, 23], "list_of_str": 3, "listedcolormap": 1, "lists_path": 10, "liter": [1, 2, 3, 4], "load": [1, 3, 9, 15, 23], "load_dataset": 1, "load_h5ad": [0, 3, 5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22], "load_token": 3, "loc": 6, "local": [2, 3, 7, 18], "locat": [3, 5, 17], "log": [1, 2, 3, 4, 12, 13, 14], "log10": 1, "log1p": 2, "log1p_mean_count": 17, "log1p_n_gen": 2, "log1p_total_count": [2, 17], "log_anndata": [0, 3, 13], "log_fil": [12, 13, 14], "log_idf": 2, "log_tf": 2, "log_tfidf": 2, "logarithm": 2, "logger": 3, "logit": 13, "long": 1, "longest": 3, "longest_common_suffix": [0, 3], "look": [1, 3, 8, 19], "looso": 2, "lot": 2, "louvain": [1, 2, 8, 19], "low": [7, 17], "lower": [1, 3, 7, 8, 19], "lsi": [0, 2, 4], "lst": 3, "lw_multipli": [2, 21], "m": [2, 3], "macrophag": 20, "mai": [1, 7, 17, 18], "main": [3, 10, 18], "mainli": 4, "make": [1, 2, 4, 7, 8, 18, 19], "male": 2, "mamplan": [11, 15], "mamplan_cr": 12, "mamplan_filenam": 12, "mampok": [4, 12], "manangemnt": 12, "mandatori": [1, 6], "mango": 1, "mani": 7, "manifold": [8, 19], "manno": 22, "manual": [2, 4, 6], "manual_threshold": 2, "map": [1, 2, 3, 5, 6, 14, 16, 17], "map_async": 3, "mark": 14, "marker": [1, 2, 3, 4, 11, 12, 15, 23], "marker_dict": 3, "marker_gen": [0, 4, 17], "marker_gene_clust": [0, 1], "marker_genes_": 20, "marker_genes_dict": [1, 2], "marker_list": 10, "marker_pred_celltyp": 12, "marker_repo": 10, "marker_styl": 20, "marker_t": 20, "marker_table_col": 20, "marker_velocity_genes_": 22, "markerrepo": 10, "mask": 2, "mask_rank_gen": [0, 2], "match": [1, 2, 3, 4, 5], "matplotlib": [1, 2, 6, 7, 17, 19, 20], "matplotlib_venn": 1, "matric": 2, "matrix": [1, 2, 3, 16], "matrx": 4, "matter": 2, "max": [1, 2, 3, 6, 7, 8, 10, 12, 13, 14, 17, 18, 19, 20, 21, 22], "max_cal": 3, "max_cel": [2, 6], "max_clust": 1, "max_column": [5, 10, 13, 16], "max_dim": 2, "max_disp": 18, "max_dist": 2, "max_iter": 2, "max_limit": 18, "max_mean": 18, "max_out_group_fract": [2, 20], "max_queue_s": 2, "max_row": [5, 16], "max_siz": 3, "max_threshold": 1, "maxi": 3, "maximum": [1, 2, 3, 7, 18, 20], "mc": 12, "mean": [2, 3, 6, 17, 20, 22], "mean_inserts": 2, "mean_threshold": [6, 17], "measur": [1, 2, 6, 7, 14, 17, 22], "median": 1, "mediat": 21, "mem_limit": 12, "mem_request": 12, "membran": 17, "memori": [2, 3, 12], "merg": [2, 3, 22], "merge_anndata": [0, 2], "mertk": 20, "messag": [2, 3], "messeng": 22, "metacol": 7, "metadata": [3, 4, 12], "method": [1, 2, 3, 4, 7, 8, 13, 14, 18, 19, 20], "method_kwarg": 2, "metric": [2, 4, 6, 7, 18], "middl": 1, "might": 3, "mileston": 14, "million": 2, "min": [1, 2, 6, 21], "min_bia": 2, "min_cel": [2, 6], "min_corr": 22, "min_disp": 18, "min_dist": [1, 2, 8, 19], "min_dist_rang": [1, 8, 19], "min_edge_width": 22, "min_fold_chang": [2, 20], "min_gen": [6, 17], "min_in_group_fract": [2, 20], "min_limit": 18, "min_mean": [2, 18], "min_perc": [2, 21], "min_threshold": 1, "min_valu": 2, "mind": [2, 12], "mini": 3, "minim": [2, 3], "minimum": [2, 3, 6, 17, 20, 21], "miss": 3, "mitochondir": 12, "mitochondri": [6, 17], "mitochondria": 17, "mitochrondri": [2, 17], "mix": [2, 3, 7, 18], "mm": 3, "mnn": [2, 7, 18], "mode": [2, 22], "model": 2, "moder": 13, "modifi": [2, 3, 4, 5], "modifii": 14, "modul": [1, 2, 3, 4, 5, 6], "modulenotfounderror": [10, 12], "moment": 22, "monitor": 3, "monitor_job": [0, 3], "more": [2, 3, 4, 8, 10, 12, 13, 14, 19], "most": [2, 14, 22], "mostli": 3, "motif": 2, "mous": 2, "mouse_lr_pair": 2, "move": 4, "mpl": 3, "mr": 10, "mr_ob": 10, "mrc1": 20, "mrna": [14, 17, 22], "msb": [2, 17], "mtx": 3, "multiindex": 2, "multiom": [0, 3], "multipanel": 1, "multipl": [1, 2, 3, 5, 8, 14, 16, 19], "multiplet": 6, "multipli": [2, 21], "multiprocess": [0, 2, 8, 18, 19], "muon": 2, "mus_musculu": 2, "must": [1, 2, 3, 6, 17, 20, 21], "my": 1, "my3d_umap": 1, "my_b": 1, "my_mark": 10, "n": [2, 3, 14, 18], "n_cell": 1, "n_cells_a": 2, "n_cells_aft": 17, "n_cells_b": 2, "n_cells_barplot": [0, 1, 8, 19], "n_cells_befor": 17, "n_cells_by_count": [1, 2, 6, 17], "n_col": 13, "n_column": 13, "n_comp": 2, "n_compon": [1, 2, 7, 8, 18, 19], "n_count": 1, "n_featur": [6, 7], "n_fragment": 2, "n_gene": [1, 2, 4, 7, 17, 18, 20], "n_genes_deg_plot": 20, "n_genes_markerplot": 20, "n_job": [14, 22], "n_map": 14, "n_marker": 20, "n_neighbor": [7, 18], "n_pc": [1, 2, 7, 18, 22], "n_pcs_heatmap": [7, 18], "n_rep": 13, "n_sim": 13, "n_thread": 14, "n_thresh": [1, 7, 18], "na": [2, 3, 20], "name": [1, 2, 3, 5, 6, 7, 8, 10, 14, 16, 17, 18, 19, 20, 21, 22], "name_col": 3, "nan": [1, 2, 4, 6, 20], "narrowpeak": 2, "nativ": 6, "natur": 22, "ncol": [1, 6, 8, 14, 17, 18, 19, 20, 22], "ncorrel": 7, "ndarrai": [1, 2, 3, 4], "ndims_rep": 14, "nearest": 18, "necessari": [10, 20], "need": [1, 2, 3, 5, 6, 7, 8, 12, 14, 16, 19, 22], "negativ_shift": 2, "neighbor": [8, 19, 22], "neighborhood": [2, 18], "neither": 2, "nerd": 12, "nest": [1, 3], "net_color_max": 21, "net_color_min": 21, "net_interaction_perc": 21, "net_interaction_scor": 21, "net_min_perc": 21, "net_out_suffix": 21, "net_restrict_to": 21, "net_show_count": 21, "net_titl": 21, "network": 2, "new": [1, 2, 3, 4, 5, 6, 12, 16], "newer": [8, 19], "next": 10, "nicer": 12, "nmeth": 7, "node": [2, 14], "node_size_scal": 22, "nois": 6, "non_m": 6, "non_xi": 6, "none": [1, 2, 3, 5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22], "norm_correct": [0, 4], "norm_method": 7, "normal": [1, 2, 4, 20, 21], "normalis": 2, "normalize_adata": [0, 2, 7, 18], "normalize_tot": [2, 20], "note": [1, 2, 3, 8, 12, 18, 19, 20, 21, 22], "notebook": [1, 3, 5, 6, 10, 12, 13, 14, 15, 16, 17, 20, 21, 22], "noth": 3, "now": 4, "np": [1, 2, 3, 6, 7, 20], "nproc": 2, "npt": [1, 2, 3], "nrg2825": 7, "nrow": 1, "nuclear": 6, "nucleosom": [4, 5], "numba": 4, "number": [1, 2, 3, 5, 6, 7, 8, 10, 13, 14, 17, 18, 19, 21, 22], "numer": [1, 2, 14], "numpi": [1, 2, 4, 6, 7, 20], "ob": [1, 2, 3, 4, 5, 6, 7, 8, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22], "obj": [1, 3], "obj_nam": 3, "object": [1, 2, 3, 4, 6, 7, 9, 10, 12, 15, 17, 21, 23], "obs_cell_typ": 1, "obs_column": [1, 6, 14, 17], "obs_figur": [6, 17], "obs_group": 1, "obs_metr": [17, 18, 19], "obs_slider_dict": [6, 17], "observ": [1, 2, 7, 14, 18], "obsm": [1, 2, 3], "obsm_kei": 2, "obtain": 2, "occurr": 3, "offer": 13, "often": [2, 12], "old": 3, "old_nam": [5, 16], "om_col": 1, "omit": [1, 2], "onc": 3, "one": [1, 2, 3, 5, 6, 7, 8, 10, 12, 14, 16, 18, 19], "onli": [1, 2, 3, 4, 5, 12, 16, 17, 18, 21], "onlin": 12, "only_automatic_threshold": [2, 6], "onteract": 2, "onto": 14, "open": 2, "open_bam": [0, 2], "optim": 6, "option": [1, 2, 3, 7, 10, 12, 14, 20], "option_context": [5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22], "orang": 1, "order": [1, 2, 3, 14], "org": [2, 3, 7, 14, 22], "organ": [2, 3, 10, 12], "orient": 1, "origin": [2, 3, 6, 20], "orphan": 4, "oryzias_latip": 2, "other": [1, 2, 4, 7, 8, 12, 18, 19, 20, 22], "other_column": 2, "otherwis": [1, 2, 3, 5], "our": 7, "out": [2, 6, 13, 17], "out_group_fract": [2, 20], "out_path": [2, 3], "outdat": 4, "outdir": 2, "outer": 2, "outlier": [3, 17], "output": [1, 2, 3, 7, 16, 18, 21], "output_prefix": 2, "outsid": 1, "over": [2, 3], "overal": 2, "overhaul": 4, "overlai": 12, "overlap": [2, 22], "overlap_and_aggreg": [0, 2], "overlap_fragments_in_region": 2, "overlay_previ": 1, "overrid": 1, "overview": [1, 4, 5, 16], "overwrit": [2, 3, 10, 21], "overwriten": 3, "owner": 12, "p": [1, 2, 13], "packag": [2, 3, 12, 22], "package1": 3, "package2": 3, "page": 4, "pair": [2, 21], "pairwis": [1, 2, 6, 17], "pairwise_rank_gen": [0, 1, 2], "pairwise_scatt": [0, 1, 4, 6, 17], "pairwise_t": 1, "palett": [1, 3], "panda": [1, 2, 3, 5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22], "panglao": 10, "panglao_scor": 10, "panglao_two_column": 10, "papaya": 1, "parallel": [2, 3], "param": 2, "paramert": 1, "paramet": [1, 2, 3, 4, 6, 7, 8, 13, 17, 18, 19, 21, 22], "parameter1": 1, "parameter2": 1, "pars": 12, "part": [2, 10], "partli": 4, "pass": [1, 2, 3], "path": [1, 2, 3, 4, 5, 6, 10, 13, 14, 16, 17, 21, 22], "path_h5ad": [5, 16], "path_mtx": [5, 16], "path_quant": [5, 16], "path_rd": [5, 16], "pathlib": 3, "pattern": [3, 5], "pbmc3k": 2, "pbmc3k_process": 1, "pbmc68k_reduc": [1, 2], "pc": [2, 22], "pca": [1, 2, 3, 4, 14], "pca_": 7, "pca_correlation_ob": 18, "pca_correlation_var": 18, "pca_embed": 18, "pca_overview": 7, "pca_variance_final_select": 18, "pca_variance_proposed_select": 18, "pcna": 1, "pct_counts_": 17, "pct_counts_is_mito": [7, 12, 18], "pct_counts_is_ribo": [7, 12, 18], "pct_fragments_in_": 2, "pct_fragments_in_promot": [0, 2], "pd": [1, 2, 3, 4, 5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22], "pdf": [6, 7, 8, 13, 14, 17, 18, 19, 20, 21, 22], "pdfpage": 20, "peak": [2, 3, 5, 6, 17], "peak_annot": 0, "peak_chr": 3, "peak_end": 3, "peak_start": 3, "peakqc": [4, 5], "peaks_thr": 5, "pearson": [1, 2], "pearsonr": [1, 2], "penalti": 14, "per": [1, 2, 3, 6, 10, 13, 16], "per_base_tss": 2, "perc_thresh": [2, 7, 18], "percent": [2, 21], "percent_mito": 1, "percent_top": 2, "percentag": [1, 2, 21], "percentil": [2, 3, 7, 18, 21], "percentile_rang": [2, 3], "perform": [2, 3, 5, 7, 10, 14, 17, 18, 20], "perform_batch_correct": [7, 18], "period": 3, "perplex": [1, 2, 8, 19], "perplexity_rang": [1, 8, 19], "person": 12, "phase": [1, 2, 7, 12, 18], "pick": 14, "pin": 4, "pineappl": 1, "pip": 2, "pipelin": [3, 4, 5], "pl": [1, 2, 3, 4, 6, 7, 8, 14, 17, 18, 19, 20, 22], "pl_method": 20, "place": 2, "placehold": 2, "plai": 21, "pleas": [1, 2, 5, 10, 12, 14, 16, 22], "plot": [0, 2, 3, 4, 5, 6, 10, 13, 15, 17, 22], "plot_3d_umap": [0, 1, 8, 19], "plot_comparison": 2, "plot_correct": 2, "plot_differential_gen": [0, 1], "plot_embed": [0, 1, 4, 14], "plot_gene_correl": [0, 1], "plot_group_embed": [0, 1, 19], "plot_inserts": [0, 1], "plot_pca_correl": [0, 1, 4, 7, 18], "plot_pca_vari": [0, 1, 4, 7, 18], "plot_starsolo_qu": [0, 1, 4, 6, 17], "plot_starsolo_umi": [0, 1, 4, 6, 17], "plot_suffix": 13, "plot_valu": 1, "plot_venn": [0, 1, 2, 4], "plotli": 1, "plt": [6, 7, 17, 19], "png": [1, 6, 17, 19], "point": [1, 2, 8, 14, 19, 22], "pool": 3, "poor": 14, "popul": [8, 19], "posit": [1, 2], "positiv_shift": 2, "possibilti": 2, "possibl": [3, 4, 12], "post": 4, "potenti": [7, 18, 20], "powderblu": [5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22], "pp": [2, 6, 7, 18, 20, 22], "ppt": 14, "ppt_lambda": 14, "ppt_nstep": 14, "ppt_sigma": 14, "pre": [6, 17], "preced": 5, "preceed": 16, "precis": 22, "pred_celltyp": 10, "predefin": 2, "predict": [1, 2, 12], "predict_cell_cycl": [0, 2, 18], "predict_sex": [0, 2, 4, 17], "predicted_doublet": [2, 6, 17], "predominantli": 20, "prefix": [1, 2, 3, 10], "preliminari": [7, 18], "prepar": [3, 4, 7, 11, 15], "prepare_atac_anndata": [0, 3, 5], "prepare_for_cellxgen": [0, 3, 4, 12], "prepare_for_cellxgene_log": 12, "preprocess": [0, 3, 5], "present": [2, 3, 12], "preserv": [8, 19], "pretti": 1, "preussner": 2, "prevent": [4, 18], "previou": [1, 6, 17], "previous": 2, "princip": [2, 7, 14, 18], "principi": 14, "principl": 7, "print": [1, 2, 3, 6, 10, 14, 17, 20, 22], "prior": [20, 21, 22], "prioriti": 2, "privat": 3, "prmt2": 1, "probabilist": 14, "problemat": 12, "proce": [7, 14], "process": [3, 5, 7, 10, 12, 14, 16, 18, 21], "processed_data": [2, 21], "processs": 2, "profil": 2, "progress": [1, 2, 3], "progress_violin": [0, 2], "project": [8, 12, 19], "project_id": 12, "promot": [2, 5], "promoters2000": 5, "prop": 13, "prop_cycl": 3, "properti": [5, 7], "proport": [2, 11, 15], "proportion_analysi": [4, 13], "propos": [2, 7, 18], "propose_pc": [0, 2, 4, 7, 18], "provid": [2, 3, 5, 6, 7, 10, 14, 16, 18, 19, 20, 21, 22], "pseudo": [1, 13], "pseudobulk": [2, 3, 20], "pseudobulk_t": [0, 2, 3], "pseudotim": [1, 4, 11, 15], "pseudotime_analysi": 4, "pseudotime_analysis_log": 14, "pseudotime_heatmap": [0, 1], "pseudotime_trajectory_": 14, "psuedobulk": 3, "public": [5, 12, 16], "pubm": 12, "pubmedid": 12, "pull": 2, "pumc": 2, "push": 4, "pvalu": [1, 2], "pvalue_threshold": 1, "py": 4, "pygenometrack": 1, "pyplot": [1, 2, 6, 7, 17, 19], "pysam": 2, "pysam_thread": 2, "python": [2, 3, 4], "python_path": 2, "qc": [1, 2, 4, 5, 9, 15, 18, 19, 23], "qc_column": 19, "qc_filter": [0, 5, 6, 17], "qc_var": 17, "qualiti": [1, 2, 4, 14], "quality_violin": [0, 1, 6, 17], "quality_violinplot": 4, "quant": [1, 3], "quant_fold": [6, 17], "quantil": 1, "queri": [2, 10], "question": [3, 7, 18], "queue": 2, "r": [2, 3, 4, 5, 22], "r_home": [3, 4], "ragi": [2, 4], "rainbow": [2, 21], "rais": [1, 2, 3, 5, 7, 10, 12, 14, 16, 18, 20, 22], "ram": 2, "randint": 1, "random": [1, 3], "rang": [1, 2, 3, 8, 14, 19, 21], "rank": [2, 10], "rank_gen": 20, "rank_genes_": 20, "rank_genes_column": 10, "rank_genes_group": [1, 2], "rank_genes_groups_dotplot": 1, "rank_genes_groups_filt": 1, "rank_genes_groups_matrixplot": 1, "rank_genes_kei": 20, "rank_genes_plot": [0, 1, 20], "rank_tabl": 1, "rank_velocity_gen": 22, "ranking_method": 20, "rapidli": 18, "raspberri": 1, "rat": 2, "rate": [2, 3, 14, 22], "ratio": [1, 2, 6, 22], "rattus_norvegicu": 2, "raw": [2, 3, 4, 5, 6, 18, 20], "raw_norm": 20, "rcparam": 3, "rd": [3, 5, 16], "re": [2, 5, 14, 16], "reach": 3, "read": [1, 2, 3, 6, 22], "read_csv": 3, "read_list_fil": [0, 3], "read_mtx": 3, "read_tag": 2, "reader": 2, "reader_thread": 2, "readi": 3, "readm": 4, "readthedoc": [1, 2, 4], "recalcul": [6, 17], "receiv": 3, "receptor": [2, 4, 15, 23], "receptor_clust": [2, 21], "receptor_cluster_col": [2, 21], "receptor_cluster_s": 2, "receptor_col": [2, 21], "receptor_column": [2, 21], "receptor_gen": [2, 4, 21], "receptor_gene_symbol": [2, 21], "receptor_hu": [2, 21], "receptor_ligand": [0, 21], "receptor_perc": [2, 21], "receptor_s": [2, 21], "receptor_scor": [2, 21], "reclust": [0, 2, 4], "recommend": [8, 19], "red": 1, "reduc": [7, 14], "reduct": [1, 2, 6, 8, 19], "ref_gen": [1, 2], "refactor": [3, 4], "refer": [1, 2, 8, 10, 15], "reference_ob": 10, "reformat": 3, "refrenc": 2, "regex": [2, 3, 4], "region": [1, 2, 3, 5, 6], "region_nam": 5, "regions_fil": [2, 5], "regions_nam": [2, 5], "regplot": 1, "regress": 13, "regular": 14, "rel": 1, "relat": [1, 3, 4, 5, 6, 7, 10, 12, 18], "relationship": [8, 19], "releas": 4, "remain": 22, "remov": [1, 2, 3, 4, 5, 7, 12, 16, 17, 18, 20], "remove_bool": [2, 6, 17], "remove_fil": [0, 3], "remove_prefix": [0, 3], "remove_suffix": [0, 3], "remove_temp": 2, "renam": [3, 4, 12, 16], "rename_categori": [0, 3, 8, 19], "rename_ob": [3, 5, 12, 16], "rename_var": [3, 5, 12, 16], "reorganis": 4, "replac": [1, 2, 3, 4, 12], "replic": 13, "repo": [3, 4], "repo_path": 10, "report": [2, 20], "repositori": [3, 12], "repres": [1, 2, 3, 17, 18], "represent": [8, 19], "request": 12, "requir": [1, 2, 3, 7, 12, 14, 18, 20, 21, 22], "rerun": 17, "reset_index": 1, "resolut": [1, 2, 8, 19], "resolution_rang": 1, "resourc": [7, 12], "respect": [1, 2, 3, 17, 19, 21], "respons": [3, 12], "restrict": [2, 21], "restrict_to": [2, 21], "restructur": 4, "result": [1, 2, 6, 7, 12, 13, 14, 17, 18], "return": [1, 2, 3, 20], "return_fig": 20, "revers": 2, "review": 20, "revis": 4, "ribosom": [2, 12, 17], "right": 1, "right_label": 1, "right_valu": 1, "rl": 21, "rl_": 21, "rm_dir": 3, "rm_tmp": [0, 3], "rna": [3, 6, 15, 22], "robj": [3, 5, 16], "robust": [2, 14], "role": 21, "room_air": [3, 16], "root": [3, 14], "rotat": 1, "row": [1, 2, 3, 6, 7, 8, 10, 12, 13, 14, 17, 18, 19, 20, 21, 22], "row_clust": 1, "rpy2": [2, 3], "rst": 4, "rule": 4, "run": [1, 2, 3, 5, 6, 7, 14, 16, 17, 18, 20, 21, 22], "run_cmd": [0, 3], "run_deseq2": [0, 2, 4, 20], "run_rank_gen": [0, 2, 4, 20], "run_scsa": [0, 2], "runtim": [2, 4, 12, 18], "s13059": [2, 6], "s41467": 6, "s41586": 22, "s41592": 7, "s41598": 2, "s_gene": [2, 18], "s_score": 1, "said": 20, "sam": 2, "same": [1, 2, 3, 18], "sampl": [1, 2, 3, 5, 6, 7, 8, 12, 13, 16, 18, 19, 20], "sample1": [2, 3, 16], "sample2": [2, 3, 16], "sample_col": [2, 13, 20], "sample_dict": 6, "samples_col": 13, "sanit": 3, "sanitize_str": [0, 3], "satisfi": 5, "save": [1, 2, 3, 4, 21], "save_dens": 5, "save_excel": [2, 20], "save_h5ad": [0, 3, 5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 22], "save_overview": 5, "save_token": 3, "savefig": 20, "sc": [1, 2, 3, 6, 7, 8, 14, 18, 19, 20], "sc_colormap": [0, 1, 20], "sc_framework": 3, "scale": [1, 2, 3, 6, 8, 19], "scale_embed": 2, "scale_factor": 2, "scale_path": 14, "scale_valu": [0, 3], "scane": 16, "scanorama": [2, 7, 18], "scanpi": [1, 2, 3, 4, 6, 7, 8, 12, 14, 18, 19, 21], "scanpro_analysis_log": 13, "scanpro_uns_dict": 13, "scatac": 2, "scatac_pbmc": 5, "scatac_pbmc_frag": 5, "scatter": [1, 22], "scatter3d": 1, "scatter_hvf_distribut": [0, 1], "scatterplot": [1, 20], "scf": 14, "scfate": 4, "scheme": 16, "scientif": 2, "scikit": 4, "scipi": 2, "score": [1, 2, 3, 4, 7, 10, 18, 21], "score_gen": [0, 2], "score_nam": 2, "scr": 6, "script": 4, "scrna": [12, 14, 22], "scrub": 6, "scrub_doublet": 6, "scrublet": [2, 6, 17], "scsa": [2, 10, 12], "scsa_ob": 10, "scsa_pred_celltyp": [2, 12, 13], "sct": 2, "sctoolbox": [1, 2, 3, 4, 5, 6, 7, 8, 10, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22], "scv": 22, "scvelo": 4, "se": 10, "seaborn": 1, "search": [2, 8, 10, 19], "search_clustering_paramet": [0, 1, 8, 19], "search_paramet": [8, 19], "search_tsne_paramet": [0, 1, 8, 19], "search_umap_paramet": [0, 1, 8, 19], "second": [1, 2, 3, 10, 19], "second_round": 14, "section": [2, 4, 7, 10, 18, 20], "see": [1, 2, 3, 12], "seed": [1, 3, 14], "seg": 14, "segment": 14, "sel_col": 1, "selction": [8, 19], "select": [1, 2, 3, 4, 7, 10, 12, 14, 17, 20, 21], "select_highly_vari": 6, "selected_pc": [7, 18], "selection_color": 20, "semant": 2, "sep": 2, "separ": [1, 2, 6], "seq": [2, 3, 6], "sequenc": [3, 7, 18], "seri": 3, "server": 3, "set": [1, 2, 3, 4, 5, 6, 7, 10, 12, 16, 17, 18, 20, 21, 22], "set_figure_param": [7, 8, 14, 18, 19], "set_index": [3, 5], "set_opt": [10, 13], "set_titl": 7, "settings_from_config": [5, 6, 7, 8, 16, 17, 18, 19, 20, 21, 22], "setup": [3, 4], "setup_experi": [0, 3], "setup_r": [0, 3], "seurat": 3, "sever": [7, 18], "sex": [2, 6], "shape": [1, 6, 14, 22], "share": 1, "sheet": [2, 3], "sheet_name1": 3, "sheet_name2": 3, "shell": 4, "shift": 2, "shift_minu": 2, "shift_plu": 2, "should": [1, 2, 3, 5, 7, 10, 12, 14, 16, 18, 21], "show": [1, 2, 7, 8, 13, 14, 18, 19, 20, 21, 22], "show_attribut": 2, "show_bord": 1, "show_col_dendrogram": 1, "show_comparison": 10, "show_config": [0, 1], "show_contour": 1, "show_count": [1, 2, 21], "show_ct_tabl": 10, "show_cumul": 1, "show_grid": 1, "show_list": 10, "show_median": 1, "show_plot": [0, 1, 10], "show_row_dendrogram": 1, "show_titl": 1, "show_umap": 1, "shown": [1, 2, 3, 4, 7, 10, 12, 13, 17, 18, 20, 21], "shrink": 1, "shrink_cbar": 1, "shrink_colorbar": 1, "shuffl": 3, "shuffle_cel": [0, 3], "sigma": [5, 14], "sign": 1, "signal": [5, 6, 21], "signific": [1, 2, 13], "significan": 13, "significance_threshold": 13, "significant_chang": 13, "significantli": 22, "simga": 14, "similar": [2, 8, 19], "simpl": 1, "simplemamplan": 12, "simpleppt": 14, "simpli": 2, "simplyfi": 12, "simpson": 7, "simul": 13, "sinc": [1, 2, 3], "singel": 1, "singl": [1, 2, 3, 6, 8, 17, 19, 20, 21, 22], "singlecellexperi": 3, "singular": 2, "sinto": [2, 4], "site": 5, "size": [1, 2, 3, 20, 21, 22], "skin": 10, "skip": [6, 8, 17, 19, 20], "slider": [1, 6], "slider_dict": [1, 6, 17], "slow": [2, 18], "small": [1, 3, 4, 8, 12, 19, 20], "smaller": [1, 2, 4], "sn": 1, "snakemak": 3, "snapatac": 3, "sne": [8, 19], "so": 18, "soft": 2, "soldatov": 22, "solo": [1, 3], "solorun1": 1, "solorun2": 1, "solv": [2, 4], "some": [7, 14, 18], "sometim": [2, 14], "sort": [1, 2], "sort_bam": 2, "sortbi": 1, "sourc": [1, 2, 3, 7, 10, 14, 17, 18, 22], "space": 2, "spacer": 1, "spars": 2, "spearman": 1, "spearmanr": [1, 2, 7, 18], "speci": [2, 3, 17, 18], "speciefi": 3, "specif": [2, 10, 12, 14, 21, 22], "specifi": [1, 2, 3, 10, 13, 14, 20], "specific_clust": 13, "specific_cond": 13, "speed": [2, 22], "splice": 14, "split": [1, 2, 3, 8, 19], "split_": 2, "split_bam_clust": [0, 2], "split_list": [0, 3], "split_list_s": [0, 3], "spmatrix": 2, "spread": [1, 8, 19], "spread_rang": [1, 8, 19], "squar": 1, "src": [8, 19], "stabl": 2, "stack": 1, "stackoverflow": 3, "stage": [4, 10], "standard": [1, 2, 3, 6, 7, 16, 18], "star": 1, "starsolo": [1, 3], "starsolo_cell_select": [6, 17], "starsolo_output": [1, 6, 17], "starsolo_qu": [6, 17], "start": [2, 3, 4, 5, 14], "start_nam": 3, "starts_with": 3, "startswith": [6, 12, 20], "state": 22, "static": 1, "statist": 13, "statu": 3, "stderr": 2, "stdout": 2, "step": [1, 2, 3, 5, 6, 7, 14, 18, 22], "still": [1, 17], "stochast": [8, 19, 22], "stop": [2, 3, 5], "store": [2, 3, 5, 6, 10, 16, 17, 18, 21], "stout": 3, "str": [1, 2, 3, 5, 7, 12, 14, 18], "strawberri": 1, "stream": 22, "stream_arrow_s": 22, "strength": [7, 21], "stretch": 14, "string": [1, 2, 3, 4, 10, 14], "stringenc": [7, 18], "stripplot": 13, "structur": [2, 3, 4, 8, 16, 18, 19], "studi": 21, "style": [1, 4, 10, 20, 22], "sub": 3, "subdirectori": 4, "submit": 3, "submitt": 10, "subplot": [2, 7], "subprocess": 3, "subsequ": [1, 3, 5, 6, 7], "subset": [1, 2, 6, 17], "subset_bam": [0, 2], "subset_pc": [7, 18], "subset_pca": [0, 2, 7, 18], "substr": 3, "suffic": 20, "suffici": [10, 20], "suffix": [2, 3, 21], "suit": 18, "suitabl": [5, 14], "sum": [2, 3, 5, 6, 16, 18], "summar": 1, "summari": 1, "sumo3": 1, "support": [1, 2, 3, 4, 8, 12, 19], "sus_scrofa": 2, "svd": 2, "svd_solver": 18, "sy": 2, "symbol": [2, 10], "system": [1, 21], "t": [2, 3, 8, 14, 19, 20, 22], "tab": 12, "tab_interaction_perc": 21, "tab_interaction_scor": 21, "tab_min_perc": 21, "tab_out_suffix": 21, "tabix": 2, "tabl": [0, 1, 2, 5, 17, 20], "table_dict": 3, "table_dir": [2, 20], "table_zscor": [0, 3], "tad_classif": 1, "tag": [2, 4, 10], "take": [2, 7, 13, 18], "taken": [2, 6, 14, 22], "target": [2, 3, 10], "target_sum": [2, 18], "task": [2, 8, 19], "taxonomi": 10, "tcm": [2, 21], "technic": [7, 18], "temp": [1, 2, 3], "temp_dir": [2, 3, 4, 5, 8], "temp_fil": 3, "tempdir": [2, 3], "templat": 19, "temporari": [1, 2, 3], "term": [2, 10], "term_frequ": 2, "test": [1, 3, 4, 12, 13, 20], "test_data": [5, 8, 10, 16], "test_typ": 3, "text": 4, "tf": 2, "tf_type": 2, "tfidf": [0, 2, 7], "tfidf_norm": [0, 2], "than": [1, 2, 17], "the_10x_yml": 16, "thei": [2, 7, 10, 18], "therefor": [6, 7, 18], "theshold": 2, "thi": [1, 2, 3, 4, 5, 6, 7, 8, 10, 12, 13, 14, 16, 18, 20, 21, 22], "third": [2, 10], "those": [8, 19], "thread": [1, 2, 4, 5, 6, 7, 8, 17, 18, 19, 22], "three": [2, 10, 14, 16, 17], "threshold": [1, 2, 7, 13, 18, 21, 22], "threshold_dict": [1, 2], "thresholds_as_t": [0, 2, 6, 17], "through": [7, 16, 17, 18], "throw": 2, "thu": [14, 17], "time": [1, 3, 14, 22], "timepoint": [2, 7, 16, 18], "tip": 14, "tissu": [2, 10], "tissues_two_column": 10, "titan": 1, "titl": [1, 2, 6, 7, 17, 18, 20, 21], "tl": [1, 2, 7, 8, 14, 18, 19, 22], "tmp": 8, "to_dict": 13, "to_excel": [3, 20], "to_record": 20, "tobia": [0, 4], "tobias_output": 2, "todo": [2, 3, 6], "token": 3, "tolist": 2, "tool": [0, 1, 3, 4, 6, 7, 8, 12, 15, 17, 18, 19, 20, 21], "top": [1, 2, 7, 10, 18, 19, 20], "total": [1, 2, 3, 6, 7, 18], "total_count": [2, 7, 18], "total_read": 2, "tqdm": 3, "track": 1, "tracklabelfract": 1, "trajectori": [11, 15], "trajectory_": 14, "trajectory_infer": 14, "trajectory_tree_segments_": 14, "trajectory_tree_segments_dendogram": 14, "tran": 13, "transcript": 5, "transform": [2, 8, 13, 19], "transitions_confid": 22, "transpos": 3, "treat1": 1, "treat2": 1, "treat3": 1, "treat4": 1, "tree": 14, "tri": 2, "trigger": 4, "true": [1, 2, 3, 5, 6, 7, 8, 10, 12, 14, 16, 17, 18, 19, 20, 21], "try": [1, 10, 12, 14, 17], "tsne": [1, 2, 3, 4, 14, 20, 22], "tsne_parameter_search": [8, 19], "tss": 2, "tss_agg": 2, "tss_list": 2, "tsse": [0, 4], "tsse_df": 2, "tsse_scor": [0, 2], "tsv": [3, 21], "tune": 10, "tupl": [1, 2, 3], "two": [1, 2, 3, 4, 10, 19, 20], "two_column": 10, "txt": [1, 2, 3, 12, 13, 14, 21], "type": [1, 2, 3, 4, 7, 8, 11, 14, 15, 16, 18, 19, 20], "type_default": 1, "typeerror": [3, 14], "typic": [7, 18], "typo": 4, "u": 14, "umap": [1, 2, 3, 4, 10, 14, 18, 20, 22], "umap_3d_": [8, 19], "umap_marker_overview": [0, 1], "umap_parameter_search": [8, 19], "umap_pub": 4, "umi": 1, "umipercellsort": 1, "un": [1, 2, 3, 4, 10, 13, 18, 19, 20, 21, 22], "unabl": 4, "unari": 2, "under": 1, "underli": [7, 18], "undesir": [7, 18], "unexpect": 3, "unexpectedli": 17, "unifi": 3, "unified_column": 3, "unified_nam": 3, "uniform": [8, 19], "unify_genes_column": [0, 3], "union": [1, 2, 3, 22], "uniqu": [1, 6], "unit": 2, "univers": 2, "unknown": 3, "unmap": 14, "unnessesari": 12, "unrepl": 13, "unsplic": 14, "unsupport": 12, "until": [2, 3], "unwant": [7, 18], "up": [1, 2, 3, 16], "upcom": 7, "updat": [1, 4, 6, 13], "upper": 3, "upstream": 2, "url": 2, "uropa": [2, 8], "us": [1, 2, 3, 4, 5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 21, 22], "usag": 2, "use_bam": 5, "use_condition_column": 6, "use_default_threshold": 6, "use_highly_vari": [2, 7, 18], "use_native_scrublet": 6, "use_rep": 14, "use_sampl": 3, "user": [2, 3, 4, 10, 12], "user_db": 2, "util": [0, 2, 4, 5, 6, 7, 8, 10, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22], "v": [2, 6, 22], "valid": [1, 2, 3, 4, 12, 14], "validate_region": [0, 3], "validate_threshold_dict": [0, 2], "valu": [1, 2, 3, 4, 6, 7, 8, 12, 13, 14, 16, 17, 18, 19, 21], "value_count": 20, "valueerror": [1, 2, 3, 5, 14, 16, 20, 22], "var": [1, 2, 3, 4, 5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22], "var_column": [2, 17], "var_figur": 17, "var_index_from": [0, 3], "var_list": [1, 7], "var_method": 2, "var_nam": [6, 14], "var_slider_dict": 17, "var_typ": 6, "variabel": 6, "variability_scor": 1, "variabl": [1, 2, 3, 4, 5, 7, 10, 14, 16], "varianc": [1, 2, 7, 18], "variance_ratio": 2, "variant": 2, "variat": [1, 13, 18], "varieti": 12, "variou": [7, 18, 21], "varm": 2, "vdata": 22, "vdata_path": 22, "vector": 22, "vector_friendli": [7, 8, 14, 18, 19], "velo_graph_threshold": 22, "veloc": [0, 4, 14, 15, 23], "velocity_confid": 22, "velocity_embed": 22, "velocity_embedding_stream": 22, "velocity_embeding_stream": 22, "velocity_gen": 22, "velocity_goi": 22, "velocity_graph": 22, "velocity_length": 22, "velocity_proport": 22, "velocity_pseudotim": 22, "venn": [1, 2], "venn2": 1, "venn2_exampl": 1, "venn3": 1, "venn3_exampl": 1, "verbos": [1, 2, 4, 10], "verison": 4, "version": [2, 3, 4, 10, 12], "vertic": 1, "vhighlight": 1, "via": [1, 2, 6, 10, 12], "viabl": [6, 17], "view": 21, "violin": [1, 2, 6, 17], "violin_figs": 21, "violin_hvf_distribut": [0, 1, 6], "violin_min_perc": 21, "violin_out_suffix": 21, "violinplot": [0, 1], "vip": 12, "viridi": 1, "visual": [1, 7, 21], "vlag": 1, "vline": 1, "w": [6, 17], "wa": [1, 3, 4, 5, 6, 7, 14, 16, 17], "want": 13, "warn": [2, 3, 7, 14, 18], "wavelength": 5, "we": [2, 3, 6, 13, 17, 18], "well": [2, 7, 18], "were": 2, "what": [3, 14], "when": [1, 2, 3, 4, 6, 8, 10, 18, 19, 20], "where": [1, 2, 3, 4, 5, 14], "whether": [1, 2, 3, 6, 7, 17, 18], "which": [1, 2, 3, 4, 5, 6, 7, 12, 14, 17, 18, 20], "while": [8, 19], "white": 1, "whose": 17, "widget": [1, 6, 17], "width": [1, 2, 8, 19, 21], "wiki": [2, 12, 14], "wikipedia": [2, 14], "wilson": 2, "within": [1, 2, 3, 7, 8, 18, 19], "without": [1, 3, 4, 10], "won": 3, "word": 3, "work": [1, 2, 3, 10, 18], "would": [3, 7, 18, 20, 22], "wrap_batch_evalu": [0, 2, 7, 18], "wrap_correct": [0, 2, 7, 18], "wrap_umap": [0, 2, 7, 18], "wrapper": [1, 2, 4, 6, 10], "write": [1, 2, 3, 4, 20], "write_excel": [0, 3], "write_list_fil": [0, 3], "write_tobias_config": [0, 2, 4], "write_tss_b": [0, 2], "writeabl": 3, "writer": 2, "writer_thread": 2, "written": [2, 6, 17], "wrong": 2, "wspace": [2, 21], "www": 3, "x": [1, 2, 3, 4, 5, 6, 16, 18, 20], "x_": [1, 2, 8, 14, 19, 20], "x_pca": 2, "x_rot": 1, "x_umap": [1, 2], "xenopus_tropicali": 2, "xist": [2, 17], "xlsx": 20, "xticklabel": 1, "y": [1, 2, 7], "y_valu": 1, "yaml": [2, 5, 6, 7, 8, 12, 16, 17, 18, 19, 20, 21, 22], "yet": 3, "yield": 3, "ylabel": 1, "you": [2, 5, 7, 8, 10, 13, 14, 16, 19], "yy": 3, "z": [1, 2, 3], "zebrafish": 2, "zeisel": 22, "zero": 2, "zero_bool": [6, 17], "zju": [2, 21], "zscore": 2}, "titles": ["sctoolbox API reference", "Plotting", "Tools", "Utils", "Changelog", "01 - Assembling or loading anndata object", "02 - QC and filtering", "03 - Batch effect correction and comparisons", "04 - Embedding and clustering", "ATAC notebooks", "Cell type annotation and marker list assembly", "General notebooks", "Preparing adata for cellxgene / MaMPlan creation", "Proportion Analysis", "Trajectory and Pseudotime Analysis", "Welcome to the documentation of the single cell framework!", "01 - Assembling or loading anndata object", "02 - QC and filtering", "03 - Batch effect correction and comparisons", "04 - Embedding and clustering", "0A - Marker gene computation and analysis", "0B - Receptor-Ligand Analysis", "0C - Velocity Analysis", "RNA notebooks"], "titleterms": {"0": 4, "01": [5, 16], "02": [6, 17], "03": [4, 7, 18], "04": [4, 8, 19], "05": 4, "08": 4, "0a": 20, "0b": 21, "0c": 22, "1": [4, 5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22], "10": [8, 18, 19, 22], "11": [4, 18, 22], "12": 22, "2": [4, 5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22], "2023": 4, "22": 4, "23": 4, "24": 4, "27": 4, "28": 4, "3": [4, 5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22], "30": 4, "31": 4, "3d": [8, 19], "4": [4, 5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22], "5": [4, 5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22], "6": [4, 5, 6, 7, 8, 10, 12, 13, 14, 17, 18, 19, 20, 21, 22], "7": [4, 7, 8, 10, 13, 17, 18, 19, 21, 22], "8": [7, 8, 18, 19, 20, 21, 22], "9": [8, 18, 19, 22], "across": [8, 19], "adapt": 6, "adata": [3, 6, 7, 8, 10, 12, 13, 14, 17, 18, 19, 20, 21, 22], "add": [5, 12], "addit": [2, 6, 17], "after": [6, 17], "ambigu": 22, "an": [5, 16], "analysi": [2, 13, 14, 20, 21, 22], "anndata": [5, 6, 7, 8, 12, 13, 14, 16, 17, 18, 19, 22], "annot": [8, 10], "api": 0, "appli": [6, 17], "assembl": [3, 5, 10, 16], "assembli": 10, "atac": [4, 5, 9], "automat": [17, 20], "bam": 2, "barcod": [5, 16], "batch": [7, 18], "best": [8, 19], "between": 20, "binar": 6, "bioutil": 3, "calc_overlap_fc": 2, "calcul": [5, 6, 7, 8, 14, 17, 18, 19, 22], "cell": [6, 8, 10, 15, 17, 18, 19], "celltype_annot": 2, "cellxgen": 12, "chang": 4, "changelog": 4, "check": [5, 22], "checker": 3, "choos": [7, 18], "clean": 12, "cluster": [1, 2, 8, 19, 20], "column": [5, 12], "comparison": [7, 18], "comput": [20, 21], "condit": 20, "confid": 22, "connect": 21, "content": 15, "control": 5, "convert": [5, 16], "coordin": [8, 19], "correct": [7, 18], "correl": 7, "creat": [5, 16], "create_multiple_marker_list": 10, "creation": 12, "creator": 3, "custom": [17, 20], "cycl": 18, "data": [5, 6, 7, 10, 16, 17, 18], "databas": 21, "decor": 3, "deg": 20, "delet": 5, "dendogram": 14, "descript": [5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22], "dim_reduct": 2, "distribut": [5, 8, 19], "document": 15, "doublet": [6, 17], "download": 21, "effect": [7, 18], "embed": [1, 2, 8, 19, 22], "empti": 6, "essenti": 10, "estim": 17, "express": [18, 20], "featur": 6, "file": 16, "filter": [6, 17], "final": [8, 17, 18, 19], "find": [8, 18, 19], "folder": 16, "format": 5, "fragment": 5, "framework": 15, "frip": 2, "from": [5, 16], "gene": [6, 8, 16, 17, 18, 20, 22], "gene_correl": 2, "gener": [1, 3, 7, 8, 11, 12, 13, 18, 19, 21, 22], "genometrack": 1, "get": 21, "graph": [14, 22], "group": 20, "h5ad": [5, 16], "highli": [6, 18], "highly_vari": [1, 2], "html": [8, 19], "indic": 5, "initi": 17, "input": [7, 8, 10, 12, 13, 14, 18, 19, 21, 22], "inserts": 2, "interact": 21, "interest": 22, "io": 3, "jupyt": 3, "label": 17, "layer": 22, "leiden": 12, "length": [5, 22], "ligand": 21, "list": [10, 20], "load": [5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22], "mamplan": 12, "mark": [6, 17], "marker": [10, 20], "marker_gen": [1, 2], "matrix": 6, "measur": 19, "metric": [5, 17], "mtx": 16, "multiom": 2, "multiprocess": 3, "neighbor": [7, 18], "network": 21, "next": [7, 8, 18, 19], "norm_correct": 2, "normal": [7, 18], "note": 14, "notebook": [4, 7, 8, 9, 11, 18, 19, 23], "ob": 10, "object": [5, 8, 16, 18, 19], "option": [5, 6, 16, 17, 18, 19], "output": [13, 14], "overlap": 5, "overview": [7, 10, 18], "packag": [13, 14, 20], "paga": 22, "paramet": [10, 12, 14], "pc": [7, 18], "pca": [7, 18], "peak_annot": 2, "per": [17, 20], "pipelin": 16, "plot": [1, 7, 8, 14, 18, 19, 20, 21], "predict": [17, 18], "prepar": [5, 12, 16], "preprocess": [2, 16, 22], "process": 6, "proport": [13, 22], "pseudotim": [14, 22], "qc": [6, 17], "qc_filter": [1, 2], "qualiti": [5, 6, 17, 19], "quant": 16, "r": 16, "rank": 20, "rank_genes_group": 20, "read": [5, 16], "receptor": 21, "receptor_ligand": 2, "reclust": [8, 19], "refer": 0, "region": 8, "remov": 6, "renam": 5, "rna": [4, 23], "run_annot": 10, "sampl": 17, "save": [5, 6, 7, 8, 10, 12, 13, 14, 16, 17, 18, 19, 20, 22], "scanpi": 20, "scanpro": 13, "scfate": 14, "score": [5, 17], "sctoolbox": 0, "scvelo": 22, "select": [6, 18], "set": [8, 13, 14, 19], "setup": [5, 6, 7, 8, 10, 12, 16, 17, 18, 19, 21, 22], "seurat": [5, 16], "sex": 17, "show": [6, 10, 17], "singl": 15, "slider": 17, "solo": 16, "sourc": 5, "specif": 5, "splice": 22, "star": 16, "starsolo": [6, 17], "subset": [7, 18], "tabl": [3, 10, 21], "tag": 5, "threshold": [6, 17], "tobia": 2, "tool": 2, "trajectori": 14, "tsne": [8, 19], "tsse": 2, "tsv": 16, "type": 10, "umap": [7, 8, 19], "uncorrect": [7, 18], "unsplic": 22, "up": 12, "us": 20, "util": 3, "variabl": [6, 18], "veloc": [1, 22], "via": 17, "violin": 21, "visual": 22, "welcom": 15, "wrap": [7, 10], "write": 12}})