<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Changelog &mdash; SC FRAMEWORK  documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="_static/plot_directive.css" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=5929fcd5"></script>
        <script src="_static/doctools.js?v=9a2dae69"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="RNA notebooks" href="rna-notebooks/index.html" />
    <link rel="prev" title="Welcome to the documentation of the single cell framework!" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            SC FRAMEWORK
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Changelog</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#id1">0.7.0 (23-04-24)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#changes-to-notebooks">Changes to notebooks</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id2">0.6.1 (28-03-24)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id3">0.6 (27-03-24)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id4">Changes to notebooks</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id5">0.5 (04-03-24)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id6">Changes to notebooks</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id7">0.4 (31-1-24)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id8">Changes to notebooks</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id9">0.3 (30-11-2023)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id12">Changes to notebooks</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id13">0.2 (30-08-2023)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#changes-to-rna-notebooks">Changes to RNA notebooks</a></li>
<li class="toctree-l3"><a class="reference internal" href="#changes-to-atac-notebooks">Changes to ATAC notebooks</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id14">0.1.1 (24-05-2023)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id15">0.1 (22-05-2023)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="rna-notebooks/index.html">RNA notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="atac-notebooks/index.html">ATAC notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="general-notebooks/index.html">General notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="API/index.html">sctoolbox API reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">SC FRAMEWORK</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Changelog</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/CHANGES.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="changelog">
<h1>Changelog<a class="headerlink" href="#changelog" title="Link to this heading"></a></h1>
<section id="id1">
<h2>0.7.0 (23-04-24)<a class="headerlink" href="#id1" title="Link to this heading"></a></h2>
<ul class="simple">
<li><dl class="simple">
<dt>Added code examples for tools and utils (#140)</dt><dd><ul>
<li><p>recluster</p></li>
<li><p>group_heatmap</p></li>
<li><p>plot_venn</p></li>
<li><p>in_range</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>Fix notebooks in readthedocs documentation (#220)</p></li>
<li><p>Removed custom_marker_annotation script</p></li>
<li><p>Disintegrated FLD scoring and added PEAKQC to setup.py (#233)</p></li>
<li><p>fixed PCA-var plot not fitting into anndata_overview (#232)</p></li>
</ul>
<section id="changes-to-notebooks">
<h3>Changes to notebooks<a class="headerlink" href="#changes-to-notebooks" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>Overhaul RNA &amp; ATAC notebooks structure (includes #207)</p></li>
<li><p>Revise RNA notebook 4 recluster section (#201)</p></li>
</ul>
</section>
</section>
<section id="id2">
<h2>0.6.1 (28-03-24)<a class="headerlink" href="#id2" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Fix release pages by renaming the release-pages: job to pages:</p></li>
<li><p>refactor move clean-orphaned-tags to new stage .post (#229)</p></li>
</ul>
</section>
<section id="id3">
<h2>0.6 (27-03-24)<a class="headerlink" href="#id3" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Fix unable to determine R_HOME error (#190)</p></li>
<li><p>implemented propose_pcs to automatically select PCA components (#187)</p></li>
<li><p>add correlation barplot to plot_pca_variance</p></li>
<li><p>created correlation_matrix method by restructuring plot_pca_correlation</p></li>
<li><p>Fix beartype issue with Lists and Iterables containing Literals (#227)</p></li>
<li><p>CICD overhaul (#191)</p></li>
<li><p>fixed notebook version in the env to 6.5.2 (#199, partly #44)</p></li>
</ul>
<section id="id4">
<h3>Changes to notebooks<a class="headerlink" href="#id4" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>Move proportion_analysis notebooks to general notebooks (#195 and #214)</p></li>
<li><p>replace scanpy pseudotime with scFates in pseudotime_analysis notebook</p></li>
<li><p>prepare_for_cellxgene: Adapt to new mampok verison 2.0.9</p></li>
<li><p>prepare_for_cellxgene: Allows the user to set an analyst manually (#213)</p></li>
<li><p>rna 03_batch revision (#209, #202, #200, #152)</p></li>
<li><p>05_marker_genes: Complete Overhaul (#181)</p></li>
</ul>
</section>
</section>
<section id="id5">
<h2>0.5 (04-03-24)<a class="headerlink" href="#id5" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>add receptor_genes &amp; ligand_genes parameters to connectionPlot and decreased runtime</p></li>
<li><p>readme update(#188)</p></li>
<li><p>Fix error when writing adata converted from an R object (#205, #180)</p></li>
<li><p>Marker Repo integration (#162)</p></li>
<li><p>Set scvelo version to &gt;=0.3.1 (#193)</p></li>
<li><p>Added fa2 as dependency for pseudotime analysis</p></li>
<li><p>anndata_overview: fix issue where colorbars for continuous data was not shown</p></li>
<li><p>added ability to use highly variable features using the lsi() function (#165)</p></li>
<li><p>removed deprecated group_heatmap, umap_pub (replaced by gene_expression_heatmap, plot_embedding)</p></li>
<li><p>add doku page</p></li>
<li><p>start change log</p></li>
</ul>
<section id="id6">
<h3>Changes to notebooks<a class="headerlink" href="#id6" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>rna assembly: refactor</p></li>
<li><p>prepare_for_cellxgene: Added BN_public as possible deployment cluster (#192)</p></li>
<li><p>14_velocity_analysis: Remove duplicate parameter (#194)</p></li>
<li><p>pseudotime_analysis: Save generated plots (#211)</p></li>
<li><p>rna 03_batch: added qc metrics to overview plot</p></li>
</ul>
</section>
</section>
<section id="id7">
<h2>0.4 (31-1-24)<a class="headerlink" href="#id7" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Fix get_rank_genes_tables for groups without marker genes (#179)</p></li>
<li><p>Bugfixes for CI jobs</p></li>
<li><p>Fix check_changes pipeline</p></li>
<li><p>Fix typos (#173 &amp; #174)</p></li>
<li><p>Include kwargs in utils.bioutils._overlap_two_bedfiles(#177)</p></li>
<li><p>Implemented _add_path() to automatically add python path to environment</p></li>
<li><p>added tests for _add_path() and _overlap_two_bedfiles() (#177)</p></li>
<li><p>constraint ipywidgets version to 7.7.5 to fix the quality_violinplot() (#151)(#143)</p></li>
<li><p>Add temp_dir to calc_overlap_fc.py (#167) and revised related functions</p></li>
<li><p>more testing (mainly sctoolbox.tools) (#166)</p></li>
<li><p>gerneral text revisions</p></li>
</ul>
<section id="id8">
<h3>Changes to notebooks<a class="headerlink" href="#id8" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>Add pseudotime &amp; velocity analysis notebooks (#164)</p></li>
<li><p>Update receptor-ligand notebook (#176)</p></li>
<li><p>Refactored annotate_genes() from ATAC-notebook 05 to 04 and removed 05 (#175)</p></li>
</ul>
</section>
</section>
<section id="id9">
<h2>0.3 (30-11-2023)<a class="headerlink" href="#id9" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Add parameter type hinting including runtime type checking (#46)</p></li>
<li><p>Fixed prepare_for_cellxgene color issue (#145, #146)</p></li>
<li><p>Add CI/CD container build pipeline for testing (#135)</p></li>
<li><p>Fixed example for gene_expression_heatmap and smaller bugfixes related to marker genes (#124)</p></li>
<li><p>Removed pl.group_heatmap as it is fully covered by pl.gene_expression_heatmap</p></li>
<li><p>Removed ‘sinto’ as dependency and added code in ‘create_fragment_file’ to create fragment file internally (solves #147)</p></li>
<li><p>The function ‘create_fragment_file’ was moved to bam tools.</p></li>
<li><p>Added “n_genes” parameter to tools.marker_genes.get_rank_genes_tables, and set the default to 200 (#153)</p></li>
<li><p>Fixed CI/CD build job rules. Only trigger build job when files changed or triggered manually</p></li>
<li><p>Add parameter to plot_pca_correlation to plot correlation with UMAP components (#157)</p></li>
<li><p>Handle NaN values for plot_pca_correlation (#156)</p></li>
<li><p>implemented prepare_for_cellxgene</p></li>
<li><p>Added pl.embedding.plot_embedding() function to plot embeddings with different styles, e.g. hexbin and density (#149)</p></li>
<li><p>Modified pl.embedding.plot_embedding() to plot different embedding dimensions</p></li>
<li><p>Deprecated pl.umap_pub as this is now covered by pl.plot_embedding</p></li>
<li><p>changed typing to beartype.typing</p></li>
<li><p>Added GenomeTracks plotting</p></li>
<li><p>Fix batch evaluation for small datasets (#148)</p></li>
<li><p>Added <a href="#id10"><span class="problematic" id="id11">**</span></a>kwargs to functions which are wrappers for other functions</p></li>
<li><p>added RAGI cluster validation to clustering.py (!201)</p></li>
<li><p>started disintegrating fld scoring (!201)</p></li>
<li><p>reorganised ATAC-notebooks (!201)</p></li>
</ul>
<section id="id12">
<h3>Changes to notebooks<a class="headerlink" href="#id12" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>Added prepare for cellxgene notebook (#139)</p></li>
<li><p>Added plot of highly expressed genes to RNA notebook 03 (#43)</p></li>
<li><p>Changed structure of notebooks in directory; added “notebooks” subdirectories for RNA and ATAC</p></li>
</ul>
</section>
</section>
<section id="id13">
<h2>0.2 (30-08-2023)<a class="headerlink" href="#id13" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>fix error in prepare_for_cellxgene caused by .uns[_color] not matching .obs column. (#176)</p></li>
<li><p>implemented prepare_for_cellxgene (#147)</p></li>
<li><p>fixed raw value copy issue in rna/02-batch notebook</p></li>
<li><p>Added parameters for the TOBIAS flags in the config file to write_TOBIAS_config()</p></li>
<li><p>Added logging verbose and decorator to ATAC related functions</p></li>
<li><p>Fix “shell not found” error for CI pipeline (#129)</p></li>
<li><p>Pinned scikit-learn to version &lt;=1.2.2 (#128)</p></li>
<li><p>Added script for gene correlation and comparison between two conditions</p></li>
<li><p>Added check for marker gene lists (#103)</p></li>
<li><p>Keep notebook metadata on push to prevent deleting kernel information</p></li>
<li><p>Added sctoolbox as default kernel to RNA &amp; ATAC notebooks</p></li>
<li><p>Added check of column validity to tools.marker_genes.run_DESeq2() (#134)</p></li>
<li><p>Increase test coverage for plotting functions (#126)</p></li>
<li><p>Apply fixes to bugs found by increasing the test coverage.</p></li>
<li><p>Added type hinting to functions.</p></li>
<li><p>Revised doc-strings.</p></li>
<li><p>run_rank_genes() auto converts groupby column to type ‘category’ (#137)</p></li>
<li><p>Fix parameter for gene/cell filtering (#136)</p></li>
<li><p>Add Check to _filter_object() if column contains only boolean (#110)</p></li>
<li><p>Add support of matrx and numpy.ndarray type of adata.X for predict_sex (#111)</p></li>
<li><p>Add method to get pd.DataFrame columns with list of regex (#90)</p></li>
<li><p>Added ‘pairwise_scatter’ method for plotting QC metrics (#54)</p></li>
<li><p>Add ATAC quality metrics TSSe (ENCODE), FRiP</p></li>
<li><p>Revised FLD density plotting</p></li>
<li><p>Adjusted style of default values in docs (#33)</p></li>
<li><p>Added ‘plot_pca_correlation’ for plotting PCA correlation with obs/var columns (#118)</p></li>
<li><p>Removed outdated normalization methods.</p></li>
<li><p>Changed all line endings to LF (#138)</p></li>
<li><p>Disabled threads parameter for tSNE (#130)</p></li>
<li><p>Added ‘plot_starsolo_quality’ and ‘plot_starsolo_UMI’ to plotting module (#78)</p></li>
<li><p>Fixed issues with clustered dotplot with new code (#122)</p></li>
</ul>
<section id="changes-to-rna-notebooks">
<h3>Changes to RNA notebooks<a class="headerlink" href="#changes-to-rna-notebooks" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>Added display of 3D UMAP html in notebook 04 (#119)</p></li>
</ul>
</section>
<section id="changes-to-atac-notebooks">
<h3>Changes to ATAC notebooks<a class="headerlink" href="#changes-to-atac-notebooks" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>Fixed assembling atac notebook 01</p></li>
<li><p>Fixed get_atac_thresholds_wrapper and renamed it to get_thresholds_wrapper</p></li>
<li><p>Added custome cwt implementation</p></li>
<li><p>Added additional parameters to add_insertsize_metrics</p></li>
<li><p>Revised nucleosomal score scoring</p></li>
</ul>
</section>
</section>
<section id="id14">
<h2>0.1.1 (24-05-2023)<a class="headerlink" href="#id14" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Fixed import issue</p></li>
<li><p>Make version accessible</p></li>
<li><p>Added check for CHANGES.rst in gitlab-ci</p></li>
<li><p>Pinned numba==0.57.0rc1 due to import error (#117)</p></li>
<li><p>Fixed bug in tools.norm_correct.atac_norm</p></li>
<li><p>Added check for sctoolbox/_version.py file in gitlab-ci</p></li>
</ul>
</section>
<section id="id15">
<h2>0.1 (22-05-2023)<a class="headerlink" href="#id15" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>First version</p></li>
</ul>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="Welcome to the documentation of the single cell framework!" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="rna-notebooks/index.html" class="btn btn-neutral float-right" title="RNA notebooks" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Loosolab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>